---
title: "Compare Relative Abundance Assignment Methods"
author: "Hanna Berman"
date: "2/06/2019"
output: html_document
---

Compare microbiome relative abundance measures from 16S, Metaphlan2, and Kraken2/Bracken. Taxa of interest (TOI) including *Lactobacillus crispatus*, *Lactobacillus iners*, *Lactobacillus gasseri*, *Lactobacillus jensenii*, *Gardnerella vaginalis*, *Mycoplasma hominis*, *Prevotella bivia*, *Atopobium vaginae* (or only *Lactobacillus* at species level and the four others at genus level)

# Setup
##Load packages
```{r libraries, warning=FALSE}
library(tidyverse); packageVersion("tidyverse")
library(phyloseq); packageVersion("phyloseq")
library(ggpubr) ; packageVersion("ggpubr")
#`%!in%` = Negate(`%in%`)
```

## Define file paths
```{r}
input_dir <- "/Users/hlberman/Documents/GitHub/pregnancy_metagenome/"
PS_path <- file.path(input_dir, "methods_comparisons","PS_RData")
metaphlanPS <- file.path(PS_path, "metaphlan_PS.RData")
brackenPS <- file.path(PS_path, "bracken_PS.RData")
sixteenSPS <- file.path(PS_path, "16S_PS.RData")
output_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/abundanceMethodsComparisons/analysisOutput"
df_path <- file.path(input_dir, "shotgunMetadata.tsv")
```

## load PS objects and dataframe
```{r}
load(file=metaphlanPS)
load(file=brackenPS)
load(file=sixteenSPS)
df <- read.table(df_path, sep="\t", header=TRUE, stringsAsFactors = FALSE)
```

## Plot subject traces
### 16S vs metaphlan w. other
```{r}
ggplot() + 
  geom_line(data=sixteenS_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=metaphlLG_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella  ", "Prevotella  ", "Atopobium  ", "Mycoplasma  ", "Other  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#777777", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Proportions of taxa of interest\n16S (solid) vs. Metaphlan (dashed)", x="Gestation Weeks at Collection",y= "Proportion", color = "Taxon")
#ggsave(file.path(output_dir, "Abundance_metaphlV16S_other.png"), width = 20, height = 15)
```

### 16S vs metaphlan w.o other
```{r}
ggplot() + 
  geom_line(data = sixteenS_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data = metaphlLG_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella  ", "Prevotella  ", "Atopobium  ", "Mycoplasma  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Proportions of taxa of interest\n16S (solid) vs. Metaphlan (dashed)", x="Gestation Weeks at Collection", y="Proportion", color = "Taxon")
#ggsave(file.path(output_dir, "Abundance_metaphlV16S.png"), width = 20, height = 15)
```

### 16S vs bracken w. other
```{r}
ggplot() + 
  geom_line(data= sixteenS_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=brackenLG_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella  ", "Prevotella  ", "Atopobium  ", "Mycoplasma  ", "Other  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#777777", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Proportions of taxa of interest\n16S (solid) vs. Bracken (dashed)", x="Gestation Weeks at Collection", y="Proportions", color = "Taxon") 
#ggsave(file.path(output_dir, "Abundance_brackenV16S_other.png"), width = 20, height = 15)
```

### 16S vs bracken w.o other
```{r}
ggplot() + 
  geom_line(data= sixteenS_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=brackenLG_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella  ", "Prevotella  ", "Atopobium  ", "Mycoplasma  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Proportions of taxa of interest\n16S (solid) vs. Bracken (dashed)", x="Gestation Weeks at Collection", y="Proportion", color = "Taxon") 
#ggsave(file.path(output_dir, "Abundance_brackenV16S.png"), width = 20, height = 15)
```

### metaphlan vs bracken w. other
```{r}
ggplot() + 
  geom_line(data= metaphlA_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=brackenA_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella vaginalis", "Prevotella bivia", "Atopobium vaginae", "Mycoplasma hominis", "Other  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#777777", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Proportions of taxa of interest\nMetaphlan (solid) vs. Bracken (dashed)", x="Gestation Weeks at Collection", y="Proportion", color = "Taxon")
#ggsave(file.path(output_dir, "Abundance_brackenVmetaphlan_other.png"), width = 20, height = 15)
```

### metaphlan vs bracken w.o other
```{r}
ggplot() + 
  geom_line(data= metaphlA_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=brackenA_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella vaginalis", "Prevotella bivia", "Atopobium  vaginae", "Mycoplasma hominis"),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Proportions of taxa of interest\nMetaphlan (solid) vs. Bracken (dashed)", x="Gestation Weeks at Collection", "Proportion", color = "Taxon")
#ggsave(file.path(output_dir, "Abundance_brackenVmetaphlan.png"), width = 20, height = 15)
```

## Scatterplot by taxa

### Make merged data frame 
```{r}
#data frames with taxa not in toi listed as "other"
meta_brack_16S_other <- metaphlLG_toi_o_df %>%
  select(SampleID_a, PaperCohort, RNAlater, Genus, Species, Abundance) %>%
  full_join(brackenLG_toi_o_df[,c("SampleID_a", "Genus", "Species", "Abundance","RNAlater", "PaperCohort")], by=c("SampleID_a", "Genus", "Species", "RNAlater", "PaperCohort")) %>%
  full_join(sixteenS_toi_o_df[,c("SampleID_a", "Genus", "Species", "Abundance", "RNAlater", "PaperCohort")], by=c("SampleID_a", "Genus", "Species", "RNAlater", "PaperCohort")) %>%
  left_join(df[,c("SampleID", "SampleID_a")], by="SampleID_a") %>%
  rename(Abundance_metaphl=Abundance.x, Abundance_brack=Abundance.y, Abundance_16S=Abundance) %>%
  replace_na(list(Abundance_metaphl=0, Abundance_brack=0, Abundance_16S=0)) %>% #make undetected values 0
  mutate(Abundance_16S=case_when(!is.na(PaperCohort) ~ na_if(Abundance_16S, TRUE)), #make sure samples that do not have 16S data do not have 0 values for 16S abundance
         Taxon=paste(Genus, Species),
         PaperCohort=as.character(PaperCohort))    
         
meta_brack_other <- metaphlA_toi_o_df %>%
  select(SampleID, SampleID_a, PaperCohort, RNAlater,Genus, Species, Abundance) %>%
  full_join(brackenA_toi_o_df[,c("SampleID", "SampleID_a", "Genus", "Species", "Abundance", "RNAlater", "PaperCohort")], by=c("SampleID", "SampleID_a", "Genus", "Species", "RNAlater", "PaperCohort")) %>%
  rename(Abundance_metaphl=Abundance.x, Abundance_brack=Abundance.y) %>%
  replace_na(list(Abundance_metaphl=0, Abundance_brack=0, Abundance_16S=0)) %>%
  mutate(Taxon=paste(Genus, Species))

#data frames with toi taxa renormalized to 1
meta_brack_16S <- metaphlLG_toi_df %>%
  select(SampleID_a, PaperCohort, RNAlater,Genus, Species, Abundance) %>%
  full_join(brackenLG_toi_df[,c("SampleID_a", "Genus", "Species", "Abundance","RNAlater", "PaperCohort")], by=c("SampleID_a", "Genus", "Species", "RNAlater", "PaperCohort")) %>%
  full_join(sixteenS_toi_df[,c("SampleID_a", "Genus", "Species", "Abundance", "RNAlater", "PaperCohort")], by=c("SampleID_a", "Genus", "Species", "RNAlater", "PaperCohort")) %>%
  left_join(df[,c("SampleID", "SampleID_a")], by="SampleID_a") %>%
  rename(Abundance_metaphl=Abundance.x, Abundance_brack=Abundance.y, Abundance_16S=Abundance) %>%
  replace_na(list(Abundance_metaphl=0, Abundance_brack=0, Abundance_16S=0)) %>%
  mutate(Abundance_16S=case_when(!is.na(PaperCohort) ~ na_if(Abundance_16S, TRUE)),
         Taxon=paste(Genus, Species),
         PaperCohort=as.character(PaperCohort))

meta_brack <- metaphlA_toi_df %>%
  select(SampleID, SampleID_a, PaperCohort, RNAlater,Genus, Species, Abundance) %>%
  full_join(brackenA_toi_df[,c("SampleID", "SampleID_a", "Genus", "Species", "Abundance", "RNAlater", "PaperCohort")], by=c("SampleID", "SampleID_a", "Genus", "Species", "RNAlater", "PaperCohort")) %>%
  rename(Abundance_metaphl=Abundance.x, Abundance_brack=Abundance.y) %>%
  replace_na(list(Abundance_metaphl=0, Abundance_brack=0, Abundance_16S=0)) %>%
  mutate(Taxon=paste(Genus, Species),
         PaperCohort=as.character(PaperCohort))
```

### 16 vs. Metaphlan w. other
```{r}
ggplot(data= meta_brack_16S_other) + 
  geom_point(aes(x = Abundance_metaphl, y = Abundance_16S, color=RNAlater), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Proportions:\n16S Amplicon vs. Metaphlan", x="Proportion (Metaphlan)", y="Proportion (16S Amplicon)", color = "RNA Later")
#ggsave(file.path(output_dir, "Metaphlan16S_byTaxon_o_rnalater.png"), height=8, width=8)

ggplot(data= meta_brack_16S_other) + 
  geom_point(aes(x = Abundance_metaphl, y = Abundance_16S, color=PaperCohort), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Proportions:\n16S Amplicon vs. Metaphlan", x="Proportion (Metaphlan)", y="Proportion (16S Amplicon)", color = "Paper Cohort ")
#ggsave(file.path(output_dir, "Metaphlan16S_byTaxon_o_papercohort.png"), height=8, width=8)
```

### 16 vs. Bracken w. other
```{r}
ggplot(data= meta_brack_16S_other) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_16S, color=RNAlater), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Porportions:\n16S Amplicon vs. Bracken", x="Proportions (Bracken)", y="Proportions (16S Amplicon)", color = "RNA Later")
#ggsave(file.path(output_dir, "Bracken16S_byTaxon_o_rnalater.png"), height=8, width=8)

ggplot(data= meta_brack_16S_other) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_16S, color=PaperCohort), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Porportions:\n16S Amplicon vs. Bracken", x="Proportions (Bracken)", y="Proportions (16S Amplicon)", color = "Paper Cohort") 
#ggsave(file.path(output_dir, "Bracken16S_byTaxon_o_papercohort.png"), height=8, width=8)
```

### Metaphlan vs. Bracken w. other
```{r}
ggplot(data= meta_brack_other) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_metaphl), color="#000000", stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Proportions:\nMetaphlan vs. Bracken", x="Proportion (Bracken)", y="Proportion (Metaphlan)")
#ggsave(file.path(output_dir, "BrackenMetaphl_byTaxon_o.png"), height=8, width=8)
```

### 16 vs. Metaphlan w.o other
```{r}
ggplot(data= meta_brack_16S) + 
  geom_point(aes(x = Abundance_metaphl, y = Abundance_16S, color=RNAlater), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Proportions:\n16S Amplicon vs. Metaphlan", x="Proportion (Metaphlan)", y="Proportion (16S Amplicon)", color = "RNA Later")
#ggsave(file.path(output_dir, "Metaphlan16S_byTaxon_rnalater.png"), height=8, width=8)
  
ggplot(data= meta_brack_16S) + 
  geom_point(aes(x = Abundance_metaphl, y = Abundance_16S, color=PaperCohort), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Proportions:\n16S Amplicon vs. Metaphlan", x="Proportion (Metaphlan)", y="Proportion (16S Amplicon)", color = "Paper Cohort")
#ggsave(file.path(output_dir, "Metaphlan16S_byTaxon_papercohort.png"), height=8, width=8)
```

### 16 vs. Bracken w.o other
```{r}
ggplot(data= meta_brack_16S) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_16S, color=RNAlater), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Proportions:\n16S Amplicon vs. Bracken", x="Proportion (Bracken)", y="Proportion (16S Amplicon)", color = "RNA Later")
#ggsave(file.path(output_dir, "Bracken16S_byTaxon_rnalater.png"), height=8, width=8)

ggplot(data= meta_brack_16S) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_16S, color=PaperCohort), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Proportions:\n16S Amplicon vs. Bracken", x="Proportion (Bracken)", y="Proportion (16S Amplicon)", color = "Paper Cohort")
#ggsave(file.path(output_dir, "Bracken16S_byTaxon_papercohort.png"), height=8, width=8)
```

### Metaphlan vs. Bracken w.o other
```{r}
ggplot(data= meta_brack) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_metaphl), color="#000000", stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Proportions:\nMetaphlan vs. Bracken", x="Proportion (Bracken)", y="Proportion (Metaphlan)")
#ggsave(file.path(output_dir, "BrackenMetaphl_byTaxon.png"), height=8, width=8)
```

# Abundance vs. presence absence
## Add presence absence informtation to tables
```{r}
#for vs. 16S
meta_brack_16S_other <- meta_brack_16S_other %>%
  mutate(metaphl_yn=case_when(Abundance_metaphl == 0 ~ "No", 
                              Abundance_metaphl > 0 ~ "Yes"), 
         brack_yn=case_when(Abundance_brack == 0 ~ "No",
                            Abundance_brack > 0 ~ "Yes"),
         sixteenS_yn=case_when(Abundance_16S == 0 ~ "No", 
                               Abundance_16S > 0 ~ "Yes"))

#for metaphlan vs. bracken only
meta_brack_other <- meta_brack_other %>%
  mutate(metaphl_yn=case_when(Abundance_metaphl == 0 ~ "No", 
                              Abundance_metaphl > 0 ~ "Yes"), 
         brack_yn=case_when(Abundance_brack == 0 ~ "No",
                            Abundance_brack > 0 ~ "Yes"))
```

## Plot 
### By RNA Later
```{r}
#metaphlan vs. 16S
ggplot(meta_brack_16S_other)+ 
  geom_point(aes(x=Abundance_metaphl, y=sixteenS_yn, color=RNAlater)) +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\n16S Amplicon vs. Metaphlan", x="Proportion (Metaphlan)", y="Detected (16S Amplicon)") 
#ggsave(file.path(output_dir, "presVprop_16SVMetaphl_rnalater.png"), height = 9, width = 9)

ggplot(meta_brack_16S_other)+ 
  geom_point(aes(x=Abundance_16S, y=metaphl_yn, color=RNAlater)) +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\nMetaphlan vs. 16S Amplicon", x="Proportion (16S Amplicon)", y="Detected (Metaphlan)") 
#ggsave(file.path(output_dir, "presVprob_metaphlV16S_rnalater.png"), height = 9, width = 9)

#bracken vs. 16S
ggplot(meta_brack_16S_other)+ 
  geom_point(aes(x=Abundance_brack, y=sixteenS_yn, color=RNAlater)) +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\n16S Amplicon vs. Bracken", x="Proportion (Bracken)", y="Detected (16S Amplicon)") 
#ggsave(file.path(output_dir, "presVprop_16SVbrack_rnalater.png"), height = 9, width = 9)

ggplot(meta_brack_16S_other)+ 
  geom_point(aes(x=Abundance_16S, y=brack_yn, color=RNAlater)) +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\nBracken vs. 16S Amplicon", x="Proportion (16S Amplicon)", y="Detected (Bracken)") 
#ggsave(file.path(output_dir, "presVprop_brackV16S_rnalater.png"), height = 9, width = 9)

#bracken vs. metaphlan
ggplot(meta_brack_other)+ 
  geom_point(aes(x=Abundance_brack, y=metaphl_yn),  color = "#000000") +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\nMetaphlan vs. Bracken", x="Proportion (Bracken)", y="Detected (Metaphlan)") 
#ggsave(file.path(output_dir, "presVprop_metaphlVbrack.png"), height = 9, width = 9)

ggplot(meta_brack_other)+ 
  geom_point(aes(x=Abundance_metaphl, y=brack_yn), color="#000000") +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\nBracken vs. Metaphlan", x="Proportion (Metaphlan)", y="Detected (Bracken)") 
#ggsave(file.path(output_dir, "presVprop_brackVmetaphl.png"), height = 9, width = 9)
```

### By Paper Cohort (for 16S)
```{r}
#metaphlan vs. 16S
ggplot(meta_brack_16S_other)+ 
  geom_point(aes(x=Abundance_metaphl, y=sixteenS_yn, color=PaperCohort)) +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\n16S Amplicon vs. Metaphlan", x="Proportion (Metaphlan)", y="Detected (16S Amplicon)")
#ggsave(file.path(output_dir, "presVprop_16SVMetaphl_papercohort.png"), height = 9, width = 9)

ggplot(meta_brack_16S_other)+ 
  geom_point(aes(x=Abundance_16S, y=metaphl_yn, color=PaperCohort)) +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\nMetaphlan vs. 16S Amplicon", x="Proportion (16S Amplicon)", y="Detected (Metaphlan)")
#ggsave(file.path(output_dir, "presVprob_metaphlV16S_papercohort.png"), height = 9, width = 9)

#bracken vs. 16S
ggplot(meta_brack_16S_other)+ 
  geom_point(aes(x=Abundance_brack, y=sixteenS_yn, color=PaperCohort)) +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\n16S Amplicon vs. Bracken", x="Proportion (Bracken)", y="Detected (16S Amplicon)")
#ggsave(file.path(output_dir, "presVprop_16SVbrack_papercohort.png"), height = 9, width = 9)

ggplot(meta_brack_16S_other)+ 
  geom_point(aes(x=Abundance_16S, y=brack_yn, color=PaperCohort)) +
  facet_wrap(~Taxon) +
  labs(title = "Presence vs. Proportion\nBracken vs. 16S Amplicon", x="Proportion (16S Amplicon)", y="Detected (Bracken)")
#ggsave(file.path(output_dir, "presVprop_brackV16S_papercohort.png"), height = 9, width = 9)
```

# Log Ratios of Taxa Pairs
##Compute
```{r}
#get ratio comparison conditions
taxaLG <- c("Lactobacillus iners", "Lactobacillus gasseri", "Lactobacillus crispatus", "Lactobacillus jensenii", "Gardnerella  ", "Mycoplasma  ", "Prevotella  ", "Atopobium  ")
pairLG <- combn(taxaLG, 2) %>%
  as.data.frame() %>%
  t()

taxaS <- c("Lactobacillus iners", "Lactobacillus gasseri", "Lactobacillus crispatus", "Lactobacillus jensenii", "Gardnerella vaginalis", "Mycoplasma hominis", "Prevotella bivia", "Atopobium vaginae")
pairS <- combn(taxaS, 2) %>%
  as.data.frame() %>%
  t()

#log ratios (lactobacillus at species level, only)
ratiosLG <- cbind(SampleID_a=rep(df$SampleID_a, nrow(pairLG)), #make data_frame with list of all taxa pairs for each sample
              Taxon=rep(pairLG[,1], nrow(df)),
              Taxon2=rep(pairLG[,2], nrow(df))) %>%
  as.data.frame(stringsAsFactors=FALSE) %>%
  mutate(SampleID_a=as.numeric(as.character(SampleID_a))) %>%
  left_join(meta_brack_16S_other[,c("SampleID_a", "RNAlater", "PaperCohort", "Taxon", "Abundance_16S", "Abundance_metaphl", "Abundance_brack")], by=c("SampleID_a", "Taxon")) %>% #add abundance for first taxon of each pair
  rename(Taxon1=Taxon, #rename columns
         ab_16S1=Abundance_16S,
         ab_metaphl1=Abundance_metaphl,
         ab_brack1=Abundance_brack,
         Taxon=Taxon2) %>%
  left_join(meta_brack_16S_other[,c("SampleID_a", "Taxon", "Abundance_16S", "Abundance_metaphl", "Abundance_brack")], by=c("SampleID_a", "Taxon")) %>% #add abundance for second taxon of each pair
  rename(Taxon2=Taxon, #rename columns
         ab_16S2=Abundance_16S,
         ab_metaphl2=Abundance_metaphl,
         ab_brack2=Abundance_brack) %>%
  mutate(Taxa=paste0(Taxon1, ":", Taxon2), #compute log ratios
         logRatio_16S=log10(ab_16S1/ab_16S2),
         logRatio_metaphlan=log10(ab_metaphl1/ab_metaphl2),
         logRatio_Braken=log10(ab_brack1/ab_brack2))
ratiosLG[ratiosLG=="Inf"|ratiosLG=="-Inf"|ratiosLG=="NaN"] <- NA

#log ratios (all taxa at species level) 
ratiosS <- cbind(SampleID_a=rep(df$SampleID_a, nrow(pairLG)), #make data_frame with list of all taxa pairs for each sample
              Taxon=rep(pairS[,1], nrow(df)),
              Taxon2=rep(pairS[,2], nrow(df))) %>%
  as.data.frame(stringsAsFactors=FALSE) %>%
  mutate(SampleID_a=as.numeric(as.character(SampleID_a))) %>%
  left_join(meta_brack_other[,c("SampleID_a", "RNAlater", "PaperCohort", "Taxon",  "Abundance_metaphl", "Abundance_brack")], by=c("SampleID_a", "Taxon")) %>% #add abundance for first taxon of each pair
  rename(Taxon1=Taxon,
         ab_metaphl1=Abundance_metaphl,
         ab_brack1=Abundance_brack,
         Taxon=Taxon2) %>%
  left_join(meta_brack_other[,c("SampleID_a", "Taxon", "Abundance_metaphl", "Abundance_brack")], by=c("SampleID_a", "Taxon")) %>% #add abundance for second taxon of each pair
  rename(Taxon2=Taxon,
         ab_metaphl2=Abundance_metaphl,
         ab_brack2=Abundance_brack) %>%
  mutate(Taxa=paste0(Taxon1, ":", Taxon2), #compute log ratios
         logRatio_metaphlan=log10(ab_metaphl1/ab_metaphl2),
         logRatio_Braken=log10(ab_brack1/ab_brack2)) 
ratiosS[ratiosS=="Inf"|ratiosS=="-Inf"|ratiosS=="NaN"] <- NA
```

## Plot
### By RNA Later
```{r}
ggplot(data=ratiosLG) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_16S, color=RNAlater), na.rm = TRUE) +
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S), se=FALSE, color="#000000", method = "lm", na.rm=TRUE) + # all
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S, group=RNAlater, color=RNAlater), se=FALSE, method = "lm", show.legend = FALSE, na.rm=TRUE) + # by RNA later 
        labs(title = "Log ratios of taxa:\n16S Amplicon vs. Metaphlan", x="Metaphlan", y="16S Amplicon") + 
        theme(text = element_text(size=7),
              axis.title = element_text(size = 19),
              title=element_text(size=20),
              legend.text = element_text(size=15),
              strip.text = element_text(size=9),
              axis.text = element_text(size=10)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) + 
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~")), label.y = -5, na.rm = TRUE, formula=y~x) +
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~"), color=RNAlater, group=RNAlater), label.y = c(-3,-4), na.rm = TRUE, formula=y~x, show.legend = FALSE)
#ggsave(file.path(output_dir, "logRatio_16SVmetaphl_rnalater.png"),height = 15, width = 18.5)

ggplot(data=ratiosLG) +
        geom_point(aes(x=logRatio_bracken, y=logRatio_16S, color=RNAlater), na.rm = TRUE) +
        geom_smooth(aes(x=logRatio_bracken, y=logRatio_16S), se=FALSE, color="#000000", method = "lm", na.rm=TRUE) + # all
        geom_smooth(aes(x=logRatio_bracken, y=logRatio_16S, group=RNAlater, color=RNAlater), se=FALSE, method = "lm", show.legend = FALSE, na.rm=TRUE) + # by RNA later 
        labs(title = "Log ratios of taxa:\n16S Amplicon vs. Bracken", x="Bracken", y="16S Amplicon") + 
        theme(text = element_text(size=7),
              axis.title = element_text(size = 19),
              title=element_text(size=20),
              legend.text = element_text(size=15),
              strip.text = element_text(size=9),
              axis.text = element_text(size=10)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) +
        stat_regline_equation(aes(x=logRatio_bracken, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~")), label.y = -5, na.rm = TRUE, formula=y~x) +
        stat_regline_equation(aes(x=logRatio_bracken, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~"), color=RNAlater, group=RNAlater), label.y = c(-3,-4), na.rm = TRUE, formula=y~x, show.legend = FALSE)
#ggsave(file.path(output_dir, "logRatio_16SVbrack_RNAlater.png"),height = 15, width = 18.5)

ggplot(data=ratiosS) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_bracken), color = "#000000", na.rm = TRUE) +
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_bracken), se=FALSE, color="#000000", method = "lm", na.rm=TRUE) + # all
        labs(title = "Log ratios of taxa:\nBracken vs. Metaphlan", x="Metaphlan", y="Bracken") + 
        theme(text = element_text(size=7),
              axis.title = element_text(size = 19),
              title=element_text(size=20),
              legend.text = element_text(size=15),
              strip.text = element_text(size=9),
              axis.text = element_text(size=10)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) +
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_bracken, label = paste(..eq.label.., ..rr.label.., sep = "~~~")), label.y = -5, na.rm = TRUE, formula=y~x, color = "#000000") 
#ggsave(file.path(output_dir, "logRatio_brackVmetaphl.png"),height = 15, width = 18.5)
```

### By Paper Cohort
```{r}
ggplot(data=ratiosLG) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_16S, color=PaperCohort), na.rm = TRUE) +
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S), se=FALSE, color="#000000", method = "lm", na.rm=TRUE) + # all
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S, group=PaperCohort, color=PaperCohort), se=FALSE, method = "lm", show.legend = FALSE, na.rm=TRUE) + # by RNA later 
        labs(title = "Log ratios of taxa:\n16S Amplicon vs. Metaphlan", x="Metaphlan", y="16S Amplicon") + 
        theme(text = element_text(size=7),
              axis.title = element_text(size = 19),
              title=element_text(size=20),
              legend.text = element_text(size=15),
              strip.text = element_text(size=9),
              axis.text = element_text(size=10)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) + 
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~")), label.y = -5, na.rm = TRUE, formula=y~x) +
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~"), color=PaperCohort, group=PaperCohort), label.y = c(-3,-4), na.rm = TRUE, formula=y~x, show.legend = FALSE) #+
#ggsave(file.path(output_dir, "logRatio_16SVmetaphl_papercohort.png"),height = 15, width = 18.5)

ggplot(data=ratiosLG) +
        geom_point(aes(x=logRatio_bracken, y=logRatio_16S, color=PaperCohort), na.rm = TRUE) +
        geom_smooth(aes(x=logRatio_bracken, y=logRatio_16S), se=FALSE, color="#000000", method = "lm", na.rm=TRUE) + # all
        geom_smooth(aes(x=logRatio_bracken, y=logRatio_16S, group=PaperCohort, color=PaperCohort), se=FALSE, method = "lm", show.legend = FALSE, na.rm=TRUE) + # by RNA later 
        labs(title = "Log ratios of taxa:\n16S Amplicon vs. Bracken", x="Bracken", y="16S Amplicon") + 
        theme(text = element_text(size=7),
              axis.title = element_text(size = 19),
              title=element_text(size=20),
              legend.text = element_text(size=15),
              strip.text = element_text(size=9),
              axis.text = element_text(size=10)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) +
        stat_regline_equation(aes(x=logRatio_bracken, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~")), label.y = -5, na.rm = TRUE, formula=y~x) +
        stat_regline_equation(aes(x=logRatio_bracken, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~"), color=PaperCohort, group=PaperCohort), label.y = c(-3,-4), na.rm = TRUE, formula=y~x, show.legend = FALSE) #+
#ggsave(file.path(output_dir, "logRatio_16SVbrack_papercohort.png"),height = 15, width = 18.5)
```

## Filtering
Mean and spread of logratios
```{r}
sapply(list(S16=ratiosLG$logRatio_16S, metaphl=ratiosLG$logRatio_metaphlan, brack=ratiosLG$logRatio_bracken), summary)
ratiosLG0 <- ratiosLG %>%
  gather(Method, "logRatio", 3:5) %>%
  mutate(Method=case_when(Method=="logRatio_metaphlan"~"Metaphlan2",
                          Method=="logRatio_bracken"~"Bracken",
                          Method=="logRatio_16S"~"16S"))
ggplot(data=ratiosLG0) +
  geom_boxplot(aes(x=Method, y=logRatio)) +
  scale_y_continuous(name="Log Ratio", limits=c(-5,5), breaks = seq(from=-5, to=5, by=5)) +
  coord_flip() +
  theme(text=element_text(size=5), 
        axis.text = element_text(size = 8), 
        axis.title = element_text(size=10)) +
  facet_wrap(~Taxa)
#ggsave(file.path(output_dir, "ratioBoxplots.png"))
```

### Filter based on 16S and redo log ratios 
```{r}
#find greatest value of 16S not in metaphlan
S16values <- meta_brack_16S_other$Abundance_16S[meta_brack_16S_other$metaphl_yn=="No"]
max(S16values, na.rm = TRUE)

ratiosLG0 <- ratiosLG %>%
  filter(ab_16S1>max(S16values, na.rm = TRUE),
         ab_16S2>max(S16values, na.rm = TRUE)) #filter on the 0.0395 value
```

Plot
```{r}
#by RNA later
ggplot(data=ratiosLG0) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_16S, color=RNAlater), na.rm = TRUE) +
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S), se=FALSE, color="#000000", method = "lm", na.rm=TRUE) + # all
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S, group=RNAlater, color=RNAlater), se=FALSE, method = "lm", show.legend = FALSE, na.rm=TRUE) + # by RNA later 
        labs(title = "Log ratios of taxa:\n16S Amplicon vs. Metaphlan (filter 16S<0.0395)", x="Metaphlan", y="16S Amplicon") + 
        theme(text = element_text(size=7),
              axis.title = element_text(size = 19),
              title=element_text(size=20),
              legend.text = element_text(size=15),
              strip.text = element_text(size=9),
              axis.text = element_text(size=10)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) + 
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~")), label.y = -5, na.rm = TRUE, formula=y~x) +
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~"), color=RNAlater, group=RNAlater), label.y = c(-3,-4), na.rm = TRUE, formula=y~x, show.legend = FALSE)
ggsave(file.path(output_dir, "logRatio_16SVmetaphl_rnalater_0.0395.png"),height = 15, width = 18.5)

ggplot(data=ratiosLG0) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_16S, color=PaperCohort), na.rm = TRUE) +
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S), se=FALSE, color="#000000", method = "lm", na.rm=TRUE) + # all
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S, group=PaperCohort, color=PaperCohort), se=FALSE, method = "lm", show.legend = FALSE, na.rm=TRUE) + # by RNA later 
        labs(title = "Log ratios of taxa:\n16S Amplicon vs. Metaphlan (filter 16S<0.0395)", x="Metaphlan", y="16S Amplicon") + 
        theme(text = element_text(size=7),
              axis.title = element_text(size = 19),
              title=element_text(size=20),
              legend.text = element_text(size=15),
              strip.text = element_text(size=9),
              axis.text = element_text(size=10)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) + 
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~")), label.y = -5, na.rm = TRUE, formula=y~x) +
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~"), color=PaperCohort, group=PaperCohort), label.y = c(-3,-4), na.rm = TRUE, formula=y~x, show.legend = FALSE)
ggsave(file.path(output_dir, "logRatio_16SVmetaphl_papercohort_0.0395.png"),height = 15, width = 18.5)
```

```{r}
ratiosLG1 <- ratiosLG %>% #filter on 0.02 (arbitrary)
  filter(ab_16S1>0.02,
         ab_16S2>0.02)
```


Plot
```{r}
#by RNA later
ggplot(data=ratiosLG1) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_16S, color=RNAlater), na.rm = TRUE) +
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S), se=FALSE, color="#000000", method = "lm", na.rm=TRUE) + # all
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S, group=RNAlater, color=RNAlater), se=FALSE, method = "lm", show.legend = FALSE, na.rm=TRUE) + # by RNA later 
        labs(title = "Log ratios of taxa:\n16S Amplicon vs. Metaphlan (filter 16S<0.02)", x="Metaphlan", y="16S Amplicon") + 
        theme(text = element_text(size=7),
              axis.title = element_text(size = 19),
              title=element_text(size=20),
              legend.text = element_text(size=15),
              strip.text = element_text(size=9),
              axis.text = element_text(size=10)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) + 
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~")), label.y = -5, na.rm = TRUE, formula=y~x) +
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~"), color=RNAlater, group=RNAlater), label.y = c(-3,-4), na.rm = TRUE, formula=y~x, show.legend = FALSE)
ggsave(file.path(output_dir, "logRatio_16SVmetaphl_rnalater_0.02.png"),height = 15, width = 18.5)

ggplot(data=ratiosLG1) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_16S, color=PaperCohort), na.rm = TRUE) +
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S), se=FALSE, color="#000000", method = "lm", na.rm=TRUE) + # all
        geom_smooth(aes(x=logRatio_metaphlan, y=logRatio_16S, group=PaperCohort, color=PaperCohort), se=FALSE, method = "lm", show.legend = FALSE, na.rm=TRUE) + # by RNA later 
        labs(title = "Log ratios of taxa:\n16S Amplicon vs. Metaphlan (filter 16S<0.02)", x="Metaphlan", y="16S Amplicon") + 
        theme(text = element_text(size=7),
              axis.title = element_text(size = 19),
              title=element_text(size=20),
              legend.text = element_text(size=15),
              strip.text = element_text(size=9),
              axis.text = element_text(size=10)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) + 
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~")), label.y = -5, na.rm = TRUE, formula=y~x) +
        stat_regline_equation(aes(x=logRatio_metaphlan, y=logRatio_16S, label = paste(..eq.label.., ..rr.label.., sep = "~~~"), color=PaperCohort, group=PaperCohort), label.y = c(-3,-4), na.rm = TRUE, formula=y~x, show.legend = FALSE)
ggsave(file.path(output_dir, "logRatio_16SVmetaphl_papercohort_0.02.png"),height = 15, width = 18.5)
```

# Session Info
```{r}
sessionInfo()
```