---
title: "Gardnerella Core Genome Phylogeny"
author: "HLB"
date: "October 2020"
output: 
  html_document:
    toc: true
    toc_float: true
---
# Set up
## Packages
```{r, warning=FALSE}
library(tidyverse)
```
```{r, warning=FALSE, message=FALSE}
library(ggrepel)
library(treeio)
library(ggtree)
library(ape)
library(Biostrings)
library(formattable)
library(ggdendro)
library(kableExtra)
library(magick)
`%!in%` <- negate(`%in%`)
```

## File Paths
```{r}
dataPath <- "/Volumes/GoogleDrive/My\ Drive/Callahan\ Lab/metagenome_gardnerella/gardPhylogeny"

# Metadata
gardDF <- "./GardnerellaMetadata.csv" %>%
  read_csv() %>%
  filter(in_phylogeny==TRUE)

# phylogeny outputs
prokkaStatsDir <- file.path(dataPath, "prokka_annotated_genomes", "txt")

roaryPresAbs <- "./build_phylogeny/20201002_Roary95/gene_presence_absence.csv" %>%
  read_csv

pangenomeStats <- "./build_phylogeny/20201002_Roary95/summary_statistics.txt" %>%
  read_delim(delim="\t", col_names = c("gene_type", "definition", "count")) %>%
  filter(gene_type!="Total genes") %>%
  mutate(gene_type=factor(gene_type, levels=c("Core genes", "Soft core genes", "Shell genes", "Cloud genes")))

gardTree <- "./build_phylogeny/20201005_ml_phylogeny/RAxML_originalLabelledTree.rootedGardTree"
rootedGardTree <- "./build_phylogeny/20201005_ml_phylogeny/RAxML_labelledTree.rootedGardTree"

ffnDir <- file.path(dataPath, "prokka_annotated_genomes", "ffn")

# whole genome mash distances
load(file.path(dataPath, "assembly_quality_control/completeness_quality_dereplication/mashDists.Rdata"))
assemblyDistDF <- mashDists
remove(mashDists)

# output paths
figureOut <- "/Volumes/GoogleDrive/My\ Drive/Callahan\ Lab/metagenome_gardnerella/FIGURES/V3_manuscript_figures/gard_phyl"
coreGeneFastasDir <- file.path(dataPath, "core_gene_fastas")
assemblyCoreGenesDir <- file.path(dataPath, "assembly_core_genes")
```

## Read in genomes
```{r}
refAssemblies <- ffnDir %>%
  list.files(pattern = ".ffn", full.names = TRUE) %>%
  map(~readDNAStringSet(.x))

names(refAssemblies) <- ffnDir %>%
  list.files(pattern = ".ffn") %>%
  str_extract(".*(?=\\.ffn)")

# take a look at what sequence names look like
head(names(refAssemblies$`00703Bmash`))
```

## ggplot
```{r}
# theme
theme_set(theme_bw()+
          theme(panel.grid = element_blank(),
                axis.text = element_text(size=10),
                axis.title = element_text(size=15),                                                                                                  
                legend.text = element_text(size=12),
                legend.title = element_text(size=14)))

# Colors
cladeColors <- list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))),
                     scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))))

binaryColors <- list(scale_color_manual(values=c("#0072B2", "#D55E00")),
                     scale_fill_manual(values=c("#0072B2", "#D55E00")))

okabeIto <- c("#E69F00", "#F0E442", "#0072B2", "#009E73", "#56B4E9", "#D55E00", "#CC79A7", "#000000")
```

# Gardnerella Assembly Summary Stats
## Genome sizes
```{r}
meanGenomeLength <- mean(gardDF$total_length)
meanGenomeLength
sd(gardDF$total_length)
range(gardDF$total_length)

gardDF %>%
  ggplot(aes(total_length)) +
  geom_histogram() +
  geom_vline(xintercept = meanGenomeLength) +
  labs(x="Genome Length (Mbp)", y="Count")
#ggsave(file.path(figureOut, "20201009_genomeLengthHist.png"))
```

## Genes identified by Prokka
```{r, message=FALSE}
prokkaStats <- prokkaStatsDir %>%
  list.files(pattern=".txt", full.names = TRUE) %>%
  map(~read_delim(.x, delim="\n", col_names = c("category"))) %>%
  map(~separate(.x, category, into = c("category", "result"), sep=": ")) %>%
  map(~spread(.x, category, result)) %>%
  purrr::reduce(full_join) %>%
  separate(organism, c("Genus", "Species", "Strain"), sep=" ") %>%
  select(Strain, Genus, bases, contigs, gene, mRNA, CDS, rRNA, tRNA, tmRNA, repeat_region)
  
head(prokkaStats)
```

## Core and Pangenome
  Total number of genes in the pangenome:
```{r}
sum(pangenomeStats$count)
```

  Representation:
```{r}
# Donut chart
pangenomeStats %>%
  mutate(fraction=count/sum(count), # compute percentages
         ymax=cumsum(fraction), # cumulative percentages (top of each rectangle)
         ymin=c(0, head(ymax, n=-1)), # bottom of each rectangle
         labelPosition=((ymax+ymin)/2), # label position
         label=paste(gene_type, definition, count, sep="\n")) %>% 
  ggplot(aes(ymax=ymax, ymin=ymin, xmax=20, xmin=1, fill=gene_type)) +
  geom_rect() +
  #geom_text_repel(x=21, aes(y=labelPosition, label=label), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values=okabeIto[1:4], labels=paste0(pangenomeStats$gene_type, "\n", pangenomeStats$definition, ":\n", pangenomeStats$count)) +
  coord_polar(theta="y") +
  xlim(c(-20, 20)) +
  theme_void() +
  theme(legend.position = "right",
        legend.text = element_text(size=13)) +
  labs(fill=NULL)
ggsave(file.path(figureOut, "20210819_corePanGenesFraction.png"))
```

# Gardnerella Phylogeny
```{r}
tree <- "./build_phylogeny/20201005_ml_phylogeny/RAxML_bipartitionsBranchLabels.bootstrappedGardTree" %>%
  read.raxml()

root <- "./build_phylogeny/20201005_ml_phylogeny/RAxML_labelledTree.rootedGardTree" %>%
  treeio::read.newick()

placement <- "./build_phylogeny/20201005_ml_phylogeny/RAxML_portableTree.rootedGardTree.jplace" %>%
  read.jplace()
```

```{r, fig.width=7, fig.height=7}
outgroupRootTree <- root %>%
  root(outgroup="QUERY___B_longum_51A")

outgroupRootTree %>%
  ggtree(branch.length = "none") +
  geom_tiplab() +
  geom_nodelab(mapping=aes(label=node)) +
  geom_treescale()  

outgroupRootTree %>%
  drop.tip("QUERY___B_longum_51A") %>%
  ggtree() + 
  geom_tiplab() +
  geom_nodelab(mapping=aes(label=node)) +
  geom_treescale()
```
# Define Clade and Genomospecies Dataframe
## Clades
```{r}
C1 <- outgroupRootTree %>%
  tree_subset("JCP7275", levels_back=2) %>%
  .$tip.label
C1

C2 <- outgroupRootTree %>%
   tree_subset("W11", levels_back=2) %>%
  .$tip.label
C2

C3 <- outgroupRootTree %>%
  tree_subset("NR010", levels_back=2) %>%
  .$tip.label
C3

C4 <- outgroupRootTree %>%
  tree_subset("6420B", levels_back=2) %>%
  .$tip.label
C4

C5 <- c("KA00735", "CMW7778B")

C6 <- "KA00225"
```

```{r}
# testClades <- function(clade, cladeList) {
#   a <- test %>%
#     filter(Clade==clade) %>%
#     .$Strain
#   b <- setdiff(a, cladeList)
#   #b <- setdiff(cladeList, a)
#   return(b)
# }
# 
# map2(list("C1", "C2", "C3", "C4", "C5", "C6"),
#      list(C1, C2, C3, C4, C5, C6),
#      testClades)
```


## Genomospecies
### Unclear genomospecies NR010? W11? N160?
  *W11*
    W11 genomospecies identity is unclear by placement on phylogeny. It was labeled genomospecies 3 by Khan et al., 2020 "A generalist lifestyle allows rare Gardnerella spp. to persist at low levels in the vaginal microbiome" *Microbial Ecology*.
```{r}
# closest whole genome mash distance function
smallestANI <- function(STRAIN){
  DF <- assemblyDistDF %>%
    filter(assembly1==gardDF$GenBank_accession[gardDF$Strain==STRAIN]|assembly2==gardDF$GenBank_accession[gardDF$Strain==STRAIN]) %>%
    filter(distance==min(distance)) %>%
    mutate(strain1=gardDF$Strain[gardDF$GenBank_accession==assembly1],
           strain2=gardDF$Strain[gardDF$GenBank_accession==assembly2],
           ANI=1-distance) %>%
    select(strain1, strain2, distance, ANI)
  return(DF)  
}

smallestANI("W11")
```
  Closest by approximate whole genome ANI is 00703C2mash (GS3) at about 98%.
  
  *NR010*
```{r}
smallestANI("NR010")
```
  Smallest approximate ANI is ~90%. Genomospecies 14.
  *N160*
  N160 appears to be in GS9 based on this phylogeny, by Khan et al. appears to place it in GS10. Unsure their method, but possibly by *cpn60* sequence.
```{r}
smallestANI("N160")
```
  Closest by whole genome mash distance is strain 6119V5 (GS9) with ~98% ANI. 

  Check *cpn60* sequence
```{r}
cpn60s <- refAssemblies %>%
  map(~.x[str_detect(names(.x), "60 kDa chaperonin")])

names(cpn60s) <- NULL #remove list names to flatten string set

cpn60s <- cpn60s %>% # flatten string set
  do.call(c,.)

names(cpn60s) <- str_extract(names(cpn60s), ".*(?=_[0-9]{5}.*)") # remove annotation from DNA String set names to get just strain names

cpn60s# print
```
  Calculate distances
```{r}
cpn60Dists <- cpn60s %>%
  rMSA::distAlignment(cpn60s) %>%
  as.matrix() %>%
  data.frame(check.names = FALSE) %>%
  rownames_to_column("strain1") %>%
  gather("strain2", "dist", 2:ncol(.)) %>% #flatten matrix
  filter(strain1!=strain2) #remove diagonal
```

  Get minimum distance:
```{r}
cpn60Dists %>%
  filter(strain1=="N160") %>%
  filter(dist==min(dist))
```
  Using full-length *cpn60* sequence N160 would be placed into GS10, as its *cpn60* sequence is closest to N160.
  Compare distances to 1500E (GS10) and 6119V5 (GS9)
```{r}
cpn60Dists %>%
  filter(strain1=="N160",
         strain2 %in% c("6119V5", "1500E"))
```
  The sequences are close, and they are also close phylogenetically.
  
  Check placement based on the UT region, which would be used in PCR:
```{r}
cpn60UTDists <- cpn60s %>%
  substr(241, 847) %>% #substring to UT region
  rMSA::distAlignment(cpn60s) %>%
  as.matrix() %>%
  data.frame(check.names = FALSE) %>%
  rownames_to_column("strain1") %>%
  gather("strain2", "dist", 2:ncol(.)) %>% #flatten matrix
  filter(strain1!=strain2) #remove diagonal
```

  Get minimum distance:
```{r}
cpn60Dists %>%
  filter(strain1=="N160") %>%
  filter(dist==min(dist))
```
  Same result of placement in GS10.

  Compare distances to 1500E (GS10) and 6119V5 (GS9)
```{r}
cpn60Dists %>%
  filter(strain1=="N160",
         strain2 %in% c("6119V5", "1500E"))
```
  Again, same result as full-length.

### Define genomospecies identities
```{r}
GS1 <- outgroupRootTree %>%
  tree_subset("N165", levels_back=1) %>%
  .$tip.label
GS2 <-c("JCP7275", "41V", "55152", "1400E", "JCP8108")
GS3 <- outgroupRootTree %>%
  tree_subset("00703C2mash", levels_back=4) %>%
  .$tip.label
GS4 <- outgroupRootTree %>%
  tree_subset("JCP8070", levels_back=1) %>%
  .$tip.label
GS5 <- outgroupRootTree %>%
  tree_subset("6420B", levels_back=1) %>%
  .$tip.label
GS6 <- outgroupRootTree %>%
  tree_subset("UMB0264", levels_back=1) %>%
  .$tip.label
GS7 <- c("JCP8481A", "PSS_7772B")
GS8 <- c("101", "00703Dmash", "UMB1686")
GS9 <- c("6119V5", "N160")
GS10 <- "1500E"
GS11 <- "GED7760B"
GS12 <- c("KA00735", "CMW7778B")
GS13 <- "KA00225"
GS14 <- "NR010"
```


## Clade and genomoospecies dataframe
```{r}
cladeGSDF <- gardDF %>%
  select(Strain) %>%
  mutate(Genomospecies=case_when(Strain %in% GS1 ~ "GS1",
                                 Strain %in% GS2 ~ "GS2",
                                 Strain %in% GS3 ~ "GS3",
                                 Strain %in% GS4 ~ "GS4",
                                 Strain %in% GS5 ~ "GS5",
                                 Strain %in% GS6 ~ "GS6",
                                 Strain %in% GS7 ~ "GS7",
                                 Strain %in% GS8 ~ "GS8",
                                 Strain %in% GS9 ~ "GS9",
                                 Strain %in% GS10 ~ "GS10",
                                 Strain %in% GS11 ~ "GS11",
                                 Strain %in% GS12 ~ "GS12",
                                 Strain %in% GS13 ~ "GS13",
                                 Strain %in% GS14 ~ "GS14"),
         Clade=case_when(Strain %in% C1 ~ "C1",
                         Strain %in% C2 ~ "C2",
                         Strain %in% C3 ~ "C3",
                         Strain %in% C4 ~ "C4",
                         Strain %in% C5 ~ "C5",
                         Strain %in% C6 ~ "C6"))
  
# save clade and genomospecies dataframe
# write_csv(cladeGSDF, "../refGardnerellaCladesGenomos.csv")
```

# Single-Copy Core Genes
```{r}
coreGenes <- roaryPresAbs %>%
  filter(`No. isolates`==85)

# check single copy
coreGenes$`Avg sequences per isolate` %>%
  table
```

## Testing core genome annotation counts in pangenome
  Use strict definition of annotation of gene family appearing only once, to ensure only single-copy core genes are used in the 
  All replicated annotations:
```{r}
replicatedGeneFamilies <- roaryPresAbs %>%
  group_by(Annotation) %>%
  summarise(n=n()) %>%
  filter(n>1) %>%
  .$Annotation

length(replicatedGeneFamilies)
```

  Observe core genes with replicated annotations and get single-copy core genes
```{r}
replicatedCoreGeneFamilies <- coreGenes$Annotation[coreGenes$Annotation %in% replicatedGeneFamilies]
replicatedCoreGeneFamilies
length(replicatedCoreGeneFamilies)

singleCopyCoreGenes <- coreGenes %>%
  filter(Annotation %!in% replicatedGeneFamilies)
  
nrow(singleCopyCoreGenes)

#sanity check
(nrow(singleCopyCoreGenes)+length(replicatedCoreGeneFamilies)) == length(coreGenes$Annotation)
```

# Dereplicate sequences within clades/genomospeices
  Remove replicate sequences within clades or genomospecies by identifying replicate sequences
## Make list of single-copy core genes for each reference genome assembly
```{r}
strains <- gardDF$Strain
  
# Gene ID lists by Strain
coreGenesByStrain <- singleCopyCoreGenes %>%
  select(all_of(strains)) %>%
  as.list() %>%
  map(~paste(.x, collapse = "|")) %>%
  .[names(refAssemblies)] # reorder to match order of ffn files

head(coreGenesByStrain)[1:2]

# Gene ID lists by gene family
coreGenesByFamily <- singleCopyCoreGenes %>%
  select(Gene, all_of(strains)) %>%
  column_to_rownames(var = "Gene") %>%
  t() %>%
  as.data.frame() %>%
  as.list()

head(coreGenesByFamily)[1:2]
length(coreGenesByFamily)
```

## select only core genomes
```{r}
coreGeneSeqs <- refAssemblies %>%
  map2(coreGenesByStrain, ~.x[str_detect(names(.x), .y)])

# length test
lengthTest <- map(coreGeneSeqs, length) %>%
  unlist
names(lengthTest) <- NULL
table(lengthTest)
```
  70 single-copy core genes for all 85 genome assemblies.

### Save fasta files of core genomes of each assembly
```{r}
#map2(coreGeneSeqs, names(refAssemblies), ~writeXStringSet(.x, filepath = file.path(assemblyCoreGenesDir, paste0(.y, ".fasta")), format="fasta"))
```

# Select representatives
  Select one representative core genome per genomospecies
## Group based on genomospecies and select representative
```{r}
names(coreGeneSeqs) <- names(refAssemblies)

coreSelectDF <- coreGeneSeqs %>%
  .[names(coreGeneSeqs) %!in% c("N165", "JCP7275")] %>% # remove assemblies we are not using for mapping
  as.list() %>%
  map_df(., width) %>% #core genome lengths
  colSums() %>%
  as.data.frame() %>%
  dplyr::rename(core_genome_length=".") %>%
  rownames_to_column(var="Strain") %>%
  left_join(cladeGSDF, by="Strain") %>%
  left_join(gardDF[,c("Strain", "assembly_level", "contig_N50", "contig_L50")], by="Strain") %>%
  replace_na(list(contig_N50=9999999, contig_L50=1)) # replace values in complete assemblies

coreSelectDF %>%
  ggplot(aes(core_genome_length)) +
  geom_histogram(bins = 50)

coreSelectDF %>%
  group_by(assembly_level) %>%
  summarise(mean=mean(core_genome_length),
            min=min(core_genome_length),
            firstQ=quantile(core_genome_length, c(0.25)),
            median=median(core_genome_length),
            thirdQ=quantile(core_genome_length, c(0.75)),
            max=max(core_genome_length))

# all sizes
coreSelectDF %>%
  .$core_genome_length %>%
  summary
```
  Use median complete assembly genome core genome length of 77574 as reference length. Is also the median of all core genome lengths
  
  Select one core genome per genomospecies
```{r}
includeCoreGenomes <- coreSelectDF %>%
  group_by(Genomospecies) %>%
  filter(contig_N50==max(contig_N50)) %>% # first criterion is assembly quality
  filter(abs(core_genome_length-77574)==min(abs(core_genome_length-77574))) %>% # second criterion is core genome length
  filter(Strain!="GV37") %>% # still a duplicate in GS 6, keep 409-05
  .$Strain
```

  Make sure each clade is represented and each genomospecies is represented
```{r}
cladeGSDF %>%
  filter(Strain %in% includeCoreGenomes) %>%
  group_by(Clade) %>%
  tally()

cladeGSDF %>%
  filter(Strain %in% includeCoreGenomes) %>%
  group_by(Genomospecies) %>%
  tally()
```

  Save mapping core genome database
```{r}
# Sequences
coreRefDB <- coreGeneSeqs[names(coreGeneSeqs) %in% includeCoreGenomes] %>%
  `names<-`(NULL) %>%
   do.call(c, .)
# remove annotations from name
names(coreRefDB) <- str_extract(names(coreRefDB), ".*_[0-9]{5}") 

singleCopyCoreGenes0 <- singleCopyCoreGenes %>%
  select(Gene, Annotation, strains) %>%
  gather("Strain", "reference", strains) %>%
  dplyr::rename(gene=Gene,
                annotation=Annotation)

# Create reference data frame
coreRefDB_DF <- coreRefDB %>%
  names() %>%
  as.data.frame() %>%
  dplyr::rename(reference=".") %>%
  left_join(singleCopyCoreGenes0, by="reference") %>%
  left_join(cladeGSDF, by="Strain")

# save files
#writeXStringSet(coreRefDB, filepath = "../clade_assignments/ref_genes/refGenes.fasta") # fasta of sequences
#write_csv(coreRefDB_DF, "../clade_assignments/ref_genes/refGenesDF.csv") #data frame
```

# Table of Reference Gardnerella Genomes
```{r}
gardTable <- gardDF %>%
  select(Strain, Genus, GenBank_accession) %>%
  left_join(cladeGSDF, by="Strain") %>%
  mutate(Species=case_when(Genomospecies=="GS1"~"vaginalis",
                           Genomospecies=="GS4"~"piotii",
                           Genomospecies=="GS5"~"leopoldii",
                           Genomospecies=="GS6"~"swidsinskii",
                           Genomospecies %!in% c("GS1", "GS4", "GS5", "GS6")~paste0("sp. ", str_extract(Genomospecies, "(?<=GS)[0-9]+"))),
         `In Mapping DB`=(Strain %in% includeCoreGenomes),
         Genomospecies=factor(Genomospecies, levels=c("GS1", "GS2", "GS3", "GS4", "GS5", "GS6", "GS7", "GS8", "GS9", "GS10", "GS11", "GS12", "GS13", "GS14"))) %>%
  dplyr::rename(Accession=GenBank_accession) %>%
  select(Strain, Genus, Species, Genomospecies, Clade, `In Mapping DB`, Accession) %>%
  arrange(Genomospecies) %>%
  kbl(caption = "Table 1. Reference Gardnerella Genomes") %>%
  kable_classic(full_width=TRUE, html_font = "Arial") %>%
  row_spec(0, bold = TRUE)
gardTable
save_kable(gardTable, file = file.path(figureOut, paste(Sys.Date(), "GardRefTable.png", sep="_")))
```


# Session Info
```{r}
sessionInfo()
```
