---
title: "Prokka Assembly Annotations"
author: "HLB"
date: "October 2020"
output: html_document
---

Annotate Genome Assemblies with Prokka

# Set Up
## Packages
```{r}
library(tidyverse)
```

```{r, message=FALSE}
library(R.utils)
```

## Paths
```{r}
repoPath <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/pregnancy_metagenome/"
dataPath <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/gardPhylogeny/"

gardDFpath <- file.path(repoPath, "gard_phylogeny/GardnerellaMetadata.csv")
gardDF0 <- gardDFpath  %>%
  read_csv

fastaDir <- file.path(dataPath,  "_ncbi_downloads/genomes/")
outDir <- file.path(dataPath, "prokka_annotated_genomes/")
#dir.create(outDir)
```

## Organize

```{r}
annotateAssemblies <- gardDF %>%
  filter(in_phylogeny==TRUE) %>%
  select(Strain, GenBank_accession) %>%
  as.list()
```

  unzip files
```{r}
fastaDir %>%
  list.files(pattern = "genomic_genbank.fna.gz", full.names = TRUE)
  map(gunzip)
```

# Prokka
```{r}
annotateAssemblies %>%
  pmap_chr(~paste0("prokka --outdir ", outDir, "", .x,
               " --prefix ", .x, 
               " --locustag ", .x,
               " --prefix ", .x,
               " --genus Gardnerella --species vaginalis --strain ", .x,
               " --addgenes --addmrna --usegenus --cpu 2 ",
               fastaDir, .y, "_genomic_genbank.fna")) %>%
  writeLines("./prokka_annotate.sh")

system("perlbrew use 5.32.0")
cat(Sys.getenv("PERL5LIB"), "\n")
system("prokka -v")
```

# Re-organize files
## Re organize prokka output files

## Re-zip genome assemblies
```{r}
fastaDir %>%
  list.files(pattern = "genomic_genbank.fna", full.names = TRUE) %>%
  map(gzip)
```

```{bash}
prokka -v
```


# Session Info
```{r}
sessionInfo()
```