---
title: "Sample and Subject Metadata"
author: "HLB"
output:
  html_document: default
  html_notebook: default
---

The following script is to update metadata with detailed sample metadata and describe the sampling. 

# Setup
Packages
```{r}
library(tidyverse)
library(formattable)
set.seed(7334)
```

  Current metadata table:
```{r}
currentsgDF <- read_tsv("../shotgunMetadata.tsv")
```

## MGS sample data table
  Data table from DG with sample information for shotgun metagenomic sequenced samples 
```{r}
sgDF0 <- "./Tables_4_Ben.txt" %>%
  read_tsv() %>%
  dplyr::rename(SampleIndex=`Sample Index`,
                SampleType=`Sample Type`,
                SubjectID=Subject,
                SampleID=`Sample ID`,
                GDcoll=GD_Coll,
                GWdel=GW_Dell,
                TermPre=`Term/Pre`,
                TotalReads=`Total reads`,
                TotalPairs=`Total pairs`,
                Sickle100_pe=`Sickle_100 (pe)`,
                Sickle_pairs=Sickle_pe,
                Human_pairs=Human_pe,    #number of human pairs from DAG, not final number human pairs.
                PctTrim=`Pct Trim`,
                PctHuman=`Pct Human`) %>%
  mutate(SampleID=as.character(SampleID),
         SubjectID=as.character(SubjectID),
         PctTrim=as.numeric(as.character(str_extract(PctTrim, ".*(?=\\%)")))/100,
         PctHuman=as.numeric(as.character(str_extract(PctHuman, ".*(?=\\%)")))/100,
         GWcoll=GDcoll/7,
         #SampleID_a=paste0(substr(SampleID, 1, 5), "01", substr(SampleID, 8, 10)),
         RNAlater=case_when(str_detect(SampleID, "[0-9]{5}01[0-9]{3}")~"No",
                            str_detect(SampleID, "[0-9]{5}35[0-9]{3}")~"Yes"))
head(sgDF0)
```

iew
## Amplicon sample data
  Sample metadata tables from amplicon sequencing form 2015 and 2017 papers with more demographic and other information
```{r, message=FALSE}
df2015 <- "./sampledata_VaginalSwabs_Pregnancy_DiGiulio2015.tsv" %>%
  read_tsv()

df2017 <- "./MAP_MANUSCRIPT_IL02-IL09_ToBen_20170404_fixed.txt" %>%
  read_tsv()

figureOut <- "../../FIGURES/metadata"

theme_set(theme_bw() +
          theme(panel.grid = element_blank()))
```
  Add cohort information to 2015 paper and standardize how cohorts are labeled.
```{r}
PS15 <- df2015 %>%
  dplyr::rename(GDcoll=GDColl) %>%
  mutate(SubjectID= as.character(SubjectID),
         Cohort="Stanford",
         Method="Amplicon",
         PaperCohort="2015") %>%
  select(SubjectID, GDcoll, Cohort, Method, PaperCohort)

PS17 <- df2017 %>%
  mutate(SubjectID= as.character(SubjectID),
         Cohort=case_when(UAB_Sample==0~"Stanford",
                                 UAB_Sample==1~"UAB"),
         Method="Amplicon",
         PaperCohort="2017") %>%
  select(SubjectID, GDcoll, Cohort, Method, PaperCohort)

sgDF1 <- sgDF0 %>%
  left_join(AMP[,c("SubjectID", "Cohort", "PaperCohort")], by="SubjectID") %>% # add cohort information
  filter(SampleID %in% sgDF0$SampleID) %>%
  mutate(Cohort=case_when(SubjectID == "10604"~"Stanford",
                          SubjectID != "10604"~Cohort)) %>% # add cohort info for 10604
  unique.data.frame()
```

# Add information to metadata table
## Trimesters
```{r}
sgDF2 <- sgDF1 %>%
  group_by(SubjectID) %>%
  mutate(sampleOrder=rank(GDcoll),
         sampleTrimester=case_when(GDcoll<=91~"First",
                                   GDcoll>=92&GDcoll<=189~"Second",
                                   GDcoll>=190~"Third")) %>%
  ungroup() 
```

## Demographics
  Note, no demographics data for 10604
```{r}
colnames(df2015)
colnames(df2017)
unique(df2015$Race)
unique(df2017$MaternalRace)
df2015 %>%
  select(SubjectID, Race)

df2017 %>%
  select(SubjectID, MaternalRace, SpecifyOTHERrace) %>%
  filter(SubjectID %in% sgDF2$SubjectID) %>%
  unique
```

  Create table with subject demographics
```{r}
demo2015 <- df2015 %>%
  select(SubjectID, Race, Ethnicity)  %>%
  filter(SubjectID %in% sgDF2$SubjectID) %>%
  unique %>%
  mutate(SubjectID=as.character(SubjectID),
         Cohort="Stanford")

demo2017 <- df2017%>%
  select(SubjectID, UAB_Sample, MaternalRace, SpecifyOTHERrace, MaternalEthnicity, AgeAtConsent, BMI, NumberOfPriorPregnancies) %>%
  filter(SubjectID %in% sgDF2$SubjectID) %>%
  unique %>%
  dplyr::rename(Race=MaternalRace,
                Ethnicity=MaternalEthnicity) %>%
  mutate(SubjectID= as.character(SubjectID),
         Cohort=case_when(UAB_Sample==0~"Stanford",
                          UAB_Sample==1~"UAB")) %>%
  select(-UAB_Sample)
#Race levels for matching
unique(demo2015$Race)
unique(demo2017$Race)
#Ethnicity levels for matching
unique(demo2015$Ethnicity)
unique(demo2017$Ethnicity)
```
  
  Race categories based on US Census and collapsing Native American and Native Hawaiian to Other due to small n.
```{r}
demo <- demo2015 %>%
  full_join(demo2017, by = c("SubjectID", "Race", "Ethnicity", "Cohort")) %>%
  bind_rows(., c(SubjectID="10604", Cohort="Stanford")) %>% # add subject without demo data
  mutate(Ethnicity=case_when(Ethnicity=="Hispanic"~"Hispanic",
                             Ethnicity=="Non-Hispanic"~"Non-Hispanic",
                             Ethnicity=="NonHispanic"~"Non-Hispanic"),
         Race=case_when(Race=="White"~"White",
                        Race=="American Indian"~"Other",
                        Race=="AmericanIndian"~"Other",
                        Race=="Other (Specify below)"~"Other",
                        Race=="Other"~"Other",
                        Race=="AsianChinese"~"Asian",
                        Race=="Indian"~"Asian",
                        Race=="Black"~"Black",
                        Race=="Hawaiian"~"Other")) %>%
  replace_na(list(Ethnicity="Unknown", Race="Other"))
```

## Reads
### Get read counts:
```{r}
readCountDF0 <- read_delim("../filter_host/fastq_counts_bbmap.txt", delim = "\ ") %>%
  mutate(Sample=as.character(Sample)) %>%
  mutate_if(is.numeric, ~.x/4)
colnames(readCountDF0) 
head(readCountDF0)
```

```{r}
readCountDF <- readCountDF0 %>%
  dplyr::rename(SampleID=Sample,
                bbmapFiltered_pairs=filteredCount1,
                bowtieFiltered_pairs=uploadCount1) %>%
  mutate(SampleID=as.character(SampleID))
  

sgDF <- sgDF2 %>%
    select(-PctTrim, -Sickle100_pe, -Human_pairs, -Human_sngl, -PctHuman) %>% # remove unnessecary columns
    left_join(readCountDF[,c("SampleID", "bowtieFiltered_pairs", "bbmapFiltered_pairs")], by="SampleID") %>%
    left_join(demo[,c("SubjectID", "Race", "Ethnicity")], by="SubjectID")
#write_tsv(sgDF, "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/pregnancy_metagenome/sumgsDF.tsv")
```

# Collection figure
## MGS
```{r}
AMP <- rbind(PS15, PS17) %>%
  unique.data.frame()


ggplot() +
  geom_point(data=AMP, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, color=Cohort, size=Method), alpha=.5) +
   geom_point(data=SHOT, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, size=Method), color="black") +
  xlim(0,280) +
     labs(x = "Gestation Days",
          y = "Subject") +
     scale_color_manual(values=c("olivedrab", "steelblue", "black"))+
     scale_shape_manual(values=c(19, 15)) +
     scale_size_manual(values=c(1,2)) +
     #theme_classic()+
     theme(axis.ticks.y = element_blank(),
     #axis.text.x = element_text(size = 10),
     axis.text.y  = element_blank()
     #axis.title.x = element_text(size=20),
     #axis.title.y = element_text(size=20)
     )
     #scale_x_continuous(breaks = c(0,50,100,150,200, 250))
#ggsave("./collectionPlot.svg", height=8, width=8.5) 
```

## paired sequencing
```{r}
SHOT2 <- sgDF %>%
  select(SubjectID, GDcoll, Cohort, PaperCohort, RNAlater) %>%
  filter(PaperCohort=="2015"|PaperCohort=="2017") %>%
  mutate(Method=case_when(RNAlater=="No"~"Paired amplicon and shotgun",
                          RNAlater=="Yes"~"Paired meta-omic")) %>%
  unique.data.frame() 

SHOT <- sgDF1 %>% 
  select(SubjectID, GDcoll, Cohort, PaperCohort) %>%
  filter(PaperCohort=="2015"|PaperCohort=="2017") %>%
  mutate(Method="Shotgun") %>%
  unique.data.frame()

ggplot() +
  geom_point(data=AMP, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, color=Cohort, size=Method), alpha=.5) +
  geom_point(data=SHOT2, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, size=Method), color="black") +
  xlim(0,280) +
     labs(x = "Gestation Days",
          y = "Subject") +
     scale_color_manual(values=c("olivedrab", "steelblue", "black"))+
     scale_shape_manual(values=c(19, 15,0)) +
     scale_size_manual(values=c(1,2,2)) +
     #theme_classic()+
     theme(axis.ticks.y = element_blank(),
     axis.text.y  = element_blank())
#ggsave("/Users/hlberman/Desktop/prelimPlot.svg", height=8, width=8.5) 
```


# Session Info
```{r}
sessionInfo()
```