---
title: "Sample and Subject Metadata"
author: "HLB"
output:
  html_document: default
  html_notebook: default
---

The following script is to update metadata with detailed sample metadata and describe the sampling. 

# Setup
Packages
```{r}
library(tidyverse)
set.seed(7334)
```


```{r}
sgDF0 <- "./Tables_4_Ben.txt" %>%
  read_tsv() %>%
  dplyr::rename(SampleIndex=`Sample Index`,
                SampleType=`Sample Type`,
                SubjectID=Subject,
                SampleID=`Sample ID`,
                GDcoll=GD_Coll,
                GWdel=GW_Dell,
                TermPre=`Term/Pre`,
                TotalReads=`Total reads`,
                TotalPairs=`Total pairs`,
                Sickle100_pe=`Sickle_100 (pe)`,
                Sickle_pairs=Sickle_pe,
                Human_pairs=Human_pe,    #number of human pairs from DAG, not final number human pairs.
                PctTrim=`Pct Trim`,
                PctHuman=`Pct Human`) %>%
  mutate(SampleID=as.character(SampleID),
         SubjectID=as.character(SubjectID),
         PctTrim=as.numeric(as.character(str_extract(PctTrim, ".*(?=\\%)")))/100,
         PctHuman=as.numeric(as.character(str_extract(PctHuman, ".*(?=\\%)")))/100,
         GWcoll=GDcoll/7,
         SampleID_a=paste0(substr(SampleID, 1, 5), "01", substr(SampleID, 8, 10)),
         RNAlater=case_when(str_detect(SampleID, "[0-9]{5}01[0-9]{3}")~"No",
                            str_detect(SampleID, "[0-9]{5}35[0-9]{3}")~"Yes"))
head(sgDF0)
```

```{r, message=FALSE}
oldDF <- read_tsv("/Users/hlberman/Desktop/2021-6-4_shotgunMetadata.tsv")

df2015 <- "./sampledata_VaginalSwabs_Pregnancy_DiGiulio2015.tsv" %>%
  read_tsv()

df2017 <- "./MAP_MANUSCRIPT_IL02-IL09_ToBen_20170404_fixed.txt" %>%
  read_tsv()

figureOut <- "../../FIGURES/metadata"

theme_set(theme_bw() +
          theme(panel.grid = element_blank()))
```


```{r}
PS15 <- df2015 %>%
  dplyr::rename(GDcoll=GDColl) %>%
  mutate(SubjectID= as.character(SubjectID),
         Cohort="Stanford",
         Method="Amplicon",
         PaperCohort="2015") %>%
  select(SubjectID, GDcoll, Cohort, Method, PaperCohort)

PS17 <- df2017 %>%
  mutate(SubjectID= as.character(SubjectID),
         Cohort=case_when(UAB_Sample==0~"Stanford",
                                 UAB_Sample==1~"UAB"),
         Method="Amplicon",
         PaperCohort="2017") %>%
  select(SubjectID, GDcoll, Cohort, Method, PaperCohort)

AMP <- rbind(PS15, PS17) %>%
  unique.data.frame()

sgDF1 <- sgDF0 %>%
  left_join(AMP[,c("SubjectID", "Cohort", "PaperCohort")], by="SubjectID") %>%
  filter(SampleID %in% sgDF0$SampleID) %>%
  mutate(Cohort=case_when(SubjectID == "10604"~"Stanford",
                          SubjectID != "10604"~Cohort)) %>% # add cohort info for 10604
  unique.data.frame()


SHOT <- sgDF1 %>%
  select(SubjectID, GDcoll, Cohort, PaperCohort) %>%
  filter(PaperCohort=="2015"|PaperCohort=="2017") %>%
  mutate(Method="Shotgun") %>%
  unique.data.frame()
```

# Trimesters
```{r}
sgDF2 <- sgDF1 %>%
  group_by(SubjectID) %>%
  mutate(sampleOrder=rank(GDcoll),
         sampleTrimester=case_when(GDcoll<=97~"First",
                                   GDcoll>=98&GDcoll<=195~"Second",
                                   GDcoll>=196~"Third")) %>%
  ungroup() 

#save updated dataframe
#write_tsv(sgDF2, "../shotgunMetadata.tsv") # save to main path of repo
```

  Total 107 Samples. 
  From Stanford: 62 samples from 20 subjects (11 term and 9 preterm). Average 3.1 samples per subject.
  From UAB: 45 samples from 15 subjects (5 term and 10 preterm). Average 3 samples per subject.

```{r}
colnames(df2015)
colnames(df2017)
unique(df2015$Race)
unique(df2017$MaternalRace)
df2015 %>%
  select(SubjectID, Race)

df2017 %>%
  select(SubjectID, MaternalRace, SpecifyOTHERrace) %>%
  filter(SubjectID %in% sgDF$SubjectID) %>%
  unique
```


## Demographics
Note, no demographics data for 10604
```{r}
demo2015 <- df2015 %>%
  select(SubjectID, Race, Ethnicity)  %>%
  filter(SubjectID %in% sgDF$SubjectID) %>%
  unique %>%
  mutate(SubjectID=as.character(SubjectID),
         Cohort="Stanford")

demo2017 <- df2017%>%
  select(SubjectID, UAB_Sample, MaternalRace, SpecifyOTHERrace, MaternalEthnicity, AgeAtConsent, BMI, NumberOfPriorPregnancies) %>%
  filter(SubjectID %in% sgDF$SubjectID) %>%
  unique %>%
  dplyr::rename(Race=MaternalRace,
                Ethnicity=MaternalEthnicity) %>%
  mutate(SubjectID= as.character(SubjectID),
         Cohort=case_when(UAB_Sample==0~"Stanford",
                          UAB_Sample==1~"UAB")) %>%
  select(-UAB_Sample)
#Race levels for matching coding
unique(demo2015$Race)
unique(demo2017$Race)
#Ethnicity levels for matching coding
unique(demo2015$Ethnicity)
unique(demo2017$Ethnicity)
```
  
  Fix levels to match and assess
```{r}
demo <- demo2015 %>%
  full_join(demo2017) %>%
  bind_rows(., c(SubjectID="10604", Cohort="Stanford")) %>% # add subject without demo data
  mutate(Ethnicity=case_when(Ethnicity=="Hispanic"~"Hispanic",
                             Ethnicity=="Non-Hispanic"~"Non Hispanic",
                             Ethnicity=="NonHispanic"~"Non Hispanic"),
         Race=case_when(Race=="White"~"White",
                        Race=="American Indian"~"American Indian",
                        Race=="AmericanIndian"~"American Indian",
                        Race=="Other (Specify below)"~"Other",
                        Race=="Other"~"Other",
                        Race=="AsianChinese"~"Asian",
                        Race=="Indian"~"Asian",
                        Race=="Black"~"Black",
                        Race=="Hawaiian"~"Native Hawaiian")) %>%
  replace_na(list(Ethnicity="Unknown", Race="Unknown"))

# Gest age
gestage <- demo %>%
  left_join(unique(sgDF[,c("SubjectID", "TermPre", "GWdel")]), by="SubjectID") %>%
  group_by(Cohort, TermPre) %>%
  summarize(across(GWdel, list(mean = ~mean(.x, na.rm = TRUE), sd = ~sd(.x, na.rm = TRUE)))) %>%
  ungroup %>%
  mutate(GWdel_mean=round(GWdel_mean, 1), GWdel_sd=round(GWdel_sd, 1),
        x=paste0(GWdel_mean, " (", GWdel_sd, ")"),
        cat=paste(Cohort, TermPre),
        Variable="Mean gestational age in weeks at delivery (Â±SD)") %>%
  select(Variable, cat, x) %>%
  spread(cat, x) %>%
  select(Variable, `Stanford T`, `Stanford P`, `UAB T`, `UAB P`)

# Don't have AgeAtConsent, BMI, or prior number of pregnancies for nulliparity for 2015 data

#race
race <- demo %>%
  left_join(unique(sgDF[,c("SubjectID", "TermPre")]), by="SubjectID") %>%
  select(SubjectID, Cohort, TermPre, Race) %>%
  group_by(Cohort, TermPre, Race) %>%
  tally %>%
  group_by(Cohort, TermPre) %>%
  mutate(percent=round(n/sum(n), 2)*100,
         x=paste0(n, " (", percent, "%)"), 
         cat=paste(Cohort, TermPre)) %>%
  ungroup %>%
  select(Race, cat, x) %>%
  spread(cat, x) %>%
  mutate(across(everything(), ~replace_na(.x, "0")), 
         Race=factor(Race, levels=c("White", "Black", "Asian", "American Indian", "Pacific Islander", "Unknown", "Other"))) %>%
  select(Race, `Stanford T`, `Stanford P`, `UAB T`, `UAB P`) %>%
  arrange(Race) %>%
  bind_rows(data_frame(Race="Race", `Stanford T`="", `Stanford P`="", `UAB T`="", `UAB P`=""), .) %>%
  dplyr::rename(Variable=Race)

#ethnicity
ethn <- demo %>%
  left_join(unique(sgDF[,c("SubjectID", "TermPre")]), by="SubjectID") %>%
  select(SubjectID, Cohort, TermPre, Ethnicity) %>%
  group_by(Cohort, TermPre, Ethnicity) %>%
  tally %>%
  group_by(Cohort, TermPre) %>%
  mutate(percent=round(n/sum(n), 2)*100,
         x=paste0(n, " (", percent, "%)"), 
         cat=paste(Cohort, TermPre)) %>%
  ungroup %>%
  select(Ethnicity, cat, x) %>%
  spread(cat, x) %>%
  mutate(across(everything(), ~replace_na(.x, "0")), 
         Ethnicity=factor(Ethnicity, levels=c("Non Hispanic", "Hispanic", "Unknown"))) %>%
  select(Ethnicity, `Stanford T`, `Stanford P`, `UAB T`, `UAB P`) %>%
  arrange(Ethnicity) %>%
  bind_rows(data_frame(Ethnicity="Ethnicity", `Stanford T`="", `Stanford P`="", `UAB T`="", `UAB P`=""), .) %>%
  dplyr::rename(Variable=Ethnicity)

demoTable <- gestage %>%
  full_join(race) %>%
  full_join(ethn)
demoTable

#write_csv(demoTable, file.path(figureOut, "20210118_demoTable.csv"))
```


## Reads
### Get read counts:
```{r}
readCountDF0 <- read_delim("../filter_host/fastq_counts_bbmap.txt", delim = "\ ") %>%
  mutate(Sample=as.character(Sample)) %>%
  mutate_if(is.numeric, ~.x/4)
colnames(readCountDF0) 
head(readCountDF0)
```

```{r}
readCountDF <- readCountDF0 %>%
  dplyr::rename(SampleID=Sample,
                bbmapFiltered_pairs=filteredCount1) %>%
  mutate(SampleID=as.character(SampleID))
  
testDF <- read_tsv("/Users/hlberman/Desktop/2021-6-4_shotgunMetadata.tsv")
View(testDF)

sgDF2 %>%
  left_join(readCountDF[,c("SampleID", "uploadCount1")]) %>%
  mutate(test=Sickle_pairs-Human_pairs) %>%
  select(SampleID, Sickle_pairs, Human_pairs, uploadCount1, test) %>%
  View
  .$test
  table()

sgDF <- sgDF2 %>%
    left_join(readCountDF[,c("SampleID", "bbmapFiltered_pairs")], by="SampleID") %>%
    left_join(demo[,c("SubjectID", "Race")], by="SubjectID") %>%
    mutate(bacLoad=bbmapFiltered_pairs/(Sickle_pairs-bbmapFiltered_pairs),
         pctHuman=((Sickle_pairs-bbmapFiltered_pairs)/Sickle_pairs)*100)
#write_tsv(sgDF, "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/pregnancy_metagenome/shotgunMetadata.tsv")
```

# Collection figure
## MGS
```{r}
ggplot() +
  geom_point(data=AMP, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, color=Cohort, size=Method), alpha=.5) +
   geom_point(data=SHOT, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, size=Method), color="black") +
  xlim(0,280) +
     labs(x = "Gestation Days",
          y = "Subject") +
     scale_color_manual(values=c("olivedrab", "steelblue", "black"))+
     scale_shape_manual(values=c(19, 15)) +
     scale_size_manual(values=c(1,2)) +
     #theme_classic()+
     theme(axis.ticks.y = element_blank(),
     #axis.text.x = element_text(size = 10),
     axis.text.y  = element_blank()
     #axis.title.x = element_text(size=20),
     #axis.title.y = element_text(size=20)
     )
     #scale_x_continuous(breaks = c(0,50,100,150,200, 250))
#ggsave("./collectionPlot.svg", height=8, width=8.5) 
```

## paired sequencing
```{r}
SHOT2 <- sgDF %>%
  select(SubjectID, GDcoll, Cohort, PaperCohort, RNAlater) %>%
  filter(PaperCohort=="2015"|PaperCohort=="2017") %>%
  mutate(Method=case_when(RNAlater=="No"~"Paired amplicon and shotgun",
                          RNAlater=="Yes"~"Paired meta-omic")) %>%
  unique.data.frame() 

ggplot() +
  geom_point(data=AMP, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, color=Cohort, size=Method), alpha=.5) +
  geom_point(data=SHOT2, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, size=Method), color="black") +
  xlim(0,280) +
     labs(x = "Gestation Days",
          y = "Subject") +
     scale_color_manual(values=c("olivedrab", "steelblue", "black"))+
     scale_shape_manual(values=c(19, 15,0)) +
     scale_size_manual(values=c(1,2,2)) +
     #theme_classic()+
     theme(axis.ticks.y = element_blank(),
     axis.text.y  = element_blank())
#ggsave("/Users/hlberman/Desktop/prelimPlot.svg", height=8, width=8.5) 
```


# Session Info
```{r}
sessionInfo()
```