---
title: "Untitled"
author: "Hanna Berman"
date: "3/15/2021"
output: html_document
---

```{r, message=FALSE}
library(tidyverse)
library(formattable)

`%!in%` <- negate(`%in%`)

theme_set(theme_bw()+
          theme(panel.grid = element_blank()))
```

# Phenotype Files
```{r}
# SRA run table from dbGaP
mgsRunTable <- read_csv("/Users/hlberman/mnt/cluster/home6/callah_momspi/SRAruntable.txt") %>%
  filter(body_site=="Vagina",
         LibrarySource=="METAGENOMIC")
#write_lines(mgsRunTable$Run, "/Users/hlberman/Desktop/MGSsras.txt")

phenFilePath <- "/Users/hlberman/mnt/cluster/home6/callah_momspi/Phenotypefiles" # directory for phenotype files
phenFileTxts <- list.files(phenFilePath, pattern = "txt", full.names = TRUE) # list of phenotype (metadata) files

# read in phenotype dataframes and store object as list
txts <- phenFileTxts %>%
  map(~read_tsv(.x, comment="#"))

# name dataframes with filenames
names(txts) <- phenFileTxts <- list.files(phenFilePath, pattern = "txt")
#preview dataframes
map(txts, head)
```

## Get sample attributes
```{r}
#dataframe of sample attributes from MGS and MTS samples
vsamples <- txts$`phs001523.v1.pht007515.v1.p1.c1.MOMS_PI_Sample_Attributes.DS-PREG-COM-IRB-PUB-MDS.txt` %>%
  filter(HISTOLOGICAL_TYPE=="Vagina", omic!="16S")

# Reported trimester from subject at each visit
reported_ga <- txts$`phs001523.v1.pht007510.v1.p1.c1.phenotype_clinical.DS-PREG-COM-IRB-PUB-MDS.txt` %>%
  select(dbGaP_Subject_ID, SUBJECT_ID, visit_ID, visit_num, reported_ga)

# add 
attrDF <- vsamples %>% # dataframe of sample attributes
  left_join(reported_ga, by = c("visit_ID", "visit_num"))

mgsRunTable0 <- mgsRunTable %>%
  dplyr::rename(SAMPLE_ID=biospecimen_repository_sample_id)
```

```{r}
n_distinct(attrDF$SAMPLE_ID)
n_distinct(attrDF$SUBJECT_ID)

attrDF %>%
  group_by(SUBJECT_ID) %>%
  summarize(n_visits=n_distinct(visit_num)) %>%
  ggplot(aes(x=n_visits)) +
  geom_histogram()

attrDF %>%
  count(omic) %>%
  formattable()
```
  MGS/MTS 1,227 samples from 234 subjects with 1 to 8 visits each.
  297 MTS samples and 930 MGS samples. 

  Reported Trimester of samples
```{r}
attrDF %>%
  ggplot(aes(x=reported_ga, y=SUBJECT_ID, color=omic)) +
  geom_jitter(alpha=0.5) +
  theme(axis.text.y = element_blank()) +
  labs(x="Trimester", y="Subject ID")
```

# Metagenomic sample and subject metadata
  Lists of samples and subjects IDs in MGS data for convenience
```{r}
#MGS samples
MGSsamps <- attrDF %>%
  filter(omic=="Metagenomics") %>%
  .$SAMPLE_ID %>%
  unique
#MGS Subjects
MGSsubs <- attrDF %>%
  filter(omic=="Metagenomics") %>%
  .$SUBJECT_ID %>%
  unique
#Visits where MGS samples were collected to match dataframes of surveys taken per visit to samples
MGSvisits <- attrDF %>%
  filter(omic=="Metagenomics") %>%
  .$visit_ID %>%
  unique
```

  Assess a few variables from clinical and health history surveys
## Health History Survey
```{r}
hxDF <- txts$`phs001523.v1.pht007512.v1.p1.c1.phenotype_healthHistorySurvey.DS-PREG-COM-IRB-PUB-MDS.txt` %>%
  filter(SUBJECT_ID %in% MGSsubs)
#preview table
head(hxDF)
colnames(hxDF)
#number of rows
nrow(hxDF)
#number of subjects represented in the health history
n_distinct(hxDF$SUBJECT_ID)
```
  There are 236 entries from 233 subjects out of 234 total subjects that have shotgun sequencing
  Who is duplicated:
```{r}
hxDF %>%
  group_by(SUBJECT_ID) %>%
  tally %>%
  filter(n>1) %>%
  .$SUBJECT_ID
```

  Who is missing:
```{r}
setdiff(MGSsubs, hxDF$SUBJECT_ID)
```

  Add all subjects
```{r}
hxDF <- hxDF %>%
  full_join(attrDF[,c("dbGaP_Subject_ID", "SUBJECT_ID")], by=c("dbGaP_Subject_ID", "SUBJECT_ID"))
```

### Maternal race
  Get reported maternal race and create column standardized with Stanford and UAB cohorts. Categories based on US census race categories, with Native American and Native Hawaiian listed as Other due to small n
```{r}
# create race data frame
raceDF0 <- hxDF %>%
  mutate(Ethnicity=case_when(hispanic_or_latino=="Yes"~"Hispanic",
                             hispanic_or_latino=="No"~"Non-Hispanic"),
    Race=case_when(african_american=="Yes"&american_indian_or_alaska_native=="No"&asian=="No"&caucasian=="No"&native_hawaiian=="No"~"Black",
                        african_american=="No"&american_indian_or_alaska_native=="No"&asian=="No"&caucasian=="Yes"&native_hawaiian=="No"~"White",
                        african_american=="No"&american_indian_or_alaska_native=="No"&asian=="Yes"&caucasian=="No"&native_hawaiian=="No"~"Asian",
                        african_american=="No"&american_indian_or_alaska_native=="Yes"&asian=="No"&caucasian=="No"&native_hawaiian=="No"~"Other",
                        african_american=="No"&american_indian_or_alaska_native=="No"&asian=="No"&caucasian=="No"&native_hawaiian=="Yes"~"Other",
                        african_american=="No"&american_indian_or_alaska_native=="No"&asian=="No"&caucasian=="No"&native_hawaiian=="No"~"Other",
                        is.na(african_american)&is.na(american_indian_or_alaska_native)&is.na(asian)&is.na(caucasian)&is.na(hispanic_or_latino)&is.na(native_hawaiian)~"Other")) %>%
  replace_na(list(Ethnicity="Unknown", Race="Other")) %>%
  select(dbGaP_Subject_ID, Race, Ethnicity) %>%
  dplyr::rename(SubjectID=dbGaP_Subject_ID) %>%
  unique

# check that there is only one race category per subject
nrow(raceDF0)==n_distinct(raceDF0$SubjectID)
```
  Number of rows in the race dataframe do not equal the number of distinct subjects. There is at least one subject with multiple races. Check which subject(s) and why:
  
```{r}
multiRaceSubject <- raceDF0 %>%
  count(SubjectID) %>%
  filter(n>1) %>%
  .$SubjectID
multiRaceSubject


hxDF %>%
  select(dbGaP_Subject_ID, visit_ID, african_american, american_indian_or_alaska_native, asian, hispanic_or_latino, native_hawaiian) %>%
  filter(dbGaP_Subject_ID %in% multiRaceSubject) %>%
  unique
```
  One subject has indicated races differently at two different visits. Indicate subject as other race (for unknown).
  Create a dataframe with subject race:
```{r}
raceDF <- raceDF0 %>%
  mutate(Race=case_when(SubjectID==multiRaceSubject~"Other",
                   SubjectID!=multiRaceSubject~Race)) %>%
  unique()

# test again that there is only one entry per subject
nrow(raceDF)==n_distinct(raceDF$SubjectID)
```
  Observe percent in each category:
```{r}
raceDF %>%
  count(Race) %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

### Gestational age at collection and delivery
  Additional tables provided from Greg Buck for gestational age at sample collection and delivery.
```{r}
gaacGaad0 <- read_tsv("/Users/hlberman/mnt/cluster/home6/callah_momspi/gaAtCollecNDel_4_PTB_042121.txt")
head(gaacGaad0)
colnames(gaacGaad0)
#Number of MGS samples/subjects that have gaac and gaad information
gaacGaad0 %>%
  filter(visitID %in% MGSvisits) %>%
  summarize_at(c("participantID", "visitID"),n_distinct)
```
  503 samples of 930, from 135 subjects of 234

  Clean up entries to match other DF
```{r}
# fix ga at sampling and delivery
gaacGaad <- gaacGaad0 %>%
  mutate(ga_at_sampling=na_if(ga_at_sampling, "2NA"),
         ga_at_sampling=na_if(ga_at_sampling, "3NA"),
         ga_at_sampling=na_if(ga_at_sampling, "4NA"),
         ga_at_delivery=na_if(ga_at_delivery, "2NA"),
         ga_at_delivery=na_if(ga_at_delivery, "3NA"),
         ga_at_delivery=na_if(ga_at_delivery, "4NA")) %>%
  separate(ga_at_sampling, c("GWcoll", "GDcoll"), sep="w") %>%
  separate(ga_at_delivery, c("GWdel", "GDdel"), sep="w") %>%
  mutate_at(c("GDcoll", "GDdel"), ~str_extract(.x, "[0-9]*")) %>%
  mutate_at(c("GWcoll", "GDcoll", "GWdel", "GDdel"), as.numeric) %>%
  mutate_at(c("GWcoll", "GWdel"), ~.x*7) %>%
  mutate(GDcoll=GWcoll+GDcoll, 
         GWcoll=GDcoll/7,
         GDdel=GWdel+GDdel,
         GWdel=GDdel/7, 
         TermPre=case_when(`case/control`=="case"~"P",
                           `case/control`=="control"~"T")) %>%
  dplyr::rename(visit_ID=visitID) %>%
  select(-`case/control`, -participantID)
head(gaacGaad)
```
  First pass check how reported_ga column (reported trimester) matches up with gestational age at collection 
```{r}
attrDF %>%
  select(visit_ID, reported_ga) %>%
  left_join(gaacGaad[,c("visit_ID", "GDcoll")], by="visit_ID") %>%
  filter(!is.na(GDcoll)) %>%
  filter(!is.na(reported_ga)) %>%
  ggplot(aes(x=reported_ga, y=GDcoll)) +
  geom_jitter() +
  geom_hline(aes(yintercept=91), linetype="dashed") +
  geom_hline(aes(yintercept=189), linetype="dashed")
```
  Separate trimesters at 91 days and 189 days and change trimesters to match GDcoll in samples where reported_ga and gestational age at collection in days (GDcoll) do not match based on those trimester cutoffs. For samples where GDcoll is not provided, use reported_ga for trimester.

# Download MGS files
## Download .sra files
  Attempt to download files using cart file method on 3/23/2021 but not all files downloaded.
```{r}
# mgssraDir <- "/Users/hlberman/mnt/cluster/home6/callah_momspi/mgs_files/sra" 
# downloadedmgsSRAfiles <- list.files(mgssraDir)
# head(downloadedmgsSRAfiles)
# length(downloadedmgsSRAfiles)
# 
# downloadedmgsSRAs <- downloadedmgsSRAfiles %>%
#   str_extract("SRR[0-9]{7}")
# head(downloadedmgsSRAs)
# length(downloadedmgsSRAs)
```
  441 files were downloaded.
  698 after the second round.
  821 after third round.
  875 after fourth round.
  904 after fifth round.
  916 after sixth round.
  926 after seventh round.
  930 after eigth round!

  What's missing?  
```{r}
# missing <- mgsRunTable %>%
#   filter(!(Run %in% downloadedmgsSRAs))
# 
# downloaded <- mgsRunTable %>%
#   filter(Run %in% downloadedmgsSRAs)
# 
# missing <- mgsRunTable %>%
#   filter(Run %!in% downloadedmgsSRAs)
# 
# #write_lines(missing$Run, "/Users/hlberman/Desktop/missingSRAs.txt")
```
 Copy list of missing SRAs into download script `/Users/hlberman/mnt/cluster/home6/callah_momspi/download_scripts/download_mgs.sh`

## Decrypt and get fastq files
  List of all MGS SRAs
```{r}
# mgsSRAs <- mgsRunTable$Run
# #write_lines(mgsSRAs, "/Users/hlberman/Desktop/mgsSRAs.txt")
```
  Copy list of all SRAs to `/Users/hlberman/mnt/cluster/home6/calla_momspi/download_scripts/decrypt_mgs.sh`
  Check which fastq files have been downloaded:
```{r}
# downloadedmgsFastqfiles <- list.files("/Users/hlberman/mnt/cluster/home6/callah_momspi/mgs_files/fastq")
# head(downloadedmgsFastqfiles)
# length(downloadedmgsFastqfiles)
# 930*2 # number of fastq files there should be
```

  What's missing?
  Repeat script for any SRAs that are missing.
```{r}
# downloadedmgsFastqSRAs <- str_extract(downloadedmgsFastqfiles, "SRR[0-9]*") %>%
#   unique
# head(downloadedmgsFastqSRAs)
# # test if each SRA has two files:
# length(downloadedmgsFastqSRAs)==(length(downloadedmgsFastqfiles)/2)
```

```{r}
# missingmgsFastq <- setdiff(mgsSRAs, downloadedmgsFastqSRAs)
# length(missingmgsFastq)
# missingmgsFastq
```


# Filtering MGS fastq files
## Filter and Trim
  Filter and trim with sickle. Check progress:
```{r}
trimmgsFastqDir <- "/Users/hlberman/mnt/cluster/home6/callah_momspi/filter_trim/hmp2mgs_filt_fastq"
timmgsFastq <- list.files(trimmgsFastqDir) %>%
  str_extract("SRR[0-9]+") %>%
  unique
length(timmgsFastq)
```

## Filter human 
  Filter human reads from fastq files. Check filtering progress.
```{r}
filtmgsFastqDir <- "/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/bbmap_human/hmp2mgs_filt_fastq"
filtmgsFastq <- list.files(filtmgsFastqDir) %>%
  str_extract("SRR[0-9]+") %>%
  unique
length(filtmgsFastq)
```



# Count reads in filtered and unfiltered MGS fastq files
```{r, warning=FALSE}
# read in count fastq file:
readCountDF0 <- read_delim("/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/bbmap_human/hmp2mgs_fastq_counts_bbmap.txt", delim = "\ ") %>%
  mutate_if(is.numeric, ~(.x/4))
  
nrow(readCountDF0)
```

  Do number of forward and reverse downloaded reads match?
```{r}
readCountDF0 %>%
  mutate(test=uploadCount1==uploadCount2) %>%
  .$test %>%
  table
```

  Do number of forward and reverse trimmed reads match?
```{r}
readCountDF0 %>%
  mutate(test=trimmedCount1==trimmedCount2) %>%
  .$test %>%
  table
```

  Do number of forward and reverse filtered reads match?
```{r}
readCountDF0 %>%
  mutate(test=filteredCount1==filteredCount2) %>%
  .$test %>%
  table
```

  Make dataframe to add readcount information to metadata
```{r}
readCountDF <- readCountDF0 %>%
  dplyr::rename(TotalPairs=uploadCount1,
                Sickle_pairs=trimmedCount1,
                bbmapFiltered_pairs=filteredCount1) %>%
  select(SampleID, TotalPairs, Sickle_pairs, bbmapFiltered_pairs)
```


# Save Metagenome metadata
  Create and save metadata table
```{r}
mgsMetadata <- attrDF %>%
  right_join(mgsRunTable0[,c("SAMPLE_ID", "Run")], by="SAMPLE_ID") %>%
  dplyr::rename(SampleID=Run,
                SubjectID=dbGaP_Subject_ID,
                hmp2_SampleID=SAMPLE_ID,
                hmp2_SubjectID=SUBJECT_ID,
                reportedTrimester=reported_ga) %>%
  left_join(gaacGaad, by="visit_ID") %>%
  left_join(readCountDF, by="SampleID") %>%
  mutate(reportedTrimester=str_extract(reportedTrimester, "[:alpha:]*(?=[:space:]Trimester)"),
        sampleTrimester=case_when(!is.na(GDcoll)~case_when(GDcoll<=91~"First", # add trimester information based on GDcoll if present or reported trimester if not
                                   GDcoll>=92&GDcoll<=189~"Second",
                                   GDcoll>=190~"Third"),
                                   is.na(GDcoll)~reportedTrimester)) %>%
  select(-reportedTrimester) %>%
  group_by(SubjectID) %>%
  mutate(sampleOrder=str_extract(visit_num, "[0-9]"),
         sampleOrder=rank(sampleOrder)) %>%
  ungroup(SubjectID) %>%
  left_join(raceDF, by="SubjectID") %>%
  mutate(SampleID=str_extract(SampleID, "[0-9]{7}"),
         SampleID=as.character(SampleID))
#write_tsv(mgsMetadata, "../hmp2mgsDF.tsv")
```

# Map Gardnerella variants
## Bowtie output
```{r}
bowtiemgsfastq <- list.files("/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/gard_map/hmp2mgs_gard_fastq") %>%
   str_extract("SRR[0-9]{7}") %>%
    unique
length(bowtiemgsfastq)
```

## Usearch output
```{r}
usearchmgsAlns <- list.files("/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/gard_map/hmp2mgs_usearch_alns") %>%
  str_extract("SRR[0-9]{7}") %>%
  unique
length(usearchmgsAlns)
```

# Run Metaphlan
  (Ran in humann2)
  
  Check progress
```{r}
humann2outs <- list.files("/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/humann2/hmp2mgs_output") %>%
  str_extract("SRR[0-9]{7}")
length(humann2outs)
```

# MTS
Visits with paired MGS/MTS
```{r}
MGSvisits <- attrDF %>%
  filter(omic=="Metagenomics") %>%
  .$visit_ID

MTSvisits <- attrDF %>%
  filter(omic=="Metatranscriptomics") %>%
  .$visit_ID

matchedVisits <- intersect(MGSvisits, MTSvisits)
length(matchedVisits)

attrDF %>%
  filter(visit_ID %in% matchedVisits) %>%
  .$SUBJECT_ID %>%
  n_distinct
```
277 visits from 174 subjects have matched samples

Trimesters from matched samples
```{r}
attrDF %>%
  filter(visit_ID %in% matchedVisits) %>%
  select(visit_ID, reported_ga, SUBJECT_ID) %>%
  unique %>%
  ggplot(aes(x=reported_ga, y=SUBJECT_ID)) +
  geom_jitter(alpha=0.5) +
  theme(axis.text.y = element_blank()) +
  labs(x="", y="Subject ID")
```
Most matched samples are in the third trimester.

## Download MTS files
  Download using cart file on 8/10/2021. Check if all files downloaded against list of runs in run table:
```{r, message=FALSE}
mtssraDir <- "/Users/hlberman/mnt/cluster/scratch/hlberman/mts_files/sra" #directory path for sra files. Directory location is encrypted and access limited
downloadedmtsSRAfiles <- list.files(mtssraDir)
head(downloadedmtsSRAfiles)
length(downloadedmtsSRAfiles)

mtsRunTable <- read_csv("/Users/hlberman/mnt/cluster/home6/callah_momspi/SRAruntable.txt") %>% # file path for run table 
  filter(body_site=="Vagina",
         LibrarySource=="METATRANSCRIPTOMIC")
```

```{r}
downloadedmtsSRAs <- downloadedmtsSRAfiles %>%
  str_extract("SRR[0-9]{7}")
head(downloadedmtsSRAs)
length(downloadedmtsSRAs)

missingMTS <- mtsRunTable %>%
  filter(!(Run %in% downloadedmtsSRAs))

downloaded <- mtsRunTable %>%
  filter(Run %in% downloadedmtsSRAs)

missing <- mtsRunTable %>%
  filter(Run %!in% downloadedmtsSRAs)
length(missing$Run)

#write_lines(missing$Run, "/Users/hlberman/Desktop/missingSRAs.txt") 
```

  All sra files were downloaded. 

## Decrypt and get fastq files
  Check which fastq files have been downloaded:
```{r}
downloadedmtsFastqfiles <- list.files("/Users/hlberman/mnt/cluster/scratch/hlberman/mts_files/fastq")
head(downloadedmtsFastqfiles)
length(downloadedmtsFastqfiles)
length(downloadedmtsFastqfiles)/2
```

  What fastq files are missing?
```{r}
downloadedmtsFastqs <- downloadedmtsFastqfiles %>%
  str_extract("SRR[0-9]{7}") %>%
  unique
setdiff(downloadedmtsSRAs, downloadedmtsFastqs)
```
  All fastq files are downloaded
  
## Filter and trim with sickle
```{r}

```


```{r}
sessionInfo()
```