---
title: "Untitled"
author: "Hanna Berman"
date: "3/15/2021"
output: html_document
---

```{r}
library(tidyverse)
library(xml2)
library(formattable)
library(ggExtra)

`%!in%` <- negate(`%in%`)

theme_set(theme_bw()+
          theme(panel.grid = element_blank()))
```


```{r}
mgsRunTable <- read_csv("/Users/hlberman/mnt/cluster/home6/callah_momspi/SRAruntable.txt") %>%
  filter(body_site=="Vagina",
         LibrarySource=="METAGENOMIC")

#write_lines(mgsRunTable$Run, "/Users/hlberman/Desktop/MGSsras.txt")

phenFilePath <- "/Users/hlberman/mnt/cluster/home6/callah_momspi/Phenotypefiles"
phenFileTxts <- list.files(phenFilePath, pattern = "txt", full.names = TRUE)

txts <- phenFileTxts %>%
  map(~read_tsv(.x, comment="#"))

names(txts) <- phenFileTxts <- list.files(phenFilePath, pattern = "txt")

map(txts, head)
```

```{r}
vsamples <- txts$`phs001523.v1.pht007515.v1.p1.c1.MOMS_PI_Sample_Attributes.DS-PREG-COM-IRB-PUB-MDS.txt` %>%
  filter(HISTOLOGICAL_TYPE=="Vagina", omic!="16S")
reported_ga <- txts$`phs001523.v1.pht007510.v1.p1.c1.phenotype_clinical.DS-PREG-COM-IRB-PUB-MDS.txt` %>%
  select(dbGaP_Subject_ID, SUBJECT_ID, visit_ID, visit_num, reported_ga)

df <- vsamples %>%
  left_join(reported_ga, by = c("visit_ID", "visit_num"))

mgsRunTable0 <- mgsRunTable %>%
  dplyr::rename(SAMPLE_ID=biospecimen_repository_sample_id)
```

```{r}
n_distinct(df$SAMPLE_ID)
n_distinct(df$SUBJECT_ID)

df %>%
  group_by(SUBJECT_ID) %>%
  summarize(n_visits=n_distinct(visit_num)) %>%
  ggplot(aes(x=n_visits)) +
  geom_histogram()

df %>%
  group_by(omic) %>%
  tally %>%
  formattable()
```
MGS/MTS 1,227 samples from 234 subjects with 1 to 8 visits each.
297 MTS samples and 930 MGS samples. 

Trimester of samples
```{r}
df %>%
  ggplot(aes(x=reported_ga, y=SUBJECT_ID, color=omic)) +
  geom_jitter(alpha=0.5) +
  theme(axis.text.y = element_blank()) +
  labs(x="Trimester", y="Subject ID")
```


Visits with paired MGS/MTS
```{r}
MGSvisits <- df %>%
  filter(omic=="Metagenomics") %>%
  .$visit_ID

MTSvisits <- df %>%
  filter(omic=="Metatranscriptomics") %>%
  .$visit_ID

matchedVisits <- intersect(MGSvisits, MTSvisits)
length(matchedVisits)

df %>%
  filter(visit_ID %in% matchedVisits) %>%
  .$SUBJECT_ID %>%
  n_distinct
```
277 visits from 174 subjects have matched samples

Trimesters from matched samples
```{r}
df %>%
  filter(visit_ID %in% matchedVisits) %>%
  select(visit_ID, reported_ga, SUBJECT_ID) %>%
  unique %>%
  ggplot(aes(x=reported_ga, y=SUBJECT_ID)) +
  geom_jitter(alpha=0.5) +
  theme(axis.text.y = element_blank()) +
  labs(x="", y="Subject ID")
```
Most matched samples are in the third trimester.

Remaining questions:
1. Term vs. Preterm?
2. Gestational age at collection beyond trimester

## Subject metadata
  Lists of samples and subjects IDs in MGS data for convenience
```{r}
#MGS samples
MGSsamps <- df %>%
  filter(omic=="Metagenomics") %>%
  .$SAMPLE_ID %>%
  unique
#MGS Subjects
MGSsubs <- df %>%
  filter(omic=="Metagenomics") %>%
  .$SUBJECT_ID %>%
  unique
```


  Assess a few variables from clinical and health history surveys
### Health History
```{r}
hxDF <- txts$`phs001523.v1.pht007512.v1.p1.c1.phenotype_healthHistorySurvey.DS-PREG-COM-IRB-PUB-MDS.txt` %>%
  filter(SUBJECT_ID %in% MGSsubs)
#preview table
head(hxDF)
colnames(hxDF)
#number of rows
nrow(hxDF)
#number of subjects represented in the health history
n_distinct(hxDF$SUBJECT_ID)
```
  There are 236 entries from 233 subjects out of 234 total subjects that have shotgun sequencing
  Who is duplicated:
```{r}
hxDF %>%
  group_by(SUBJECT_ID) %>%
  tally %>%
  filter(n>1) %>%
  .$SUBJECT_ID
```

  Who is missing:
```{r}
setdiff(MGSsubs, hxDF$SUBJECT_ID)
```

  Add all subjects
```{r}
hxDF <- hxDF %>%
  full_join(df[,c("dbGaP_Subject_ID", "SUBJECT_ID")], by=c("dbGaP_Subject_ID", "SUBJECT_ID"))
```

**Maternal Age**
```{r}
hxDF %>%
  count(age) %>%
  mutate(n=round((n/sum(n)*100), 2),
         age=factor(age, levels=c("Below_18", "18", "18_to_28", "29-38", "Above_38"))) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

**Maternal Race**
Stanford and UAB levels: White, Black, Asian, American Indian, Pacific Islander, Unknown, Other
```{r}
# create race data frame
raceDF0 <- hxDF %>%
  mutate(Race=case_when(african_american=="Yes"&american_indian_or_alaska_native=="No"&asian=="No"&caucasian=="No"&hispanic_or_latino=="No"&native_hawaiian=="No"~"Black",
                        african_american=="No"&american_indian_or_alaska_native=="No"&asian=="No"&caucasian=="Yes"&hispanic_or_latino=="No"&native_hawaiian=="No"~"White",
                        african_american=="No"&american_indian_or_alaska_native=="No"&asian=="Yes"&caucasian=="No"&hispanic_or_latino=="No"&native_hawaiian=="No"~"Asian",
                        african_american=="No"&american_indian_or_alaska_native=="Yes"&asian=="No"&caucasian=="No"&hispanic_or_latino=="No"&native_hawaiian=="No"~"American Indian",
                        african_american=="No"&american_indian_or_alaska_native=="No"&asian=="No"&caucasian=="No"&hispanic_or_latino=="No"&native_hawaiian=="Yes"~"Native Hawaiian",
                        african_american=="No"&american_indian_or_alaska_native=="No"&asian=="No"&caucasian=="No"&hispanic_or_latino=="Yes"&native_hawaiian=="No"~"Hispanic or Latino",
                        is.na(african_american)&is.na(american_indian_or_alaska_native)&is.na(asian)&is.na(caucasian)&is.na(hispanic_or_latino)&is.na(native_hawaiian)~"Unknown")) %>%
  replace_na(list(Race="Other")) %>%
  select(dbGaP_Subject_ID, Race) %>%
  dplyr::rename(SubjectID=dbGaP_Subject_ID) %>%
  unique


# check that there is only one race category per subject
nrow(raceDF0)==n_distinct(raceDF0$SubjectID)
```
  There is at least one subject with multiple races. Check which subject(s) and why:
  
```{r}
multiRaceSubject <- raceDF0 %>%
  count(SubjectID) %>%
  filter(n>1) %>%
  .$SubjectID
multiRaceSubject


hxDF %>%
  select(dbGaP_Subject_ID, visit_ID, african_american, american_indian_or_alaska_native, asian, hispanic_or_latino, native_hawaiian) %>%
  filter(dbGaP_Subject_ID %in% multiRaceSubject) %>%
  unique
```
  One subject has indicated races differently at two different visits. Indicate subject as unknown race.

```{r}
raceDF <- raceDF0 %>%
  mutate(Race=case_when(SubjectID==multiRaceSubject~"Unknown",
                   SubjectID!=multiRaceSubject~Race)) %>%
  unique()

# test again that there is only one entry per subject
nrow(raceDF)==n_distinct(raceDF$SubjectID)
```

```{r}
raceDF %>%
  count(Race) %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

**BV**
```{r}
hxDF %>%
  group_by(bv) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

**BV history**
```{r}
hxDF %>%
  group_by(bv_history) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

**BV Lifetime Number**
```{r}
hxDF %>%
  group_by(bv_lifetime_number) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

**Yeast**
```{r}
hxDF %>%
  group_by(yeast) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

**Yeast History**
```{r}
hxDF %>%
  group_by(yeast_infection_history) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

**Yeast Infection Lifetime Number**
```{r}
hxDF %>%
  group_by(yeast_infection_lifetime_number) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```


  **Dataframe with age and race**
```{r}
df2015 <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/pregnancy_metagenome/metadata_files/sampledata_VaginalSwabs_Pregnancy_DiGiulio2015.tsv" %>%
  read_tsv()
df2017 <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/pregnancy_metagenome/metadata_files/MAP_MANUSCRIPT_IL02-IL09_ToBen_20170404_fixed.txt" %>%
  read_tsv()

ageDF <- hxDF %>%
  select(SUBJECT_ID, visit_ID, age)
```

### Clinical
```{r}
clinDF <- txts$`phs001523.v1.pht007510.v1.p1.c1.phenotype_clinical.DS-PREG-COM-IRB-PUB-MDS.txt` %>%
  filter(visit_ID %in% MGSvisits)
colnames(clinDF)
nrow(clinDF)
n_distinct(clinDF$SUBJECT_ID)
n_distinct(clinDF$visit_ID)
```

**vaginal pH**
```{r}
clinDF %>%
  group_by(vaginal_pH) %>%
  tally %>%
  #mutate(n=round((n/sum(n)*100), 2)) %>%
  #dplyr::rename(`%`=n) %>%
  #formattable()
  ggplot(aes(x=vaginal_pH, y=n)) +
  geom_col(stat="identity")

count(hxDF, mom_delivery_method)
```

### Review of Systems Survey
```{r}
rssDF <- txts$`phs001523.v1.pht007511.v1.p1.c1.phenotype_reviewOfSystemsSurvey.DS-PREG-COM-IRB-PUB-MDS.txt` %>%
  filter(visit_ID %in% MGSvisits)
nrow(rssDF)
```

  Have this survey for 660/930 visits where MGS samples were collected
```{r}
rssDF %>%
  group_by(antibiotics_current) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

```{r}
rssDF %>%
  group_by(antibiotics_since_last_visit) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

**BV since last visit**
```{r}
rssDF %>%
  group_by(bv_since_last_visit) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

**UTI since last visit**
```{r}
rssDF %>%
  group_by(yeast_infection_since_last_visit) %>%
  tally %>%
  mutate(n=round((n/sum(n)*100), 2)) %>%
  dplyr::rename(`%`=n) %>%
  formattable()
```

```{r}
txts %>%
  map(colnames) %>%
  map(~str_detect(.x, "Weight")) %>%
  map(unique)
```


### additional information from Greg Buck
```{r}
gaacGaad0 <- read_tsv("/Users/hlberman/mnt/cluster/home6/callah_momspi/gaAtCollecNDel_4_PTB_042121.txt")
head(gaacGaad0)
colnames(gaacGaad0)
#Number of MGS samples/subjects that have gaac and gaad information
gaacGaad0 %>%
  filter(visitID %in% MGSvisits) %>%
  summarize_at(c("participantID", "visitID"),n_distinct)
```
  503 samples of 930, from 135 subjects of 234

  Clean up to match other DF
```{r}
# fix ga at sampling and delivery
gaacGaad <- gaacGaad0 %>%
  mutate(ga_at_sampling=na_if(ga_at_sampling, "2NA"),
         ga_at_sampling=na_if(ga_at_sampling, "3NA"),
         ga_at_sampling=na_if(ga_at_sampling, "4NA"),
         ga_at_delivery=na_if(ga_at_delivery, "2NA"),
         ga_at_delivery=na_if(ga_at_delivery, "3NA"),
         ga_at_delivery=na_if(ga_at_delivery, "4NA")) %>%
  separate(ga_at_sampling, c("GWcoll", "GDcoll"), sep="w") %>%
  separate(ga_at_delivery, c("GWdel", "GDdel"), sep="w") %>%
  mutate_at(c("GDcoll", "GDdel"), ~str_extract(.x, "[0-9]*")) %>%
  mutate_at(c("GWcoll", "GDcoll", "GWdel", "GDdel"), as.numeric) %>%
  mutate_at(c("GWcoll", "GWdel"), ~.x*7) %>%
  mutate(GDcoll=GWcoll+GDcoll, 
         GWcoll=GDcoll/7,
         GDdel=GWdel+GDdel,
         GWdel=GDdel/7, 
         TermPre=case_when(`case/control`=="case"~"P",
                           `case/control`=="control"~"T")) %>%
  dplyr::rename(visit_ID=visitID) %>%
  select(-`case/control`, -participantID)
head(gaacGaad)
```


# Download MGS files
## Download .sra files
  Attempt to download files using cart file method on 3/23/2021 but not all files downloaded.
```{r}
# mgssraDir <- "/Users/hlberman/mnt/cluster/home6/callah_momspi/mgs_files/sra" 
# downloadedmgsSRAfiles <- list.files(mgssraDir)
# head(downloadedmgsSRAfiles)
# length(downloadedmgsSRAfiles)
# 
# downloadedmgsSRAs <- downloadedmgsSRAfiles %>%
#   str_extract("SRR[0-9]{7}")
# head(downloadedmgsSRAs)
# length(downloadedmgsSRAs)
```
  441 files were downloaded.
  698 after the second round.
  821 after third round.
  875 after fourth round.
  904 after fifth round.
  916 after sixth round.
  926 after seventh round.
  930 after eigth round!

  What's missing?  
```{r}
# missing <- mgsRunTable %>%
#   filter(!(Run %in% downloadedmgsSRAs))
# 
# downloaded <- mgsRunTable %>%
#   filter(Run %in% downloadedmgsSRAs)
# 
# missing <- mgsRunTable %>%
#   filter(Run %!in% downloadedmgsSRAs)
# 
# #write_lines(missing$Run, "/Users/hlberman/Desktop/missingSRAs.txt")
```
 Copy list of missing SRAs into download script `/Users/hlberman/mnt/cluster/home6/callah_momspi/download_scripts/download_mgs.sh`

## Decrypt and get fastq files
  List of all MGS SRAs
```{r}
# mgsSRAs <- mgsRunTable$Run
# #write_lines(mgsSRAs, "/Users/hlberman/Desktop/mgsSRAs.txt")
```
  Copy list of all SRAs to `/Users/hlberman/mnt/cluster/home6/calla_momspi/download_scripts/decrypt_mgs.sh`
  Check which fastq files have been downloaded:
```{r}
# downloadedmgsFastqfiles <- list.files("/Users/hlberman/mnt/cluster/home6/callah_momspi/mgs_files/fastq")
# head(downloadedmgsFastqfiles)
# length(downloadedmgsFastqfiles)
# 930*2 # number of fastq files there should be
```

  What's missing?
  Repeat script for any SRAs that are missing.
```{r}
# downloadedmgsFastqSRAs <- str_extract(downloadedmgsFastqfiles, "SRR[0-9]*") %>%
#   unique
# head(downloadedmgsFastqSRAs)
# # test if each SRA has two files:
# length(downloadedmgsFastqSRAs)==(length(downloadedmgsFastqfiles)/2)
```

```{r}
# missingmgsFastq <- setdiff(mgsSRAs, downloadedmgsFastqSRAs)
# length(missingmgsFastq)
# missingmgsFastq
```

# Filtering fastq files
## Filter and Trim
  Filter and trim with sickle. Check progress:
```{r}
trimmgsFastqDir <- "/Users/hlberman/mnt/cluster/home6/callah_momspi/filter_trim/hmp2mgs_filt_fastq"
timmgsFastq <- list.files(trimmgsFastqDir) %>%
  str_extract("SRR[0-9]+") %>%
  unique
length(timmgsFastq)
```

## Filter human 
  Filter human reads from fastq files. Check filtering progress.
```{r}
filtmgsFastqDir <- "/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/bbmap_human/hmp2mgs_filt_fastq"
filtmgsFastq <- list.files(filtmgsFastqDir) %>%
  str_extract("SRR[0-9]+") %>%
  unique
length(filtmgsFastq)
```


## Count reads in filtered and unfiltered fastq files
```{r, warning=FALSE}
# read in count fastq file:
readCountDF0 <- read_delim("/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/bbmap_human/hmp2mgs_fastq_counts_bbmap.txt", delim = "\ ") %>%
  mutate_if(is.numeric, ~(.x/4))
  
nrow(readCountDF0)
```

  Do number of forward and reverse downloaded reads match?
```{r}
readCountDF0 %>%
  mutate(test=uploadCount1==uploadCount2) %>%
  .$test %>%
  table
```

  Do number of forward and reverse trimmed reads match?
```{r}
readCountDF0 %>%
  mutate(test=trimmedCount1==trimmedCount2) %>%
  .$test %>%
  table
```

  Do number of forward and reverse filtered reads match?
```{r}
readCountDF0 %>%
  mutate(test=filteredCount1==filteredCount2) %>%
  .$test %>%
  table
```

  Distribution sequenced, trimmed, and human filtered reads
```{r}
summary(readCountDF0$uploadCount1)
summary(readCountDF0$trimmedCount1)
summary(readCountDF0$filteredCount1)
```


### Test alternative QC filtering stats
```{r}
# altQCCounts <- read_delim("/Users/hlberman/mnt/cluster/home6/callah_momspi/filter_trim/mgs_qc_filter_test_fastq_counts_bbmap.txt", delim=" ", col_names = c("SampleID", "q15Count1", "q15Count2", "l50Count1", "l50Count2")) %>%
#   mutate_if(is.numeric, ~.x/4)
# nrow(altQCCounts)
# 
# altQCCounts$q15Count1==altQCCounts$q15Count2
# 
# altQCCounts$l50Count1==altQCCounts$l50Count2
# 
# compareReadsDF %>%
#   spread(seqStep, pairs) %>%
#   right_join(altQCCounts, by="SampleID") %>%
#   select(-q15Count2, -l50Count2) %>%
#   dplyr::rename(q15_pairs=q15Count1,
#                 l50_pairs=l50Count1) %>%
#   gather("sickle_params", "pairs", c(TotalPairs, Sickle_pairs, q15_pairs, l50_pairs)) %>%
#   mutate(sickle_params=factor(sickle_params, levels=c("TotalPairs", "Sickle_pairs", "q15_pairs", "l50_pairs"), labels=c("Total Pairs", "q20, l100", "q15, l100", "q20, l50"))) %>%
#   ggplot(aes(x=sickle_params, y=pairs, color=sickle_params)) +
#   geom_boxplot() +
#   labs(x=NULL, y="Read Pairs", color="Sickle Parameters")
# 
# compareReadsDF %>%
#   spread(seqStep, pairs) %>%
#   right_join(altQCCounts, by="SampleID") %>%
#   select(-q15Count2, -l50Count2) %>%
#   dplyr::rename(q15_pairs=q15Count1,
#                 l50_pairs=l50Count1) %>%
#   mutate_at(c("Sickle_pairs", "q15_pairs", "l50_pairs"), ~.x/TotalPairs) %>%
#   summarise_at(c("Sickle_pairs", "q15_pairs", "l50_pairs"), median) %>%
#   dplyr::rename(`q20, l100`=Sickle_pairs,
#                 `q15, l100`=q15_pairs,
#                 `q20, l50`=l50_pairs) %>%
#   formattable()
```


```{r}
readCountDF <- readCountDF0 %>%
  dplyr::rename(TotalPairs=uploadCount1,
                Sickle_pairs=trimmedCount1,
                bbmapFiltered_pairs=filteredCount1) %>%
  mutate(pctHuman=((Sickle_pairs-bbmapFiltered_pairs)/Sickle_pairs)*100,
         bacLoad=bbmapFiltered_pairs/(Sickle_pairs-bbmapFiltered_pairs)) %>%
  select(SampleID, TotalPairs, Sickle_pairs,bbmapFiltered_pairs, pctHuman, bacLoad)

summary(readCountDF$bacLoad)

readCountDF %>%
  ggplot(aes(x=bacLoad)) +
  geom_histogram()
```


# Save Metagenome metadata
```{r}
mgsMetadata <- df %>%
  right_join(mgsRunTable0[,c("SAMPLE_ID", "Run")], by="SAMPLE_ID") %>%
  dplyr::rename(SampleID=Run,
                SubjectID=dbGaP_Subject_ID,
                hmp2_Sample_ID=SAMPLE_ID,
                hmp2_Subject_ID=SUBJECT_ID) %>%
  left_join(gaacGaad, by="visit_ID") %>%
  left_join(readCountDF, by="SampleID") %>%
  left_join(raceDF, by="SubjectID") %>%
  mutate(SampleID=str_extract(SampleID, "[0-9]{7}"),
         SampleID=as.character(SampleID))
#write_tsv(mgsMetadata, "../hmp2mgsDF.tsv")
```

# Map Gardnerella variants
## Bowtie output
```{r}
bowtiemgsfastq <- list.files("/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/gard_map/hmp2mgs_gard_fastq") %>%
   str_extract("SRR[0-9]{7}") %>%
    unique
length(bowtiemgsfastq)
```

## Usearch output
```{r}
usearchmgsAlns <- list.files("/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/gard_map/hmp2mgs_usearch_alns") %>%
  str_extract("SRR[0-9]{7}") %>%
  unique
length(usearchmgsAlns)
```

# Run Metaphlan
  (Ran in humann2)
  
  Check progress
```{r}
humann2outs <- list.files("/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/humann2/hmp2mgs_output") %>%
  str_extract("SRR[0-9]{7}")
length(humann2outs)
```


# Download MTS files
  Download using cart file on 3/31/2021. Check if all files downloaded:
```{r}
mtssraDir <- "/Users/hlberman/mnt/cluster/home6/callah_momspi/mts_files/sra" 
downloadedmtsSRAfiles <- list.files(mtssraDir)
head(downloadedmtsSRAfiles)
length(downloadedmtsSRAfiles)

downloadedmtsSRAs <- downloadedmtsSRAfiles %>%
  str_extract("SRR[0-9]{7}")
head(downloadedmtsSRAs)
length(downloadedmtsSRAs)
```

```{r}
mtsRunTable <- read_csv("/Users/hlberman/mnt/cluster/home6/callah_momspi/SRAruntable.txt") %>%
  filter(body_site=="Vagina",
         LibrarySource=="METATRANSCRIPTOMIC")

missingMTS <- mtsRunTable %>%
  filter(!(Run %in% downloadedmtsSRAs))

downloaded <- mtsRunTable %>%
  filter(Run %in% downloadedmtsSRAs)

missing <- mtsRunTable %>%
  filter(Run %!in% downloadedmtsSRAs)

#write_lines(missing$Run, "/Users/hlberman/Desktop/missingSRAs.txt") 
```
  Copy list of missing SRAs into download script `/Users/hlberman/mnt/cluster/home6/callah_momspi/download_scripts/download_mts.sh`

  List of all MTS SRAs
```{r}
mtsSRAs <- mtsRunTable$Run
#write_lines(mtsSRAs, "/Users/hlberman/Desktop/mtsSRAs.txt")
```
  Copy list of all SRAs to `/Users/hlberman/mnt/cluster/home6/calla_momspi/download_scripts/decrypt_mts.sh`

  Check that all SRAs were downloaded
```{r}
mtssraDir <- "/Users/hlberman/mnt/cluster/home6/callah_momspi/mts_files/sra" 
downloadedmtsSRAfiles <- list.files(mtssraDir)
head(downloadedmtsSRAfiles)
length(downloadedmtsSRAfiles)

downloadedmtsSRAs <- downloadedmtsSRAfiles %>%
  str_extract("SRR[0-9]{7}")
head(downloadedmtsSRAs)
length(downloadedmtsSRAs)
```
  All sra files were downloaded. 

## Decrypt and get fastq files
  Check which fastq files have been downloaded:
```{r}
downloadedmtsFastqfiles <- list.files("/Users/hlberman/mnt/cluster/home6/callah_momspi/mts_files/fastq")
head(downloadedmtsFastqfiles)
length(downloadedmtsFastqfiles)
length(downloadedmtsFastqfiles)/2
```

  What fastq files are missing?
```{r}
downloadedmtsFastqs <- downloadedmtsFastqfiles %>%
  str_extract("SRR[0-9]{7}") %>%
  unique
setdiff(downloadedmtsSRAs, downloadedmtsFastqs)
```
  Rerun for missing fastq files

```{r}
sessionInfo()
```