---
title: "Assign clades or genomospecies to short reads"
subtitle: "One reference strain per genomospecies"
author: "Hanna Berman"
output: html_document
---

# Setup
##Load packages
```{r libraries, warning=FALSE}
library(tidyverse)
library(Biostrings); packageVersion("Biostrings")
options(scipen=10000)
`%!in%` <- negate(`%in%`)
```

## Define file paths
### Reference genes and usearch output
```{r}
# Directories
repo_dir <- "../../"
gard_tree_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/Gard WGS"
output_dir <- file.path(repo_dir, "gard_mapping/clade_assignments/")
gdrive_path <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/gardMapping/clade_assignments/20200820_1Strain1GenomospeciesOutput"
 
# Usearch alignments 
usearchAlns <- "/Users/hlberman/mnt/cluster/home/hlberman/VMMG/gard_map/usearch_alns_1Strain1Genomospecies" #path to blast6 tables from usearch alignments

# Gardnerella reference genes
ref_db <- "./ref_genes/refGenes.fasta" %>%
  readDNAStringSet(); head(ref_db)[1:5]

# reference genes annotations
refGenesDF <- "./ref_genes/refGenesDF.csv" %>%
  read_csv(); head(refGenesDF)

# Gardnerella reference strains metadata
refDF <- file.path(repo_dir, "GardnerellaMetadata.csv") %>%
  read_csv() %>%
  filter(In_Directory==TRUE); head(refDF)

#13 genomospecies of reference strains
referenceGenomospecies <- file.path(repo_dir, "72GardStrainGenomospecies.csv") %>%
  read_csv(); head(referenceGenomospecies)

# shotgun metadata
sgDF <- file.path(repo_dir, "shotgunMetadata.tsv") %>%
  read_delim(delim="\t") %>%
  mutate(SampleID=as.character(SampleID), 
         SubjectID=as.character(SubjectID)); head(sgDF)
```

# Asses alignments by usearch
## set up (load data)
```{r}
#alns <- list.files(usearchAlns, all.files = FALSE, full.names = TRUE) %>%
#   map(read_delim, delim="\t", col_names=c("query", "reference", "pcntid", "length", "mismatch", "gap_open", "qstart", "qend", "rstart", "rend", "evalue", "bitscore")) #read in blast tables
# names(alns) <-  list.files(usearchAlns, all.files = FALSE, full.names = FALSE) %>%
#   str_extract("[0-9]+")
#save(alns, file=file.path(repo_dir, "gard_mapping", "clade_assignments", "alns_1Strain1Genomospecies.Rdata"))
load(file.path(repo_dir, "gard_mapping", "clade_assignments", "alns_1Strain1Genomospecies.Rdata"))
```

## Set up (formatting)
```{r}
#remove strains where no gard mapped
nrows <- map(alns, nrow) 
alns <- alns[nrows>0]

#add annotation info
refAnnotations <- refGenesDF %>%
  select(Annotation, refDF$Strain[refDF$In_Directory==TRUE]) %>%
  gather("strain", "reference", 2:ncol(.))
alns0 <- alns %>%
  map2(., names(alns), ~mutate(.x, SampleID=.y)) %>%
  map(~left_join(.x, refAnnotations[,c("Annotation", "reference")], by="reference")) %>%
  map(~mutate(.x, Strain=str_sub(.x$reference, 1L, -7L))) %>% #remove gene identifiers so that only strains appear in Strain column
  map(~left_join(.x, refDF[,c("Strain", "Clade")], by="Strain")) %>% # add reference gene/strain clade information
  map(~left_join(.x, referenceGenomospecies, by="Strain")) %>% # add genomospecies information
  map(~left_join(.x, sgDF[,c("SampleID", "SubjectID", "RNAlater", "GWcoll", "PaperCohort")], by="SampleID")) #add sample metadata
```

## Some descriptives
```{r}
HitsPerSample <- alns %>% #count number of hits coming back in each sample 
  map(~nrow(.x)) %>%
  as_tibble() %>%
  t() %>%
  as_tibble()
ggplot(data = HitsPerSample, aes(V1)) +
  geom_histogram() +
  theme_classic() +
  labs(title = "'Hits' per sample", x = "'Hits'")
#ggsave(file.path(gdrive_path, "figures", "hitsPerSample.pdf"))

ReadsPerSample <- alns0 %>%
  map(~group_by(.x, SampleID)) %>%
  map(~summarise(.x, readspersample=n_distinct(query))) %>%
  map(~ungroup(.x)) %>%
  purrr::reduce(full_join, by=c("SampleID", "readspersample"))
ggplot(data = ReadsPerSample, aes(readspersample)) +
  geom_histogram(bins=50) +
  theme_classic() +
  labs(title = "Reads per sample", x = "# Reads")
#ggsave(file.path(gdrive_path, "figures", "readsPerSample.pdf"))

ggplot(data = ReadsPerSample, aes(readspersample)) +
  geom_histogram(bins=50) +
  theme_classic() +
  labs(title = "Reads per sample", x = "# Reads")
#ggsave(file.path(gdrive_path, "figures", "readsPerSampleMax2000.pdf"))
  
QueriesPerAnnotation <- alns0 %>% #number of reads that map to each annotation
  map(~group_by(.x, Annotation)) %>%
  map(~summarize(.x, queriesbyannotation=n_distinct(query))) %>%
  map(~ungroup(.x)) %>%
  map2(., names(alns0), ~mutate(.x, sampleID=.y)) %>%  
  purrr::reduce(full_join, by=c("Annotation", "queriesbyannotation", "sampleID"))
ggplot(data=QueriesPerAnnotation, aes(x=Annotation, y=queriesbyannotation)) +
  geom_violin() +
  labs(title="Reads per annotation", x="Annotation", y="Number of reads aligning") +
  theme(axis.text.x = element_text(angle=45, hjust=1, size=6))
#ggsave(file.path(gdrive_path, "figures", "queriesPerAnnotation.pdf"), width=10, height=8)
 
ReferencesPerQuery <- alns0 %>%  #number of references that each query maps to
  map(~group_by(.x, query)) %>%
  map(~summarize(.x, referencesbyquery=n_distinct(reference))) %>%
  map(~ungroup(.x)) %>%
  map2(., names(alns0), ~mutate(.x, sampleID=.y)) %>%
  purrr:: reduce(full_join, by=c("query", "referencesbyquery", "sampleID"))
ggplot(data = ReferencesPerQuery, aes(referencesbyquery)) +
  geom_histogram() +
  theme_classic() +
  labs(title = "Number of reference genes that each query maps to", x = "# Reference Genes") 
#ggsave(file.path(gdrive_path, "figures", "refsPerQuery.pdf")) 

QueriesPerReference <- alns0 %>% #number of queries that map to each reference (per sample)
  map(~group_by(.x, reference)) %>%
  map(~summarize(.x, queriesbyreference=n_distinct(query))) %>%
  map(~ungroup(.x)) %>%
  map2(., names(alns0), ~mutate(.x, sampleID=.y)) %>%
  purrr:: reduce(full_join, by=c("reference", "queriesbyreference", "sampleID"))
ggplot(data = QueriesPerReference, aes(queriesbyreference)) +
  geom_histogram(bins=50) +
  theme_classic() +
  labs(title = "Number of queries(unique reads) that map to each reference (per sample)", x = "# queries (reads)") 
#ggsave(file.path(gdrive_path,"figures", "queriesPerRef.pdf"))

StrainsPerSample <- alns0 %>%  #get count of strains mapped in each sample
  map_df(~summarise(.x, nSamples=n_distinct(.x$Strain))) %>%
  mutate(SampleID=names(alns))
ggplot(data = StrainsPerSample, aes(nSamples)) +
  geom_histogram(bins=50) +
  theme_classic() +
  labs(title = "Number of strains found in each sample", x = "# strains") 
#ggsave(file.path(gdrive_path, "figures", "strainsPerSample.pdf"))

CladesPerSample <- alns0 %>%  #get count of clades mapped in each sample
  map_df(~summarise(.x, nClades=n_distinct(Clade))) %>%
  mutate(SampleID=names(alns))
ggplot(data = CladesPerSample, aes(nClades)) +
  geom_histogram() +
  theme_classic() +
  labs(title = "Number of clades found in each sample", x = "# clades")  
#ggsave(file.path(gdrive_path, "figures", "cladesPerSample.pdf"))
 
SpeciesPerSample <- alns0 %>% #get count of species mapped in each sample
  map_df(~summarise(.x, nSpecies=n_distinct(SpeciesNumber))) %>%
  mutate(SampleID=names(alns))
ggplot(data = SpeciesPerSample, aes(nSpecies)) +
  geom_histogram() +
  theme_classic() +
  scale_x_continuous(breaks=seq(0, 13, by=1)) +
  labs(title = "Number of genomospecies found in each sample", x = "# genomospecies")  
#ggsave(file.path(gdrive_path, "figures", "speciesPerSample.pdf")) 
 
sumStats1 <- sapply(list(Reads_per_sample=ReadsPerSample$readspersample,
                         Hits_per_sample=HitsPerSample$V1,
                         Queries_per_annotation=QueriesPerAnnotation$queriesbyannotation,
                         References_per_query=ReferencesPerQuery$referencesbyquery,
                         Queries_per_reference=QueriesPerReference$queriesbyreference,
                         Strains_per_sample=StrainsPerSample$nSamples, 
                         Clades_per_sample=CladesPerSample$nClades,
                         Species_per_sample=SpeciesPerSample$nSpecies), 
                     summary) %>%
  as.data.frame() %>%
  rownames_to_column("stat")
sumStats1
#write_csv(sumStats1, file.path(gdrive_path, "usearchAlignmentSummaryStats1.csv"))
```

## Assess top alignmnet percent identities
```{r}
allHits <- alns0 %>%
  map2(., names(alns), ~mutate(.x, sampleID=.y)) %>%
  map(~select(.x, sampleID, query, reference, Annotation, Strain, Clade, SpeciesNumber, SpeciesName, pcntid, length)) %>%
  purrr::reduce(full_join, by=c("sampleID", "query", "reference", "Annotation", "Strain", "Clade", "SpeciesNumber", "SpeciesName", "pcntid", "length")) %>%
  mutate(sampleID=as.character(sampleID))
#reference by annotation
#violin plot of each ref gene by annotation
ggplot(data = allHits, aes(pcntid)) +
  geom_histogram(binwidth = 1) +
  theme_classic() +
  labs(title = "Percent ID of all hits", x = "Percent ID") 
#ggsave(file.path(gdrive_path, "figures", "hitID_hist.pdf"))

ggplot(data=allHits, aes(x=Annotation, y=pcntid))+
  geom_violin() +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  scale_y_continuous(breaks = seq(0, 100, by = 2.5)) +
  labs(y="Percent ID")
#ggsave(file.path(gdrive_path, "figures","hitID_violinByAnno.pdf"), height = 8, width = 11)

ggplot(data=allHits, aes(x=sampleID, y=pcntid)) +
  geom_violin() +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  scale_y_continuous(breaks = seq(0, 100, by = 2.5)) +
  labs(y="Percent ID")
#ggsave(file.path(gdrive_path, "figures", "hitID_violinBySample.pdf"), height = 8, width = 11)

maxByQuery <- alns %>%  #max percent of ID of alignments per query
  map(~group_by(.x, query)) %>%
  map(~summarise(.x, maxbyquery=max(pcntid))) %>%
  map(~ungroup(.x)) %>%
  map2(., names(alns), ~mutate(.x, sampleID=.y)) %>%
  purrr::reduce(full_join, by=c("query", "maxbyquery", "sampleID"))
ggplot(data = maxByQuery, aes(maxbyquery)) +
  geom_histogram(binwidth = 1) +
  theme_classic() +
  labs(title = "Max percent ID per input query", x = "Percent Identity")
#ggsave(file.path(gdrive_path, "figures","maxIDperQuery_hist.pdf"))

maxByReference <- alns %>%  #max percent ID of alignments per reference (within each sample)
  map(~group_by(.x, reference)) %>%
  map(~summarise(.x, maxbyreference =max(pcntid))) %>%
  map(~ungroup(.x)) %>%
  map2(., names(alns), ~mutate(.x, sampleID=.y)) %>%
  purrr::reduce(full_join, by=c("reference", "maxbyreference", "sampleID"))
ggplot(data = maxByReference, aes(maxbyreference)) +
  geom_histogram(binwidth = 1) +
  theme_classic() +
  labs(title = "Max percent ID per reference (per sample)", x = "Percent ID")
#ggsave(file.path(gdrive_path, "figures","maxIDperRef_hist.pdf"))

maxByAnnotation <- alns0 %>% #max percentID 
  map(~group_by(.x, Annotation)) %>%
  map(~summarise(.x, maxbyannotation =max(pcntid))) %>%
  map(~ungroup(.x)) %>%
  map2(., names(alns), ~mutate(.x, sampleID=.y)) %>%
  purrr::reduce(full_join, by=c("Annotation", "maxbyannotation", "sampleID"))
ggplot(data = maxByAnnotation, aes(maxbyannotation)) +
  geom_histogram(binwidth = 1) +
  theme_classic() +
  labs(title = "Max percent ID per annotation (per sample)", x = "Percent ID")
#ggsave(file.path(gdrive_path, "figures","maxIDperAnno_hist.pdf"))

sumStats2 <- sapply(list(PercentID=allHits$pcntid,
            Max_by_query=maxByQuery$maxbyquery,
            Max_by_reference=maxByReference$maxbyreference,
            Max_by_annotation=maxByAnnotation$maxbyannotation),
            summary) %>%
  as.data.frame() %>%
  rownames_to_column("stat")
sumStats2
#write_csv(sumStats2, file.path(gdrive_path, "usearchAlignmentSummaryStats2_maxIDs.csv"))
```
Variation in hit percent ID and maxID per reference demonstrate variation over G. vaginalis. Strong mode of maximum % ID match per annotation and per query within each sample suggest that threshold of 99% or 100% ID for classfying clades may allow for strong specificty without sacrificing sensitivity.

# Take top hit above 99%
```{r}
alns99top <- alns0 %>%
  map(~filter(.x, pcntid>=99)) %>%
  map(~group_by(.x, query)) %>%
  map(~filter(.x, pcntid==max(pcntid))) %>%
  map(~ungroup(.x))

countCladesBySample99top <- alns99top %>%  #get count of clades mapped in each sample
  map_df(~summarise(.x, nClades=n_distinct(Clade))) %>%
  mutate(SampleID=names(alns)) %>%
  left_join(sgDF[,c("SampleID","SubjectID", "GWcoll")], by="SampleID")
ggplot(data = countCladesBySample99top, aes(nClades)) +
  geom_histogram() +
  theme_classic() +
  scale_y_continuous(breaks = seq(0, 100, by = 10)) +
  labs(title = "Number of clades found in each sample (filter to% ID>=99% (top hits only))", x = "# clades")
#ggsave(file.path(gdrive_path, "figures","cladesPerSample99top.pdf"))
```
# Assess duplicates
```{r}
hitsPerSample99top <- alns99top %>%
  map(~nrow(.x)) %>%
  as_tibble() %>%
  t() %>%
  as_tibble() %>%
  mutate(SampleID=names(alns99top))

ggplot(data = hitsPerSample99top, aes(V1))+
  geom_histogram(binwidth = 300) +
  labs(title="hits per sample >= 99% id (top hit only)", x="hits") +
  theme_classic()
summary(hitsPerSample99top$V1)
```
Note - many samples have few reads (<300) that map to G. vaginalis after filtering.

## Clade
```{r}
duplicateReferences99top <- alns99top %>%  # dataframes with multiples only
  map(~group_by(.x, query)) %>%
  map(~filter(.x, n()>1)) %>%
  map(~ungroup(.x))

diffClades99top_100 <- duplicateReferences99top %>%
  map(~filter(.x, pcntid==100)) %>%
  map(~select(.x, query, SampleID, Annotation, Clade)) %>%
  purrr::reduce(full_join, by=c("query", "SampleID", "Annotation", "Clade")) %>%
  group_by(query) %>%
  filter(n_distinct(Clade)>1) %>%
  ungroup()
ambiguous99top_100 <-table(diffClades99top_100$Annotation, diffClades99top_100$Clade)
print("ambiguous annotations with 100% id")
ambiguous99top_100
# ambiguous99top_100 %>%
#   unclass() %>%
#   as.data.frame() %>%
#   write.csv(file.path(gdrive_path, "cladeAmbiguousHits99top_100.csv"))

diffClades99top_99 <- duplicateReferences99top %>%
  map(~filter(.x, pcntid!=100)) %>%
  map(~select(.x, query, SampleID, Annotation, Clade)) %>%
  purrr::reduce(full_join, by=c("query", "SampleID", "Annotation", "Clade")) %>%
  group_by(query) %>%
  filter(n_distinct(Clade)>1) %>%
  ungroup()
ambiguous99top_99 <-table(diffClades99top_99$Annotation, diffClades99top_99$Clade)
print("ambiguous annotations with ~99% id")
ambiguous99top_99
# ambiguous99top_99 %>%
#   unclass() %>%
#   as.data.frame() %>%
#   write.csv(file.path(gdrive_path, "cladeAmbiguousHits99top_99.csv"))

alns099_top <- alns99top %>%
  map(~select(.x, query, pcntid, SampleID, SubjectID, GWcoll, RNAlater, PaperCohort, Annotation, Clade)) %>%
  map(~unique.data.frame(.x)) %>%
  purrr::reduce(full_join, by= c("query", "pcntid", "SampleID", "SubjectID", "GWcoll", "RNAlater", "PaperCohort", "Annotation", "Clade"))

ggplot(data=alns099_top) +
  geom_bar(aes(x=Annotation, color=as.character(Clade), fill=as.character(Clade))) +
  facet_wrap(~SampleID) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=2)) + 
  labs(title = "Hits by Annotation (top hit above 99% ID)", color="Clade", fill="Clade")
#  ggsave(file.path(gdrive_path, "figures", "cladeAnnotationBarPlot99top.pdf"), width=30, height=20)
```
## Genomopecies
```{r}
diffSpecies99top_100 <- duplicateReferences99top %>%
  map(~filter(.x, pcntid==100)) %>%
  map(~select(.x, query, SampleID, Annotation, SpeciesNumber)) %>%
  purrr::reduce(full_join, by=c("query", "SampleID", "Annotation", "SpeciesNumber")) %>%
  group_by(query) %>%
  filter(n_distinct(SpeciesNumber)>1) %>%
  ungroup()
ambiguousSpecies99top_100 <-table(diffSpecies99top_100$Annotation, diffSpecies99top_100$SpeciesNumber)
print("ambiguous annotations with 100% id")
ambiguousSpecies99top_100
 ambiguousSpecies99top_100 %>%
   unclass() %>%
   as.data.frame() %>%
   write.csv(file.path(gdrive_path, "genomospeciesAmbiguousSpeciesHits99top_100.csv"))

diffSpecies99top_99 <- duplicateReferences99top %>%
  map(~filter(.x, pcntid!=100)) %>%
  map(~select(.x, query, SampleID, Annotation, SpeciesNumber)) %>%
  purrr::reduce(full_join, by=c("query", "SampleID", "Annotation", "SpeciesNumber")) %>%
  group_by(query) %>%
  filter(n_distinct(SpeciesNumber)>1) %>%
  ungroup()
ambiguousSpecies99top_99 <-table(diffSpecies99top_99$Annotation, diffSpecies99top_99$SpeciesNumber)
print("ambiguous annotations with ~99% id")
ambiguousSpecies99top_99
 ambiguousSpecies99top_99 %>%
   unclass() %>%
   as.data.frame() %>%
   write.csv(file.path(gdrive_path, "genomospeciesAmbiguousSpeciesHits99top_99.csv"))

alns099_top_S <- alns99top %>%
  map(~select(.x, query, pcntid, SampleID, SubjectID, GWcoll, RNAlater, PaperCohort, Annotation, SpeciesNumber)) %>%
  map(~unique.data.frame(.x)) %>%
  purrr::reduce(full_join, by= c("query", "pcntid", "SampleID", "SubjectID", "GWcoll", "RNAlater", "PaperCohort", "Annotation", "SpeciesNumber"))

ggplot(data=alns099_top_S) +
  geom_bar(aes(x=Annotation, color=as.character(SpeciesNumber), fill=as.character(SpeciesNumber))) +
  facet_wrap(~SampleID) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=2)) + 
  labs(title = "Hits by Annotation (top hit above 99% ID)", color="SpeciesNumber", fill="SpeciesNumber")
 ggsave(file.path(gdrive_path, "figures", "genomospeciesAnnotationBarPlot99top.pdf"), width=30, height=20)

duplicateSpeciesAnnos <- c(diffSpecies99top_99$Annotation, diffSpecies99top_100$Annotation) %>%
  unique() 
length(duplicateSpeciesAnnos)

alns099_top_Species <- alns99top %>%
  map(~select(.x, query, pcntid, SampleID, SubjectID, GWcoll, RNAlater, PaperCohort, Annotation, Clade, SpeciesNumber, SpeciesName)) %>%
  map(~unique.data.frame(.x)) %>%
  purrr::reduce(full_join, by= c("query", "pcntid", "SampleID", "SubjectID", "GWcoll", "RNAlater", "PaperCohort", "Annotation", "Clade", "SpeciesNumber", "SpeciesName"))
```
There are 43 ambiguous genomospecies orthologues and 17 ambiguous clade orthologues

# Write and save dataframes with alignment and relative abundance information 
## Clade
```{r}
duplicateCladeAnnos <- c("30S ribosomal protein S13", 
                         "30S ribosomal protein S15",
                         "50S ribosomal protein L31",
                         "50S ribosomal protein L11",
                         "Aspartyl/glutamyl-tRNA(Asn/Gln) amidotransferase subunit B",
                         "DNA-directed RNA polymerase subunit beta",
                         "Elongation factor G",
                         "CTP synthase",
                         "Glutamyl-tRNA(Gln) amidotransferase subunit A",
                         "GTPase Der",
                         "Ribonucleoside-diphosphate reductase subunit alpha 1",
                         "Glutamyl-tRNA(Gln) amidotransferase subunit C",
                         "Phosphoglucomutase",
                         "Transketolase",
                         "Pyruvate formate-lyase 1-activating enzyme" ,
                         "Ribonucleoside-diphosphate reductase subunit beta nrdF2",
                         "Valine--tRNA ligase")

alignmentsFinal <- alns099_top %>%
  filter(Annotation %!in% duplicateCladeAnnos)
#write_delim(alignmentsFinal, "./alingmentsFinal_ref1strain1genomospecies.tsv", delim="\t")

gardRelativeAbundance <-  alignmentsFinal %>%
  group_by(SampleID) %>%
  count(Clade) %>%
  mutate(propClade=(n/sum(n)),
              Clade=as.character(Clade))
#write_delim(gardRelativeAbundance, "./gardRelativeAbundance_ref1strain1genomospecies.tsv", delim="\t")
```

## Genomospecies
```{r}
speciesAlignmentsFinal <- alns099_top_Species %>%
  filter(Annotation %!in% duplicateSpeciesAnnos)
#write_delim(speciesAlignmentsFinal, "./genomospeciesAlingmentsFinal_ref1strain1genomospecies.tsv", delim="\t")

gardSpeciesRelativeAbundance <-  speciesAlignmentsFinal %>%
  group_by(SampleID) %>%
  count(SpeciesNumber) %>%
  mutate(propGenomospecies=(n/sum(n)),
              Genomospecies=as.character(SpeciesNumber))
#write_delim(gardSpeciesRelativeAbundance, "./gardGenomospeciesRelativeAbundance_ref1strain1genomospecies.tsv", delim="\t")
```

# Compare reads: clades vs. genomospecies
```{r}
duplicateSpeciesOnly <- setdiff(duplicateSpeciesAnnos, duplicateCladeAnnos)

alignmentsFinal %>%
  group_by(Annotation, SampleID) %>%
  summarise(matches=n_distinct(query)) %>%
  mutate(Resolution=case_when(Annotation %in% duplicateSpeciesOnly~"Resolves to clade only",
                              Annotation %!in% duplicateSpeciesOnly~"Resolves to clade and species")) %>%
  ggplot(aes(x=Annotation, y=matches, color=Resolution)) +
  geom_violin() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust=1, size=5)) +
  labs(x="", y="n reads/sample")

alignmentsFinal %>%
  mutate(Resolution=case_when(Annotation %in% duplicateSpeciesOnly~"Resolves to clade only",
                              Annotation %!in% duplicateSpeciesOnly~"Resolves to clade and species")) %>%
  ggplot(aes(x=Annotation, fill=Resolution)) +
  geom_histogram(stat = "count") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust=1, size=5)) +
  labs(x="", y="n reads")
```

```{r}
alignmentsFinal %>%
  filter(SampleID %in% c("1062101108", "1062101238", "2050501168")) %>%
  mutate(Resolution=case_when(Annotation %in% duplicateSpeciesOnly~"Resolves to clade only",
                              Annotation %!in% duplicateSpeciesOnly~"Resolves to clade and species"), 
         Resolution=factor(Resolution, levels = c("Resolves to clade only", "Resolves to clade and species"))) %>%
  
  ggplot(aes(x=Annotation, fill=Clade, alpha=Resolution)) +
  geom_histogram(stat = "count") +
  theme_bw() +
  scale_fill_manual(values=c("salmon", "indianred",  "blue3")) +
  scale_alpha_manual(values=c(0.4, 1)) +
  facet_wrap(~SampleID, nrow=1) +
  theme(axis.text.x = element_text(angle = 90, hjust=1, size=5)) +
  labs(x="", y="n reads")
```

## Hits and Reference Gene lengths
```{r}
# DNA string set of reference genes
 refGenes <- "./ref_genes/refGenes.fasta" %>%
  readDNAStringSet()
# table of gene IDs, withds, clades, and genomospecies
refGenesLengths <- refGenes %>%
  as.data.frame() %>%
  rownames_to_column("reference") %>%
  mutate(seqwidth=nchar(x)) %>%
  select(-x) %>%
  left_join(refAnnotations, by="reference") %>%
  dplyr::rename(Strain=strain) %>%
  left_join(refDF[,c("Strain", "Clade")], by="Strain") %>%
  left_join(referenceGenomospecies, by="Strain")
```

  Matches by gene length
```{r}
seqWidths <- refGenesLengths %>%
  group_by(Annotation) %>%
  summarise(`Mean Gene Length`=mean(seqwidth))

alignmentsFinal %>%
  group_by(Annotation) %>%
  summarise(Matches=n_distinct(query)) %>%
  mutate(Resolution=case_when(Annotation %in% duplicateSpeciesOnly~"Resolves to clade only",
                                       Annotation %!in% duplicateSpeciesOnly~"Resolves to clade and species")) %>%
  left_join(seqWidths) %>%
  ggplot(aes(y=Matches, x=`Mean Gene Length`, color=Resolution)) +
  geom_point() +
  theme_bw()
```

# Session Info
```{r}
sessionInfo()
```