---
title: "Gardnerella mapping vs. MetaPhlAn Results"
author: "Hanna L. Berman"
output: 
  html_document:
    toc: true
    toc_float: true
---
  
  Compare our custom core genome mapping method to MetaPhlAn4 results. Determine whether the number of *Gardnerella* core genes in our database found in each sample correlates to the number of *Gardnerella* core genes found by MetaPhlAn4 in the MetaPhlAn4 database.
  
# Set up
## Packages
```{r, warning=FALSE}
library(tidyverse)
```

```{r, warning=FALSE, message=FALSE}
library(formattable)
library(ggpubr)
library(cowplot)
theme_set(theme_bw())
```

## File Paths
```{r, message=FALSE, warning=FALSE}
figureOut <- "../../FIGURES/resubmission_manuscript_figures"
```


## Abundance
### Stanford UAB Data
```{r, message=FALSE, warning=FALSE}
#Metaphlan Aabundance:
suMetaphlanPath <- "../metaphlan4/stuab_mergedMetaphlanAbundance.tsv"
sumDF <- suMetaphlanPath %>%
  read_tsv(comment = "#") %>%
  filter(str_detect(clade_name, "s__"), !str_detect(clade_name, "t__")) %>%
  mutate(Species = str_extract(clade_name, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -clade_name) %>%
  dplyr::rename_at(2:ncol(.), ~str_extract(.x, "[0-9]{10}")) %>%
  column_to_rownames(var="Species") %>%
  t %>%
  as.data.frame() %>%
  mutate_all(funs(./100)) %>%
  rownames_to_column(var="SampleID")
sumDF[1:6,1:5]

sumetaphlanGardAbundance <- sumDF %>%
  select(SampleID, Gardnerella_vaginalis) %>%
  dplyr::rename(metaphlanGardAbundance=Gardnerella_vaginalis)

# Mapping Abundance
sumapCladeAbundance <- "../gardCladeRelativeAbundance.tsv" %>%
  read_tsv()
head(sumapCladeAbundance)
```

### HMP2 data
```{r, message=FALSE, warning=FALSE}
#Metaphlan Aabundance:
hmp2MetaphlanPath <- "../metaphlan4/hmp2mgs_mergedMetaphlanAbundance.tsv"
hmp2mDF <- hmp2MetaphlanPath %>%
 read_tsv(comment = "#") %>%
  filter(str_detect(clade_name, "s__"), !str_detect(clade_name, "t__")) %>%
  mutate(Species = str_extract(clade_name, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -clade_name) %>%
  as.data.frame() %>%
  dplyr::rename_at(2:ncol(.), ~str_extract(.x, "[0-9]{7}")) %>%
  column_to_rownames(var="Species") %>%
  t %>%
  as.data.frame() %>%
  mutate_all(funs(./100)) %>%
  rownames_to_column(var="SampleID")
hmp2mDF[1:6,1:5]

hmp2metaphlanGardAbundance <- hmp2mDF %>%
  select(SampleID, Gardnerella_vaginalis) %>%
  dplyr::rename(metaphlanGardAbundance=Gardnerella_vaginalis)

# Mapping Abundance
hmp2mapCladeAbundance <- "../hmp2gardCladeRelativeAbundance.tsv" %>%
  read_tsv()
head(hmp2mapCladeAbundance)
```

### Join
```{r, message=FALSE, warning=FALSE}
mapCladeAbundance <- full_join(sumapCladeAbundance, hmp2mapCladeAbundance, by = c("SampleID", "Clade", "n", "propClade"))

metaphlanGardAbundance <- full_join(sumetaphlanGardAbundance, hmp2metaphlanGardAbundance, by = c("SampleID", "metaphlanGardAbundance"))
```


### Alignments
  Mapping
```{r, message=FALSE, warning=FALSE}
suMappingAlignments <- read_tsv("cladeAlignmentsFinal.tsv")

hmp2MappingAlignments <- read_tsv("./hmp2cladeAlignmentsFinal.tsv")

mapCladeCounts <- suMappingAlignments %>%
  full_join(hmp2MappingAlignments) %>%
  group_by(SampleID) %>%
  count(Clade)
head(mapCladeCounts)

mapGardCounts <- mapCladeCounts %>%
  group_by(SampleID) %>%
  summarise(mapReads=sum(n)) %>%
  mutate(SampleID=as.character(SampleID))
head(mapGardCounts)  
```

  Metaphlan
```{r, message=FALSE}
hmp2metaphlanGardCounts <- "../metaphlan4/hmp2mgs_gardMarkerCounts.csv" %>%
  read_csv %>%
  mutate(SampleID=str_extract(SampleID, "(?<=SRR)[0-9]{7}"))


sumetaphlanGardCounts <- "../metaphlan4/stuab_gardMarkerCounts.csv" %>%
  read_csv %>%
  mutate(SampleID=as.character(SampleID))

metaphlanGardCounts <- full_join(sumetaphlanGardCounts, hmp2metaphlanGardCounts) %>%
  dplyr::rename(metaphlanGardReads=gardReads)
nrow(metaphlanGardCounts)
```

```{r, message=FALSE, warning=FALSE}
compareDF <- metaphlanGardCounts %>%
  full_join(mapGardCounts, by="SampleID") %>%
  full_join(metaphlanGardAbundance, by="SampleID") %>%
  replace_na(list(metaphlanGardReads=0, metaphlanGardAbundance=0, mapReads=0))
head(compareDF)

anyNA(compareDF$metaphlanGardAbundance)
```

# Compare alignments
 Reads vs. Reads
```{r, message=FALSE, warning=FALSE}
compareDF %>%
  ggplot(aes(x=mapReads, y=metaphlanGardReads)) +
  geom_abline(color="grey", linetype=2) +
  geom_point(alpha=0.75, size=2) +
  geom_smooth(method="lm", se=FALSE, color="blue", linetype=1) +
  stat_cor(method = "pearson", alternative = "greater", label.sep = "\n") + 
  xlim(0,1000000) +
  coord_fixed() +
  labs(x=bquote('Aligned'~italic("Gardnerella") ~'Marker Reads - Our Method'), y=bquote('Aligned'~italic("Gardnerella") ~'Marker Reads - MetaPhlAn4'))
```

  
```{r, message=FALSE, warning=FALSE}
readsVreadsPlot <- compareDF %>%
  ggplot(aes(x=mapReads/1000, y=metaphlanGardReads/1000)) +
  geom_abline(color="grey", linetype=2) +
  geom_point(alpha=0.5, size=2) +
  geom_smooth(method="lm", se=FALSE, color="blue", linetype=1) +
  stat_cor(method = "pearson", alternative = "greater") + 
  labs(x="Our Method (Thousands)", y="MetaPhlAn4 (Thousands)")
readsVreadsPlot
```
  
  Reads vs Gardnerella abundance
```{r, message=FALSE, warning=FALSE}
abVreadsPlot <- compareDF %>%
  ggplot(aes(x=mapReads, y=metaphlanGardAbundance)) +
  geom_point(alpha=0.5, size=2) +
  geom_smooth(method="lm", se=FALSE, color="blue", linetype=1) +
  stat_cor(method = "pearson", alternative = "greater", label.sep = "\n", label.x = 20000, label.y = 0.125)  +
  ylim(0,1) +
  labs(x=bquote('Aligned'~italic("Gardnerella") ~'Marker Reads - Our Method'), y=bquote(italic("Gardnerella") ~'Relative Abundance - MetaPhlAn4'))
abVreadsPlot
```

## Figure 2D: Gardnerella mapping vs MetaPhlAn
```{r, message=FALSE, warning=FALSE} 
#, fig.width=1.5, fig.height=1.2
plot_grid(readsVreadsPlot, nrow=1, labels = c("D"), label_size = 15)
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "Figure2D_mappingVsMetaphlan.png", sep="_")))
```

# Session info
```{r}
sessionInfo()
```