---
title: "Build Bowtie DB"
author: "Hanna Berman"
date: "8/13/2020"
output: html_document
---

# Setup
##Load packages
```{r}
library(tidyverse)
library(Biostrings)
set.seed(999)
```

## Define file paths
```{r}
repo_dir <- "../../"
gard_tree_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/Gard WGS"
roary_path <- file.path(gard_tree_dir, "20181008_GvRoary95strains", "gene_presence_absence.csv")
ref_ffns <- file.path(gard_tree_dir, "annotated_genomes_byext", "ffn")
ref_tsvs <- file.path(gard_tree_dir, "annotated_genomes_byext", "tsv")
output_dir <- "./ref_genes_1seq"
df_path <- file.path(repo_dir, "shotgunMetadata.tsv")

gardDF <- file.path(repo_dir, "GardnerellaMetadata.csv") %>%
  read_csv
  
gardGenomoDF <- file.path(repo_dir, "72GardStrainGenomospecies.csv") %>%
  read_csv()
```

# Get Reference Genes
```{r}
strains <- str_extract(list.files(ref_ffns, full.names = FALSE, all.files = FALSE, pattern = ".ffn"), 
                       ".*(?=\\.ffn)")

#roary gene presence absence file and select core genes that are definitely unique
gnpa <- read_csv(roary_path)
refGenes <- gnpa %>%
  add_count(Annotation) %>%
  filter(`No. isolates`==72,
         n==1) %>%
  select(-n)

refSeqNames <- refGenes %>% #make a vector of all the names of reference genes for which we want to extract sequences
  select(strains) %>%
  c() %>%
  unlist
names(refSeqNames) <- NULL

#fasta seqs
allSeqs <- DNAStringSetList(lapply(list.files(ref_ffns, full.names = TRUE, all.files = FALSE, pattern = ".ffn"), 
            readDNAStringSet))
names(allSeqs) <- strains
allSeqs <- unlist(allSeqs) #make one DNA string set
allSeqs@ranges@NAMES <- str_extract(allSeqs@ranges@NAMES, "(?<=\\.).*_[0-9]*(?=\\ .*)") #remove extra strain name label and annotation from each DNA string name


cladeRefSeqs <- allSeqs[allSeqs@ranges@NAMES %in% refSeqNames]
length(cladeRefSeqs) == nrow(refGenes)*length(strains) #check that there are the appropriate amount of reference sequences
```

# Choose only one strain per clade or one strain per 
## One strain per genomospecies
```{r}
genomoStrainsDF <- gardDF %>%
  filter(In_Directory==TRUE) %>%
  select(Strain, Clade) %>%
  left_join(gardGenomoDF[,c("Strain", "SpeciesNumber")], by="Strain") %>%
  group_by(SpeciesNumber) %>%
  sample_n(1) %>%
  ungroup()

genomoStrainsDF$Strain

genomoStrains <- genomoStrainsDF %>%
  .$Strain %>%
  paste(collapse = "|")
genomoStrains

refSeqs1StrainGenomospecies <- cladeRefSeqs[str_detect(cladeRefSeqs@ranges@NAMES, genomoStrains)]
refSeqs1StrainGenomospecies
#test number of genes
length(refSeqs1StrainGenomospecies)==nrow(refGenes)*nrow(genomoStrainsDF)
```

## One strain per clade
```{r}
cladeStrainsDF <- genomoStrainsDF %>%
  group_by(Clade) %>%
  sample_n(1) %>%
  ungroup()

cladeStrainsDF$Strain

cladeStrains <- cladeStrainsDF %>%
  .$Strain %>%
  paste0(collapse = "|")
cladeStrains

refSeqs1StrainClade <- cladeRefSeqs[str_detect(cladeRefSeqs@ranges@NAMES, cladeStrains)]
refSeqs1StrainClade
#test number of genes
length(refSeqs1StrainClade)==nrow(refGenes)*nrow(cladeStrainsDF)
```
# Save sequences
```{r}
#writeXStringSet(refSeqs1StrainGenomospecies, file.path(output_dir, "refSeqs1StrainGenomospecies.fasta"), format="fasta")
#writeXStringSet(refSeqs1StrainClade, file.path(output_dir, "refSeqs1StrainClade.fasta"), format = "fasta")
```

# Session Info
```{r}
sessionInfo()
```