---
title: "Measure Gardnerella clades and species in shotgun data"
subtitle: "MOMS-PI Cohort"
author: "Hanna L. Berman"
output: html_document
---

# Setup
##Load packages
```{r libraries, warning=FALSE}
library(tidyverse)
library(Biostrings); packageVersion("Biostrings")
options(scipen=10000)
`%!in%` <- negate(`%in%`)
```

## Define file paths
### Reference genes and usearch output
```{r}
# Directories
repoDir <- ".."
gardTreeDir <- file.path(repoDir, "gard_phylogeny")
dataOut <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/cladeAssignments"
figureOut <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/FIGURES/clade_assignments"

# Usearch alignments 
usearchAlns <- "/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/gard_map/hmp2mgs_usearch_alns" #path to blast6 tables from usearch alignments

# reference genes annotations
refGenesDF <- "./ref_genes/refGenesDF.csv" %>%
  read_csv(); head(refGenesDF)

#reference assembly clades and genomospecies
refGardDF <- file.path(repoDir, "refGardnerellaCladesGenomos.csv") %>%
  read_csv(); head(refGardDF)

# hmp2 metadata
sgDF <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/pregnancy_metagenome/hmp2mgsDF.tsv" %>%
  read_tsv %>%
  mutate(SampleID=str_extract(SampleID, "[0-9]{7}"),
         SampleID=as.character(SampleID), 
         SubjectID=as.character(SubjectID)); head(sgDF)

#uncharacterized samples
#unchGardSamples <- read_lines("/Users/hlberman/Desktop/unchGardSamples.txt")

#metaphlan
#load("./metaphlanAbundances.Rdata")
```

## ggplot
```{r}
theme_set(theme_bw()+
          theme(panel.grid = element_blank()))
cladeColors <- list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))),
                     scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))))

binaryColors <- list(scale_color_manual(values=c("#0072B2", "#D55E00")),
                     scale_fill_manual(values=c("#0072B2", "#D55E00")))

okabeIto <- c("#E69F00", "#F0E442", "#0072B2", "#009E73", "#56B4E9", "#D55E00", "#CC79A7", "#000000")
```

# Asses alignments by usearch
## Load alignment data
```{r, message=FALSE}
alns <- list.files(usearchAlns, all.files = FALSE, full.names = TRUE) %>%
   map(read_delim, delim="\t", col_names=c("query", "reference", "pcntid", "length", "mismatch", "gap_open", "qstart", "qend", "rstart", "rend", "evalue", "bitscore")) #read in blast tables
names(alns) <-  list.files(usearchAlns, all.files = FALSE, full.names = FALSE) %>%
   str_extract("[0-9]+(?=\\.txt)") # add sample names to alns object
save(alns, file="./hmp2mgsAlns.Rdata")
#load("./hmp2mgsAlns.Rdata")
```

## Format
```{r}
#remove strains where no gard mapped
nrows <- map(alns, nrow) 
alns <- alns[nrows>0]

#number of samples where Gardnerella reads were found
length(alns)
```
```{r}
alns0 <- alns %>%
  map2(., names(alns), ~mutate(.x, SampleID=.y)) %>%
  map(~left_join(.x, refGenesDF, by="reference")) %>%
  map(~left_join(.x, sgDF[,c("SampleID", "SubjectID", "reported_ga")], by="SampleID")) #add sample metadata
```

## Some descriptives
```{r}
HitsPerSample <- alns %>% #count number of hits coming back in each sample 
  map(~nrow(.x)) %>%
  as_tibble() %>%
  t() %>%
  as_tibble()

ggplot(data = HitsPerSample, aes(V1)) +
  geom_histogram() +
  #scale_x_continuous(breaks = seq(0, 900000, by = 100000)) +
  #scale_y_continuous(breaks = seq(0, 40, by = 5)) +
  labs(title = "'Hits' per sample", x = "'Hits'")

ReadsPerSample <- alns0 %>%
  map(~group_by(.x, SampleID)) %>%
  map(~summarise(.x, readspersample=n_distinct(query))) %>%
  map(~ungroup(.x)) %>%
  purrr::reduce(full_join, by=c("SampleID", "readspersample"))
  
QueriesPerAnnotation <- alns0 %>% #number of reads that map to each annotation
  map(~group_by(.x, annotation)) %>%
  map(~summarize(.x, queriesbyannotation=n_distinct(query))) %>%
  map(~ungroup(.x)) %>%
  map2(., names(alns0), ~mutate(.x, sampleID=.y)) %>%  
  purrr::reduce(full_join, by=c("annotation", "queriesbyannotation", "sampleID"))
ggplot(data=QueriesPerAnnotation, aes(x=annotation, y=queriesbyannotation)) +
  geom_violin() +
  labs(title="Reads per annotation", x="Annotation", y="Number of reads aligning") +
  theme(axis.text.x = element_text(angle=45, hjust=1, size=6))

ReferencesPerQuery <- alns0 %>%  #number of references that each query maps to
  map(~group_by(.x, query)) %>%
  map(~summarize(.x, referencesbyquery=n_distinct(reference))) %>%
  map(~ungroup(.x)) %>%
  map2(., names(alns0), ~mutate(.x, sampleID=.y)) %>%
  purrr:: reduce(full_join, by=c("query", "referencesbyquery", "sampleID"))

sumStats1 <- sapply(list(Reads_per_sample=ReadsPerSample$readspersample,
                         Hits_per_sample=HitsPerSample$V1,
                         Queries_per_annotation=QueriesPerAnnotation$queriesbyannotation,
                         References_per_query=ReferencesPerQuery$referencesbyquery),
                     summary) %>%
  as.data.frame()
sumStats1
#write_csv(sumStats1, file.path(dataOut, "stats", "usearchAlignmentSummaryStats1.csv"))
```
  
## Assess top alignmnet percent identities
```{r}
# allHits <- alns0 %>%
#   map2(., names(alns), ~mutate(.x, sampleID=.y)) %>%
#   map(~select(.x, sampleID, query, reference, gene, annotation, Strain, Clade, Genomospecies, pcntid, length)) %>%
#   purrr::reduce(full_join, by=c("sampleID", "query", "reference", "gene", "annotation", "Strain", "Clade", "Genomospecies", "pcntid", "length")) %>%
#   mutate(sampleID=as.character(sampleID))
# save(allHits, file=file.path(dataOut, "hmp2mgs_allHits.Rdata"))
load(file.path(dataOut, "hmp2mgs_allHits.Rdata"))
```

```{r}
#reference by annotation
#violin plot of each ref gene by annotation
ggplot(data = allHits, aes(pcntid)) +
  geom_histogram(binwidth = 1) +
  theme_classic() +
  labs(title = "Percent ID of all hits", x = "Percent ID") 

# alignment percent identities by gene family (annotation)
ggplot(data=allHits, aes(x=annotation, y=pcntid))+
  geom_violin() +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  #scale_y_continuous(breaks = seq(0, 100, by = 2.5)) +
  labs(y="Percent ID")

#max percent of ID of alignments per query
maxByQuery <- alns %>%  
  map(~group_by(.x, query)) %>%
  map(~summarise(.x, maxbyquery=max(pcntid))) %>%
  map(~ungroup(.x)) %>%
  map2(., names(alns), ~mutate(.x, sampleID=.y)) %>%
  purrr::reduce(full_join, by=c("query", "maxbyquery", "sampleID"))
ggplot(data = maxByQuery, aes(maxbyquery)) +
  geom_histogram(binwidth = 1) +
  labs(title = "Max percent ID per input query", x = "Percent Identity")
#ggsave(file.path(figureOut, "maxIDperQuery_hist.png"))


sumStats2 <- sapply(list(PercentID=allHits$pcntid,
            Max_by_query=maxByQuery$maxbyquery),
            summary) %>%
  as.data.frame()
sumStats2
#write_csv(sumStats2, file.path(dataOut, "stats", "usearchAlignmentSummaryStats2_maxIDs.csv"))
```
Variation in hit percent ID and maxID per reference demonstrate variation over G. vaginalis. Strong mode of maximum % ID match per annotation and per query within each sample suggest that threshold of 99% ID for classfying clades would allow for strong specificity without sacrificing sensitivity.

# Clade and Genomospecies Assignments
  Take the top hit above 99% and then remove ambiguous alignments
## Top alignment â‰¥99%
```{r}
alns99top <- alns0 %>%
   map(~filter(.x, pcntid>=99)) %>%
   map(~group_by(.x, query)) %>%
   map(~filter(.x, pcntid==max(pcntid))) %>%
   map(~ungroup(.x))

# observe remaining alignments per sample
hitsPerSample99top <- alns99top %>%
  map(~nrow(.x)) %>%
  as_tibble() %>%
  t() %>%
  as_tibble() %>%
  mutate(SampleID=names(alns99top))

ggplot(data = hitsPerSample99top, aes(V1))+
  geom_histogram() +
  labs(title="hits per sample >= 99% id (top hit only)", x="hits")
summary(hitsPerSample99top$V1)
```
## Assess duplicates
  Some gene families may not resolve reads to the clade or genomospecies level. In this case, reads will map to references from multiple clades or genomospecies. 

### Clade level
```{r}
duplicateReferences99top <- alns99top %>%  # dataframes with ambiguous best alignments only
  map(~group_by(.x, query)) %>%
  map(~filter(.x, n()>1)) %>%
  map(~ungroup(.x))

ambiguousCladeAlns_100 <-duplicateReferences99top %>%
  map(~filter(.x, pcntid==100)) %>%
  map(~select(.x, query, SampleID, annotation, Clade)) %>%
  purrr::reduce(full_join, by=c("query", "SampleID", "annotation", "Clade")) %>%
  group_by(query) %>%
  filter(n_distinct(Clade)>1) %>%
  ungroup %>%
  with(table(annotation, Clade)) %>%
  unclass %>%
  as.data.frame %>%
  rownames_to_column(var="annotation") %>%
  mutate_if(is.numeric, ~case_when(.x!=0~"X",
                                   .x==0~""))

ambiguousCladeAlns_100

ambiguousCladeAlns_99 <- duplicateReferences99top %>%
  map(~filter(.x, pcntid!=100)) %>%
  map(~select(.x, query, SampleID, annotation, Clade)) %>%
  purrr::reduce(full_join, by=c("query", "SampleID", "annotation", "Clade")) %>%
  group_by(query) %>%
  filter(n_distinct(Clade)>1) %>%
  ungroup %>%
  with(table(annotation, Clade)) %>%
  unclass %>%
  as.data.frame() %>%
  rownames_to_column(var="annotation") %>%
  mutate_if(is.numeric, ~case_when(.x!=0~"X",
                                      .x==0~""))

ambiguousCladeAlns_99

nrow(ambiguousCladeAlns_100)
nrow(ambiguousCladeAlns_99)

duplicateCladeAnnos <- c(ambiguousCladeAlns_99$annotation, ambiguousCladeAlns_100$annotation) %>%
  unique
length(duplicateCladeAnnos)
```
13/70 gene families do not resolve to the clade level

### Genomopecies level
```{r}
ambiguousSpeciesAlns_100 <- duplicateReferences99top %>%
  map(~filter(.x, pcntid==100)) %>%
  map(~select(.x, query, SampleID, annotation, Genomospecies)) %>%
  purrr::reduce(full_join, by=c("query", "SampleID", "annotation", "Genomospecies")) %>%
  group_by(query) %>%
  filter(n_distinct(Genomospecies)>1) %>%
  ungroup() %>%
  with(table(annotation, Genomospecies)) %>%
  unclass %>%
  as.data.frame() %>%
  rownames_to_column(var="annotation") %>%
  mutate_if(is.numeric, ~case_when(.x!=0~"X",
                                      .x==0~""))

ambiguousSpeciesAlns_100

ambiguousSpeciesAlns_99 <- duplicateReferences99top %>%
  map(~filter(.x, pcntid!=100)) %>%
  map(~select(.x, query, SampleID, annotation, Genomospecies)) %>%
  purrr::reduce(full_join, by=c("query", "SampleID", "annotation", "Genomospecies")) %>%
  group_by(query) %>%
  filter(n_distinct(Genomospecies)>1) %>%
  ungroup() %>%
  with(table(annotation, Genomospecies)) %>%
  unclass %>%
  as.data.frame() %>%
  rownames_to_column(var="annotation") %>%
  mutate_if(is.numeric, ~case_when(.x!=0~"X",
                                      .x==0~""))

ambiguousSpeciesAlns_99

nrow(ambiguousSpeciesAlns_100)
nrow(ambiguousSpeciesAlns_99)

duplicateSpeciesAnnos <- c(ambiguousSpeciesAlns_99$annotation, ambiguousSpeciesAlns_100$annotation) %>%
  unique() 
length(duplicateSpeciesAnnos)
```
40/70 gene families do not resolve to the genomospecies level 

# Write and save dataframes with alignment and relative abundance information 
## Clade
```{r}
alns99top_flat <- alns99top %>%
  map(~select(.x, query, pcntid, SampleID, SubjectID, reported_ga, annotation, Clade, Genomospecies)) %>%
  map(~unique.data.frame(.x)) %>%
  purrr::reduce(full_join, by= c("query", "pcntid", "SampleID", "SubjectID", "reported_ga", "annotation", "Clade", "Genomospecies"))

cladeAlignmentsFinal <- alns99top_flat %>%
  select(-Genomospecies) %>%
  unique %>%
  filter(annotation %!in% duplicateCladeAnnos)
#write_delim(cladeAlignmentsFinal, "./hmp2cladeAlignmentsFinal.tsv", delim="\t")

gardCladeRelativeAbundance <-  cladeAlignmentsFinal %>%
  group_by(SampleID) %>%
  count(Clade) %>%
  mutate(propClade=(n/sum(n)),
              Clade=as.character(Clade))

nrow(gardCladeRelativeAbundance)

gardCladeRelativeAbundance %>%
  filter(n<10) %>%
  group_by(n) %>%
  tally

gardCladeRelativeAbundance2 <- gardCladeRelativeAbundance %>%
  filter(n>=2) %>%
  mutate(propClade=n/sum(n))

n_distinct(gardCladeRelativeAbundance$SampleID)
n_distinct(gardCladeRelativeAbundance2$SampleID)

#write_delim(gardCladeRelativeAbundance2, "../hmp2gardCladeRelativeAbundance.tsv", delim="\t")
```

## Genomospecies
```{r}
speciesAlignmentsFinal <- alns99top_flat %>%
  filter(annotation %!in% duplicateSpeciesAnnos)
write_delim(speciesAlignmentsFinal, "./hmp2genomospeciesAlignmentsFinal.tsv", delim="\t")

gardSpeciesRelativeAbundance <-  speciesAlignmentsFinal %>%
  group_by(SampleID) %>%
  count(Genomospecies) %>%
  mutate(propGenomospecies=(n/sum(n)),
              Genomospecies=as.character(Genomospecies))

nrow(gardSpeciesRelativeAbundance)

gardSpeciesRelativeAbundance %>%
  filter(n<10) %>%
  group_by(n) %>%
  tally

gardSpeciesRelativeAbundance2 <- gardSpeciesRelativeAbundance %>%
  filter(n>=2) %>%
  mutate(propGenomospecies=(n/sum(n)))

n_distinct(gardSpeciesRelativeAbundance$SampleID)
n_distinct(gardSpeciesRelativeAbundance2$SampleID)

write_delim(gardSpeciesRelativeAbundance2, "../hmp2gardGenomospeciesRelativeAbundance.tsv", delim="\t")
```

# Compare reads: clades vs. genomospecies
```{r}
duplicateSpeciesOnly <- setdiff(duplicateSpeciesAnnos, duplicateCladeAnnos)

cladeAlignmentsFinal %>%
  group_by(annotation, SampleID) %>%
  summarise(matches=n_distinct(query)) %>%
  mutate(Resolution=case_when(annotation %in% duplicateSpeciesOnly~"Resolves to clade only",
                              annotation %!in% duplicateSpeciesOnly~"Resolves to clade and species")) %>%
  ggplot(aes(x=annotation, y=matches, color=Resolution)) +
  geom_violin() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust=1, size=5)) +
  labs(x="", y="n reads/sample")

cladeAlignmentsFinal %>%
  mutate(Resolution=case_when(annotation %in% duplicateSpeciesOnly~"Resolves to clade only",
                              annotation %!in% duplicateSpeciesOnly~"Resolves to clade and species")) %>%
  ggplot(aes(x=annotation, fill=Resolution)) +
  geom_histogram(stat = "count") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust=1, size=5)) +
  labs(x="", y="n reads")
```

# Session Info
```{r}
sessionInfo()
```