---
title: "Genome Distances of Reference Genes"
author: "Hanna Berman"
output: html_document
---

# Setup
```{r}
library(tidyverse)
library(Biostrings)
```

```{r}
refDF <- "./ref_genes/refGenesDF.csv" %>%
  read_csv

refGenes <- "./ref_genes/refGenes.fasta" %>%
  readDNAStringSet

gardDF <- "../../GardnerellaMetadata.csv" %>%
  read_csv %>%
  filter(In_Directory==TRUE)

gardGenomos <- "../../72GardStrainGenomospecies.csv" %>%
  read_csv()

minHashDir <- "./minhash_files"
```
# Organize genes
## Convenience objects
```{r}
gardStrains <- gardDF$Strain

#geneLists <- gardStrains %>%
#  map(~select(refDF, .x)) %>%
#  map(as_vector) %>%
#  map(unname)
```

```{r}
strainSets<- gardStrains %>%
  map(., ~refGenes[str_detect(names(refGenes), paste0(.x, "_[0-9]{5}"))])
strainSets[1:3]

# strainSets %>%
#   map2(., gardStrains, ~writeXStringSet(.x, paste0("./minhash_files/", .y, ".fasta")))
```

# Run Minhash
```{r}
gardStrainsCladesGenomos <- gardDF %>%
  select(Strain, Clade) %>%
  mutate(Clade=str_extract(Clade, "[0-9]"),
         Clade=as.numeric(Clade)) %>%
  left_join(gardGenomos[,c("Strain", "SpeciesNumber")]) %>%
  arrange(SpeciesNumber) %>%
  arrange(Clade) %>%
  mutate(order=c(1:nrow(.)))

strainCombs <- combn(gardStrainsCladesGenomos$Strain, 2) %>%
  t %>%
  as.data.frame() %>%
  as.list()
strain1 <- strainCombs$V1
names(strain1) <- c(1:length(strain1))
strain2 <- strainCombs$V2

system("mash -h")
mashDists <- map2_dfc(strain1, strain2, ~system(paste0("mash dist ./minhash_files/",
                                                       .x,
                                                       ".fasta  ./minhash_files/",
                                                       .y,
                                                       ".fasta"), intern=TRUE)) %>%
  t %>%
  as.data.frame %>%
  separate(V1,into = c("strain1", "strain2", "distance", "p", "Matching_hashes"), sep="\t") %>%
  mutate(strain1=str_extract(strain1, "(?<=\\.\\/minhash_files\\/).*(?=\\.fasta)"),
         strain2=str_extract(strain2, "(?<=\\.\\/minhash_files\\/).*(?=\\.fasta)"), 
         distance=as.numeric(distance),
         p=as.numeric(p))

head(mashDists)
```

```{r}
df1 <- gardStrainsCladesGenomos %>%
  dplyr::rename(strain1=Strain,
                strain1_Clade=Clade,
                strain1_Genomospecies=SpeciesNumber,
                strain1_Order=order)

df2 <- gardStrainsCladesGenomos %>%
  dplyr::rename(strain2=Strain,
                strain2_Clade=Clade,
                strain2_Genomospecies=SpeciesNumber, 
                strain2_Order=order)


mashDists0 <- mashDists %>%
  left_join(df1, by="strain1") %>%
  left_join(df2, by="strain2")
```

```{r, warning=FALSE}
mashDists0 %>%
  ggplot(aes(x=reorder(strain2, strain2_Order), y=reorder(strain1, -strain1_Order),  fill=distance)) +
  geom_tile() +
  theme_bw() +
  scale_x_discrete(position="top") +
  scale_y_discrete(position ="right") +
  theme(axis.text.x = element_text(angle=90, hjust=0, size=4),
        axis.text.y = element_text(size=4),
        panel.grid = element_blank(),
        legend.position = "left") +
  labs(x="", y="")
```

```{r}
mashDists0 %>%
  filter(strain1_Clade==strain2_Clade) %>%
  select(-strain2_Clade) %>%
  split(.$strain1_Clade) %>%
  map(~ggplot(.x, aes(x=strain1, y=strain2, fill=distance)) +
  geom_tile() +
  facet_wrap(~strain1_Clade, scales = "free") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust=1),
        panel.grid = element_blank())+
  labs(x="", y=""))
```

```{r}
mashDists0 %>%
  filter(strain1_Genomospecies==strain2_Genomospecies) %>%
  select(-strain2_Genomospecies) %>%
  split(.$strain1_Genomospecies) %>%
  map(~ggplot(.x, aes(x=strain1, y=strain2, fill=distance)) +
  geom_tile() +
  facet_wrap(~strain1_Genomospecies, scales = "free") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust=1),
        panel.grid = element_blank())+
  labs(x="", y=""))
```

# Session Info
```{r}
sessionInfo()
```