---
title: "Shotgun Vs. 16S Gardnerella variants"
author: "Hanna Berman"
date: "12/10/2019"
output: html_document
---
# Set up
## Packages
```{r}
library(tidyverse)
library(phyloseq)
library(Biostrings)
library(formattable)
`%!in%` <- negate(`%in%`)
```
## Paths
```{r}
repoDir <- ".."
dataDir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/"
figureOut <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/metagenome_gardnerella/FIGURES/V1_manuscript_figures/clade_assignments"

#shotgun metadata
sgDF <- file.path(repoDir, "sumgsDF.tsv") %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID),
         SubjectID=as.character(SubjectID))

#ASVs from Callahan et al., 2017 PNAS Replication and Refinement
S16_path <- file.path(dataDir, "16S_data/RepRefine_Scripts/input/processed.rda")
load(S16_path)

#Gardnerella 16S fasta sequences from reference WGS strains
rRNAs16S <- file.path(dataDir, "gardPhylogeny/prokka_annotated_genomes/ffn") %>%
  list.files(full.names = TRUE, pattern = "ffn") %>%
  map(readDNAStringSet) %>%
  do.call("c", .) %>%
  .[str_detect(names(.), "16S ribosomal RNA")]


#gardnerella metadata
refDF <- file.path(repoDir, "refGardnerellaCladesGenomos.csv") %>%
  read_csv

#Gardnerella clade abundance from cladeAssignments.Rmd
sgCladeDF <- file.path(repoDir, "gardCladeRelativeAbundance.tsv") %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

#Gardnerella genomospecies abundance from cladeAssignments.Rmd
sgGenomoDF <- file.path(repoDir, "gardGenomospeciesRelativeAbundance.tsv") %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))
```
## ggplot themes
```{r}
theme_set(theme_bw())
# variant colors for clades vs V4
ggplotCladeV4 <- list(scale_size_manual(values=c(1.5, 3)),
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "Clade C5 (shotgun)", "Clade C6 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")),
  scale_shape_manual(values = c(16, 1)))

# clade, genomospecies, and V4 variant colors
ggplotAll <- list(scale_size_manual(values=c(1.5, 2, 3)),
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen",
                              "salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "blue4", "blue", "lightblue1", "lightblue2", "lightblue3", "lightblue4",   "darkorchid2", "yellowgreen", "cadetblue1",
                              "red", "blue1", "grey48", "darkorchid2", "yellowgreen"),
                     labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "Clade C5 (shotgun)", "Clade C6 (shotgun)",
                              "Genomospecies 1, G. vaginalis (shotgun)",
                              "Genomospecies 2 (shotgun)",
                              "Genomospecies 3 (shotgun)",
                              "Genomospecies 4, G. piotti (shotgun)",
                              "Genomospecies 11 (shotgun)",
                              "Genomospecies 5, G. leopoldii (shotgun)",
                              "Genomospecies 6, G. swidsinskii (shotgun)",
                              "Genomospecies 7 (shotgun)",
                              "Genomospecies 8 (shotgun)",
                              "Genomospecies 9 (shotgun)",
                              "Genomospecies 10 (shotgun)",
                              "Genomospecies 12 (shotgun)",
                              "Genomospecies 13 (shotgun)",
                              "Genomospecies 14 (shotgun)",
                              "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indetereminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")),
  scale_shape_manual(values = c(16, 5, 1)))  

# clade and genomospecies variant colors  
ggplotShotgunOnly <- list(scale_size_manual(values=c(1.5, 2)),
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen",
                              "salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "blue4", "blue", "lightblue1", "lightblue2", "lightblue3", "lightblue4",   "darkorchid2", "yellowgreen", "cadetblue1"),
                     labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6",
                              "Genomospecies 1, G. vaginalis",
                              "Genomospecies 2",
                              "Genomospecies 3",
                              "Genomospecies 4, G. piotti",
                              "Genomospecies 11",
                              "Genomospecies 5, G. leopoldii",
                              "Genomospecies 6, G. swidsinskii",
                              "Genomospecies 7",
                              "Genomospecies 8",
                              "Genomospecies 9",
                              "Genomospecies 10",
                              "Genomospecies 12",
                              "Genomospecies 13",
                              "Genomospecies 14")),
  scale_shape_manual(values = c(16, 5)))
```

## Set up abundance table from MetaPhlAn2 results
```{r}
#metaphlan data
mDF0 <- file.path(dataDir, "/humann2/merged_tables/mergedMetaphlanAbundance.tsv") %>%
  read_tsv() %>%
  filter(str_detect(ID, "s__"), !str_detect(ID, "t__")) %>%
  mutate(Species = str_extract(ID, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -ID) %>%
  as.data.frame()

#make col and row names
samps <- colnames(mDF0[2:ncol(mDF0)]) %>%
  str_extract("[0-9]{10}")
samps
colnames(mDF0) <- c("Species", samps)
rownames(mDF0) <- mDF0$Species

#Final edits to table
mDF <- mDF0 %>%
  select(-Species) %>%
  t() %>%
  as_tibble() %>%
  mutate_all(funs(./100)) %>%
  mutate(SampleID=samps) %>% 
  select(SampleID, everything())
mDF[1:6,1:5]
```

# Organize method abundances
## Get V4 variants
  Get V4 Variants that appear in these samples and determine corresponding clades using 16S rRNA sequences from 
```{r}
# all sample IDs that would have corresponding ASV data
studyIDs <- sgDF %>%
  filter(PaperCohort=="2017",
         RNAlater=="No") %>%
  .$SampleID

# phyloseq object of data from Callahan et al., 2017 PNAS  Replication and Refinement
PS17 <- phyloseq(otu_table(st, taxa_are_rows=FALSE), sample_data(df), tax_table(tax))
  #transform_sample_counts(function(x) {x/sum(x)}) #make read counts into relative abundances

has.shotgun <- PS17@sam_data$SampleID %in% studyIDs
PS16S_170 <- PS17 %>%
  subset_samples(has.shotgun) #keep only samples of interest in OTU table

S16gard <- PS16S_170 %>%
  psmelt %>% 
  filter(Genus=="Gardnerella", 
         Abundance>0)
```
  Note: there are 9 variants among the entire sample set for 2017, but only 5 within the shotgun samples samples
  
  Are there any samples in which *Gardnerella* was not found by 16S?
```{r}
noGard16S <- setdiff(studyIDs, unique(S16gard$SampleID))
noGard16S

PS16S_170 %>%
  psmelt %>%
  filter(Genus=="Gardnerella",
         SampleID %in% noGard16S) %>%
  select(SampleID, Abundance) %>%
  unique
```
  Two from subject 10623. Was *Gardnerella* found by MetaPhlAn2?
```{r}
mDF %>%
  filter(SampleID %in% noGard16S) %>%
  select(SampleID, Gardnerella_vaginalis)
```
  *Gardnerella* was not found by MetaPhlAn2 or 16S in these two samples.

### Map variants to clades
  Determine which V4 variants are in each reference WGS isolate
```{r}
# get list of ASVs from 16S amplicon data
vars <- S16gard %>%
  select(OTU) %>%
  unique %>%
  mutate(var=paste0("G", c(1:nrow(.)))) %>%
  dplyr::rename(seq="OTU") %>%
  select(var, seq)

# Determine which clades each V4 ASV maps to in Gardnerella WGS reference sequences
varClades <- rRNAs16S %>%
  as.data.frame() %>%
  rownames_to_column("Strain") %>%
  dplyr::rename(seq=x) %>%
  mutate(Strain=str_extract(Strain, ".*(?=_[0-9]{5} 16S ribosomal RNA$)"),
         seq=str_extract(seq, paste(vars$seq, collapse = "|"))) %>%
  filter(!is.na(seq)) %>%
  left_join(refDF, by="Strain") %>%
  left_join(vars, by="seq") %>%
  select(var, Clade) %>%
  unique
  
varClades
```

**NOTE: THESE LABELS DO NOT MATCH Callahan 2017 G1/G2 ARE FLIPPED**

## Prganize sample abundances for 
  Get abundances of variants in each sample that has 16S and shotgun sequencing and format tables
```{r}
vars0 <- vars %>%
  spread(var, seq)

S16gardDF <- S16gard %>%
  mutate(Var=case_when(OTU==vars0$G1~"G2: Clades 1/2", #change G1/G2 to match Callahan 2017 
                       OTU==vars0$G2~"G1: Clades 3/4",
                       OTU==vars0$G3~"G3: indetereminant",
                       OTU==vars0$G4~"G4: Clade 5",
                       OTU==vars0$G5~"G5: Clade 6")) %>%
  group_by(SampleID) %>%
  mutate(propGard=Abundance/sum(Abundance)) %>%
  ungroup() %>%
  select(OTU, SampleID, SubjectID, Abundance, propGard, GWcoll, Var) #RNAlater

S16gardDF$Var <- factor(S16gardDF$Var, levels = c("G2: Clades 1/2", "G1: Clades 3/4", "G3: indetereminant", "G4: Clade 5", "G5: Clade 6")) #Keep order G2, G1, G3, G4, G5
```


## Organize dataframes
  Place clade, genomospecies, and V4 variant abundances into one dataframe
```{r}
S16gardDF0 <- S16gardDF %>%
  select(SampleID, Var, propGard) %>%
  mutate(Method="V4 region 16S amplicon")

sgCladeDF0 <- sgCladeDF %>%
  filter(SampleID %in% studyIDs) %>%
  dplyr::rename(Var=Clade,
                propGard=propClade) %>%
  mutate(Var=case_when(Var=="C1"~"Clade 1",
                       Var=="C2"~"Clade 2",
                       Var=="C3"~"Clade 3",
                       Var=="C4"~"Clade 4",
                       Var=="C5"~"Clade 5",
                       Var=="C6"~"Clade 6"),
         Method="Shotgun Clade")

sgGenomoDF0 <- sgGenomoDF %>%
  filter(SampleID %in% studyIDs) %>%
  dplyr::rename(Var=Genomospecies,
                propGard=propGenomospecies) %>%
  mutate(Var=as.character(Var),
          Var=case_when(Var=="GS1"~"Genomospecies 1, G. vaginalis",
                        Var=="GS2"~"Genomospecies 2",
                        Var=="GS3"~"Genomospecies 3",
                        Var=="GS4"~"Genomospecies 4, G. piotti",
                        Var=="GS5"~"Genomospecies 5, G. leopoldii",
                        Var=="GS6"~"Genomospecies 6, G. swidsinskii",
                        Var=="GS7"~"Genomospecies 7",
                        Var=="GS8"~"Genomospecies 8",
                        Var=="GS9"~"Genomospecies 9",
                        Var=="GS10"~"Genomospecies 10",
                        Var=="GS11"~"Genomospecies 11",
                        Var=="GS12"~"Genomospecies 12",
                        Var=="GS13"~"Genomospecies 13", 
                        Var=="GS14"~"Genomospecies 14"),
         Method="Shotgun Genomospecies")

gardAllVarDF <- S16gardDF0 %>%
  full_join(sgCladeDF0, by=c("SampleID", "Var", "propGard", "Method")) %>%
  full_join(sgGenomoDF0) %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "GWcoll")], by="SampleID") %>%
  mutate(Var=factor(Var, levels = c("Clade 1", "Clade 2", "Clade 3", "Clade 4", "Clade 5", "Clade 6",
                                     "Genomospecies 1, G. vaginalis", "Genomospecies 2", "Genomospecies 3", "Genomospecies 4, G. piotti", "Genomospecies 11", "Genomospecies 5, G. leopoldii", "Genomospecies 6, G. swidsinskii", "Genomospecies 7", "Genomospecies 8", "Genomospecies 9", "Genomospecies 10",  "Genomospecies 12", "Genomospecies 13", "Genomospecies 14",
                                     "G2: Clades 1/2", "G1: Clades 3/4", "G3: indetereminant", "G4: Clade 5", "G5: Clade 6")))

#gardCladeVarDF$Var <- factor(gardCladeVarDF$Var, levels=c("Clade 1", "Clade 2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "G2: Clades 1/2", "G1: Clades 3/4", "G3: indetereminant", "G4: Clade 5", "G5: Clade 6"))  
  
head(gardAllVarDF)
unique(gardAllVarDF$Var)  
```

## View assignments
  First pass assignmentby mapping reads to variants without imposing any read count thresholds. 
```{r, fig.width=5, fig.height=4}
gardAllVarDF %>%
  ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  ggplotAll +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3))) +
  xlim(0,40)
#ggsave(filename = file.path(figureOut, "firstPassAllMethods.png"))
```

  Shotgun variants only with N reads on y-axis
```{r}
gardAllVarDF %>%
  filter(str_detect(Method, "Shotgun")) %>%
  ggplot(aes(x=GWcoll, y=n, shape=Method, size=Method, color=Var)) +
  geom_point() +
  ggplotShotgunOnly +
  scale_y_log10() +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3))) +
  xlim(0,40)
#ggsave(filename = file.path(figureOut, "firstPassAllMethodsN.png"))
```
  
# Test Thresholds
  Determine how many reads must map to each variant to count the variant as present.
  View variants over ranges of thresholds:
```{r, fig.height=6, fig.width=9}
# create data frames
thresholds <- c(1:49) %>% # set the threshold range
  map(~filter(gardAllVarDF, n>.x)) %>% # filter out clades below read threshold 
  map(~group_by(.x, SampleID, Method)) %>%
  map(~mutate(.x, propGard=propGard/sum(propGard))) %>% # recalculate proportions after threshold filtering
  map(ungroup) %>%
  map2(c(1:49), ~mutate(.x, Threshold=.y)) %>% # add threshold column
  purrr::reduce(full_join, by=c(colnames(gardAllVarDF), "Threshold")) # reduce to one data frame

# add threshold information to data frame
gardAllVarDFThresholds <- gardAllVarDF %>%
  mutate(Threshold=0) %>% # set zero threshold for baseline
  rbind(thresholds) %>%
  mutate(Threshold=Threshold+1) # make threshold as >=n 

# plot - thresholds 1 to 50
gardAllVarDFThresholds %>%
  ggplot(aes(x=Threshold, y=propGard, shape=Method, size=Method, color=Var)) + 
  geom_point() +
  ggplotAll +
  ylim(0,1) +
  scale_x_continuous(breaks = seq(0,50, 5)) +
  facet_wrap(~SampleID) +
  #guides(color=guide_legend(override.aes=list(shape=c(rep(16, 6), rep(5, 5)), size=c(rep(2, 6), rep(4, 5))))) +
  labs(x="Threshold (reads ≥ n)", y="Proportion of Gardnerella")
#ggsave(file.path(figureOut, "variantsByThreshold1_50.png"))

# plot - thresholds 1 to 5
gardAllVarDFThresholds %>%
  filter(Threshold<=5) %>%
  ggplot(aes(x=Threshold, y=propGard, shape=Method, size=Method, color=Var)) + 
  geom_point() +
  ggplotAll +
  ylim(0,1) +
  scale_x_continuous(breaks = seq(0,5, 1)) +
  facet_wrap(~SampleID) +
  #guides(color=guide_legend(override.aes=list(shape=c(rep(16, 6), rep(5, 5)), size=c(rep(2, 6), rep(4, 5))))) +
  labs(x="Threshold (reads ≥ n)", y="Proportion of Gardnerella")
#ggsave(file.path(figureOut, "variantsByThreshold1_5.png"))
```
  View count of variants over ranges of thresholds
```{r, fig.height=5, fig.width=6}
#fig.height=5, fig.width=6}
V4Vars <- gardAllVarDF %>%
  filter(Method=="V4 region 16S amplicon") %>%
  select(SampleID, Var, propGard) %>%
  group_by(SampleID) %>%
  summarise(V4vars=n())

sum(V4Vars$V4vars)

varCounts <- gardAllVarDFThresholds %>%
  filter(Method!="V4 region 16S amplicon") %>%
  select(SampleID, Method, Var, propGard, Threshold) %>%
  group_by(SampleID, Method, Threshold) %>%
  summarise(nVars=n()) %>%
  spread(Method, nVars) %>%
  ungroup() %>%
  full_join(map(1:50, ~mutate(V4Vars, Threshold=.x)) %>%
            purrr::reduce(full_join, by = c("SampleID", "V4vars", "Threshold")), 
            by=c("SampleID", "Threshold")) %>%
  replace_na(list(shotgunVars=0)) %>%
  gather("Method", "nVars", c(`Shotgun Clade`, `Shotgun Genomospecies`, `V4vars`))

# Thresholds 1:50
varCounts %>%
ggplot(aes(x=Threshold, y=nVars, group=Method, color=Method, linetype=Method)) +
  geom_line() +
  facet_wrap(~SampleID, scales="free_y") +
  scale_color_manual(values=c("#56B4E9", "#009E73", "#D55E00"), labels=c("Shotgun Clade", "Shotgun Genomospecies", "V4 variants")) +
  scale_color_manual(values=c("#56B4E9", "#009E73", "#D55E00"), labels=c("Shotgun Clade", "Shotgun Genomospecies", "V4 variants")) +
  labs(x="Threshold (reads≥n)", y="N Variants", color="Method")
#ggsave(filename = file.path(figureOut, "variantCountsbyThreshold.png"))

varCounts %>%
ggplot(aes(x=Threshold, y=nVars, group=Method, color=Method, linetype=Method)) +
  geom_line() +
  facet_wrap(~SampleID, scales = "free_y") +
  xlim(1,5) +
 scale_linetype_manual(values=c(1,2,3), labels=c("Shotgun Clade", "Shotgun Genomospecies", "V4 variants")) +
  scale_color_manual(values=c("#56B4E9", "#009E73", "#D55E00"), labels=c("Shotgun Clade", "Shotgun Genomospecies", "V4 variants")) +
  labs(x="Threshold (reads≥n)", y="N Variants", color="Method")
#ggsave(filename = file.path(figureOut, "variantCountsbyThreshold1_5.png"))
```

# Sensitivity and Specificity testing 
  Determine the how well variants correspond to each other. 
```{r}
# Samples we are testing
testSamples <- gardAllVarDF %>%
  .$SampleID %>%
  unique

# Number of V4 variants in these samples
nV4 <- gardAllVarDF %>% # number of v4 variants
  filter(Method=="V4 region 16S amplicon") %>%
  select(SampleID, Var) %>%
  nrow

 # Number of Clades found in all samples at each threshold up to 5
nClades <- gardAllVarDFThresholds %>%
  filter(Method=="Shotgun Clade",
         Threshold <= 5) %>%
  select(SampleID, Var, Threshold) %>%
  group_by(Threshold) %>%
  summarise(Total=n())

#  Number of Genomospecies found in all samples at each threshold up to 5
nGenomos <- gardAllVarDFThresholds %>%
  filter(Method=="Shotgun Genomospecies",
         Threshold <= 5) %>%
  select(SampleID, Var, Threshold) %>%
  group_by(Threshold) %>%
  summarise(Total=n())
```

## Sensitivity
  Sensitivity as percent of V4 variants or clades that are not represented by a clade or genomospecies
```{r}
# define function to determine number of V4 variants or clades that are not represented by a clade or genomospecies
testSens <- function(testSample, comparison, thresh) {
 if(str_detect(comparison, "V4")){
   a <- gardAllVarDF %>%
      filter(SampleID == testSample,
             Method=="V4 region 16S amplicon") %>%
      dplyr::rename(V4=Method) %>%
      select(Var, V4)
  
  if("G3: indetereminant" %in% a$Var){
    a <- a %>%
      bind_rows(data.frame(Var=c("G2: Clades 1/2", "G1: Clades 3/4"), V4=c("V4 region 16S amplicon", "V4 region 16S amplicon"))) %>%
      unique %>%
      filter(Var!="G3: indetereminant")
  }
  
  if(comparison=="V4_clade"){
     b <- gardAllVarDFThresholds %>%
      filter(SampleID == testSample,
             Threshold == thresh,
             Method=="Shotgun Clade") %>%
      mutate(Var=case_when(Var=="Clade 1"~"G2: Clades 1/2", 
                          Var=="Clade 2"~"G2: Clades 1/2",
                          Var=="Clade 3"~"G1: Clades 3/4",
                          Var=="Clade 4"~"G1: Clades 3/4",
                          Var=="Clade 5"~"G4: Clade 5",
                          Var=="Clade 6"~"G5: Clade 6"),
            Shotgun=Method) %>%
      select(Var, Shotgun) %>%
      unique 
  }
  
  if(comparison=="V4_genomospecies"){
     b <- gardAllVarDFThresholds %>%
      filter(SampleID == testSample,
             Threshold == thresh,
             Method=="Shotgun Genomospecies") %>%
      mutate(Var=case_when(Var=="Genomospecies 1, G. vaginalis"~"G2: Clades 1/2", 
                          Var=="Genomospecies 2"~"G2: Clades 1/2",
                          Var=="Genomospecies 3"~"G2: Clades 1/2",
                          Var=="Genomospecies 4, G. piotti"~"G2: Clades 1/2",
                          Var=="Genomospecies 11"~"G2: Clades 1/2",
                          Var=="Genomospecies 5, G. leopoldii"~"G1: Clades 3/4",
                          Var=="Genomospecies 6, G. swidsinskii"~"G1: Clades 3/4",
                          Var=="Genomospecies 7"~"G1: Clades 3/4",
                          Var=="Genomospecies 8"~"G1: Clades 3/4",
                          Var=="Genomospecies 9"~"G1: Clades 3/4",
                          Var=="Genomospecies 10"~"G1: Clades 3/4",
                          Var=="Genomospecies 14"~"G1: Clades 3/4",
                          Var=="Genomospecies 12"~"G4: Clade 5",
                          Var=="Genomospecies 13"~"G5: Clade 6"),
            Shotgun=Method) %>%
      select(Var, Shotgun) %>%
      unique}
 
  c <- a %>%
      full_join(b, by="Var") %>%
      group_by(Shotgun) %>%
      tally %>%
      filter(is.na(Shotgun)) %>%
      .$n
  }
  
  if(comparison=="clade_genomospecies"){
    a <- gardAllVarDFThresholds %>%
      filter(SampleID == testSample,
             Threshold == thresh,
             Method =="Shotgun Clade") %>%
      dplyr::rename(Clade=Method) %>%
      select(Var, Clade)
     
    b <- gardAllVarDFThresholds %>%
      filter(SampleID == testSample,
             Threshold == thresh,
             Method=="Shotgun Genomospecies") %>%
      mutate(Var=case_when(Var=="Genomospecies 1, G. vaginalis"~"Clade 1", 
                          Var=="Genomospecies 2"~"Clade 1",
                          Var=="Genomospecies 3"~"Clade 2",
                          Var=="Genomospecies 4, G. piotti"~"Clade 2",
                          Var=="Genomospecies 11"~"Clade 2",
                          Var=="Genomospecies 5, G. leopoldii"~"Clade 4",
                          Var=="Genomospecies 6, G. swidsinskii"~"Clade 4",
                          Var=="Genomospecies 7"~"Clade 3",
                          Var=="Genomospecies 8"~"Clade 3",
                          Var=="Genomospecies 9"~"Clade 3",
                          Var=="Genomospecies 10"~"Clade 3",
                          Var=="Genomospecies 14"~"Clade 3",
                          Var=="Genomospecies 12"~"Clade 5",
                          Var=="Genomospecies 13"~"Clade 6"),
            Genomospecies=Method) %>%
      select(Var, Genomospecies) %>%
      unique  
    
   c <- a %>%
    full_join(b, by="Var") %>%
    group_by(Genomospecies) %>%
    tally %>%
    filter(is.na(Genomospecies)) %>%
    .$n
   }
 return(c)
}
```

### V4 vs. Clade
```{r}
data.frame(testSample=rep(testSamples, 5),# set variables for testSens function
           comparison=rep("V4_clade", 135),
           thresh=rep(c(1:5), 27)) %>%
  bind_cols(z=pmap(., testSens) %>%  #run testSens to find number of missing variables for each test
              modify_if(~length(.)==0, ~0) %>%
              flatten_dbl) %>%
  group_by(thresh) %>%
  summarize(nMissing=sum(z)) %>%  # count N missing at each threshold
  dplyr::rename(Threshold=thresh) %>%  # format results table
  mutate(Total = nV4,
         `Percent Missing`=round(nMissing/Total*100, 1)) %>%
  formattable
```

### V4 vs. genomospecies
```{r}
data.frame(testSample=rep(testSamples, 5),  # set variables for testSens function
             comparison=rep("V4_genomospecies", 135),
             thresh=rep(c(1:5), 27)) %>%
  bind_cols(z=pmap(., testSens) %>%  # run testSens to find number of missing variants to each test
              modify_if(~length(.)==0, ~0) %>%
              flatten_dbl) %>%
  group_by(thresh) %>%
  summarize(nMissing=sum(z)) %>% # count N missing at each threshold
  dplyr::rename(Threshold=thresh) %>%  # format results table
  mutate(Total=nV4,  
    `Percent Missing`=round(nMissing/Total*100, 1)) %>%
  formattable
```

### Clade vs. genomospecies
```{r}
data.frame(testSample=rep(testSamples, 5), # set variables for testSens function
             comparison=rep("clade_genomospecies", 135),
             thresh=rep(c(1:5), 27)) %>%
  bind_cols(z=pmap(., testSens) %>% # run testSens to find number of missing variants to each test
              modify_if(~length(.)==0, ~0) %>%
              flatten_dbl) %>%
  group_by(thresh) %>%
  summarize(nMissing=sum(z)) %>% # count N missing at each threshold
  dplyr::rename(Threshold=thresh) %>% # format results table
  left_join(nClades, by="Threshold") %>%
  mutate(`Percent Missing`=round(nMissing/Total*100, 1)) %>%
  formattable
```

## Precision
  Precision as number of clades or genomospecies that are not supported by a V4 variant
```{r}
# define function to determine the number of clades or genomospecies that are not supported by a V4 variant
testSpec <- function(testSample, comparison, thresh) {
  a <- gardAllVarDF %>%
    filter(SampleID == testSample,
           Method=="V4 region 16S amplicon") %>%
    dplyr::rename(V4=Method) %>%
    select(Var, V4)

  if("G3: indetereminant" %in% a$Var){
    a <- a %>%
      bind_rows(data.frame(Var=c("G2: Clades 1/2", "G1: Clades 3/4"), V4=c("V4 region 16S amplicon", "V4 region 16S amplicon"))) %>%
      unique %>%
      filter(Var!="G3: indetereminant")
  }
  
  if(comparison=="V4_clade"){
     b <- gardAllVarDFThresholds %>%
      filter(SampleID == testSample,
             Threshold == thresh,
             Method=="Shotgun Clade") %>%
      mutate(Var=case_when(Var=="Clade 1"~"G2: Clades 1/2", 
                          Var=="Clade 2"~"G2: Clades 1/2",
                          Var=="Clade 3"~"G1: Clades 3/4",
                          Var=="Clade 4"~"G1: Clades 3/4",
                          Var=="Clade 5"~"G4: Clade 5",
                          Var=="Clade 6"~"G5: Clade 6"),
            Shotgun=Method) %>%
      select(Var, Shotgun) %>%
      unique 
  }
  
  if(comparison=="V4_genomospecies"){
     b <- gardAllVarDFThresholds %>%
      filter(SampleID == testSample,
             Threshold == thresh,
             Method=="Shotgun Genomospecies") %>%
      mutate(Var=case_when(Var=="Genomospecies 1, G. vaginalis"~"G2: Clades 1/2", 
                          Var=="Genomospecies 2"~"G2: Clades 1/2",
                          Var=="Genomospecies 3"~"G2: Clades 1/2",
                          Var=="Genomospecies 4, G. piotti"~"G2: Clades 1/2",
                          Var=="Genomospecies 11"~"G2: Clades 1/2",
                          Var=="Genomospecies 5, G. leopoldii"~"G1: Clades 3/4",
                          Var=="Genomospecies 6, G. swidsinskii"~"G1: Clades 3/4",
                          Var=="Genomospecies 7"~"G1: Clades 3/4",
                          Var=="Genomospecies 8"~"G1: Clades 3/4",
                          Var=="Genomospecies 9"~"G1: Clades 3/4",
                          Var=="Genomospecies 10"~"G1: Clades 3/4",
                          Var=="Genomospecies 14"~"G1: Clades 3/4",
                          Var=="Genomospecies 12"~"G4: Clade 5",
                          Var=="Genomospecies 13"~"G5: Clade 6"),
            Shotgun=Method) %>%
      select(Var, Shotgun) %>%
      unique
     }
  
   c <- a %>%
    full_join(b, by="Var") %>%
    group_by(V4) %>%
    tally %>%
    filter(is.na(V4)) %>%
    .$n
   
return(c)}
```

### V4 vs. Clade
```{r}
data.frame(testSample=rep(testSamples, 5),  # set variables for testSpec function
             comparison=rep("V4_clade", 135),
             thresh=rep(c(1:5), 27)) %>%
  bind_cols(z=pmap(., testSpec) %>%  # run testSpec to find number of missing variants to each test
              modify_if(~length(.)==0, ~0) %>%
              flatten_dbl) %>%
  group_by(thresh) %>%
  summarize(nUnsupported=sum(z)) %>%  # count N unsupported variants at each threshold
  dplyr::rename(Threshold=thresh) %>% # format results table
  left_join(nClades, by="Threshold") %>%
  mutate(`Percent Unsupported`=round(nUnsupported/Total*100, 1)) %>%
  formattable
```

### V4 vs. Genomospecies
```{r}
data.frame(testSample=rep(testSamples, 5),  # set variables for testSpec function
             comparison=rep("V4_genomospecies", 135),
             thresh=rep(c(1:5), 27)) %>%
  bind_cols(z=pmap(., testSpec) %>%  # run testSpec to find number of missing variants at each test
              modify_if(~length(.)==0, ~0) %>%
              flatten_dbl) %>%
  group_by(thresh) %>%
  summarize(nUnsupported=sum(z)) %>%  # count N unsupported variants at each threshold
  dplyr::rename(Threshold=thresh) %>% # format results table
  left_join(nGenomos, by="Threshold") %>%
  mutate(`Percent Unsupported`=round(nUnsupported/Total*100, 1)) %>%
  formattable
```
  Choose a threshold of ≥2 reads for presents. Ignore all one-read assignments

```{r}
gardAllVarDF_withThreshold <- gardAllVarDFThresholds %>%
  filter(str_detect(Method, "Shotgun")&Threshold==2|str_detect(Method, "V4")&Threshold==1) %>%
  select(-Threshold)
```

# View Variants in Samples
## All variants
```{r, fig.width=6, fig.height=3.5}
gardAllVarDF_withThreshold %>%
ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  ggplotAll +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3))) +
  xlim(0,40)
#ggsave(file.path(figureOut, "allVariants.png"), width=10, height=6)
```

## Clades vs. 16S
```{r}
gardAllVarDF_withThreshold %>%
  filter(Method!="Shotgun Genomospecies") %>%
ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  ggplotCladeV4 +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3))) +
  xlim(0,40)
#ggsave(file.path(figureOut, "16SvsClade.png"), width=8, height=6)
```

## Shotgun methods only
```{r, fig.width=4.5, fig.height=3.5}
gardAllVarDF_withThreshold %>%
  filter(Method!="V4 region 16S amplicon") %>%
  ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  ggplotShotgunOnly +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3), ncol=2)) +
  xlim(0,40)
#ggsave(filename = file.path(figureOut, "shotgunMethods.png"))
```

# Comparing methods alternate figures
## Clade vs. 16S
```{r}
# samples with variant G3
g3Samples <- gardAllVarDF_withThreshold %>%
  filter(Var=="G3: indetereminant") %>%
  .$SampleID

gardAllVarDF_withThreshold %>%
  filter(Method!="Shotgun Genomospecies",
         SampleID %!in% g3Samples) %>%
  mutate(Var=as.character(Var),
         Var=case_when(Var=="Clade 1"~"G2: Clades 1/2",
                   Var=="Clade 2"~"G2: Clades 1/2",
                   Var=="Clade 3"~"G1: Clades 3/4",
                   Var=="Clade 4"~"G1: Clades 3/4",
                   Var=="Clade 5"~"G4: Clade 5",
                   Var=="Clade 6"~"G5: Clade 6",
                   str_detect(Var, "G")~Var)) %>%
  group_by(SampleID, Method, Var) %>%
  summarise(propGard=sum(propGard)) %>%
  ungroup %>%
  spread(Method, propGard) %>%
  replace_na(list(`V4 region 16S amplicon`=0, 
                  `Shotgun Clade`=0)) %>%
  ggplot(aes(x=`V4 region 16S amplicon`, y=`Shotgun Clade`, color=Var)) +
  geom_point(alpha=0.5)
```


# Samples where we lose all Gardnerella
In subject 10608: No Gardnerella was detected. Shotgun sequencing can demonstrate less sensitivity than amplicon sequencing.  
**Subject 10608**
```{r}
#gardnerella found by metaphlan in subject 10608
gardAllVarDF %>%
  filter(SubjectID==10608) %>%
  select(SampleID, Var, propGard, Method)

samples10608 <- gardAllVarDF %>%
  filter(SubjectID==10608) %>%
  .$SampleID %>%
  unique
samples10608

#percent Gardnerella in these samples
mDF %>%
  select(SampleID, Gardnerella_vaginalis) %>%
  filter(SampleID %in% samples10608)

#read depth of these samples
sgDF %>%
  filter(SampleID %in% samples10608) %>%
  select(SampleID, TotalPairs, bbmapFiltered_pairs, finalPctHuman_po)

ggplot(data=sgDF) +
  geom_density(aes(x=TotalPairs)) +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID==samples10608[1]], color="red") +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID==samples10608[2]], color="blue")

ggplot(data=sgDF) +
  geom_density(aes(x=bbmapFiltered_pairs)) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID==samples10608[1]], color="red") +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID==samples10608[2]], color="blue")
```

**Sample 1050801318**
```{r}
#No gardnerella found by metaphlan in sample 1050801318
gardAllVarDF %>%
  filter(SubjectID=="10508") %>%
  select(SampleID, Var, propGard, Method)

#read depth of all sub samples
sgDF %>%
  filter(SampleID=="1050801318") %>%
  select(SampleID, TotalPairs, bbmapFiltered_pairs, PctHuman, finalPctHuman_po)

#Bacterial reads in all subject 10508 samples
ggplot(data=sgDF) +
  geom_density(aes(x=bbmapFiltered_pairs)) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID=="1050801318"], color="blue", show.legend = TRUE) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID=="1050801308"], color="red", show.legend = TRUE) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID=="1050835178"], color="green", show.legend = TRUE)
```

# Compare to Gardnerella found by MetaPhlAn2
## Gardnerella in MetaPhlAn2 but not this method
```{r}
sgCladeDF %>%
  filter(n>=2) %>%
  group_by(SampleID) %>%
  summarise(n=sum(n)) %>%
  full_join(mDF[,c("SampleID", "Gardnerella_vaginalis")], by="SampleID") %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
No uncharacterized *Gardnerella*!

```{r}
sgGenomoDF %>%
  filter(n>=2) %>%
  group_by(SampleID) %>%
  summarise(n=sum(n)) %>%
  full_join(mDF[,c("SampleID", "Gardnerella_vaginalis")], by="SampleID") %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
One sample with uncharacterized *Gardnerella* by genomospecies.

# Session Info
```{r}
sessionInfo()
```

