                     ---
title: "Gardnerella variants in Pregnancy and Preterm Birth"
author: "Hanna Berman"
output: html_document
---

# Set up
## Load Packages
```{r}
library(tidyverse)
library(kableExtra)
library(broom)
library(formattable)
library(reshape2)
library(vegan)
library(ggbeeswarm)
library(ggExtra)
 library(coin)
library(cowplot)
library(rstatix)
library(ggpubr)
`%!in%` <- negate(`%in%`)
set.seed(7334)
```

## Define file paths
```{r, message=FALSE, warning=FALSE}
figureOut <- "../FIGURES/V5_manuscript_figures/cohorts_and_analyses"

tmpFigureOut <- "../FIGURES/V5_manuscript_figures/tmp" # folder for storing temp files for figure formatting

suDF <- "./sumgsDF.tsv" %>% # metadata for Stanford and UAB
  read_tsv %>%
  mutate(SampleID=as.character(SampleID))

hmp2DF <- "./hmp2mgsDF.tsv" %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID),
         Cohort="MOMS-PI")

# sample data
sgDF0 <- suDF %>%
  select("SampleID", "SubjectID", "Cohort", "sampleTrimester", "GDcoll", "GWdel", "Age", "Income", "Education", "deliveryMode", "Race", "Ethnicity", "TermPre", "TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs") %>%
  mutate(Cohort=paste(Cohort, "Enriched")) %>%
  full_join(hmp2DF[,c("SampleID", "SubjectID", "Cohort", "sampleTrimester", "GDcoll", "GWdel", "Age", "Income", "Education", "deliveryMode", "Race", "Ethnicity", "TermPre", "TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs")], by = c("SampleID", "SubjectID", "Cohort", "sampleTrimester", "GDcoll", "GWdel", "Age", "Income", "Education", "deliveryMode", "Race", "Ethnicity", "TermPre", "TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs")) %>%
  mutate(Cohort=factor(Cohort, levels = c("Stanford Enriched", "UAB Enriched", "MOMS-PI")), # Stanford and UAB samples were selected for enrichment in Gardnerella, label them has Enriched.
         SampleID=as.character(SampleID), 
         SubjectID=as.character(SubjectID),
         pctHuman=((Sickle_pairs-bbmapFiltered_pairs)/Sickle_pairs)*100, # calculate percent human reads
         bacLoad=bbmapFiltered_pairs/(Sickle_pairs-bbmapFiltered_pairs)) # calculate microbial load
```

 **Taxa abundance tables**
 Stanford and UAB cohorts:
```{r, message=FALSE, warning=FALSE}
suMetaphlanPath <- "../humann2/merged_tables/mergedMetaphlanAbundance.tsv"
suGardCladePath <- "./gardCladeRelativeAbundance.tsv"
suGardGenomoPath <- "./gardGenomospeciesRelativeAbundance.tsv"
```
  MOMS-PI (HMP2) cohort
```{r, message=FALSE, warning=FALSE}
hmp2MetaphlanPath <- "../humann2/merged_tables/hmp2mgs_mergedMetaphlanAbundance.tsv"
hmp2GardCladePath <- "./hmp2gardCladeRelativeAbundance.tsv"
hmp2GardGenomoPath <- "./hmp2gardGenomospeciesRelativeAbundance.tsv"
```


## ggplot settings
  Themes:
```{r, message=FALSE, warning=FALSE}
theme_set(theme_bw()+
          theme(panel.grid = element_blank(),
                axis.text = element_text(size=10),
                axis.title = element_text(size=15),
                legend.text = element_text(size=12),
                legend.title = element_text(size=14)))
```

  Clade colors:
```{r, message=FALSE, warning=FALSE}
cladeColors <-  list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73"), labels=c("C1", "C2", "C3", "C4", "C5", "C6")),
                     scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73"), labels=c("C1", "C2", "C3", "C4", "C5", "C6")))

cladeColorsUnch <-  list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Uncharacterized'~italic(Gardnerella)))),
                     scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Uncharacterized'~italic(Gardnerella)))))

cladeColorsTotal <- list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))),
                         scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))))

cladeColorsTotalUnch <- list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Uncharacterized'~italic(Gardnerella)),  bquote('Total'~italic(Gardnerella)))),
                         scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Uncharacterized'~italic(Gardnerella)),  bquote('Total'~italic(Gardnerella)))))

genomoColorsTotal <- list(scale_color_manual(values=c("salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "lightblue1", "lightblue2", "lightblue3", "lightblue4", "cadetblue1", "blue4", "blue", "#E69F00", "#009E73", "#000000"),
                     labels=c(bquote(italic("G. vaginalis")),
                              "GS2",
                              "GS3",
                              bquote(italic("G. piotti")),
                              "GS11",
                              "GS7",
                              "GS8",
                              "GS9",
                              "GS10",
                              "GS14",
                              bquote(italic("G. leopoldii")),
                              bquote(italic("G. swidsinskii")),
                              "GS12",
                              "GS13",
                              bquote('Total'~italic(Gardnerella)))),
                     scale_fill_manual(values=c("salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "lightblue1", "lightblue2", "lightblue3", "lightblue4", "cadetblue1", "blue4", "blue", "#E69F00", "#009E73", "#000000"),
                     labels=c(bquote(italic("G. vaginalis")),
                              "GS2",
                              "GS3",
                              bquote(italic("G. piotti")),
                              "GS11",
                              "GS7",
                              "GS8",
                              "GS9",
                              "GS10",
                              "GS14",
                              bquote(italic("G. leopoldii")),
                              bquote(italic("G. swidsinskii")),
                              "GS12",
                              "GS13",
                              bquote('Total'~italic(Gardnerella)))))


genomoColorsTotalUnch <- list(scale_color_manual(values=c("salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "lightblue1", "lightblue2", "lightblue3", "lightblue4", "cadetblue1", "blue4", "blue", "#E69F00", "#009E73", "#A9A9A9", "#000000"),
                     labels=c(bquote(italic("G. vaginalis")),
                              "GS2",
                              "GS3",
                              bquote(italic("G. piotti")),
                              "GS11",
                              "GS7",
                              "GS8",
                              "GS9",
                              "GS10",
                              "GS14",
                              bquote(italic("G. leopoldii")),
                              bquote(italic("G. swidsinskii")),
                              "GS12",
                              "GS13",
                              bquote('Uncharacterized'~italic(Gardnerella)),
                              bquote('Total'~italic(Gardnerella)))),
                     scale_fill_manual(values=c("salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "lightblue1", "lightblue2", "lightblue3", "lightblue4", "cadetblue1", "blue4", "blue", "#E69F00", "#009E73", "#A9A9A9", "#000000"),
                     labels=c(bquote(italic("G. vaginalis")),
                              "GS2",
                              "GS3",
                              bquote(italic("G. piotti")),
                              "GS11",
                              "GS7",
                              "GS8",
                              "GS9",
                              "GS10",
                              "GS14",
                              bquote(italic("G. leopoldii")),
                              bquote(italic("G. swidsinskii")),
                              "GS12",
                              "GS13",
                              bquote('Uncharacterized'~italic(Gardnerella)),
                              bquote('Total'~italic(Gardnerella)))))
```

  *Gardnerella* and *Lactobacillus* Colors:
```{r, message=FALSE, warning=FALSE}
gLColors <-   list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")),
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred", "darkseagreen4"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")))
```

  All Taxa Colors:
```{r, message=FALSE, warning=FALSE}
taxColors <-   list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", 
                                                 "darkgreen", "khaki1", "darkred", "darkblue",
                                                 "mediumpurple4", "#C42104",  "darkseagreen4", "#2004C2"), 
                                        labels=c("C1", "C2", "C3", "C4", "C5", "C6",
                                                 bquote(italic("L. crispatus")), bquote(italic("L. gasseri")), bquote(italic("L. jensenii")), bquote(italic("L.iners")),
                                                 bquote(italic("A. vaginae")), bquote(italic("M. hominis")), bquote(italic("P. bivia")), bquote(italic("U. parvum"))
                                        )), 
                     scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73",
                                                 "darkgreen", "khaki1", "darkred", "darkblue",
                                                 "mediumpurple4", "#C42104",  "darkseagreen4", "#2004C2"), 
                                        labels=c("C1", "C2", "C3", "C4", "C5", "C6",
                                                 bquote(italic("L. crispatus")), bquote(italic("L. gasseri")), bquote(italic("L. jensenii")), bquote(italic("L.iners")),
                                                 bquote(italic("A. vaginae")), bquote(italic("M. hominis")), bquote(italic("P. bivia")), bquote(italic("U. parvum"))
                                        )))

taxColorsOther <-   list(scale_color_manual(values=c("#808080", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", 
                                                 "darkgreen", "khaki1", "darkred", "darkblue",
                                                 "mediumpurple4", "#C42104",  "darkseagreen4", "#2004C2"), 
                                        labels=c("Other", "C1", "C2", "C3", "C4", "C5", "C6",
                                                 bquote(italic("L. crispatus")), bquote(italic("L. gasseri")), bquote(italic("L. jensenii")), bquote(italic("L.iners")),
                                                 bquote(italic("A. vaginae")), bquote(italic("M. hominis")), bquote(italic("P. bivia")), bquote(italic("U. parvum"))
                                        )), 
                     scale_fill_manual(values=c("#808080", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73",
                                                 "darkgreen", "khaki1", "darkred", "darkblue",
                                                 "mediumpurple4", "#C42104",  "darkseagreen4", "#2004C2"), 
                                        labels=c("Other", "C1", "C2", "C3", "C4", "C5", "C6",
                                                 bquote(italic("L. crispatus")), bquote(italic("L. gasseri")), bquote(italic("L. jensenii")), bquote(italic("L.iners")),
                                                 bquote(italic("A. vaginae")), bquote(italic("M. hominis")), bquote(italic("P. bivia")), bquote(italic("U. parvum"))
                                        )))
```
  
  Bi/Tri/Quad Colors:
```{r, warning=FALSE, message=FALSE}
binaryColors <- list(scale_color_manual(values = c("#D55E00", "#56B4E9")),
                     scale_fill_manual(values=c("#D55E00", "#56B4E9")))

trinaryColors <- list(scale_color_manual(values = c("#F77754", "#018790", "#0A516D")),
                      scale_fill_manual(values = c(c("#F77754", "#018790", "#0A516D"))))

quadColors <- list(scale_color_manual(values = c("#F77754", "#018790", "#0A516D", "#2B2726")),
                      scale_fill_manual(values = c(c("#F77754", "#018790", "#0A516D", "#2B2726")))) 

quadColors2 <- list(scale_color_manual(values = c("#D55E00", "#009E73", "#CC79A7", "#56B4E9")),
                     scale_fill_manual(values=c("#D55E00","#009E73", "#CC79A7", "#56B4E9")))
```

## Other shortcuts
  Create some shortcut lists for convenience.
```{r, message=FALSE, warning=FALSE}
## Other shortcuts
clades <- c("C1", "C2", "C3", "C4", "C5", "C6")
genomos <- c("GS1", "GS2", "GS3", "GS4", "GS5", "GS6", "GS7", "GS8", "GS9", "GS10", "GS11", "GS12", "GS13", "GS14")
genomoNames <- c("G. vag.", "GS2", "GS3", "G. pio.", "G. leo.", "G. swi.", "GS7", "GS8", "GS9", "GS10", "GS11", "GS12", "GS13", "GS14")
genomosCladeOrder <- c("GS1", "GS2", "GS3", "GS4", "GS11", "GS7", "GS8", "GS9", "GS10", "GS14", "GS5", "GS6", "GS12", "GS13")
genomoNamesCladeOrder <- c("G. vag.", "GS2", "GS3", "G. pio.", "GS11", "GS7", "GS8", "GS9", "GS10", "GS14", "G. leo.", "G. swi.", "GS12", "GS13")

lactos <- c("Lactobacillus_crispatus", "Lactobacillus_gasseri", "Lactobacillus_jensenii", "Lactobacillus_iners")
anaerobes <- c("Atopobium_vaginae", "Finegoldia_magna",  "Mycoplasma_hominis", "Megasphaera_genomosp_type_1", "Megasphaera_unclassified", "Prevotella_amnii", "Prevotella_bivia", "Prevotella_timonensis",  "Ureaplasma_parvum")
anaerobesSL <- c("Atopobium_vaginae", "Mycoplasma_hominis", "Prevotella_bivia", "Ureaplasma_parvum") # short list of anaerobes

abbrevs <- c("TG", clades, "Lc", "Lg", "Lj", "Li", "Av", "Fm", "Mh", "Meg1", "Megu", "Pa", "Pb", "Pt", "Up")
abbrevsSL <- c("TG", clades, "Lc", "Lg", "Lj", "Li",  "Av", "Mh", "Pb", "Up")  # short list of abbreviations

cohorts <- c("Stanford Enriched", "UAB Enriched", "MOMS-PI Enriched", "MOMS-PI")
cohortAbbrevs <- c("Stanford Enr.", "UAB Enr.", "MOMS-PI Enr.", "MOMS-PI")
```

# Remove low-read samples
  Starting number of samples and subjects:
```{r, warning=FALSE, message=FALSE}
sgDF0 %>%
  group_by(Cohort) %>%
  summarize(`N Subjects`=n_distinct(SubjectID), `N Samples`=n_distinct(SampleID)) %>%
  formattable()
```

## Library sizes
  Paired reads in total library size, after quality filtering by Sickle, and after removing human reads.
```{r, message=FALSE, warning=FALSE}
sgDF0 %>%
  gather("Step", "paired_reads", c(TotalPairs, Sickle_pairs, bbmapFiltered_pairs)) %>%
  mutate(Step=factor(Step, levels=c("TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs"), labels=c("Library Size", "Sickle Filtered", "Microbial Reads"))) %>%
  group_by(Step, Cohort) %>%
  summarise(min=min(paired_reads), mean=mean(paired_reads), max=max(paired_reads)) %>%
  mutate_if(is.numeric, ~prettyNum(.x, big.mark = ",")) %>%
  formattable()
```
  Percent filtered at each step
```{r, message=FALSE, warning=FALSE}
sgDF0 %>%
  mutate(`Sickle % Total`=Sickle_pairs/TotalPairs,
         `BBmap % Sickle`=bbmapFiltered_pairs/Sickle_pairs,
         `BBmap % Total`=bbmapFiltered_pairs/TotalPairs) %>%
  group_by(Cohort) %>%
  summarise_at(c("Sickle % Total",  "BBmap % Sickle", "BBmap % Total"), ~median(.x, na.rm=TRUE)) %>%
  formattable()
```
  
  sickle_pct_total = percent of paired reads left after sickle filtering
  bbmap_pct_sicke = percent of sickle paired reads left after removing human reads
  bbmap_pct_total = percent of all paired reads left after sickle filtering and removing human reads

  View number of paired reads after sickle and human filtering per sample and determine filtering threshold.
```{r, message=FALSE,warning=FALSE}
p <- sgDF0 %>%
  ggplot(aes(x=Sickle_pairs, y=bbmapFiltered_pairs, fill=Cohort, color=Cohort)) +
  geom_point(alpha=0.25) +
  trinaryColors +
  geom_vline(xintercept = 1e6, color="grey", linetype="dashed") + # 1 million paired reads after sickle
  scale_x_log10(n.breaks=10) +
  scale_y_log10(n.breaks=10) +
  theme_bw() +
  theme(legend.position = "bottom",
        axis.text = element_text(size=10),
        axis.title = element_text(size=15),
        legend.text = element_text(size=12),
        legend.title = element_text(size=14)) +
  labs(x="QC Filtered Paired Reads", y="Microbial Paired Reads")
ggMarginal(p, groupFill = TRUE, type="density")
```
  
  Remove the samples with fewer than 1 million paired reads after Sickle filtering and trimming.
```{r, message=FALSE, warning=FALSE}
QCfilterSamples <- sgDF0 %>%
  filter(Sickle_pairs>1e6) %>% # list of samples that pass QC filtering criteria and will be used in analyses
  .$SampleID
length(QCfilterSamples) # number of samples that will be used in analyses

sgDF1 <- sgDF0 %>%
  filter(SampleID %in% QCfilterSamples)
```
  
  Number of samples remaining in each cohort:
```{r, message=FALSE, warning=FALSE}
sgDF1 %>%
  group_by(Cohort) %>%
  summarize(`N Subjects`=n_distinct(SubjectID), `N Samples`=n_distinct(SampleID)) %>%
  formattable()
```

  Will proceed with a total of 888 samples, 62 from Stanford, 45 from UAB, and 781 from MOMS-PI

  Library sizes and percent filtered at each step post QC sample removal:
```{r, message=FALSE, warning=FALSE}
readCountsPlot <- sgDF1 %>%
  gather("Step", "paired_reads", c(TotalPairs, Sickle_pairs, bbmapFiltered_pairs)) %>%
  mutate(Step=factor(Step, levels=c("TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs"), labels=c("Library Size", "Sickle Filtered", "Microbial Reads")),
         Cohort=factor(Cohort, levels = c("Stanford Enriched", "UAB Enriched", "MOMS-PI"), labels=c("Sanford\nEnriched", "UAB\nEnriched", "MOMS-PI"))) %>%
  ggplot(aes(x=Cohort, y=paired_reads, fill=Cohort, color=Cohort)) +
  geom_quasirandom(dodge.width = 0.9, alpha=0.5, size=1, show.legend = FALSE) +
  geom_boxplot(alpha=0, position=position_dodge(width=0.9), color="black", show.legend = FALSE) +
  scale_y_continuous(n.breaks = 10, minor_breaks = waiver()) +
  facet_wrap(~Step) +
  trinaryColors +
  theme_bw() +
  theme(axis.text.x = element_text(size=10),
        axis.text.y=element_text(size=13),
        strip.text = element_text(size=13),
        axis.title = element_text(size=16), 
        aspect.ratio = 1.5) +
  labs(x=NULL, y="Paired Reads")
```

## Percent Human Reads
```{r, message=FALSE, warning=FALSE}
percentHumanHists <- sgDF1 %>%
  group_by(Cohort) %>%
  ggplot(aes(pctHuman)) +
  geom_histogram(aes(y=stat(width*density))) +
  scale_y_continuous(limits = c(0,0.4), n.breaks=9) +
  facet_wrap(~Cohort) +
  theme(axis.text = element_text(size=13),
        axis.title = element_text(size=16),
        strip.text = element_text(size=13),
        aspect.ratio = 1.5) +
  labs(x="Percent Human Reads", y="Proportion of Samples")
```

  MOMS-PI samples had greatest percent of human reads overall. They were not selected for enrichment in *Gardnerella*. Samples in UAB cohort had the smallest percent of human reads.

## Summary table of library sizes and percent human reads of QC filtered samples
```{r, message=FALSE, warning=FALSE}
readsTable <- sgDF1 %>%
  group_by(Cohort) %>%
  summarise_at(vars(TotalPairs, Sickle_pairs, bbmapFiltered_pairs, pctHuman), list(mean=mean, sd=sd)) %>%
  mutate_at(vars(pctHuman_mean, pctHuman_sd), ~round(.x, 2)) %>%
  mutate_if(is.numeric, ~prettyNum(.x, big.mark = ",")) %>%
  transmute(Cohort=Cohort,
            `Library Size`=paste0(TotalPairs_mean, " (", TotalPairs_sd, ")"),
            `Sickle Filtered`=paste0(Sickle_pairs_mean, " (", Sickle_pairs_sd, ")"),
            `Microbial Reads`=paste0(bbmapFiltered_pairs_mean, " (", bbmapFiltered_pairs_sd, ")"),
            `Percent Human Reads`=paste0(pctHuman_mean, " (", pctHuman_sd, ")")) %>%
  kbl() %>%
  kable_classic(full_width=TRUE, html_font = "Arial")

#save table to tmp file for adding to rest of read count figure

readsTablePngPath <- file.path(tmpFigureOut, paste(Sys.Date(), "readsTableComponent.png", sep="_"))

readsTable %>%
  save_kable(file = readsTablePngPath, zoom=4)
```


## Figure S3: Read counts and percent human reads
```{r, message=FALSE, warning=FALSE, fig.height=2.5, fig.width=7}
readsAB <- plot_grid(readCountsPlot, percentHumanHists, nrow = 1, align = "hv", axis="tb", labels = c("A", "B"), label_size = 15)

printReadsTable <- ggdraw() + draw_image(readsTablePngPath)

plot_grid(readsAB, printReadsTable, nrow=2, rel_heights = c(1, 0.25), align = "v", axis = "l", labels = c("", "C"), label_size = 15)

#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "FigureS3_ReadCounts.png", sep="_")))
```


# Metadata
  Description of cohorts and sample collection schemes
## Figure S2: Gestational age at collection
```{r, message=FALSE, warning=FALSE, fig.height=3, fig.width=5}
samplingSched <- sgDF1 %>%
  filter(!is.na(GDcoll)) %>%
  mutate(GWcoll=GDcoll/7,
         GWcoll=floor(GWcoll),
         Cohort=factor(Cohort, levels=c("Stanford Enriched", "UAB Enriched", "MOMS-PI"))) %>%
  split(.$Cohort) %>%
  map(., ~ggplot(.x) +
    geom_point(aes(x=ceiling(GWcoll), y=SubjectID), size=1, alpha=0.5) +
    geom_point(aes(x=(GWdel+0.5), y=SubjectID), shape=")") +
    facet_grid(.~Cohort) +
    theme(axis.text = element_text(size=13),
        axis.title = element_text(size=16),
        strip.text = element_text(size=13),
        aspect.ratio = 1.5) +
    labs(x=NULL, y=NULL))

stanfordSamplingSched <- samplingSched$`Stanford Enriched` +
  labs(y="Subject ID")

uabSamplingSched <-  samplingSched$`UAB Enriched` +
  labs(x="Gestation Weeks")

momspiSamplingSched <- samplingSched$`MOMS-PI` +
  theme(axis.text.y = element_text(size=3))

plot_grid(stanfordSamplingSched, uabSamplingSched, momspiSamplingSched, rel_widths = c(1, 1, 1), rel_heights = c(1, 1, 1), align = "hv", nrow = 1) 
#ggsave(file.path(figureOut, paste(Sys.Date(), "FigureS2_samplingSchedule.png", sep="_")))
```
```{r, message=FALSE, warning=FALSE}
sgDF1 %>%
  filter(!is.na(GDcoll)) %>%
  mutate(GWcoll=GDcoll/7,
         GWcoll=floor(GWcoll)) %>%
  group_by(Cohort) %>%
  summarize(across(GWcoll, list(median=median, `25th`=~quantile(.x, 0.25), `75th`=~quantile(.x, 0.75))))
```

## Get MOMS-PI subjects with gestational age at delivery
```{r, message=FALSE, warning=FALSE}
hmp2SubjectsWDel <- sgDF1 %>%
  filter(Cohort=="MOMS-PI",
         !is.na(TermPre)) %>%
  .$SubjectID %>%
  unique

head(hmp2SubjectsWDel)
length(hmp2SubjectsWDel)
```

## Table 2: Demographic and Clinical data
```{r, message=FALSE, warning=FALSE}
#GAAD 
gaadDF <- sgDF1 %>%
  select(SubjectID, Cohort, TermPre, GWdel) %>%
  unique() %>%
  add_count(Cohort, TermPre) %>%
  group_by(Cohort, TermPre, n) %>%
  summarize(mean=round(mean(GWdel, na.rm=TRUE), 1), 
            sd=round(sd(GWdel, na.rm = TRUE), 1)) %>%
  replace_na(list(mean="-",
                  sd="-")) %>%
  ungroup() %>%
  mutate(TermPre=case_when(TermPre=="T"~"Term",
                           TermPre=="P"~"Preterm",
                           is.na(TermPre)~"Unknown"),
         cat=paste0(Cohort, "\n", TermPre, " (n=", n, ")"),
         `Mean gestational age in weeks at delivery (±SD)`=paste0(mean, " (", sd, ")")) %>%
  replace_na(list(`Mean gestational age in weeks at delivery (±SD)`="-")) %>%
  select(cat, `Mean gestational age in weeks at delivery (±SD)`) %>%
  column_to_rownames(var="cat") %>%
  t %>%
  as.data.frame %>%
  rownames_to_column(var="x") %>%
  select("x", "Stanford Enriched\nTerm (n=11)", "Stanford Enriched\nPreterm (n=9)", "UAB Enriched\nTerm (n=5)", "UAB Enriched\nPreterm (n=10)", "MOMS-PI\nTerm (n=90)", "MOMS-PI\nPreterm (n=43)", "MOMS-PI\nUnknown (n=98)")

# Demographics
demDF <- sgDF1 %>%
  select(SubjectID, Cohort, TermPre, Age, Race, Ethnicity, Education, Income, deliveryMode) %>%
  unique() %>%
  mutate(Cohort=as.character(Cohort)) %>%
  add_count(Cohort, TermPre, name="term_count") %>%
  mutate(TermPre=case_when(TermPre=="T"~"Term",
                           TermPre=="P"~"Preterm",
                           is.na(TermPre)~"Unknown"),
         cat=paste0(Cohort, "\n", TermPre, " (n=", term_count, ")")) %>%
  select(-SubjectID, -TermPre, -Cohort, -term_count) %>%
  map2_df(., data.frame(matrix(rep(.$cat, 7), ncol=7)), ~count(data.frame(x=.x, y=.y), x, y), .id="z") %>%
  filter(z!="cat") %>%
  group_by(z,y) %>%
  mutate(n=paste0(n, " (", round(n/sum(n)*100, 0), "%)")) %>%
  ungroup %>%
  spread(y, n) %>%
  mutate_all(~replace_na(.x, "0 (0%)")) %>%
  mutate(x=paste(x,z, sep="_"), #add variable category for arranging purposes because "unknown" appears multiple times and therefore causes an error when making x a factor class.
         x=factor(x, c("Below 18_Age", "18 to 28_Age", "29 to 38_Age", "Above 38_Age", "Unknown_Age", "Asian_Race", "Black_Race", "White_Race", "Other_Race", "Hispanic_Ethnicity", "Non-Hispanic_Ethnicity", "Unknown_Ethnicity", "Less than high school_Education", "High school diploma or GED_Education", "Some college_Education", "Bachelor or undergraduate degree_Education", "Post-undergraduate degree_Education", "Unknown_Education", "No income_Income", "Under $10,000_Income", "$10,000 - $30,000_Income", "$30,000 - $50,000_Income", "$50,000 - $60,000_Income", "$80,000 - $100,000_Income", "$150,000 - $200,000_Income", "$250,000 or more_Income", "Under $15,000_Income", "$15,000 - $19,999_Income", "$20,000 - $39,999_Income", "$40,000 - $59,999_Income", "$60,000 - $79,999_Income", "$80,000 or more_Income", "Unknown_Income", "Vaginal_deliveryMode", "Cesarean_deliveryMode", "Unknown_deliveryMode"))) %>%
  arrange(x) %>% # order rows
  mutate(x=str_extract(x, ".*(?=_)")) %>% # remove category label now that rows are arranged 
  mutate_at(c("Stanford Enriched\nTerm (n=11)", "Stanford Enriched\nPreterm (n=9)", "UAB Enriched\nTerm (n=5)", "UAB Enriched\nPreterm (n=10)"), # remove 0's from income rows where they don't belong due to different levels in Stanford/UAB vs. MOMS-PI for income
            ~case_when(x %in% c("Under $15,000", "$15,000 - $19,999", "$20,000 - $39,999", "$40,000 - $59,999", "$60,000 - $79,999", "$80,000 or more")~"-",
                       x %!in% c("Under $15,000", "$15,000 - $19,999", "$20,000 - $39,999", "$40,000 - $59,999", "$60,000 - $79,999", "$80,000 or more")~.x)) %>%
  mutate_at(c("MOMS-PI\nTerm (n=90)", "MOMS-PI\nPreterm (n=43)", "MOMS-PI\nUnknown (n=98)"), # remove 0's from income rows where they don't belong due to different levels in Stanford/UAB vs. MOMS-PI for income
            ~case_when(x %in% c("No income", "Under $10,000", "$10,000 - $30,000", "$30,000 - $50,000", "$50,000 - $60,000", "$80,000 - $100,000", "$150,000 - $200,000", "$250,000 or more")~"-",
                       x %!in% c("No income", "Under $10,000", "$10,000 - $30,000", "$30,000 - $50,000", "$50,000 - $60,000", "$80,000 - $100,000", "$150,000 - $200,000", "$250,000 or more")~.x)) %>% 
  select(x, `Stanford Enriched\nTerm (n=11)`, `Stanford Enriched\nPreterm (n=9)`, `UAB Enriched\nTerm (n=5)`, `UAB Enriched\nPreterm (n=10)`, `MOMS-PI\nTerm (n=90)`, `MOMS-PI\nPreterm (n=43)`, `MOMS-PI\nUnknown (n=98)`) # order columns

demTable <- gaadDF %>%
  rbind(demDF) %>%
  dplyr::rename("  "=x) %>%
  dplyr::rename_at(2:8, ~str_extract(.x, "(?<=\\n).*")) %>%
  kbl(caption = "Table 2. Sampling Cohorts") %>%
  kable_classic(full_width=TRUE, html_font = "Arial") %>%
  add_header_above(c(" ", "Stanford Enriched" = 2, "UAB Enriched" = 2, "MOMS-PI" = 3)) %>%
  pack_rows("Age", 2, 6) %>%
  pack_rows("Race", 7, 10) %>%
  pack_rows("Ethnicity", 11, 13) %>%
  pack_rows("Education", 14, 19) %>%
  pack_rows("Income", 20, 34) %>%
  pack_rows("Delivery Mode", 35, 37)
demTable
#save_kable(demTable, file = file.path(figureOut, paste(Sys.Date(), "Table2_demTable.png", sep="_")), zoom=5)
```


# Set up taxa data frames
  Organize and set up tables from MetaPhlAn2 and *Gardnerella* mapping method.
## MetaPhlAn2
  Species relative abundances for Stanford and UAB cohorts:
```{r, message=FALSE, warning=FALSE}
sumDF <- suMetaphlanPath %>%
  read_tsv() %>%
  filter(str_detect(ID, "s__"), !str_detect(ID, "t__")) %>%
  mutate(Species = str_extract(ID, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -ID) %>%
  dplyr::rename_at(2:ncol(.), ~str_extract(.x, "[0-9]{10}")) %>%
  column_to_rownames(var="Species") %>%
  t %>%
  as.data.frame() %>%
  mutate_all(funs(./100)) %>%
  rownames_to_column(var="SampleID")
sumDF[1:6,1:5]
```
 Species relative abundances for MOMS-PI cohort:
```{r, message=FALSE, warning=FALSE}
hmp2mDF <- hmp2MetaphlanPath %>%
  read_tsv %>%
  filter(str_detect(ID, "s__"), !str_detect(ID, "t__")) %>%
  mutate(Species = str_extract(ID, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -ID) %>%
  as.data.frame() %>%
  dplyr::rename_at(2:ncol(.), ~str_extract(.x, "[0-9]{7}")) %>%
  column_to_rownames(var="Species") %>%
  t %>%
  as.data.frame() %>%
  mutate_all(funs(./100)) %>%
  rownames_to_column(var="SampleID") %>%
  filter(SampleID %in% QCfilterSamples) # keep only samples that pass quality filtering
hmp2mDF[1:6,1:5]
```
  Sanity check that relative abundances sum to 1
```{r, message=FALSE, warning=FALSE}
hmp2mDF %>%
  column_to_rownames("SampleID") %>%
  rowSums %>%
  as.data.frame() %>%
  summary

hmp2mDF %>%
  column_to_rownames("SampleID") %>%
  rowSums %>%
  as.data.frame() %>%
  ggplot(aes(x=`.`)) +
  geom_histogram()
```
  Relative abundances of most samples sum to 1.
  What is in samples that do not sum to 1
```{r, message=FALSE, warning=FALSE}
hmp2MetaphlanPath %>%
  read_tsv %>%
  dplyr::rename_at(2:ncol(.), ~str_extract(.x, "[0-9]{7}")) %>%
  select(ID, "6744527", "6744289") %>%
  filter(`6744527`>0|`6744289`>0) %>%
  formattable()
```
## *Gardnerella* Mapping
  Add relative abundance of individual clades and genomospecies in Stanford and UAB cohort samples, and check for differences in finding *Gardnerella* by MetaPhlAn2 and mapping method.
  **Stanford and UAB Clades**
```{r, message=FALSE, warning=FALSE}
suGardCladeAb0 <- suGardCladePath %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID)) %>%
  select(-propClade) %>%
  filter(n>=2) %>% #count as present if >=2 reads are mapped to that clade/genomospecies
  group_by(SampleID) %>%
  mutate(propClade=n/sum(n)) %>%
  ungroup() %>%
  full_join(sumDF[,c("SampleID", "Gardnerella_vaginalis")], by=c("SampleID")) %>%
  mutate(relativeAbundance=propClade*Gardnerella_vaginalis)

# check if there are any samples where Gard was not found by metaphlan but was found by mapping method
suGardCladeAb0 %>%
  filter(Gardnerella_vaginalis==0, n>1)

#check if there are any samples where Gard was found by metaphlan but not by mapping method
suGardCladeAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
  There is one sample where *Gardnerella* clades were found by mapping method and not by MetaPhlAn2, and no samples where *Gardnerella* was found by MetaPhlAn2 but not by mapping method
    
```{r, message=FALSE, warning=FALSE}
# make table of clade abundances
suGardCladeAb <- suGardCladeAb0 %>%
  filter(relativeAbundance>0) %>%
  select(SampleID, Clade, relativeAbundance)
```
  
  **Stanford and UAB Genomospecies**
```{r, message=FALSE, warning=FALSE}
suGardGenomoAb0 <- suGardGenomoPath %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID)) %>%
  select(-propGenomospecies) %>%
  filter(n>=2) %>%
  group_by(SampleID) %>%
  mutate(propGenomospecies=n/sum(n)) %>%
  ungroup() %>%
  full_join(sumDF[,c("SampleID", "Gardnerella_vaginalis")], by=c("SampleID")) %>%
  mutate(relativeAbundance=propGenomospecies*Gardnerella_vaginalis)

# check if there are any samples where Gard was not found by metaphlan but was found by mapping method
suGardGenomoAb0 %>%
  filter(Gardnerella_vaginalis==0, n>1)

#check if there are any samples where Gard was found by metaphlan but not by mapping method
suGardGenomoAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
  No samples where *Gardnerella* genomospecies were found by mapping method but not MetaPhlAn2 and one sample where *Gardnerella* was found by MetaPhlAn2 but not by mapping method.

```{r, message=FALSE, warning=FALSE}
# make table of genomospecies abundances
suGardGenomoAb <- suGardGenomoAb0 %>%
  filter(relativeAbundance>0) %>%
  select(SampleID, Genomospecies, relativeAbundance)
```

  **MOMS-PI Clades**
```{r, message=FALSE, warning=FALSE}
hmp2GardCladeAb0 <- hmp2GardCladePath %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID)) %>%
  select(-propClade) %>%
  filter(n>=2) %>% # 2+ reads need to map to a clade/genomospecies to be considered present
  group_by(SampleID) %>%
  mutate(propClade=n/sum(n)) %>%
  ungroup() %>%
  full_join(hmp2mDF[,c("SampleID", "Gardnerella_vaginalis")], by=c("SampleID")) %>%
  mutate(relativeAbundance=propClade*Gardnerella_vaginalis) %>%
  filter(SampleID %in% QCfilterSamples)

# check if there are any samples where Gard was not found by metaphlan but was found by mapping method
hmp2GardCladeAb0 %>%
  filter(Gardnerella_vaginalis==0, n>1)

#check if there are any samples where Gard was found by metaphlan but not by mapping method
hmp2GardCladeAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
  There are 58 samples where *Gardnerella*  was found by MetaPhlan2 but no clades found by mapping method, and 9 samples where *Gardnerella* was found by mapping method but not by MetaPhlan2

```{r, message=FALSE, warning=FALSE}
# make abundance table
hmp2GardCladeAb <- hmp2GardCladeAb0 %>%
  filter(relativeAbundance>0) %>%
  select(SampleID, Clade, relativeAbundance)
```

  **MOMS-PI Genomospecies**
```{r, message=FALSE, warning=FALSE}
hmp2GardGenomoAb0 <- hmp2GardGenomoPath %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID)) %>%
  select(-propGenomospecies) %>%
  filter(n>=2) %>% # 2+ reads mapped to each clade/genomospecies to be considered present
  group_by(SampleID) %>%
  mutate(propGenomospecies=n/sum(n)) %>%
  ungroup() %>%
  full_join(hmp2mDF[,c("SampleID", "Gardnerella_vaginalis")], by=c("SampleID")) %>%
  mutate(relativeAbundance=propGenomospecies*Gardnerella_vaginalis)

# check if there are any samples where Gard was not found by metaphlan but was found by mapping method
hmp2GardGenomoAb0 %>%
  filter(Gardnerella_vaginalis==0, n>1)

#check if there are any samples where Gard was found by metaphlan but not by mapping method
hmp2GardGenomoAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
  127 samples where *Gardnerella* was found by MetaPhlan2 but no genomospecies were found by mapping method and 2 samples where *Gardnerella* was found by mapping method but not by MetaPhlan2

```{r}
#make abundance table
hmp2GardGenomoAb <- hmp2GardGenomoAb0 %>%
  filter(relativeAbundance>0) %>%
  select(SampleID, Genomospecies, relativeAbundance)
```

### Further assess uncharacterized samples:
  Asses samples where *Gardnerella* was uncharacterized at the clade level. (Where *Gardnerella* was found by MetaPhlAn2 but no clades were identified using the mapping method).
```{r}
unchGardSamples <- hmp2GardCladeAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0) %>%
  .$SampleID
```

  Comparison of number of microbial paired reads and proportion *Gardnerella* in samples with uncharacterized *Gardnerella* vs. samples where *Gardnerella* was found with both methods.
```{r}
q <- hmp2mDF %>%
  select(SampleID, Gardnerella_vaginalis) %>%
  mutate(unchGard=case_when(SampleID %in% unchGardSamples~TRUE,
                            !(SampleID %in% unchGardSamples)~FALSE)) %>%
  left_join(sgDF1[,c("SampleID", "bbmapFiltered_pairs", "bacLoad")], by="SampleID") %>%
  ggplot(aes(x=Gardnerella_vaginalis, y=bbmapFiltered_pairs, color=unchGard, shape=unchGard)) +
  geom_point(alpha=0.5) +
  theme(legend.position = "bottom") +
  binaryColors +
  labs(x="Proportion Gardnerella", y="Microbial Reads", color="Uncharacterized Gardnerella", shape="Uncharacterized Gardnerella")

ggMarginal(q, groupFill = TRUE, type="density")
```
  Samples with uncharacterized *Gardnerella* have fewer microbial reads and a lower proportion of *Gardnerella*

## Join abundance tables
```{r, message=FALSE, warning=FALSE}
# Gardnerella abundance
gardCladeAb0 <- full_join(suGardCladeAb, hmp2GardCladeAb, by = c("SampleID", "Clade", "relativeAbundance"))
gardGenomoAb0 <- full_join(suGardGenomoAb, hmp2GardGenomoAb, by = c("SampleID", "Genomospecies", "relativeAbundance"))

# metaphlan
mDF0 <- sumDF %>%
  full_join(hmp2mDF)

#number of speacies that appear in only one vs. both samples
testOverlap <- summarise_all(mDF0, anyNA) %>%
  gather("Species", "anyNA", 2:ncol(.)) %>%
  select(-SampleID)

nrow(testOverlap)
table(testOverlap$anyNA)
```
  364 total species and 209 appear in both studies (Stanford/UAB vs. MOMS-PI).
  Number of species in each study:
```{r, message=FALSE, warning=FALSE}
ncol(sumDF)-1
ncol(hmp2mDF)-1
```
  
  Replace NAs with zeros for abundance values for species that did not appear in each respective study. 
```{r, message=FALSE, warning=FALSE}
mDF1 <- mDF0 %>%
  mutate_if(is.numeric, ~replace_na(.x, 0)) %>%
  left_join(sgDF1[,c("SampleID", "Cohort")], by="SampleID") %>%
  select(SampleID, Cohort, everything())
anyNA(mDF1) # check for remaining NAs
```

## Select enriched *Gardnerella* MOMS-PI subset
    Proportion *Gardnerella* in samples by study:
```{r, message=FALSE, warning=FALSE}
mDF1 %>%
  select(SampleID, Cohort, Gardnerella_vaginalis) %>%
  mutate(Cohort=case_when(Cohort=="Stanford Enriched"|Cohort=="UAB Enriched"~"Stanford & UAB Enriched",
                          Cohort=="MOMS-PI"~"MOMS-PI"),
         Cohort=factor(Cohort, levels = c("Stanford & UAB Enriched", "MOMS-PI"))) %>%
  ggplot(aes(x=Gardnerella_vaginalis, color=Cohort, fill=Cohort, group=Cohort)) +
  geom_density(alpha=0.5) +
  binaryColors +
  labs(x="Proportion Gardnerella")

gardAbundance2Cohort <- mDF1 %>%
  select(SampleID, Cohort, Gardnerella_vaginalis) %>%
  left_join(sgDF1[,c("SampleID", "SubjectID")], by="SampleID") %>%
  mutate(Cohort=case_when(Cohort=="Stanford Enriched"|Cohort=="UAB Enriched"~"Stanford & UAB Enriched",
                          Cohort=="MOMS-PI"~"MOMS-PI"),
         Cohort=factor(Cohort, levels = c("Stanford & UAB Enriched", "MOMS-PI")))
gardAbundance2Cohort %>%
  group_by(Cohort) %>%
  summarize(min=min(Gardnerella_vaginalis),
            `1stQu`=summary(Gardnerella_vaginalis)[2],
            median=median(Gardnerella_vaginalis),
            mean=mean(Gardnerella_vaginalis),
            `3rdQu`=summary(Gardnerella_vaginalis)[5],
            max=max(Gardnerella_vaginalis)) %>%
  formattable()
```
  
  Proportion *Gardnerella* by as subject averages by study:
```{r, message=FALSE, warning=FALSE}
subjectGardAbundance2Cohort <- gardAbundance2Cohort %>%
  group_by(SubjectID, Cohort) %>%
  summarize(Gardnerella_vaginalis=mean(Gardnerella_vaginalis)) %>%
  ungroup
  
subjectAvgPropGard <- subjectGardAbundance2Cohort %>%
  ggplot(aes(x=Gardnerella_vaginalis, color=Cohort, fill=Cohort)) +
  geom_density(alpha=0.5) +
  binaryColors +
  theme(legend.position = "bottom", 
        legend.text = element_text(size=10), 
        legend.title = element_text(size=11)) +
  labs(x="Subject Average Proportion Gardnerella")
subjectAvgPropGard
```

  MGS sequenced samples from Stanford and UAB were selected based on enrichment of *Gardnerella* at the subject level. Create a cohort of a subset of samples in the MOMS-PI cohort that are also selected for enrichment of *Gardnerella* at the subject level for better comparison.
```{r, message=FALSE, warning=FALSE}
ntiles <- subjectGardAbundance2Cohort %>%
  filter(Cohort=="Stanford & UAB Enriched") %>%
  mutate(decile=ntile(Gardnerella_vaginalis, 3)) %>%
  group_by(decile) %>%
  filter(Gardnerella_vaginalis==min(Gardnerella_vaginalis)) %>%
  .$Gardnerella_vaginalis %>%
  sort %>%
  unique
ntiles

# subjectGardAbundance2Cohort0 <- subjectGardAbundance2Cohort %>%
#   filter(Cohort=="MOMS-PI",
#          SubjectID %in% hmp2SubjectsWDel) %>% # *** For selecting subjects that have gestational age at delivery  ***
#   mutate(Ntile=case_when(between(Gardnerella_vaginalis, ntiles[1], ntiles[2])~1,
#                         between(Gardnerella_vaginalis, ntiles[2], ntiles[3])~2,
#                         Gardnerella_vaginalis > ntiles[3]~3))
# 
# subjN <- min(count(subjectGardAbundance2Cohort0, Ntile)$n) # number of subjects from each ntile to take
# 
# HMP2highGardSubjects <- subjectGardAbundance2Cohort0 %>% 
#   group_by(Ntile) %>%
#   sample_n(subjN, replace = FALSE) %>%
#   .$SubjectID
#  
# write_lines(HMP2highGardSubjects, "./HMP2highGardSubjects.txt")
HMP2highGardSubjects <- read_lines("./HMP2highGardSubjects.txt")
```
  
  Figure S4: Distributions of subject average *Gardnerella* abundances
```{r, message=FALSE, warning=FALSE, fig.width=5, fig.height=2.5}
subjectAvgPropGardEnr <- subjectGardAbundance2Cohort %>%
  filter(Cohort=="Stanford & UAB Enriched"|(Cohort=="MOMS-PI"&SubjectID %in% HMP2highGardSubjects)) %>%
  mutate(Cohort=factor(Cohort, levels=c("Stanford & UAB Enriched", "MOMS-PI"), labels=c("Stanford & UAB Enriched", "MOMS-PI Enriched"))) %>%
  ggplot(aes(x=Gardnerella_vaginalis, color=Cohort, fill=Cohort, group=Cohort)) +
  geom_density(alpha=0.5) +
  binaryColors +
  theme(legend.position = "bottom", 
        legend.text = element_text(size=10), 
        legend.title = element_text(size=11)) +
  labs(x="Subject Average Proportion Gardnerella")

plot_grid(subjectAvgPropGard, subjectAvgPropGardEnr, nrow=1, labels = c("A", "B"), label_size = 15)
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "FigureS4_enrichedSubjectPropGardbyStudy.png", sep="_")))
```
  Distribution of *Gardnerella* abundnace in samples
```{r, message=FALSE, warning=FALSE}
HMP2HighGardSamples <- sgDF1 %>%
  filter(SubjectID %in% HMP2highGardSubjects) %>%
  .$SampleID

gardAbundance2Cohort %>%
  filter(Cohort=="Stanford & UAB Enriched"|Cohort=="MOMS-PI"&SubjectID %in% HMP2highGardSubjects) %>%
  mutate(Cohort=factor(Cohort, levels=c("Stanford & UAB Enriched", "MOMS-PI"), labels=c("Stanford & UAB Enriched", "MOMS-PI Enriched"))) %>%
  ggplot(aes(x=Gardnerella_vaginalis, color=Cohort, fill=Cohort)) +
  geom_density(alpha=0.5) +
  binaryColors +
  labs(x="Proportion of Gardnerella")
```

  **Add enriched MOMS-PI cohort to tables**
  Sample and subject metadata:
```{r, message=FALSE, warning=FALSE}
# extract enriched Gardnerella subset and full join to metadata 
highGardSgDF <- sgDF1 %>%
  filter(SubjectID %in% HMP2highGardSubjects) %>% 
  mutate(Cohort="MOMS-PI Enriched",
         SampleID=paste0(SampleID, "_E"),
         SubjectID=paste0(SubjectID, "_E"))

nrow(highGardSgDF)  

sgDF <- sgDF1 %>%
  mutate(Cohort=as.vector(Cohort),
         SubjectID=as.character(SubjectID)) %>%
  full_join(highGardSgDF, by=colnames(sgDF1)) %>%
  mutate(Cohort=factor(Cohort, levels = cohorts))

bacLoad <- sgDF %>%
  select(SampleID, bacLoad)

sgDF %>%
  group_by(Cohort) %>%
  summarize(`N Subjects`=n_distinct(SubjectID), `N Samples`=n_distinct(SampleID)) %>%
  formattable()
```

  MetaPhlAn2 species abundance tables:
```{r, message=FALSE, warning=FALSE}
highGardmDF <- mDF1 %>%
  filter(SampleID %in% HMP2HighGardSamples) %>%
  mutate(Cohort="MOMS-PI Enriched",
         SampleID=paste0(SampleID, "_E"))

mDF <- mDF1 %>%
  mutate(Cohort=as.vector(Cohort)) %>%
  rbind(highGardmDF) %>%
  mutate(Cohort=factor(Cohort, levels = c("Stanford Enriched", "UAB Enriched", "MOMS-PI Enriched", "MOMS-PI")))
```
  *Gardnerella* clade and genomospecies abundance tables:
```{r, message=FALSE, warning=FALSE}
# Clade Abundances
gardCladeAb1 <- gardCladeAb0 %>%
  left_join(sgDF1[,c("SampleID", "Cohort")], by="SampleID") %>%
  mutate(Cohort=as.vector(Cohort))
 
highGardCladeAb <- gardCladeAb1 %>%
  filter(SampleID %in% HMP2HighGardSamples) %>%
  mutate(Cohort="MOMS-PI Enriched",
         SampleID=paste0(SampleID, "_E"))

gardCladeAb <- gardCladeAb1 %>%
  rbind(highGardCladeAb) %>%
  mutate(Cohort=factor(Cohort, levels = cohorts))

# Gemomospecies Abundances
gardGenomoAb1 <- gardGenomoAb0 %>%
  left_join(sgDF1[,c("SampleID", "Cohort")], by="SampleID") %>%
  mutate(Cohort=as.vector(Cohort))
 
highGardGenomoAb <- gardGenomoAb1 %>%
  filter(SampleID %in% HMP2HighGardSamples) %>%
  mutate(Cohort="MOMS-PI Enriched", 
         SampleID=paste0(SampleID, "_E"))

gardGenomoAb <- gardGenomoAb1 %>%
  rbind(highGardGenomoAb) %>%
  mutate(Cohort=factor(Cohort, levels = cohorts))

# note these tables contain samples with at least one identified clade or genomospecies of gardnerella
```



##  Top taxa
```{r, message=FALSE, warning=FALSE}
avgTop20Abundance0 <- mDF %>%
  select(SampleID, Cohort, everything()) %>%
  gather("Species", "Abundance", 3:ncol(.)) %>%
  group_by(Cohort, Species) %>%
  summarise(avgAbundance=mean(Abundance)) 

top20taxa <- avgTop20Abundance0 %>%
  group_by(Cohort) %>%
  top_n(20, avgAbundance) %>%
  .$Species %>%
  unique

avgTop20Abundance0 %>%
  spread(Cohort, avgAbundance) %>%
  mutate(stanfordRank=rank(-`Stanford Enriched`),
         uabRank=rank(-`UAB Enriched`), 
         momspiRank=rank(-`MOMS-PI`),
         momspiERank=rank(-`MOMS-PI Enriched`),
         Stanford=round(`Stanford Enriched`, 2),
         UAB=round(`UAB Enriched`,2),
         `MOMS-PI`=round(`MOMS-PI`, 2),
         `MOMS-PI Enriched`=round(`MOMS-PI Enriched`, 2)) %>%
  filter(Species %in% top20taxa) %>%
  select(Species, `Stanford Enriched`, stanfordRank, `UAB Enriched`, uabRank, `MOMS-PI Enriched`, momspiERank, `MOMS-PI`, momspiRank) %>%
  arrange(stanfordRank) %>%
  formattable()
  
x <- mDF %>%
  select(SampleID, Cohort, top20taxa) %>%
  mutate_if(is.numeric, ~case_when(.x>0~1,
                                   .x==0~0))%>%
  group_by(Cohort) %>%
  summarize_if(is.numeric, ~sum(.x)/n()) %>%
  data.frame()
rownames(x) <- x$Cohort
x %>%
  select(-Cohort) %>%
  t %>%
  as.data.frame() %>%
  rownames_to_column("Species") %>%
  arrange(-`Stanford Enriched`) %>%
  formattable()
```

  Create abundance table with taxa of interest
```{r, message=FALSE, warning=FALSE}
# abundance table of taxa of interest with clades
mDFtoiClades <- gardCladeAb %>%
  spread(Clade, relativeAbundance) %>%
  full_join(mDF, by=c("SampleID", "Cohort")) %>%
  select(SampleID, Cohort, Gardnerella_vaginalis, clades, lactos, anaerobes) %>% # keep total Gardnerella abundance in table for now. Many operations are taxon vs. taxon. Will need to remove this column when analyses call for it.
  mutate_at(vars(c(Gardnerella_vaginalis, clades, lactos, anaerobes)), replace_na, 0) %>%
  mutate(Cohort=factor(Cohort, levels=cohorts))
```

# *Gardnerella* variant abundance
## Clades per sample
```{r, message=FALSE, warning=FALSE}
gardCladeAbU <- gardCladeAb %>%
  full_join(mDF[,c("SampleID", "Cohort", "Gardnerella_vaginalis")], by=c("SampleID","Cohort")) %>%
  spread(Clade, relativeAbundance) %>%
  select(-`<NA>`) %>% # remove "NA column" created by samples that have zero Gardnerella
  mutate(Uncharacterized_Gardnerella=case_when(Gardnerella_vaginalis>=0&is.na(C1)&is.na(C2)&is.na(C3)&is.na(C4)&is.na(C5)&is.na(C6)~Gardnerella_vaginalis,
                                               Gardnerella_vaginalis>0&!is.na(clades)~0)) %>%
  gather("var", "relativeAbundance", c(clades, Gardnerella_vaginalis, Uncharacterized_Gardnerella)) %>%
  replace_na(list(relativeAbundance=0))
```

  Average clades per sample in each cohorts:
```{r, message=FALSE, warning=FALSE}
gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nClades=C1+C2+C3+C4+C5+C6) %>%
  with_groups(Cohort, summarize, mean=mean(nClades), `standard deviation`=sd(nClades)) %>%
  formattable
```

Average clades per sample across all cohorts:
```{r, message=FALSE, warning=FALSE}
gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nClades=C1+C2+C3+C4+C5+C6) %>%
  summarize(mean=mean(nClades), `standard deviation`=sd(nClades)) %>%
  formattable
```

```{r, message=FALSE, warning=FALSE}
cladesPerSample <- gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nClades=C1+C2+C3+C4+C5+C6) %>%
  count(Cohort, nClades) %>%
  group_by(Cohort) %>%
  mutate(n=n/sum(n)) %>%
  ggplot(aes(x=nClades, y=n)) +
  geom_bar(stat="identity", fill = "black") +
  facet_wrap(~Cohort) +
  labs(x="Clades per Sample", y="Proportion of Samples", fill=bquote('Uncharacterized'~italic(Gardnerella))) +
  theme(axis.text.x = element_text(size=9),
        strip.text = element_text(size=13),
        aspect.ratio = 0.5)
```

## Genomospecies per sample
```{r, message=FALSE, warning=FALSE}
gardGenomoAbU <- gardGenomoAb %>%
  spread(Genomospecies, relativeAbundance) %>%
  full_join(mDF[,c("SampleID", "Cohort", "Gardnerella_vaginalis")], by=c("SampleID", "Cohort")) %>%
  mutate(Uncharacterized_Gardnerella=case_when(Gardnerella_vaginalis>=0&is.na(GS1)&is.na(GS2)&is.na(GS3)&is.na(GS4)&is.na(GS5)&is.na(GS6)&is.na(GS7)&is.na(GS8)&is.na(GS9)&is.na(GS10)&is.na(GS11)&is.na(GS12)&is.na(GS13)&is.na(GS14)~Gardnerella_vaginalis,
                                               Gardnerella_vaginalis>0&!is.na(genomos)~0)) %>%
  gather("var", "relativeAbundance", c(genomos, Gardnerella_vaginalis, Uncharacterized_Gardnerella)) %>%
  replace_na(list(relativeAbundance=0))
```
  
  Average clades per sample in each cohorts:
```{r, message=FALSE, warning=FALSE}
 gardGenomoAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_at(genomos, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nGenomos=GS1+GS2+GS3+GS4+GS5+GS6+GS7+GS8+GS9+GS10+GS11+GS12+GS13+GS14) %>%
  with_groups(Cohort, summarize, mean=mean(nGenomos), `standard deviation`=sd(nGenomos)) %>%
  formattable
```
 Number of samples with all 14 genomospecies
```{r, message=FALSE, warning=FALSE}
 gardGenomoAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_at(genomos, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nGenomos=GS1+GS2+GS3+GS4+GS5+GS6+GS7+GS8+GS9+GS10+GS11+GS12+GS13+GS14) %>%
  count(Cohort, nGenomos) %>%
  group_by(Cohort) %>%
  mutate(pcnt=n/sum(n)*100) %>% # convert to fraction of sample
  filter(nGenomos==14) %>%
  formattable
```

  Average clades per sample across all cohorts:
```{r, message=FALSE, warning=FALSE}
 gardGenomoAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_at(genomos, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nGenomos=GS1+GS2+GS3+GS4+GS5+GS6+GS7+GS8+GS9+GS10+GS11+GS12+GS13+GS14) %>%
  summarize(mean=mean(nGenomos), `standard deviation`=sd(nGenomos)) %>%
  formattable
```

```{r, message=FALSE, warning=FALSE}
genomosPerSample <- gardGenomoAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_at(genomos, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nGenomos=GS1+GS2+GS3+GS4+GS5+GS6+GS7+GS8+GS9+GS10+GS11+GS12+GS13+GS14) %>%
  count(Cohort, nGenomos) %>%
  group_by(Cohort) %>%
  mutate(n=n/sum(n), # convert to fraction of sample
         nGenomos=factor(nGenomos, c("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14"))) %>% 
  ggplot(aes(x=nGenomos, y=n)) +
  geom_bar(stat="identity", fill = "black") +
  facet_wrap(~Cohort) +
  labs(x="Genomospecies per Sample", y="Proportion of Samples") +
  theme(axis.text.x = element_text(size=9),
        strip.text = element_text(size=13),
        aspect.ratio = 0.5)
```

## Variant Relative abundance
### Clades
  Calculate clade prevalences
```{r, message=FALSE, warning=FALSE}
#p = # times clade is present, n = number samples, m=mean relative abundance
prevTable <- gardCladeAbU %>%
  with_groups(c(Cohort, var), summarize, p = sum(relativeAbundance > 0), n = n(), m=mean(relativeAbundance), sd=sd(relativeAbundance)) %>% 
  mutate(prevalence=round(p/n, 2),
         var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"), labels=c(clades, "Unch.", "Total")))
```

  Sample relative abundances with prevalence
```{r, fig.width=7, fig.height=3.5}
gardCladeAbU %>%
  filter(var!="Uncharacterized_Gardnerella") %>%
  mutate(var=factor(var, levels = c(clades,  "Gardnerella_vaginalis"), labels=c(clades, "Total"))) %>%
  left_join(prevTable[,c("Cohort", "var", "prevalence")], by=c("Cohort", "var")) %>%
  mutate(Cohort=factor(Cohort, levels = cohorts, labels=cohortAbbrevs)) %>%
  ggplot(aes(x=var, y=relativeAbundance, color=var, fill=var)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  geom_text(aes(y=1.2, x=var, label=prevalence), color="black", size=5) +
  scale_y_continuous(breaks = c(0, 0.25, 0.50, 0.75, 1, 1.2), limits = c(0,1.3), labels=c("0.00", "0.25", "0.50", "0.75", "1.00", "prevalence")) +
  cladeColorsTotal +
  facet_grid(Cohort~.) +
  theme(legend.position = "none",
        axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size=16),
        axis.title.y = element_text(size=23),
        strip.text = element_text(size=17)) +
  labs(x=NULL, y="Relative Abundance")
```
  
  Clades differ in abundance and abundance of clades varies across cohorts. Clade 3 is less abundant in Stanford cohort but abundant in other cohorts, especially UAB. Clade 1 is more abundant in Stanford. Clade 6 is rare in Stanford but appears in UAB and MOMS-PI cohorts.  

  Prevalence vs relative abundance of *Gardnerella* clades
```{r, message=FALSE, warning=FALSE}
cladeAbvPrev <- prevTable %>%
  filter(var!="Unch.") %>%
  ggplot(aes(x=prevalence, y=m, color=var)) +
  geom_point(size=3, alpha=0.75) +
  scale_x_continuous(limits=c(0,1), breaks=seq(0,1,0.2)) +
  scale_y_continuous(limits =c(0,1), breaks=seq(0,1,0.2)) +
  facet_wrap(~Cohort) +
  cladeColorsTotal +
  coord_fixed() +
  theme(legend.position = "right",
        strip.text = element_text(size=13)) +
  labs(x="Prevalence", y="Mean Relative Abundance", color="Clade")
```

### Genomospecies
  Calculate genomospecies prevalences
```{r, message=FALSE, warning=FALSE}
genomosPrevTable <- gardGenomoAbU %>%
  with_groups(c(Cohort, var), summarize, p = sum(relativeAbundance > 0), n = n(), m=mean(relativeAbundance), sd=sd(relativeAbundance)) %>% 
  mutate(prevalence=round(p/n, 2))
```

  Sample relative abundance with prevalence
```{r, fig.height=3, fig.width=7}
gardGenomoAbU %>%
  left_join(genomosPrevTable[,c("Cohort", "var", "prevalence")], by=c("Cohort", "var")) %>%
  filter(var!="Uncharacterized_Gardnerella") %>%
  mutate(var=factor(var, levels=c(genomosCladeOrder, "Gardnerella_vaginalis"), labels = c(genomoNamesCladeOrder, "Total")),
         Cohort=factor(Cohort, levels=cohorts, labels = cohortAbbrevs)) %>%
  ggplot(aes(x=var, y=relativeAbundance, color=var, fill=var)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  geom_text(aes(y=1.2, x=var, label=prevalence), color="black", size=5) +
  scale_y_continuous(breaks = c(0, 0.25, 0.50, 0.75, 1, 1.2), limits = c(0,1.3), labels=c("0.00", "0.25", "0.50", "0.75", "1.00", "prevalence")) +
  facet_grid(Cohort~.) +
  genomoColorsTotal +
  theme(axis.text.x = element_text(size=11),
        axis.text.y = element_text(size=12),
        axis.title.y = element_text(size=20),
        strip.text = element_text(size = 11),
        legend.position="none") +
  labs(x=NULL, y="Relative Abundance")
```
 
  Prevalence vs relative abundance of *Gardnerella* genomospecies
```{r, message=FALSE, warning=FALSE}
genomoAbvPrev <- gardGenomoAbU %>%
  group_by(Cohort, var) %>%
  summarise(avgAbundance=mean(relativeAbundance)) %>%
  left_join(genomosPrevTable, by=c("Cohort", "var")) %>%
  filter(var!="Uncharacterized_Gardnerella") %>%
  dplyr::rename(Genomospecies=var) %>%
  mutate(Genomospecies=factor(Genomospecies, levels = c(genomosCladeOrder, "Gardnerella_vaginalis"))) %>%
  ggplot(aes(x=prevalence, y=avgAbundance, color=Genomospecies)) +
  geom_point(size=3, alpha=0.75) +
  scale_x_continuous(limits=c(0,1), breaks=seq(0,1,0.2)) +
  scale_y_continuous(limits =c(0,1), breaks=seq(0,1,0.2)) +
  genomoColorsTotal +
  facet_wrap(~Cohort) +
  coord_fixed() +
  theme(legend.position = "right", 
        strip.text = element_text(size=13)) +
  labs(x="Prevalence", y="Mean Relative Abundance", color="Species")
```

### Figure 3: Prevalence and abundance of Gardnerella variants across the four cohorts
```{r, message=FALSE, warning=FALSE, , fig.width = 6, fig.height=4} 
varsPerSample <- plot_grid(cladesPerSample, genomosPerSample, nrow = 1, labels = c("A", "B"), label_size = 15)
gardVarAbun <- plot_grid(cladeAbvPrev, genomoAbvPrev, nrow = 1, labels = c("C", "D"), label_size = 15)
plot_grid(varsPerSample, gardVarAbun, nrow = 2, axis = "l")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "Figure3_cladeGenomoPrevAbun.png", sep="_")))
```

## Microbial Load
  Compare microbial load across the four cohorts:
```{r, message=FALSE, warning=FALSE}
# table of descriptive statistics with outliers
sgDF %>%
  group_by(Cohort) %>%
  summarise_at("bacLoad", list(min=min, mean=mean, median=median, max=max)) %>%
  mutate_if(is.numeric, ~round(.x, 4)) %>%
  formattable()

# microbial load of outliers
sgDF %>%
  filter(bacLoad>2) %>%
  select(SampleID, Cohort, bacLoad) %>%
  formattable

# density plot without outliers (bacload ≥ 2)
sgDF %>%
  filter(bacLoad<2) %>%
  mutate(Cohort=factor(Cohort, levels=cohorts, labels=cohortAbbrevs)) %>%
  ggplot(aes(x=bacLoad, fill=Cohort, color=Cohort)) +
  geom_density(alpha=0.5) +
  quadColors2 +
  theme(legend.position = c(0.85,0.8)) +
  labs(x="Microbial Load")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "FigureS5_bacLoadDensNoOut.png", sep="_")))


# table of descriptive statistics without outliers
sgDF %>%
  filter(bacLoad<2) %>%
  group_by(Cohort) %>%
  summarise_at("bacLoad", list(min=min, mean=mean, median=median, max=max)) %>%
  mutate_if(is.numeric, ~round(.x, 4)) %>%
  formattable()
```

## Absolute abundance
### Clade aboslute abundance
  All samples absolute abundance
```{r, fig.width=7, fig.height=3.5}
gardCladeAbU %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad, 
         var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella", "Gardnerella_vaginalis"), labels = c(clades, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = cohorts, labels=cohortAbbrevs)) %>%
  ggplot(aes(x=var, y=absoluteAbundance, color=var)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  cladeColorsTotalUnch +
  scale_y_continuous(limits=c(0,1), n.breaks = 5) +
  facet_grid(Cohort~.) +
  theme(legend.position = "none",
        axis.text.x = element_text(size=16),
        axis.text.y = element_text(size=16),
        axis.title.y = element_text(size=23),
        strip.text = element_text(size=17)) +
  labs(x=NULL, y="Absolute Abundance")
```
  Clades 3 and 4 appear to show the greatest absolute abundance.

### Genomospecies aboslute abundance
  All samples absolute abundance
```{r, fig.height=3, fig.width=7}
gardGenomoAbU %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad, 
         var=factor(var, levels=c(genomosCladeOrder, "Uncharacterized_Gardnerella", "Gardnerella_vaginalis"), labels = c(genomoNamesCladeOrder, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = cohorts, labels = cohortAbbrevs)) %>%
  ggplot(aes(x=var, y=absoluteAbundance, color=var)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  facet_grid(Cohort~.) +  
  genomoColorsTotalUnch +
  theme(axis.text.x = element_text(size=11),
        axis.text.y = element_text(size=12),
        axis.title.y = element_text(size=20),
        strip.text = element_text(size = 11),
        legend.position="none") +
  labs(x=NULL, y="Absolute Abundance")
#ggsave(file.path(figureOut, file.path(paste(Sys.Date(), "genomoAbsoluteAbundanceByCohort.png", sep="_"))))
```
  Genomospecies *G. swidsinskii* and genomospecies 7 appear to show the greatest absolute abundance.


# Thresholds for presence-absence
  Determine what threshold will be used to count a taxon as present vs. absent for co-occurrence and presence absence vs. microbial load analyses
  
  MetaPhlAn2 will report all taxa that have 10% of marker genes non-zero [based on this Google group response from Nicola Segata](https://groups.google.com/g/metaphlan-users/c/0jLFs8v9nI4?pli=1). Average number of marker genes per bacterial species is 184 ± 45. (Truong et al., 2015, *Nature Mathods* "MetaPhlAn2 for enhanced metagenomic taxonomic profiling"). About 1 million markers from >7,500 species. Minimum number of reads to be detected would be 18.4. 
```{r}
sgDF1 %>%
  mutate(ntile=ntile(bbmapFiltered_pairs, n=20)) %>%
  filter(ntile==5) %>%
  filter(bbmapFiltered_pairs==min(bbmapFiltered_pairs)) %>%
  .$bbmapFiltered_pairs

sgDF1 %>%
  mutate(over30K=bbmapFiltered_pairs>100000) %>%
  count(over30K) %>% 
  mutate(freq=n/sum(n))
```
  80% of samples have greater than 100,000 read pairs. MetaPhlAn2 uses about 10% of the reads in the library, and 18.4/10,000 yields a functional limit of detection of roughly 0.1%.

```{r}
18.4/10000 
```

  A threshold of 0.1% has been used frequently in presence-absence analyses in microbiome data and will be used as a threshold here.

# Gardnerella and microbial load
## Microbial load by number of clades
```{r, message=FALSE, warning=FALSE}
loadVClades <- gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  filter(Uncharacterized_Gardnerella==0) %>%
  select(-Gardnerella_vaginalis) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  gather("var", "abundance", clades) %>%
  group_by(SampleID, Cohort) %>%
  summarize(nClades=sum(abundance)) %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  ggplot(aes(x=nClades, y=bacLoad)) +
  geom_jitter(alpha=0.25) +
  geom_smooth(method="lm", se=FALSE, color="blue", linetype=1) +
  stat_cor(method = "spearman", alternative = "greater", cor.coef.name = c("rho"), label.y = 1.5, label.x = 0, label.sep = "\n") + 
  facet_wrap(~Cohort) +
  scale_x_continuous(breaks = seq(0,6,1)) +
  labs(x="Clades per Sample", y="Microbial Load")
loadVClades
```
  Is this a function of sequencing depth?
```{r}
gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  filter(Uncharacterized_Gardnerella==0) %>%
  select(-Gardnerella_vaginalis) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  gather("var", "abundance", clades) %>%
  group_by(SampleID, Cohort) %>%
  summarize(nClades=sum(abundance)) %>%
  left_join(sgDF[,c("SampleID", "bacLoad", "Sickle_pairs")], by="SampleID") %>%
  filter(bacLoad<2) %>%
  ggplot(aes(x=nClades, y=Sickle_pairs)) +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Cohort) +
  scale_x_continuous(breaks = seq(0,6,1)) +
  labs(x="Clades per Sample", y="Library Size") # library size is QC-filtered reads
```
  
  Microbial Reads
```{r}
gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  filter(Uncharacterized_Gardnerella==0) %>%
  select(-Gardnerella_vaginalis) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  gather("var", "abundance", clades) %>%
  group_by(SampleID, Cohort) %>%
  summarize(nClades=sum(abundance)) %>%
  left_join(sgDF[,c("SampleID", "bacLoad", "bbmapFiltered_pairs")], by="SampleID") %>%
  filter(bacLoad<2) %>%
  ggplot(aes(x=nClades, y=bbmapFiltered_pairs)) +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Cohort) +
  scale_x_continuous(breaks = seq(0,6,1)) +
  labs(x="Clades per Sample", y="Microbial Reads")
```

## Microbial load by taxa presence-absence
```{r, message=FALSE, warning=FALSE}
bacloadVsTaxWilcox <- mDFtoiClades %>%  
  gather("Taxon", "relativeAbundance", c(clades, Gardnerella_vaginalis, lactos, anaerobes)) %>%
  select(SampleID, Cohort, Taxon, relativeAbundance) %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(relativeAbundance=case_when(relativeAbundance<0.001~"-",
                                      relativeAbundance>=0.001~"+"),
         relativeAbundance=factor(relativeAbundance, levels=c("+", "-")),
         alternative=case_when(Taxon %in% c("Gardnerella_vaginalis", clades, anaerobes)~"greater",
                               Taxon %in% lactos~"less"),
         Taxon=factor(Taxon, levels = c("Gardnerella_vaginalis", clades, lactos, anaerobes), labels = c(abbrevs)),
         Cohort=factor(Cohort, levels = cohorts, labels=c("Stan. Enr.", "UAB Enr.", "MP Enr.", "MOMS-PI")), 
         x=paste0(Cohort, Taxon)) %>%
  filter(x!="UAB Enr.Lg") %>%
  select(-x) %>%
  group_by(alternative) %>%
  nest %>%
  mutate(data=map(data, ~group_by(.x, Cohort, Taxon))) %>%
  mutate(wilcox_test=map2(data, alternative, ~wilcox_test(data=.x, bacLoad~relativeAbundance, alternative=.y))) %>%
  select(wilcox_test) %>%
  unnest %>%
  adjust_pvalue(method = "BH") %>%
  mutate(p.star=case_when(p.adj>0.05~"",
                          p.adj<0.05&p.adj>0.01~"*",
                          p.adj<0.01&p.adj>0.001~"**",
                          p.adj<0.001~"***"))
```

  Plot
```{r, fig.width=7, fig.height=2}
mDFtoiClades %>%  
  gather("Taxon", "relativeAbundance", c(clades, Gardnerella_vaginalis, lactos, anaerobes)) %>%
  select(SampleID, Cohort, Taxon, relativeAbundance) %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(relativeAbundance=case_when(relativeAbundance<0.001~"-",
                                      relativeAbundance>=0.001~"+"),
         Taxon=factor(Taxon, levels = c("Gardnerella_vaginalis", clades, lactos, anaerobes), labels = c(abbrevs)),
         Cohort=factor(Cohort, levels = cohorts, labels=c("Stan. Enr.", "UAB Enr.", "MP Enr.", "MOMS-PI"))) %>%
  ggplot(aes(x=relativeAbundance, y=bacLoad)) +
  geom_violin(aes(color=relativeAbundance, fill=relativeAbundance), alpha=0.5) +
  geom_text(data =bacloadVsTaxWilcox, aes(y=1, x=1.5, label=p.star), color="black", size=4, inherit.aes = FALSE) +
  binaryColors +
  facet_grid(Cohort~Taxon) +
  scale_y_log10() +
  theme(text = element_text(size=15),
        axis.text = element_text(size = 12),
        legend.position = "none") +
  labs(x="", y="Microbial Load")
#ggsave(file.path(figureOut, paste(Sys.Date(), "FigureS6_presAbsVsBacLoad.png", sep = "_")))
```

## Figure 4: Gardnerella and increases in microbial load 
  Microbial load vs. clades per sample and short list microbial load vs. presence-absence 
```{r, message=FALSE, warning=FALSE, fig.width=5, fig.height=1.5}
bacLoadpresAbs <- mDFtoiClades %>%  
  gather("Taxon", "relativeAbundance", c(clades, Gardnerella_vaginalis, lactos, anaerobesSL)) %>%
  select(SampleID, Cohort, Taxon, relativeAbundance) %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(relativeAbundance=case_when(relativeAbundance<0.001~"-",
                                      relativeAbundance>=0.001~"+"),
         Taxon=factor(Taxon, levels = c("Gardnerella_vaginalis", clades, lactos, anaerobesSL), labels = c(abbrevsSL)),
         Cohort=factor(Cohort, levels = cohorts, labels=c("Stan. Enr.", "UAB Enr.", "MP Enr.", "MOMS-PI"))) %>%
  ggplot(aes(x=relativeAbundance, y=bacLoad)) +
  geom_violin(aes(color=relativeAbundance, fill=relativeAbundance), alpha=0.5) +
  geom_text(data=subset(bacloadVsTaxWilcox, Taxon %in% abbrevsSL), aes(y=1, x=1.5, label=p.star), color="black", size=4, inherit.aes = FALSE) +
  binaryColors +
  facet_grid(Cohort~Taxon) +
  scale_y_log10() +
  theme(text = element_text(size=15),
        axis.text = element_text(size = 12),
        strip.text.y = element_text(size=8),
        legend.position = "none") +
  labs(x="", y="Microbial Load")

plot_grid(loadVClades, bacLoadpresAbs, nrow = 1, rel_widths = c(1,2.5), labels = c("A", "B"), label_size = 15)
#ggsave(file.path(figureOut, paste(Sys.Date(), "Figure4_gardAndBacLoad.png", sep="_")))
```

# Gardnerella and the vaginal microbiome
## Relative abundance of common taxa 
  Get the dominant taxon in each sample for organizing -- using clades. Define as the taxon with the greatest relative abundance. If no taxon has a relative abundance greater than 30%, specify as "no type"
```{r, message=FALSE, warning=FALSE}
dominantTax <- gardCladeAb %>%
  spread(Clade, relativeAbundance) %>%
  right_join(mDF, by=c("SampleID", "Cohort")) %>%
  select(-Gardnerella_vaginalis) %>%
  mutate_at(vars(clades), replace_na,  0) %>%
  gather("dominantTax", "abundance", !one_of(c("SampleID", "Cohort"))) %>%
  group_by(SampleID) %>%
  filter(abundance==max(abundance)) %>%
  mutate(dominantTax=case_when(abundance<.3~"No type",
                               abundance>=.3~dominantTax)) %>%
  select(-abundance) %>%
  ungroup

unique(dominantTax$dominantTax)
length(unique(dominantTax$dominantTax))
```
   Create dataframe for order of dominant types to order in barplot figure
```{r}
dominantTaxOrder <- data.frame(dominantTax=c(clades, lactos, "Atopobium_vaginae", "Prevotella_bivia", "Prevotella_amnii", "Prevotella_timonensis", "Mycoplasma_hominis", "Corynebacterium_sp_HFH0082", "Bifidobacterium_longum", "Bacteroides_fragilis", "Staphylococcus_epidermidis", "Lactobacillus_delbrueckii", "Escherichia_coli", "Haemophilus_parainfluenzae", "Lactobacillus_phage_Lv_1", "Alphapapillomavirus_9", "Staphylococcus_aureus", "Lactobacillus_helveticus", "Bradyrhizobium_sp_DFCI_1", "Enterobacter_cloacae", "Alphapapillomavirus_11", "Alphapapillomavirus_14", "Alphapapillomavirus_3", "Alphapapillomavirus_5", "No type"),
                               dominantTaxOrder=c(1:33))
```

  Stacked bar plot
```{r, message=FALSE, warning=FALSE}
# Ordered by dominant taxon 
taxaBarPlot <- mDFtoiClades %>%
  select(SampleID, Cohort, clades, lactos, anaerobesSL) %>%
  filter(SampleID!="6744034", 
         SampleID!="6744034_E") %>% # this sample is 100% uncharacterized gard and therefore cannot be displayed in this figure
  mutate(Other=1-C1-C2-C3-C4-C5-C6-Lactobacillus_crispatus-Lactobacillus_gasseri-Lactobacillus_jensenii-Lactobacillus_iners-Atopobium_vaginae-Mycoplasma_hominis-Prevotella_bivia-Ureaplasma_parvum) %>%
  gather("Taxon", "relativeAbundance", !contains(c("SampleID","Cohort"))) %>%
  mutate(relativeAbundance=na_if(relativeAbundance, 0),
         Taxon=factor(Taxon, levels=c("Other", clades, lactos, anaerobesSL))) %>%
  left_join(dominantTax, by=c("SampleID", "Cohort")) %>%
  left_join(dominantTaxOrder, by="dominantTax") %>%
  group_by(SampleID) %>%
  ggplot(aes(x=reorder(reorder(SampleID, relativeAbundance, na.rm=TRUE), dominantTaxOrder, FUN=min), y=relativeAbundance, color=Taxon, fill=Taxon)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  facet_wrap(~Cohort, scales = "free") +
  labs(x="Sample", y="Relative Abundance") +
  taxColorsOther
taxaBarPlot   
  
# ordered by microbial load
mDFtoiClades %>%
  select(SampleID, Cohort, clades, lactos, anaerobesSL) %>%
  mutate(Other=1-(C1+C2+C3+C4+C5+C6+Lactobacillus_crispatus+Lactobacillus_gasseri+Lactobacillus_jensenii+Lactobacillus_iners+Atopobium_vaginae+Mycoplasma_hominis+Prevotella_bivia+Ureaplasma_parvum)) %>%
  gather("Taxon", "relativeAbundance", !contains(c("SampleID","Cohort"))) %>%
  left_join(bacLoad, by="SampleID") %>%
  mutate(Taxon=factor(Taxon, levels=c("Other", clades, lactos, anaerobesSL)),
         absoluteAbundance=relativeAbundance*bacLoad) %>%
  ggplot(aes(x=reorder(SampleID, absoluteAbundance, FUN=sum), y=relativeAbundance, color=Taxon, fill=Taxon)) +
  geom_bar(stat="identity") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  facet_wrap(~Cohort, scales="free_x") +
  labs(x="Sample", y="Relative Abundance") +
  taxColorsOther
```
## Absolute abundance of common taxa
```{r, message=FALSE, warning=FALSE}
# By absolute abundance
aaBarPlot <- mDFtoiClades %>%
  select(SampleID, Cohort, clades, lactos, anaerobesSL) %>%
  mutate(Other=1-C1-C2-C3-C4-C5-C6-Lactobacillus_crispatus-Lactobacillus_gasseri-Lactobacillus_jensenii-Lactobacillus_iners-Atopobium_vaginae-Mycoplasma_hominis-Prevotella_bivia-Ureaplasma_parvum) %>%
  gather("Taxon", "relativeAbundance", !contains(c("SampleID","Cohort"))) %>%
  left_join(bacLoad, by="SampleID") %>%
  mutate(Taxon=factor(Taxon, levels=c("Other", clades, lactos, anaerobesSL)),
         absoluteAbundance=relativeAbundance*bacLoad) %>%
  left_join(bacLoad, by="SampleID") %>%
  ggplot(aes(x=reorder(SampleID, absoluteAbundance, FUN=sum), y=absoluteAbundance, color=Taxon, fill=Taxon)) +
  geom_bar(stat="identity") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  facet_wrap(~Cohort, scales="free_x") +
  labs(x="Sample", y="Absolute Abundance") +
  taxColorsOther
aaBarPlot
```

```{r, message=FALSE, warning=FALSE}
aaBarPlot +
    coord_cartesian(ylim=c(0,1.5))
```
  What is in the outlier samples?
```{r, message=FALSE, warning=FALSE}
outlierSamples <- bacLoad %>%
  filter(bacLoad > 2) %>%
  .$SampleID
outlierSamples
```
  Get top 5 taxa from each outlier sample
```{r, message=FALSE, warning=FALSE}
mDF %>%
  filter(SampleID %in% outlierSamples) %>%
  gather("Taxon", "Abundance", !contains(c("SampleID", "Cohort"))) %>%
  group_by(SampleID) %>%
  top_n(5) %>%
  arrange(-Abundance) %>%
  split(.$SampleID)
```
  Many have *Gardnerella*, all have *Prevotella timonensis* or *P. amnii*

## Co occurrence with Jaccard distance
### All samples
#### Long List
```{r, message=FALSE, warning=FALSE, fig.height=5, fig.width=6.5}
jaccards <- mDFtoiClades %>%
  select(SampleID, Cohort, Gardnerella_vaginalis, clades, lactos, anaerobes) %>%
  mutate_if(is.numeric, ~case_when(.x>=0.001~1, 
                                   .x<0.001~0)) %>%
  split(.$Cohort) %>%
  map(~column_to_rownames(.x, var="SampleID")) %>%
  map(~select(.x, -Cohort)) %>%
  map(t) %>%
  map(~vegdist(.x, method="jaccard")) %>%
  map(as.matrix)

jaccards$`Stanford Enriched`[lower.tri(jaccards$`Stanford Enriched`, diag = TRUE)] <- NA
jaccards$`UAB Enriched`[lower.tri(jaccards$`UAB Enriched`, diag = TRUE)] <- NA
jaccards$`MOMS-PI Enriched`[lower.tri(jaccards$`MOMS-PI Enriched`, diag = TRUE)] <- NA
jaccards$`MOMS-PI`[lower.tri(jaccards$`MOMS-PI`, diag = TRUE)] <- NA

jaccards %>%
  map(melt) %>%
  map2(names(.), ~mutate(.x, cohort=.y)) %>%
  purrr::reduce(full_join) %>%
  mutate(Var1=factor(Var1, levels=c("Gardnerella_vaginalis", clades, lactos, anaerobes), labels = abbrevs), 
         Var2=factor(Var2, levels=rev(c("Gardnerella_vaginalis", clades, lactos, anaerobes)), labels = rev(abbrevs)),
         cohort=factor(cohort, levels = cohorts)) %>%
  filter(!(Var1=="TG" & Var2 %in% clades),
         !is.na(value)) %>%
  ggplot(aes(x=Var1, y=Var2, fill=value)) +
  geom_tile() +
  scale_x_discrete(drop=FALSE) +
  scale_y_discrete(drop=FALSE) +
  scale_fill_gradient2(low = "#56B4E9", mid = "gray", high = "#D55E00", midpoint = 0.5) +
  theme(axis.text.x = element_text(angle=45, hjust=1), 
        strip.text.x = element_text(size = 13)) +
  facet_wrap(~cohort) +
  labs(x="", y="", fill="Jaccard Distance")
ggsave(filename = file.path(figureOut, paste(Sys.Date(), "FigureS8_jacFourCohort.png", sep="_")), height=6, width=7.5, units="in")
```

## Figure 5: Gardnerella impacts on signatures of the vaginal microbiome
  Taxon Abundances and short list of co-occurrences
```{r, message=FALSE, warning=FALSE, fig.width=5.25, fig.height=2} 
jaccardsSL <- mDFtoiClades %>%
  select(SampleID, Cohort, Gardnerella_vaginalis, clades, lactos, anaerobesSL) %>%
  mutate_if(is.numeric, ~case_when(.x>=0.001~1, 
                                   .x<0.001~0)) %>%
  split(.$Cohort) %>%
  map(~column_to_rownames(.x, var="SampleID")) %>%
  map(~select(.x, -Cohort)) %>%
  map(t) %>%
  map(~vegdist(.x, method="jaccard")) %>%
  map(as.matrix)

jaccardsSL$`Stanford Enriched`[lower.tri(jaccardsSL$`Stanford Enriched`, diag = TRUE)] <- NA
jaccardsSL$`UAB Enriched`[lower.tri(jaccardsSL$`UAB Enriched`, diag = TRUE)] <- NA
jaccardsSL$`MOMS-PI Enriched`[lower.tri(jaccardsSL$`MOMS-PI Enriched`, diag = TRUE)] <- NA
jaccardsSL$`MOMS-PI`[lower.tri(jaccardsSL$`MOMS-PI`, diag = TRUE)] <- NA

jaccardFigure <- jaccardsSL %>%
  map(melt) %>%
  map2(names(.), ~mutate(.x, cohort=.y)) %>%
  purrr::reduce(full_join) %>%
  mutate(Var1=factor(Var1, levels=c("Gardnerella_vaginalis", clades, lactos, anaerobesSL), labels = abbrevsSL), 
         Var2=factor(Var2, levels=rev(c("Gardnerella_vaginalis", clades, lactos, anaerobesSL)), labels = rev(abbrevsSL)),
         cohort=factor(cohort, levels=cohorts)) %>%
  filter(!(Var1=="TG" & Var2 %in% clades),
         !is.na(value)) %>%
  ggplot(aes(x=Var1, y=Var2, fill=value)) +
  geom_tile() +
  scale_x_discrete(drop=FALSE) +
  scale_y_discrete(drop=FALSE) +
  scale_fill_gradient2(low = "#56B4E9", mid = "gray", high = "#D55E00", midpoint = 0.5) +
  theme(axis.text.x = element_text(angle=45, hjust=1, size=7), 
        axis.text.y = element_text(size=7)) +
  facet_wrap(~cohort) +
  coord_fixed() +
  labs(x="", y="", fill="Jaccard Distance")

plot_grid(taxaBarPlot, jaccardFigure, nrow=1, labels = c("A", "B"), label_size = 15)
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "Figure5_microbiomeAndJaccard.png", sep="_")))
```

# Gardnerella variants by PTB status
## Clade abundance by PTB status
### Relative Abundance
```{r, message=FALSE, warning=FALSE}
#plot showing clade relative abundance in term vs. preterm birth in all cohorts
allCohortsCladeRelativePTBplot <- gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre", "bacLoad")], by="SampleID") %>%
  filter(SampleID %!in% unchGardSamples, !is.na(TermPre), var!="Gardnerella_vaginalis", var!="Uncharacterized_Gardnerella") %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, abundance=mean(relativeAbundance)) %>%
  mutate(var=factor(var, levels = clades, labels = clades),
         Cohort=factor(Cohort, levels=cohorts, labels=cohortAbbrevs)) %>%
  ggplot(aes(x=TermPre, y=abundance, color=TermPre, fill=TermPre)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  binaryColors +
  facet_grid(Cohort~var, scales="free_y") +
  theme(legend.position = "none") +
  labs(x=NULL, y="Relative Abundance")

allCohortsCladeRelativePTBplot
```

```{r, message=FALSE, warning=FALSE}
# #remove samples with uncharacterized Gardnerella
# unchGardSamples <- gardCladeAbU %>%
#   filter(var=="Uncharacterized_Gardnerella", relativeAbundance>0) %>%
#   .$SampleID
```

  1. Compute differences between means for each subject and transform
```{r, message=FALSE, warning=FALSE}
# create function to mutate dataframe to create differences and then observe transformations
meanDifferenceTransformations <- function(inputDF, vars){

  #get list of pairs
  pairs <- combn(vars, 2) %>%
    t %>%
    as.data.frame %>%
    mutate(pairs=paste(V1, V2, sep="-")) %>%
    .$pairs
  
  # calculate differences and transformations of values
  diffs_trans <- inputDF %>%
    mutate(meanAbs=set_names(meanAbs, var)) %>%
    nest(diffs=c(var, meanAbs)) %>%
    mutate(diffs=map(diffs, ~abs(outer(.x$meanAbs, .x$meanAbs, `-`)))) %>% #compute differences
    mutate(diffs=map(diffs, data.frame)) %>%
    mutate(diffs=map(diffs, ~rownames_to_column(.x, var="var1"))) %>%
    mutate(diffs=map(diffs, ~gather(.x, "var2", "diff", vars))) %>%
    mutate(diffs=map(diffs, ~transmute(.x, pair=paste(var1, var2, sep="-"), diff=diff))) %>%
    mutate(diffs=map(diffs, ~filter(.x, pair %in% pairs))) %>% # filter repeats
    mutate(diffs=map(diffs, ~mutate(.x, sqrt=sqrt(diff), #square root transformation
                                          frthrt=diff^(1/4), #fourth root transformation
                                          log=log10(diff + 1e-5)))) %>% #log10 transformation with pseudocount
    unnest
  
  # plot histograms
  plots <- diffs_trans %>%
    gather("Transformation", "Value", c(diff, sqrt, frthrt, log)) %>%
    mutate(Transformation=factor(Transformation, levels=c("diff", "sqrt", "frthrt", "log"), labels=c("Difference", "Square Root", "Fourth Root", "log10 with Pseudocount"))) %>%
    ggplot(aes(x=Value)) +
    geom_histogram() +
    facet_wrap(~Transformation, scales="free")
  
  if(vars==clades){
  # perform shapiro-wilk test for normality. null hypothesis is normality.
  shapiro_results <- map(list(Difference=diffs_trans$diff, sqrt=diffs_trans$sqrt, fourthrt=diffs_trans$frthrt, log=diffs_trans$log), shapiro_test) %>% 
      map2(., names(.), ~mutate(.x, transformation=.y)) %>%
      purrr::reduce(full_join) %>%
      select(transformation, statistic, p.value)
  return(list(diffs_trans=diffs_trans, plots=plots, shapiro_results=shapiro_results))
  }
  
  if(vars==genomos){
    return(list(diffs_trans=diffs_trans, plots=plots))
  }

 }
```
  Assess and choose transformation
```{r, message=FALSE, warning=FALSE}
cladeRelativeTransfs <- gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre")], by="SampleID") %>%
  filter(!is.na(TermPre), var!="Uncharacterized_Gardnerella", var!="Gardnerella_vaginalis") %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, meanAbs=mean(relativeAbundance)) %>%
  meanDifferenceTransformations(., clades)

cladeRelativeTransfs$plots

cladeRelativeTransfs$shapiro_results %>%
  formattable()
```

  Choose fourth root transformation

  2. Perform logistic regression on mean differences to see if they vary by term or preterm birth
  Null hypothesis is that coefficients = 0
  Model: $y_{ij}=\beta_{0i} + x_{ij}\beta_{1ij}+e{ij}$
```{r, message=FALSE, warning=FALSE}
# function for performing logistic regression on mean differences of clade abundances
meanDiffLogisticRegression <- function(inputDF, transformation){
  # get cohort Ns
  coh_ns <- gardCladeAbU %>%
    left_join(sgDF[,c("SampleID", "SubjectID", "TermPre")], by="SampleID") %>%
    filter(!is.na(TermPre), var!="Uncharacterized_Gardnerella", var!="Gardnerella_vaginalis") %>%
    group_by(Cohort) %>%
    summarize(cohort_n=n_distinct(SubjectID))
  
  # perform logitic regressions tests
  logiOuts <- inputDF %>%
    gather("transformation", "difference", c(diff, sqrt, frthrt, log)) %>% # filter to choose values with appropriate transformation
    filter(transformation==transformation) %>%
    select(Cohort, TermPre, pair, difference) %>%
    mutate(TermPre=case_when(TermPre=="T"~0,
                             TermPre=="P"~1)) %>%
    split(list(.$Cohort, .$pair)) %>%
    map(~glm(data=.x, TermPre~difference, family="binomial")) %>%
    map(tidy) %>% 
    map2(., names(.), ~mutate(.x, Cohort_Pair=.y)) %>%
    map(~separate(.x, Cohort_Pair, c("Cohort", "Pair"), sep="\\.")) %>%
    purrr::reduce(full_join, by = c("term", "estimate", "std.error", "statistic", "p.value", "Cohort", "Pair")) %>%
    left_join(coh_ns, by="Cohort") %>%
    mutate(std.dev=std.error*sqrt(cohort_n))
  
  logiPlots <- logiOuts %>%
    filter(term=="difference") %>%
    mutate(Cohort=factor(Cohort, levels=cohorts, labels = cohortAbbrevs),
           Pair=factor(Pair, levels = rev(unique(logiOuts$Pair)))) %>%
    ggplot(aes(y=Pair, x=estimate, xmin=estimate-std.dev, xmax=estimate+std.dev)) +
    geom_pointrange() +
    geom_vline(aes(xintercept=0), linetype=2, color="gray", inherit.aes=FALSE) +
    facet_grid(~Cohort) 
  
  return(list(logiOuts=logiOuts, logiPlots=logiPlots))
}
```

```{r, message=FALSE, warning=FALSE}
cladeRelMeanDiffRegression <- meanDiffLogisticRegression(cladeRelativeTransfs$diffs_trans, "frthrt")

cladeRelMeanDiffRegression$logiOuts %>%
  filter(term=="difference",
         p.value<0.05) %>%
        nrow
```


  Figure S10: All cohorts *Gardnerella* and PTB logistic regression
```{r, message=FALSE, warning=FALSE, fig.height=2, fig.width=5}
plot_grid(allCohortsCladeRelativePTBplot, cladeRelMeanDiffRegression$logiPlots, nrow=1, rel_widths = c(1, 0.75), align = "v", labels = c("A", "B"), label_size = 15)
#ggsave(file.path(figureOut, paste(Sys.Date(), "Figure S10_allCladesLogisticRegressionPTB.png", sep="_")))
```
  
### Absolute Abundance
  Plot abundances
```{r, message=FALSE, warning=FALSE}
#plot showing clade absolute abundance in term vs. preterm birth in all cohorts
allCohortsCladeAbsolutePTBplot <- gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre", "bacLoad")], by="SampleID") %>%
  filter(!is.na(TermPre), var!="Gardnerella_vaginalis", var!="Uncharacterized_Gardnerella", bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, abundance=mean(absoluteAbundance)) %>%
  mutate(var=factor(var, levels = clades, labels = clades),
         Cohort=factor(Cohort, levels=cohorts, labels=cohortAbbrevs)) %>%
  ggplot(aes(x=TermPre, y=abundance, color=TermPre, fill=TermPre)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  binaryColors +
  facet_grid(Cohort~var, scales="free_y") +
  theme(legend.position = "none") +
  labs(x=NULL, y="Absolute Abundance")

allCohortsCladeAbsolutePTBplot
```


  Assess and choose transformation
```{r, message=FALSE, warning=FALSE}
cladeAbsoluteTransfs <- gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre", "bacLoad")], by="SampleID") %>%
  filter(!is.na(TermPre), var!="Gardnerella_vaginalis", var!="Uncharacterized_Gardnerella", bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, meanAbs=mean(absoluteAbundance)) %>%
  meanDifferenceTransformations(., clades)

cladeAbsoluteTransfs$plots

cladeAbsoluteTransfs$shapiro_results %>%
  formattable()
```

  Choose log10 + pseudocount transformation

  Perform logistic regression on mean differences to see if they vary by term or preterm birth
  Null hypothesis is that coefficients = 0
```{r, message=FALSE, warning=FALSE}
cladeAbsMeanDiffRegression <- meanDiffLogisticRegression(cladeAbsoluteTransfs$diffs_trans, "log")

cladeAbsMeanDiffRegression$logiOuts %>%
  filter(term=="difference",
        p.value<0.05) %>%
  nrow
```

```{r, message=FALSE, warning=FALSE, fig.height=2, fig.width=5}
plot_grid(allCohortsCladeAbsolutePTBplot, cladeAbsMeanDiffRegression$logiPlots, nrow=1, rel_widths = c(1, 0.75), align = "v")
```

# Gardnerella and PTB in MOMS-PI study Only
  Due to enrichment in *Gardnerella* in Stanford and UAB clades. The following analyses will focus on the MOMS-PI cohorts.
  Create data frame:
```{r}
momsptbDF <- gardCladeAbU %>%
    left_join(sgDF[,c("SampleID", "SubjectID", "TermPre")], by="SampleID") %>%
    filter(str_detect(Cohort, "MOMS-PI"),
        !is.na(TermPre)) %>%
  mutate(Cohort=as.character(Cohort))
```

## Variant abundance versus PTB
### Clades
  Wilcox Ranked Sum in MOMS-PI cohort ONLY
```{r}
cladeRelativeWilcoxResults <- momsptbDF %>%
  filter(Cohort=="MOMS-PI",
         var!="Uncharacterized_Gardnerella") %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, Relative=mean(relativeAbundance)) %>%
  group_by(var) %>%
  wilcox_test(Relative~TermPre, alternative="greater")
  
cladeRelativeWilcoxResults %>%
  arrange(p) %>%
  formattable()
```
  Wilcoxon rank sum test: Absolute abundance
```{r}
cladeAbsoluteWilcoxResults <- momsptbDF %>%
  left_join(sgDF[,c("SampleID", "bacLoad")], by="SampleID") %>%
  filter(Cohort=="MOMS-PI", bacLoad<2,
         var!="Uncharacterized_Gardnerella") %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, Absolute=mean(absoluteAbundance)) %>%
  group_by(var) %>%
  wilcox_test(Absolute~TermPre, alternative="greater")
  
cladeAbsoluteWilcoxResults %>%
  arrange(p) %>%
  formattable()
```
  Plot
```{r, message=FALSE, warning=FALSE}
# save p values
cladeWilcoxResults <- cladeRelativeWilcoxResults %>%
  full_join(cladeAbsoluteWilcoxResults, by = c("var", ".y.", "group1", "group2", "n1", "n2", "statistic", "p")) %>%
  mutate(p.star=case_when(p>0.05~"",
                          p<0.05&p>0.01~"*",
                          p<0.01&p>0.001~"**",
                          p<0.001~"***"),
         var=factor(var, levels = c(clades,  "Gardnerella_vaginalis"), labels = c(clades, "Total")))
```
  
  Relative Abundance Plot:
```{r, message=FALSE, warning=FALSE}
momspiCladeAbundancePTBPlot <-  momsptbDF %>%
  filter(Cohort=="MOMS-PI",
         var!="Uncharacterized_Gardnerella") %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, abundance=mean(relativeAbundance)) %>%
  mutate(var=factor(var, levels = c(clades,  "Gardnerella_vaginalis"), labels = c(clades, "Total"))) %>%
  ggplot(aes(x=TermPre, y=abundance, color=TermPre, fill=TermPre)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  geom_text(data = cladeWilcoxResults, aes(y=0.6, x=1.5, position="dodge", label=p.star), color="black", size=6, inherit.aes = FALSE) +
  binaryColors +
  facet_grid(.~var) +
  theme(legend.position = "none",
        strip.text.x = element_text(size=13),
        strip.text.y = element_text(size=9),
        axis.title=element_text(size=16)) +
  labs(x=NULL, y="Relative\nAbundance")
momspiCladeAbundancePTBPlot
```
  
  Absolute Abundance Plot:
```{r, message=FALSE, warning=FALSE}
momsptbDF %>%
  filter(Cohort=="MOMS-PI",
         var!="Uncharacterized_Gardnerella") %>%
  left_join(sgDF[,c("SampleID", "bacLoad")], by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, abundance=mean(absoluteAbundance)) %>%
  mutate(var=factor(var, levels = c(clades,  "Gardnerella_vaginalis"), labels = c(clades, "Total"))) %>%
  ggplot(aes(x=TermPre, y=abundance, color=TermPre, fill=TermPre)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  geom_text(data = cladeWilcoxResults, aes(y=0.6, x=1.5, position="dodge", label=p.star), color="black", size=6, inherit.aes = FALSE) +
  binaryColors +
  facet_grid(.~var) +
  theme(legend.position = "none",
        strip.text.x = element_text(size=13),
        strip.text.y = element_text(size=9),
        axis.title=element_text(size=16)) +
  labs(x=NULL, y="Absolute Abundance")
```

### Genomospecies
  Wilcoxon rank sum test for genomospecies, MOMS-PI cohort ONLY
```{r}
# Relative abundance
genomoRelativeWilcoxResults <- gardGenomoAbU %>%
    left_join(sgDF[,c("SampleID", "SubjectID", "TermPre")], by="SampleID") %>%
    filter(Cohort=="MOMS-PI",
           !is.na(TermPre),
           var!="Uncharacterized_Gardnerella") %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, Relative=mean(relativeAbundance)) %>%
  group_by(var) %>%
  wilcox_test(Relative~TermPre, alternative="greater")
  

# Absolute abundance
genomoAbsoluteWilcoxResults <- gardGenomoAbU %>%
    left_join(sgDF[,c("SampleID", "SubjectID", "TermPre")], by="SampleID") %>%
    filter(Cohort=="MOMS-PI",
           !is.na(TermPre),
           var!="Uncharacterized_Gardnerella") %>%
  mutate(Cohort=as.character(Cohort)) %>%
  left_join(sgDF[,c("SampleID", "bacLoad")], by="SampleID") %>%
  filter(Cohort=="MOMS-PI", bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, Absolute=mean(absoluteAbundance)) %>%
  group_by(var) %>%
  wilcox_test(Absolute~TermPre, alternative="greater")
  
genomoRelativeWilcoxResults %>%
  arrange(p) %>%
  formattable

genomoAbsoluteWilcoxResults %>%
  arrange(p) %>%
  formattable
```

```{r}
genomoWilcoxResults <- genomoRelativeWilcoxResults %>%
  full_join(genomoAbsoluteWilcoxResults, by = c("var", ".y.", "group1", "group2", "n1", "n2", "statistic", "p")) %>%
  mutate(p.star=case_when(p>0.05~"",
                          p<0.05&p>0.01~"*",
                          p<0.01&p>0.001~"**",
                          p<0.001~"***"),
         var=factor(var, levels = c(genomosCladeOrder,  "Gardnerella_vaginalis"), labels = c(genomoNamesCladeOrder, "Total")))
```
  
  Relative Abundance Plot:
```{r, message=FALSE, warning=FALSE}
momspiGenomoAbundancePTBPlot <-gardGenomoAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre")], by="SampleID") %>%
  filter(Cohort=="MOMS-PI", !is.na(TermPre),
         var!="Uncharacterized_Gardnerella") %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, abundance=mean(relativeAbundance)) %>%
  mutate(var=factor(var, levels = c(genomosCladeOrder,  "Gardnerella_vaginalis"), labels = c(genomoNamesCladeOrder, "Total"))) %>%
  ggplot(aes(x=TermPre, y=abundance, color=TermPre, fill=TermPre)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  geom_text(data = genomoWilcoxResults, aes(y=0.6, x=1.5, position="dodge", label=p.star), color="black", size=6, inherit.aes = FALSE) +
  binaryColors +
  facet_grid(.~var) +
  theme(legend.position = "none",
        strip.text.x = element_text(size=8),
        strip.text.y = element_text(size = 9),
        axis.title=element_text(size=16)) +
  labs(x=NULL, y="Relative\nAbundance")
momspiGenomoAbundancePTBPlot
```
  
  Absolute Abundance Plot:
```{r}
gardGenomoAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre", "bacLoad")], by="SampleID") %>%
  filter(Cohort=="MOMS-PI", !is.na(TermPre), bacLoad<2,
         var!="Uncharacterized_Gardnerella") %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  with_groups(c(SubjectID, Cohort, TermPre, var), summarize, abundance=mean(absoluteAbundance)) %>%
  mutate(var=factor(var, levels = c(genomosCladeOrder,  "Gardnerella_vaginalis"), labels = c(genomoNamesCladeOrder, "Total"))) %>%
  ggplot(aes(x=TermPre, y=abundance, color=TermPre, fill=TermPre)) +
  geom_quasirandom(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  geom_text(data = genomoWilcoxResults, aes(y=0.6, x=1.5, position="dodge", label=p.star), color="black", size=6, inherit.aes = FALSE) +
  binaryColors +
  facet_grid(.~var) +
  theme(legend.position = "none",
        strip.text.x = element_text(size=8),
        strip.text.y = element_text(size = 9),
        axis.title=element_text(size=16)) +
  labs(x=NULL, y="Absolute Abundance")
```

## Number of clades by PTB
  Mean clades per sample in term vs. preterm birth patients in MOMS-PI Only. 
  Calculate Wilcoxon rank sum test
```{r, message=FALSE, warning=FALSE}
nCladesWilcox <- momsptbDF %>%
  filter(Cohort=="MOMS-PI") %>%
  spread(var, relativeAbundance) %>%
  mutate_at(clades, ~case_when(.x>=0.001~1,
                               .x<0.001~0)) %>%
  mutate(nClades=C1+C2+C3+C4+C5+C6) %>%
  with_groups(c(SubjectID, Cohort, TermPre), summarize, meanNClades=mean(nClades)) %>%
  wilcox_test(meanNClades~TermPre, alternative="greater")

nCladesWilcox %>%
  formattable()
```
  
  Plot
```{r, message=FALSE, warning=FALSE}
momspiCladeCountPlot <- momsptbDF %>%
  filter(Cohort=="MOMS-PI") %>%
  spread(var, relativeAbundance) %>%
  mutate_at(clades, ~case_when(.x>=0.001~1,
                               .x<0.001~0)) %>%
  mutate(nClades=C1+C2+C3+C4+C5+C6) %>%
  with_groups(c(SubjectID, Cohort, TermPre), summarize, meanNClades=mean(nClades)) %>%
  ggplot(aes(x=TermPre, y=meanNClades, color=TermPre)) +
  geom_boxplot(alpha=0, color="black", show.legend=FALSE) +
  geom_beeswarm(alpha=0.5, cex = 3, show.legend = FALSE) +
  ylim(0,7) +
  binaryColors +
  labs(x=NULL, y="Mean Gestational\nClade Count") +
  stat_pvalue_manual(nCladesWilcox, label="p", y.position = 6.7)
momspiCladeCountPlot
```


## Assess mean number of clades of rarefying
```{r}
allUnchGardSamples <- gardCladeAbU %>%
  filter(var=="Uncharacterized_Gardnerella",
         relativeAbundance>0) %>%
  .$SampleID
#write_lines(allUnchGardSamples, "/Users/hlberman/Desktop/unchGardSamples.txt")

sgDF %>%
  mutate(isUnch=case_when(SampleID %in% allUnchGardSamples~TRUE,
                          SampleID %!in% allUnchGardSamples~FALSE)) %>%
  ggplot(aes(x=isUnch, y=bbmapFiltered_pairs, color=Cohort, shape=Cohort)) +
  geom_boxplot(alpha=0, position = position_dodge(width=1)) +
  geom_quasirandom(alpha=0.5, dodge.width = 1) +
  geom_hline(yintercept = 1e05, linetype=2, color="black") +
  geom_hline(yintercept = 16334, linetype=2, color="gray") +
  scale_y_log10() +
  labs(x="Uncharacterized Gardnerella", y="Microbial Reads")
```

```{r}
minReads <- min(sgDF$bbmapFiltered_pairs) 
minReads
```

  Rarefy to 100,000 reads per sample

  Assess read counts and get samples with 100,000 reads
```{r, message=FALSE, warning=FALSE}
rarefiedReadCountsDF <- "./rarefied_gard_counts/rarefiedReadCounts.csv" %>%
  read_csv 

k100Samples <- rarefiedReadCountsDF %>%
  filter(rarefiedReads==100000) %>%
  .$SampleID
```


  Import MetaPhlAn tables
```{r, message=FALSE, warning=FALSE}
momspiptbsamples <- momsptbDF$SampleID %>%
  str_extract(., pattern="[0-9]+") %>%
  unique %>%
  paste0("SRR", .)
  
# rarefiedMetaphlanOuts <- "/Users/hlberman/mnt/cluster/home6/hlberman/VMMG/rarefied_gard_counts/metaphlan2_outs" %>%
#   list.files(full.names = TRUE) %>%
#   .[str_detect(., pattern = paste(momspiptbsamples, collapse = "|"))] %>%
#   map(read_tsv)
# 
# rarefiedGardProportions0 <- rarefiedMetaphlanOuts %>%
#   map(~filter(.x, str_detect(`#SampleID`, pattern="Gardnerella"))) %>%
#   map(~filter(.x, str_detect(`#SampleID`, pattern="s__"))) %>%
#   map(~filter(.x, !str_detect(`#SampleID`, pattern="t__"))) %>%
#   set_names(., nm = momspiptbsamples) %>%
#   map(~dplyr::rename(.x, Species=`#SampleID`, Gardnerella_vaginalis=Metaphlan2_Analysis)) %>%
#   map2(., names(.), ~mutate(.x, SampleID=.y)) %>%
#   purrr::reduce(full_join) %>%
#   select(SampleID, Gardnerella_vaginalis)

# rarefiedGardProportions <- data.frame(SampleID=momspiptbsamples) %>%
#   full_join(rarefiedGardProportions0) %>%
#   mutate(SampleID=str_extract(SampleID, pattern="[0-9]+")) %>%
#   replace_na(list(Gardnerella_vaginalis=0))

# write_tsv(rarefiedGardProportions, "./rarefied_gard_counts/gardRelativeAbundance.tsv")

rarefiedGardProportions <- "./rarefied_gard_counts/gardRelativeAbundance.tsv" %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID))
```

  Proportion of clades
```{r, message=FALSE, warning=FALSE}
rarefiedCladeRelativeAbundance <- "./rarefied_gard_counts/gardCladeRelativeAbundance.tsv" %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID)) %>%
  select(-propClade) %>%
  filter(n>=2) %>% #count as present if >=2 reads are mapped to that clade
  group_by(SampleID) %>%
  mutate(propClade=n/sum(n)) %>%
  ungroup() %>%
  select(-n)
```

```{r, message=FALSE, warning=FALSE}
rarefiedGardDF <- rarefiedCladeRelativeAbundance %>%
  full_join(rarefiedGardProportions) %>%
  mutate(relativeAbundance=propClade*Gardnerella_vaginalis) %>%
  select(-propClade) %>%
  spread(Clade, relativeAbundance) %>%
  select(-`<NA>`) %>%
  replace_na(list(C1=0, C2=0, C3=0, C4=0, C5=0, C6=0)) %>%
  right_join(sgDF[,c("SampleID", "SubjectID", "Cohort", "TermPre")]) %>%
  filter(Cohort=="MOMS-PI", !is.na(TermPre), SampleID %in% k100Samples) %>% 
  mutate(Uncharacterized_Gardnerella=case_when(Gardnerella_vaginalis>0&(C1+C2+C3+C4+C5+C6==0)~Gardnerella_vaginalis,
                                               Gardnerella_vaginalis==0~0,
                                               Gardnerella_vaginalis>0&(C1+C2+C3+C4+C5+C6>0)~0))
```

  Perform Wilcoxon Rank Sum test
```{r, message=FALSE, warning=FALSE}
rarefiedNCladesWilcox <- rarefiedGardDF %>%
  #filter(Uncharacterized_Gardnerella==0) %>%
  mutate_at(clades, ~case_when(.x>=0.001~1,
                               .x<0.001~0)) %>%
  mutate(nClades=C1+C2+C3+C4+C5+C6) %>%
  with_groups(c(SubjectID, Cohort, TermPre), summarize, meanNClades=mean(nClades)) %>%
  wilcox_test(meanNClades~TermPre, alternative = "greater")

rarefiedNCladesWilcox %>%
  formattable()  
```

  Figure S9: Mean gestational clade richness vs. PTB
```{r, message=FALSE, warning=FALSE}
rarefiedGardDF %>%
  select(-Gardnerella_vaginalis, -Uncharacterized_Gardnerella) %>%
  gather("Clade", "Abundance", c(C1, C2, C3, C4, C5, C6)) %>%
  mutate(Abundance=case_when(Abundance>0.001~1,
                             Abundance<0.001~0)) %>%
  with_groups(c(SampleID, SubjectID, Cohort, TermPre), summarize, nClades=sum(Abundance)) %>%
  with_groups(c(SubjectID, Cohort, TermPre), summarize, meanNClades=mean(nClades)) %>%
  ggplot(aes(x=TermPre, y=meanNClades, color=TermPre)) +
  geom_boxplot(alpha=0, color="black", show.legend=FALSE) +
  geom_beeswarm(alpha=0.5, cex = 3, show.legend = FALSE) +
  ylim(0,7) +
  binaryColors +
  labs(x=NULL, y="Mean Gestational\nClade Count") +
  stat_pvalue_manual(rarefiedNCladesWilcox, label="p", y.position = 6.7)
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "FigureS9_rarefiedCladeRichnessvsPTB.png", sep="_")))
```

## Microbial load by PTB
  Perform Wilcoxon Rank Sum tests
```{r, message=FALSE, warning=FALSE}
bacloadWilcox <- momsptbDF %>%
  filter(Cohort=="MOMS-PI") %>%
  left_join(sgDF[,c("SampleID", "bacLoad")], by="SampleID") %>%
  filter(Cohort=="MOMS-PI", bacLoad<2) %>%
  with_groups(c(SubjectID, Cohort, TermPre), summarize, meanBacLoad=mean(bacLoad)) %>%
  wilcox_test(meanBacLoad~TermPre, alternative="greater")

bacloadWilcox %>%
  formattable
```

  Plot results
```{r}
momspiBacLoadPTBplot <- momsptbDF %>%
  filter(Cohort=="MOMS-PI") %>%
  left_join(sgDF[,c("SampleID", "bacLoad")], by="SampleID") %>%
  filter(Cohort=="MOMS-PI", bacLoad<2) %>%
  with_groups(c(SubjectID, Cohort, TermPre), summarize, meanBacLoad=mean(bacLoad)) %>%
  ggplot(aes(x=TermPre, color=TermPre, y=meanBacLoad)) +
  geom_boxplot(alpha=0, color="black", show.legend=FALSE) +
  geom_beeswarm(alpha=0.5, cex = 2.1, show.legend = FALSE) +
  ylim(0,0.75) +
  binaryColors +
  labs(x=NULL, y="Mean Gestational\nMicrobial Load") +
  stat_pvalue_manual(bacloadWilcox, label="p", y.position = 0.7)
```

## Figure 6: MOMS-PI Cohort *Gardnerella* and PTB
```{r, message=FALSE, warning=FALSE, fig.height=3, fig.width=5}
momspiPTBAbundancePlots <- plot_grid(momspiCladeAbundancePTBPlot, momspiGenomoAbundancePTBPlot, ncol=1, labels = c("A", "B"), label_size = 15)
momspiCladeCountBacLoadPTBPlots <- plot_grid(momspiCladeCountPlot, momspiBacLoadPTBplot, nrow=1, labels = c("C", "D"), label_size = 15)
plot_grid(momspiPTBAbundancePlots, momspiCladeCountBacLoadPTBPlots, ncol=1, rel_heights = c(1, 0.6))
#ggsave(file.path(figureOut, paste(Sys.Date(), "Figure6_momspiPTB.png", sep="_")))
```

## Library Size by PTB
```{r}
libSizeWilcox <- momsptbDF %>%
  filter(Cohort=="MOMS-PI") %>%
  left_join(sgDF[,c("SampleID", "Sickle_pairs")], by="SampleID") %>%
  with_groups(c(SubjectID, Cohort, TermPre), summarize, meanLibSize=mean(Sickle_pairs)) %>%
  wilcox_test(meanLibSize~TermPre, alternative="greater")

libSizeWilcox %>%
  formattable
```


```{r}
momsptbDF %>%
  filter(Cohort=="MOMS-PI") %>%
  left_join(sgDF[,c("SampleID", "Sickle_pairs")], by="SampleID") %>%
  with_groups(c(SubjectID, Cohort, TermPre), summarize, meanLibSize=mean(Sickle_pairs)) %>%
  ggplot(aes(x=TermPre, color=TermPre, y=meanLibSize)) +
  geom_boxplot(alpha=0, color="black", show.legend=FALSE) +
  geom_beeswarm(alpha=0.5, cex = 2.1, show.legend = FALSE) +
  binaryColors +
  theme(axis.text = element_text(size=13),
        strip.text = element_text(size=13),
        axis.title = element_text(size=16)) +
  labs(x=NULL, y="Mean Subject Library Size") +
  stat_pvalue_manual(libSizeWilcox, label="p", y.position = 3.5e+07)
```

# Gardnerella Clades by Race
  Subject average abundance by self reported race
```{r}
gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "Race")], by="SampleID") %>%
  group_by(SubjectID, Cohort, Race, var) %>%
  summarise(meanRelativeAbundance=mean(relativeAbundance)) %>%
  mutate(var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"), labels = c(clades, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = cohorts, labels=cohortAbbrevs),
         Race=factor(Race, levels=c("Asian", "Black", "White", "Other", "Unknown"))) %>%
  ggplot(aes(x=Race, y=meanRelativeAbundance, color=Race, fill=Race)) +
  geom_point(alpha=0.5, position = position_jitterdodge()) +
  geom_boxplot(color="black", alpha=0, outlier.shape = 0) +
  facet_grid(Cohort~var) +
  scale_x_discrete(labels=c("A", "B", "W", "O")) +
  theme(legend.position = "bottom") +
  labs(x=NULL, y="Relative Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "FigureS11_cladeRelativeAbundanceRace.png", sep="_")))
```
  Subject average absolute abundance by race
```{r}
gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre", "Race", "bacLoad")], by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  group_by(SubjectID, Cohort, Race, var) %>%
  summarise(meanAbsoluteAbundance=mean(absoluteAbundance)) %>%
  mutate(var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"), labels = c(clades, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = cohorts, labels=cohortAbbrevs),
         Race=factor(Race, levels=c("Asian", "Black", "White", "Other", "Unknown"))) %>%
  ggplot(aes(x=Race, y=meanAbsoluteAbundance, color=Race, fill=Race)) +
  geom_point(alpha=0.5, position = position_jitterdodge()) +
  geom_boxplot(color="black", alpha=0, outlier.shape = 0) +
  facet_grid(Cohort~var) +
  scale_x_discrete(labels=c("A", "B", "W", "O")) +
  theme(legend.position = "bottom") +
  labs(x=NULL, y="Absolute Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "cladeAbsoluteAbundanceRace.png", sep="_")))
```


# Session Info
```{r}
sessionInfo()
```