---
title: "Untitled"
author: "Hanna Berman"
date: "January 2021"
output: html_document
---

# Set up
## Load Packages
```{r}
library(tidyverse)
library(kableExtra)
library(broom)
library(formattable)
library(reshape2)
library(vegan)
library(corrplot)
library(ggExtra)
library(broom)
`%!in%` <- negate(`%in%`)
set.seed(7334)
```

## Define file paths
```{r, message=FALSE}
figureOut <- "../FIGURES/V1_manuscript_figures/cohorts_and_analyses"

suDF <- "./sumgsDF.tsv" %>% # metadata for Stanford and UAB
  read_tsv %>%
  mutate(SampleID=as.character(SampleID))

hmp2DF <- "./hmp2mgsDF.tsv" %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID),
         Cohort="HMP2")

# sample data
sgDF0 <- suDF %>%
  select("SampleID", "SubjectID", "Cohort", "sampleTrimester", "GDcoll", "GWdel", "Race", "Ethnicity", "TermPre", "TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs") %>%
  full_join(hmp2DF[,c("SampleID", "SubjectID", "Cohort", "sampleTrimester", "GDcoll", "GWdel", "Race", "Ethnicity", "TermPre", "TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs")], by = c("SampleID", "SubjectID", "Cohort", "sampleTrimester", "GDcoll", "GWdel", "Race", "Ethnicity", "TermPre", "TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs")) %>%
  mutate(Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2")),
         SampleID=as.character(SampleID), 
         SubjectID=as.character(SubjectID),
         pctHuman=((Sickle_pairs-bbmapFiltered_pairs)/Sickle_pairs)*100, # calculate percent human reads
         bacLoad=bbmapFiltered_pairs/(Sickle_pairs-bbmapFiltered_pairs)) # calculate bacterial load
```

```{r}
# # amplicon sequencing data
# PS15 <- "./metadata_files/sampledata_VaginalSwabs_Pregnancy_DiGiulio2015.tsv" %>%
#   read_tsv() %>%
#   dplyr::rename(GDcoll=GDColl) %>%
#   mutate(SubjectID= as.character(SubjectID),
#          Cohort="Stanford",
#          Method="Amplicon",
#          PaperCohort="2015")
# 
# PS17 <- "./metadata_files/MAP_MANUSCRIPT_IL02-IL09_ToBen_20170404_fixed.txt" %>%
#   read_delim(delim="\t") %>%
#   mutate(SubjectID= as.character(SubjectID),
#          Cohort=case_when(UAB_Sample==0~"Stanford",
#                                  UAB_Sample==1~"UAB"),
#          Method="Amplicon",
#          PaperCohort="2017")
```

 **Taxa abundance tables**
 Stanford and UAB cohorts:
```{r}
suMetaphlanPath <- "../humann2/merged_tables/mergedMetaphlanAbundance.tsv"
suGardCladePath <- "./gardCladeRelativeAbundance.tsv"
suGardGenomoPath <- "./gardGenomospeciesRelativeAbundance.tsv"
```
  HMP2 cohort
```{r}
hmp2MetaphlanPath <- "../humann2/merged_tables/hmp2mgs_mergedMetaphlanAbundance.tsv"
hmp2GardCladePath <- "./hmp2gardCladeRelativeAbundance.tsv"
hmp2GardGenomoPath <- "./hmp2gardGenomospeciesRelativeAbundance.tsv"
```


## ggplot settings
  Themes:
```{r}
theme_set(theme_bw()+
          theme(panel.grid = element_blank(),
                axis.text = element_text(size=10),
                axis.title = element_text(size=15),
                legend.text = element_text(size=12),
                legend.title = element_text(size=14)))
```

  Clade colors:
```{r}
cladeColors <-  list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73"), labels=c("C1", "C2", "C3", "C4", "C5", "C6")),
                     scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73"), labels=c("C1", "C2", "C3", "C4", "C5", "C6")))

cladeColorsUnch <-  list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Uncharacterized'~italic(Gardnerella)))),
                     scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Uncharacterized'~italic(Gardnerella)))))

cladeColorsTotal <- list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))),
                         scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))))

cladeColorsTotalUnch <- list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Uncharacterized'~italic(Gardnerella)),  bquote('Total'~italic(Gardnerella)))),
                         scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Uncharacterized'~italic(Gardnerella)),  bquote('Total'~italic(Gardnerella)))))
```

  Gard and Lactos Colors:
```{r}
gLColors <-   list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")),
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred", "darkseagreen4"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")))
```

  All Taxa Colors:
```{r}
taxColors <-   list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", 
                                                 "darkgreen", "khaki1", "darkred", "darkblue",
                                                 "mediumpurple4", "#C42104",  "darkseagreen4", "#2004C2"), 
                                        labels=c("C1", "C2", "C3", "C4", "C5", "C6",
                                                 bquote(italic("L. crispatus")), bquote(italic("L. gasseri")), bquote(italic("L. jensenii")), bquote(italic("L.iners")),
                                                 bquote(italic("A. vaginae")), bquote(italic("M. hominis")), bquote(italic("P. bivia")), bquote(italic("U. parvum"))
                                        )), 
                     scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73",
                                                 "darkgreen", "khaki1", "darkred", "darkblue",
                                                 "mediumpurple4", "#C42104",  "darkseagreen4", "#2004C2"), 
                                        labels=c("C1", "C2", "C3", "C4", "C5", "C6",
                                                 bquote(italic("L. crispatus")), bquote(italic("L. gasseri")), bquote(italic("L. jensenii")), bquote(italic("L.iners")),
                                                 bquote(italic("A. vaginae")), bquote(italic("M. hominis")), bquote(italic("P. bivia")), bquote(italic("U. parvum"))
                                        )))

taxColorsOther <-   list(scale_color_manual(values=c("#808080", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", 
                                                 "darkgreen", "khaki1", "darkred", "darkblue",
                                                 "mediumpurple4", "#C42104",  "darkseagreen4", "#2004C2"), 
                                        labels=c("Other", "C1", "C2", "C3", "C4", "C5", "C6",
                                                 bquote(italic("L. crispatus")), bquote(italic("L. gasseri")), bquote(italic("L. jensenii")), bquote(italic("L.iners")),
                                                 bquote(italic("A. vaginae")), bquote(italic("M. hominis")), bquote(italic("P. bivia")), bquote(italic("U. parvum"))
                                        )), 
                     scale_fill_manual(values=c("#808080", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73",
                                                 "darkgreen", "khaki1", "darkred", "darkblue",
                                                 "mediumpurple4", "#C42104",  "darkseagreen4", "#2004C2"), 
                                        labels=c("Other", "C1", "C2", "C3", "C4", "C5", "C6",
                                                 bquote(italic("L. crispatus")), bquote(italic("L. gasseri")), bquote(italic("L. jensenii")), bquote(italic("L.iners")),
                                                 bquote(italic("A. vaginae")), bquote(italic("M. hominis")), bquote(italic("P. bivia")), bquote(italic("U. parvum"))
                                        )))
```
  
  Bi/Tri/Quad Colors:
```{r}
binaryColors <- list(scale_color_manual(values = c("#D55E00", "#56B4E9")),
                     scale_fill_manual(values=c("#D55E00", "#56B4E9")))

trinaryColors <- list(scale_color_manual(values = c("#D55E00", "#009E73", "#56B4E9")),
                     scale_fill_manual(values=c("#D55E00","#009E73", "#56B4E9")))

quadColors <- list(scale_color_manual(values=c("#D55E00", "#009E73", "#CC79A7", "#56B4E9")),
                   scale_fill_manual(values=c("#D55E00", "#009E73", "#CC79A7", "#56B4E9")))
```

## Other shortcuts
  Create some shortcut lists for convenience.
```{r}
## Other shortcuts
clades <- c("C1", "C2", "C3", "C4", "C5", "C6")
genomos <- c("GS1", "GS2", "GS3", "GS4", "GS5", "GS6", "GS7", "GS8", "GS9", "GS10", "GS11", "GS12", "GS13", "GS14")
genomoNames <- c("G. vag.", "GS2", "GS3", "G. pio.", "G. leo.", "G. swi.", "GS7", "GS8", "GS9", "GS10", "GS11", "GS12", "GS13", "GS14")
lactos <- c("Lactobacillus_crispatus", "Lactobacillus_gasseri", "Lactobacillus_jensenii", "Lactobacillus_iners")
anaerobes <- c("Atopobium_vaginae", "Finegoldia_magna",  "Mycoplasma_hominis", "Megasphaera_genomosp_type_1", "Megasphaera_unclassified", "Prevotella_amnii", "Prevotella_bivia", "Prevotella_timonensis",  "Ureaplasma_parvum")
anaerobesSL <- c("Atopobium_vaginae", "Mycoplasma_hominis", "Prevotella_bivia", "Ureaplasma_parvum") # short list of anaerobes

abbrevs <- c("TG", clades, "Lc", "Lg", "Lj", "Li", "Av", "Fm", "Mh", "Meg1", "Megu", "Pa", "Pb", "Pt", "Up")
abbrevsSL <- c("TG", clades, "Lc", "Lg", "Lj", "Li",  "Av", "Mh", "Pb", "Up")  # short list of abbreviations
```

# Remove low-read samples
  Starting number of samples and subjects:
```{r}
sgDF0 %>%
  group_by(Cohort) %>%
  summarize(nSubjects=n_distinct(SubjectID), nSamples=n_distinct(SampleID)) %>%
  formattable()
```

## Library sizes
  Paired reads in total library size, after quality filtering by Sickle, and after removing human reads.
```{r, fig.height=2, fig.width=4}
sgDF0 %>%
  gather("Step", "paired_reads", c(TotalPairs, Sickle_pairs, bbmapFiltered_pairs)) %>%
  mutate(Step=factor(Step, levels=c("TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs"), labels=c("Library Size", "Sickle Filtered", "Bacterial Reads"))) %>%
  ggplot(aes(x=Step, y=paired_reads, color=Cohort)) +
  geom_boxplot() +
  trinaryColors +
  labs(x=NULL, y="Paired Reads")

sgDF0 %>%
  gather("Step", "paired_reads", c(TotalPairs, Sickle_pairs, bbmapFiltered_pairs)) %>%
  mutate(Step=factor(Step, levels=c("TotalPairs", "Sickle_pairs", "bbmapFiltered_pairs"), labels=c("Library Size", "Sickle Filtered", "Bacterial Reads"))) %>%
  group_by(Step, Cohort) %>%
  summarise(min=min(paired_reads), mean=mean(paired_reads), max=max(paired_reads)) %>%
  mutate_if(is.numeric, ~prettyNum(.x, big.mark = ",")) %>%
  formattable()
```
  Percent filtered at each step
```{r}
sgDF0 %>%
  mutate(sickle_pct_total=Sickle_pairs/TotalPairs,
         bbmap_pct_sickle=bbmapFiltered_pairs/Sickle_pairs,
         bbmap_pct_total=bbmapFiltered_pairs/TotalPairs) %>%
  group_by(Cohort) %>%
  summarise_at(c("sickle_pct_total",  "bbmap_pct_sickle", "bbmap_pct_total"), ~median(.x, na.rm=TRUE)) %>%
  formattable()
```

  sickle_pct_total = percent of paired reads left after sickle filtering
  bbmap_pct_sicke = percent of sickle paired reads left after removing human reads
  bbmap_pct_total = percent of all paired reads left after sickle filtering and removing human reads

  View number of paired reads after sickle and human filtering per sample and determine filtering threshold.
```{r}
p <- sgDF0 %>%
  ggplot(aes(x=Sickle_pairs, y=bbmapFiltered_pairs, fill=Cohort, color=Cohort)) +
  geom_point(alpha=0.25) +
  trinaryColors +
  geom_vline(xintercept = 1e6, color="grey", linetype="dashed") + # 1 million paired reads after sickle
  scale_x_log10(n.breaks=10) +
  scale_y_log10(n.breaks=10) +
  theme_bw() +
  theme(legend.position = "bottom",
        axis.text = element_text(size=10),
        axis.title = element_text(size=15),
        legend.text = element_text(size=12),
        legend.title = element_text(size=14)) +
  labs(x="QC Filtered Paired Reads", y="Bacterial Paired Reads")
#png(filename = file.path(figureOut, paste(Sys.Date(), "readFiltering.png", sep="_")), width=6.75, height=4.5, units="in", res=1080)
ggMarginal(p, groupFill = TRUE, type="density")
#dev.off()
```
  
  Remove the samples with fewer than 1 million paired reads after Sickle filtering and trimming.
```{r}
QCfilterSamples <- sgDF0 %>%
  filter(Sickle_pairs>1e6) %>% # list of samples that pass QC filtering criteria and will be used in analyses
  .$SampleID
length(QCfilterSamples) # number of samples that will be used in analyses

sgDF1 <- sgDF0 %>%
  filter(SampleID %in% QCfilterSamples)
```
  Number of samples remaining in each cohort:
```{r}
sgDF1 %>%
  group_by(Cohort) %>%
  summarize(nSubjects=n_distinct(SubjectID), nSamples=n_distinct(SampleID)) %>%
  formattable()
```

  Will proceed with a total of 888 samples, 62 from Stanford, 45 from UAB, and 781 from HMP2

## Percent Human Reads
```{r}
sgDF1 %>%
  group_by(Cohort) %>%
  ggplot(aes(pctHuman)) +
  geom_histogram(aes(y=stat(width*density))) +
  facet_wrap(~Cohort) +
  labs(x="Percent Human", y="Proportion of Samples")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "percentHumanHist.png", sep="_")))

sgDF1 %>%
  group_by(Cohort) %>%
  summarise_at("pctHuman", list(min=min, mean=mean, median=median, max=max)) %>%
  mutate_if(is.numeric, ~round(.x, 2)) %>%
  formattable()
```

  HMP2 samples had greatest percent of human reads overall. They were not selected for enrichment in *Gardnerella*. Samples in UAB cohort had the smallest percent of human reads.

# Metadata
  Description of cohorts and sample collection schemes
## Stanford and UAB MGS selection figure
  Figure demonstrating how samples from Stanford and UAB cohorts were selected from larger cohorts where amplicon sequencing had been performed
```{r}
#AMP <- rbind(PS15, PS17) %>%
#  unique.data.frame()

# sgDF0 <- sgDF %>%
#   #left_join(AMP[,c("SubjectID", "Cohort", "PaperCohort")], by="SubjectID") %>%
#   filter(SampleID %in% sgDF$SampleID) %>%
#   #mutate(Cohort=replace_na(Cohort,"Stanford")) %>%
#   unique.data.frame()
# 
# 
# SHOT <- sgDF0 %>%
#   select(SubjectID, GDcoll, Cohort, PaperCohort) %>%
#   filter(PaperCohort=="2015"|PaperCohort=="2017") %>%
#   mutate(Method="Shotgun") %>%
#   unique.data.frame()
```

```{r}
# ggplot() +
#   geom_point(data=AMP, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, color=Cohort, size=Method), alpha=.5) +
#    geom_point(data=SHOT, aes(x=GDcoll, y=as.character(SubjectID), shape=Method, size=Method), color="black") +
#   xlim(0,280) +
#      labs(x = "Gestation Days",
#           y = "Subject") +
#      scale_color_manual(values=c("olivedrab", "steelblue", "black"))+
#      scale_shape_manual(values=c(19, 15)) +
#      scale_size_manual(values=c(1,2)) +
#      #theme_classic()+
#      theme(axis.ticks.y = element_blank(),
#      #axis.text.x = element_text(size = 10),
#      axis.text.y  = element_blank()
#      #axis.title.x = element_text(size=20),
#      #axis.title.y = element_text(size=20)
#      )
#      #scale_x_continuous(breaks = c(0,50,100,150,200, 250))
# #ggsave("./collectionPlot.svg", height=8, width=8.5) 
```

## Gestational age at collection
```{r}
sgDF1 %>%
  filter(!is.na(GDcoll)) %>%
  mutate(GWcoll=GDcoll/7) %>%
  group_by(Cohort) %>%
  mutate(GWcoll=floor(GWcoll)) %>%
  summarize(across(GWcoll, list(median=median, `25th`=~quantile(.x, 0.25), `75th`=~quantile(.x, 0.75))))

samplingSched <- sgDF1 %>%
  filter(!is.na(GDcoll)) %>%
  mutate(GWcoll=GDcoll/7) %>%
  split(.$Cohort) %>%
  map2(names(.), ~ggplot(.x) +
    geom_point(aes(x=ceiling(GWcoll), y=SubjectID), size=1) +
    geom_point(aes(x=(GWdel+0.5), y=SubjectID), shape=")") +
    labs(y="Subject ID", x="GW", title = .y))
samplingSched
```

```{r, fig.height=2, fig.width=1}
samplingSched$Stanford
#ggsave(file.path(figureOut,  paste(Sys.Date(), "samplingScheduleStanford.png", sep="_")))
```

```{r, fig.height=2, fig.width=1}
samplingSched$UAB
#ggsave(file.path(figureOut, paste(Sys.Date(), "samplingScheduleUAB.png", sep="_")))
```

```{r, fig.height=2, fig.width=2}
samplingSched$HMP2 +
  theme(axis.text.y = element_text(size=3))
#ggsave(file.path(figureOut, paste(Sys.Date(), "samplingScheduleHMP2.png", sep="_")))
```

## Get HMP2 subjects with gestational age at delivery
```{r}
hmp2SubjectsWDel <- sgDF1 %>%
  filter(Cohort=="HMP2",
         !is.na(TermPre)) %>%
  .$SubjectID %>%
  unique

head(hmp2SubjectsWDel)
length(hmp2SubjectsWDel)
```

## Demographic and Clinical data
```{r}
#GAAD 
gaadDF <- sgDF1 %>%
  select(SubjectID, Cohort, TermPre, GWdel) %>%
  unique() %>%
  add_count(Cohort, TermPre) %>%
  group_by(Cohort, TermPre, n) %>%
  summarize(mean=round(mean(GWdel, na.rm=TRUE), 1), 
            sd=round(sd(GWdel, na.rm = TRUE), 1)) %>%
  replace_na(list(mean="-",
                  sd="-")) %>%
  ungroup() %>%
  mutate(TermPre=case_when(TermPre=="T"~"Term",
                           TermPre=="P"~"Preterm",
                           is.na(TermPre)~"Unknown"),
         cat=paste0(Cohort, "\n", TermPre, " (n=", n, ")"),
         `Mean gestational age in weeks at delivery (±SD)`=paste0(mean, " (", sd, ")")) %>%
  replace_na(list(`Mean gestational age in weeks at delivery (±SD)`="-")) %>%
  select(cat, `Mean gestational age in weeks at delivery (±SD)`) %>%
  column_to_rownames(var="cat") %>%
  t %>%
  as.data.frame %>%
  rownames_to_column(var="Variable") %>%
  select("Variable", "Stanford\nTerm (n=11)", "Stanford\nPreterm (n=9)", "UAB\nTerm (n=5)", "UAB\nPreterm (n=10)", "HMP2\nTerm (n=90)", "HMP2\nPreterm (n=43)", "HMP2\nUnknown (n=98)")
  
demDF <- sgDF1 %>%
  select(SubjectID, Cohort, TermPre, Race, Ethnicity) %>%
  unique() %>%
  add_count(Cohort, TermPre, name="term_count") %>%
  add_count(Cohort, TermPre, Race, name="race_count") %>%
  add_count(Cohort, TermPre, Ethnicity, name="eth_count") %>%
  select(-SubjectID) %>%
  unique %>%
  mutate(TermPre=case_when(TermPre=="T"~"Term",
                           TermPre=="P"~"Preterm",
                           is.na(TermPre)~"Unknown"),
         cat=paste0(Cohort, "\n", TermPre, " (n=", term_count, ")")) %>%
  select(cat, Race, race_count, Ethnicity, eth_count) %>%
  ungroup

# demographics
raceDF <- demDF %>%
  select(cat, Race, race_count) %>%
  unique() %>%
  group_by(cat) %>%
  mutate(pct=round((race_count/sum(race_count))*100),
         race_count=paste0(race_count, " (", pct, "%)"),
         Race=factor(Race, levels=c("Asian", "Black", "White", "Other"))) %>%
  select(-pct) %>%
  spread(cat, race_count) %>%
  mutate_all(~replace_na(.x, 0)) %>%
  mutate_if(is.numeric, ~paste0(.x, " (", round(.x/sum(.x)), "%)")) %>%
  arrange(Race) %>%
  dplyr::rename(Variable=Race) %>%
  select("Variable", "Stanford\nTerm (n=11)", "Stanford\nPreterm (n=9)", "UAB\nTerm (n=5)", "UAB\nPreterm (n=10)", "HMP2\nTerm (n=90)", "HMP2\nPreterm (n=43)", "HMP2\nUnknown (n=98)")

ethDF <- demDF %>%
  select(cat, Ethnicity, eth_count) %>%
  unique() %>%
  group_by(cat) %>%
  mutate(pct=round((eth_count/sum(eth_count))*100),
         eth_count=paste0(eth_count, " (", pct, "%)"),
         Ethnicity=factor(Ethnicity, levels=c("Hispanic", "Non-Hispanic", "Unknown"))) %>%
  select(-pct) %>%
  spread(cat, eth_count) %>%
  mutate_all(~replace_na(.x, 0)) %>%
  mutate_if(is.numeric, ~paste0(.x, " (", round(.x/sum(.x)), "%)")) %>%
  arrange(Ethnicity) %>%
  dplyr::rename(Variable=Ethnicity) %>%
  select("Variable", "Stanford\nTerm (n=11)", "Stanford\nPreterm (n=9)", "UAB\nTerm (n=5)", "UAB\nPreterm (n=10)", "HMP2\nTerm (n=90)", "HMP2\nPreterm (n=43)", "HMP2\nUnknown (n=98)")

demTable <- gaadDF %>%
  rbind(raceDF) %>%
  rbind(ethDF) %>%
  dplyr::rename_at(2:8, ~str_extract(.x, "(?<=\\n).*")) %>%
  kbl(caption = "Table 2. Sampling Cohorts") %>%
  kable_classic(full_width=TRUE, html_font = "Arial") %>%
  add_header_above(c(" ", "Stanford" = 2, "UAB" = 2, "HMP2" = 3)) %>%
  pack_rows("Race", 2, 5,) %>%
  pack_rows("Ethnicity", 6, 8)
demTable
#save_kable(demTable, file = file.path(figureOut, "demTable.png"))
```


# Set up taxa data frames
  Organize and set up tables from MetaPhlAn2 and *Gardnerella* mapping method.
## MetaPhlAn2
  Species relative abundances for Stanford and UAB cohorts:
```{r}
sumDF <- suMetaphlanPath %>%
  read_tsv() %>%
  filter(str_detect(ID, "s__"), !str_detect(ID, "t__")) %>%
  mutate(Species = str_extract(ID, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -ID) %>%
  dplyr::rename_at(2:ncol(.), ~str_extract(.x, "[0-9]{10}")) %>%
  column_to_rownames(var="Species") %>%
  t %>%
  as.data.frame() %>%
  mutate_all(funs(./100)) %>%
  rownames_to_column(var="SampleID")
sumDF[1:6,1:5]
```
 Species relative abundances for HMP2 cohort:
```{r}
hmp2mDF <- hmp2MetaphlanPath %>%
  read_tsv %>%
  filter(str_detect(ID, "s__"), !str_detect(ID, "t__")) %>%
  mutate(Species = str_extract(ID, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -ID) %>%
  as.data.frame() %>%
  dplyr::rename_at(2:ncol(.), ~str_extract(.x, "[0-9]{7}")) %>%
  column_to_rownames(var="Species") %>%
  t %>%
  as.data.frame() %>%
  mutate_all(funs(./100)) %>%
  rownames_to_column(var="SampleID") %>%
  filter(SampleID %in% QCfilterSamples) # keep only samples that pass quality filtering
hmp2mDF[1:6,1:5]
```
  Sanity check that relative abundances sum to 1
```{r}
hmp2mDF %>%
  column_to_rownames("SampleID") %>%
  rowSums %>%
  as.data.frame() %>%
  summary

hmp2mDF %>%
  column_to_rownames("SampleID") %>%
  rowSums %>%
  as.data.frame() %>%
  ggplot(aes(x=`.`)) +
  geom_histogram()
#ggsave("/Users/hlberman/Desktop/test.png")
```
  Relative abundances of most samples sum to 1.
  What is in samples that do not sum to 1
```{r}
hmp2MetaphlanPath %>%
  read_tsv %>%
  dplyr::rename_at(2:ncol(.), ~str_extract(.x, "[0-9]{7}")) %>%
  select(ID, "6744527", "6744289") %>%
  filter(`6744527`>0|`6744289`>0) %>%
  formattable()
```


## *Gardnerella* Mapping
  Add relative abundance of individual clades and genomospecies in Stanford and UAB cohort samples, and check for differences in finding *Gardnerella* by MetaPhlAn2 and mapping method.
  **Stanford and UAB Clades**
```{r}
suGardCladeAb0 <- suGardCladePath %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID)) %>%
  select(-propClade) %>%
  filter(n>=2) %>%
  group_by(SampleID) %>%
  mutate(propClade=n/sum(n)) %>%
  ungroup() %>%
  full_join(sumDF[,c("SampleID", "Gardnerella_vaginalis")], by=c("SampleID")) %>%
  mutate(relativeAbundance=propClade*Gardnerella_vaginalis)

# check if there are any samples where Gard was not found by metaphlan but was found by mapping method
suGardCladeAb0 %>%
  filter(Gardnerella_vaginalis==0, n>1)

#check if there are any samples where Gard was found by metaphlan but not by mapping method
suGardCladeAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
  There is one sample where *Gardnerella* clades were found by mapping method and not by MetaPhlAn2, and no samples where *Gardnerella* was found by MetaPhlAn2 but not by mapping method
    
```{r}
suGardCladeAb <- suGardCladeAb0 %>%
  filter(relativeAbundance>0) %>%
  select(SampleID, Clade, relativeAbundance)
```
  
  **Stanford and UAB Genomospecies**
```{r}
suGardGenomoAb0 <- suGardGenomoPath %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID)) %>%
  select(-propGenomospecies) %>%
  filter(n>=2) %>%
  group_by(SampleID) %>%
  mutate(propGenomospecies=n/sum(n)) %>%
  ungroup() %>%
  full_join(sumDF[,c("SampleID", "Gardnerella_vaginalis")], by=c("SampleID")) %>%
  mutate(relativeAbundance=propGenomospecies*Gardnerella_vaginalis)

# check if there are any samples where Gard was not found by metaphlan but was found by mapping method
suGardGenomoAb0 %>%
  filter(Gardnerella_vaginalis==0, n>1)

#check if there are any samples where Gard was found by metaphlan but not by mapping method
suGardGenomoAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
  No samples where *Gardnerella* genomospecies were found by mapping method but not MetaPhlAn2 and one sample where *Gardnerella* was found by MetaPhlAn2 but not by mapping method.

```{r}
suGardGenomoAb <- suGardGenomoAb0 %>%
  filter(relativeAbundance>0) %>%
  select(SampleID, Genomospecies, relativeAbundance)
```

  **HMP2 Clades**
```{r}
hmp2GardCladeAb0 <- hmp2GardCladePath %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID)) %>%
  select(-propClade) %>%
  filter(n>=2) %>% # 2+ reads need to map to a clade/genomospecies to be considered present
  group_by(SampleID) %>%
  mutate(propClade=n/sum(n)) %>%
  ungroup() %>%
  full_join(hmp2mDF[,c("SampleID", "Gardnerella_vaginalis")], by=c("SampleID")) %>%
  mutate(relativeAbundance=propClade*Gardnerella_vaginalis) %>%
  filter(SampleID %in% QCfilterSamples)

# check if there are any samples where Gard was not found by metaphlan but was found by mapping method
hmp2GardCladeAb0 %>%
  filter(Gardnerella_vaginalis==0, n>1)

#check if there are any samples where Gard was found by metaphlan but not by mapping method
hmp2GardCladeAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
  There are 58 samples where *Gardnerella*  was found by MetaPhlan2 but no clades not by mapping method, and 9 samples where *Gardnerella* was found by mapping method but not by MetaPhlan2

```{r}
hmp2GardCladeAb <- hmp2GardCladeAb0 %>%
  filter(relativeAbundance>0) %>%
  select(SampleID, Clade, relativeAbundance)
```

  **HMP2 Genomospecies**
```{r}
hmp2GardGenomoAb0 <- hmp2GardGenomoPath %>%
  read_tsv %>%
  mutate(SampleID=as.character(SampleID)) %>%
  select(-propGenomospecies) %>%
  filter(n>=2) %>% # 2+ reads mapped to each clade/genomospecies to be considered present
  group_by(SampleID) %>%
  mutate(propGenomospecies=n/sum(n)) %>%
  ungroup() %>%
  full_join(hmp2mDF[,c("SampleID", "Gardnerella_vaginalis")], by=c("SampleID")) %>%
  mutate(relativeAbundance=propGenomospecies*Gardnerella_vaginalis)

# check if there are any samples where Gard was not found by metaphlan but was found by mapping method
hmp2GardGenomoAb0 %>%
  filter(Gardnerella_vaginalis==0, n>1)

#check if there are any samples where Gard was found by metaphlan but not by mapping method
hmp2GardGenomoAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
  127 samples where *Gardnerella* was found by MetaPhlan2 but no genomospecies were found by mapping method and 2 samples where *Gardnerella* was found by mapping method but not by MetaPhlan2

```{r}
hmp2GardGenomoAb <- hmp2GardGenomoAb0 %>%
  filter(relativeAbundance>0) %>%
  select(SampleID, Genomospecies, relativeAbundance)
```

### Further assess uncharacterized samples:
  Asses samples where *Gardnerella* was uncharacterized at the clade level. (Where *Gardnerella* was found by MetaPhlAn2 but no clades were identified using the mapping method).
```{r}
unchGardSamples <- hmp2GardCladeAb0 %>%
  filter(is.na(n), Gardnerella_vaginalis>0) %>%
  .$SampleID
```

  Comparison of number of bacterial paired reads and proportion *Gardnerella* in samples with uncharacterized *Gardnerella* vs. samples where *Gardnerella* was found with both methods.
```{r}
q <- hmp2mDF %>%
  select(SampleID, Gardnerella_vaginalis) %>%
  mutate(unchGard=case_when(SampleID %in% unchGardSamples~TRUE,
                            !(SampleID %in% unchGardSamples)~FALSE)) %>%
  left_join(sgDF1[,c("SampleID", "bbmapFiltered_pairs", "bacLoad")], by="SampleID") %>%
  ggplot(aes(x=Gardnerella_vaginalis, y=bbmapFiltered_pairs, color=unchGard, shape=unchGard)) +
  geom_point(alpha=0.5) +
  theme(legend.position = "bottom") +
  binaryColors +
  labs(x="Proportion Gardnerella", y="Bacterial Reads")

#png(filename = file.path(figureOut, paste(Sys.Date(), "unchGard.png", sep="_")), width=6.75, height=4.5, units="in", res=1080)
ggMarginal(q, groupFill = TRUE, type="density")
#dev.off()
```
  Samples with uncharacterized *Gardnerella* have fewer bacterial reads and a lower proportion of *Gardnerella*

## Join abundance tables
```{r, message=FALSE}
# Gardnerella abundance
gardCladeAb0 <- full_join(suGardCladeAb, hmp2GardCladeAb, by = c("SampleID", "Clade", "relativeAbundance"))
gardGenomoAb0 <- full_join(suGardGenomoAb, hmp2GardGenomoAb, by = c("SampleID", "Genomospecies", "relativeAbundance"))

# metaphlan
mDF0 <- sumDF %>%
  full_join(hmp2mDF)

#number of speacies that appear in only one vs. both samples
testOverlap <- summarise_all(mDF0, anyNA) %>%
  gather("Species", "anyNA", 2:ncol(.)) %>%
  select(-SampleID)

nrow(testOverlap)
table(testOverlap$anyNA)
```
  364 total species and 209 appear in both studies (Stanford/UAB vs. HMP2).
  Number of species in each study:
```{r}
ncol(sumDF)-1
ncol(hmp2mDF)-1
```
  
  Replace NAs with zeros for abundance values for species that did not appear in each respective study. 
```{r}
mDF1 <- mDF0 %>%
  mutate_if(is.numeric, ~replace_na(.x, 0)) %>%
  left_join(sgDF1[,c("SampleID", "Cohort")], by="SampleID") %>%
  select(SampleID, Cohort, everything())
anyNA(mDF1) # check for remaining NAs
```

# Gardnerella variants
  Proportion *Gardnerella* in samples by cohort:
```{r}
mDF1 %>%
  select(SampleID, Cohort, Gardnerella_vaginalis) %>%
  ggplot(aes(x=Gardnerella_vaginalis, color=Cohort, fill=Cohort, group=Cohort)) +
  geom_density(alpha=0.5) +
  trinaryColors +
  labs(x="Proportion Gardnerella")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "samplePropGardbyCohort.png", sep="_")))

mDF1 %>%
  select(SampleID, Cohort, Gardnerella_vaginalis) %>%
  group_by(Cohort) %>%
  summarize(min=min(Gardnerella_vaginalis),
            `1stQu`=summary(Gardnerella_vaginalis)[2],
            median=median(Gardnerella_vaginalis),
            mean=mean(Gardnerella_vaginalis),
            `3rdQu`=summary(Gardnerella_vaginalis)[5],
            max=max(Gardnerella_vaginalis)) %>%
  formattable()
```
    Proportion *Gardnerella* in samples by study:
```{r}
mDF1 %>%
  select(SampleID, Cohort, Gardnerella_vaginalis) %>%
  mutate(Cohort=case_when(Cohort=="Stanford"|Cohort=="UAB"~"Stanford|UAB",
                          Cohort=="HMP2"~"HMP2"),
         Cohort=factor(Cohort, levels = c("Stanford|UAB", "HMP2"))) %>%
  ggplot(aes(x=Gardnerella_vaginalis, color=Cohort, fill=Cohort, group=Cohort)) +
  geom_density(alpha=0.5) +
  binaryColors +
  labs(x="Proportion Gardnerella")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "samplePropGardbyStudy.png", sep="_")))

gardAbundance2Cohort <- mDF1 %>%
  select(SampleID, Cohort, Gardnerella_vaginalis) %>%
  left_join(sgDF1[,c("SampleID", "SubjectID")], by="SampleID") %>%
  mutate(Cohort=case_when(Cohort=="Stanford"|Cohort=="UAB"~"Stanford|UAB",
                          Cohort=="HMP2"~"HMP2"),
         Cohort=factor(Cohort, levels = c("Stanford|UAB", "HMP2")))
gardAbundance2Cohort %>%
  group_by(Cohort) %>%
  summarize(min=min(Gardnerella_vaginalis),
            `1stQu`=summary(Gardnerella_vaginalis)[2],
            median=median(Gardnerella_vaginalis),
            mean=mean(Gardnerella_vaginalis),
            `3rdQu`=summary(Gardnerella_vaginalis)[5],
            max=max(Gardnerella_vaginalis)) %>%
  formattable()
```

  Proportion *Gardnerella* by as subject averages by study:
```{r}
subjectGardAbundance2Cohort <- gardAbundance2Cohort %>%
  group_by(SubjectID, Cohort) %>%
  summarize(Gardnerella_vaginalis=mean(Gardnerella_vaginalis)) %>%
  ungroup
  
subjectGardAbundance2Cohort %>%
  ggplot(aes(x=Gardnerella_vaginalis, color=Cohort, fill=Cohort)) +
  geom_density(alpha=0.5) +
  binaryColors +
  theme(legend.position = "left") +
  labs(x="Subject Average Proportion Gardnerella")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "subjectPropGardbyStudy.png", sep="_")))
```

## Select enriched *Gardnerella* HMP2 subset
  MGS sequenced samples from Stanford and UAB were selected based on enrichment of *Gardnerella* at the subject level. Create a cohort of a subset of samples in the HMP2 cohort that are also selected for enrichment of *Gardnerella* at the subject level for better comparison.
```{r}
ntiles <- subjectGardAbundance2Cohort %>%
  filter(Cohort=="Stanford|UAB") %>%
  mutate(decile=ntile(Gardnerella_vaginalis, 3)) %>%
  group_by(decile) %>%
  filter(Gardnerella_vaginalis==min(Gardnerella_vaginalis)) %>%
  .$Gardnerella_vaginalis %>%
  sort %>%
  unique
ntiles

# subjectGardAbundance2Cohort0 <- subjectGardAbundance2Cohort %>%
#   filter(Cohort=="HMP2",
#          SubjectID %in% hmp2SubjectsWDel) %>% # *** For selecting subjects that have gestational age at delivery  ***
#   mutate(Ntile=case_when(between(Gardnerella_vaginalis, ntiles[1], ntiles[2])~1,
#                         between(Gardnerella_vaginalis, ntiles[2], ntiles[3])~2,
#                         Gardnerella_vaginalis > ntiles[3]~3))
# 
# subjN <- min(count(subjectGardAbundance2Cohort0, Ntile)$n) # number of subjects from each ntile to take
# 
# HMP2highGardSubjects <- subjectGardAbundance2Cohort0 %>% 
#   group_by(Ntile) %>%
#   sample_n(subjN, replace = FALSE) %>%
#   .$SubjectID
#  
# write_lines(HMP2highGardSubjects, "./HMP2highGardSubjects.txt")
HMP2highGardSubjects <- read_lines("./HMP2highGardSubjects.txt")

subjectGardAbundance2Cohort %>%
  filter(Cohort=="Stanford|UAB"|(Cohort=="HMP2"&SubjectID %in% HMP2highGardSubjects)) %>%
  mutate(Cohort=factor(Cohort, levels=c("Stanford|UAB", "HMP2"), labels=c("Stanford|UAB", "HMP2 Enriched"))) %>%
  ggplot(aes(x=Gardnerella_vaginalis, color=Cohort, fill=Cohort, group=Cohort)) +
  geom_density(alpha=0.5) +
  binaryColors +
  labs(x="Subject Average Proportion Gardnerella")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "enrichedSubjectPropGardbyStudy.png", sep="_")))
```
  Distribution of *Gardnerella* abundnace in samples
```{r}
HMP2HighGardSamples <- sgDF1 %>%
  filter(SubjectID %in% HMP2highGardSubjects) %>%
  .$SampleID

gardAbundance2Cohort %>%
  filter(Cohort=="Stanford|UAB"|Cohort=="HMP2"&SubjectID %in% HMP2highGardSubjects) %>%
  ggplot(aes(x=Gardnerella_vaginalis, color=Cohort, fill=Cohort)) +
  geom_density(alpha=0.5) +
  binaryColors +
  labs(x="Proportion of Gardnerella")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "enrichedSamplePropGardbyStudy.png", sep="_")))
```

  **Add enriched HMP2 cohort to tables**
  Sample and subject metadata:
```{r}
# extract enriched Gardnerella subset and full join to metadata 
highGardSgDF <- sgDF1 %>%
  filter(SubjectID %in% HMP2highGardSubjects) %>% 
  mutate(Cohort="HMP2 Enriched",
         SampleID=paste0(SampleID, "_E"),
         SubjectID=paste0(SubjectID, "_E"))

nrow(highGardSgDF)  

sgDF <- sgDF1 %>%
  mutate(Cohort=as.vector(Cohort),
         SubjectID=as.character(SubjectID)) %>%
  full_join(highGardSgDF, by=colnames(sgDF1)) %>%
  mutate(Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2")))

bacLoad <- sgDF %>%
  select(SampleID, bacLoad)

sgDF %>%
  group_by(Cohort) %>%
  summarize(nSubjects=n_distinct(SubjectID), nSamples=n_distinct(SampleID)) %>%
  formattable()
```

  MetaPhlAn2 species abundance tables:
```{r}
highGardmDF <- mDF1 %>%
  filter(SampleID %in% HMP2HighGardSamples) %>%
  mutate(Cohort="HMP2 Enriched",
         SampleID=paste0(SampleID, "_E"))

mDF <- mDF1 %>%
  mutate(Cohort=as.vector(Cohort)) %>%
  rbind(highGardmDF) %>%
  mutate(Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2")))
```
  *Gardnerella* clade and genomospecies abundance tables:
```{r}
# Clade Abundances
gardCladeAb1 <- gardCladeAb0 %>%
  left_join(sgDF1[,c("SampleID", "Cohort")], by="SampleID") %>%
  mutate(Cohort=as.vector(Cohort))
 
highGardCladeAb <- gardCladeAb1 %>%
  filter(SampleID %in% HMP2HighGardSamples) %>%
  mutate(Cohort="HMP2 Enriched",
         SampleID=paste0(SampleID, "_E"))

gardCladeAb <- gardCladeAb1 %>%
  rbind(highGardCladeAb) %>%
  mutate(Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2")))

# Gemomospecies Abundances
gardGenomoAb1 <- gardGenomoAb0 %>%
  left_join(sgDF1[,c("SampleID", "Cohort")], by="SampleID") %>%
  mutate(Cohort=as.vector(Cohort))
 
highGardGenomoAb <- gardGenomoAb1 %>%
  filter(SampleID %in% HMP2HighGardSamples) %>%
  mutate(Cohort="HMP2 Enriched", 
         SampleID=paste0(SampleID, "_E"))

gardGenomoAb <- gardGenomoAb1 %>%
  rbind(highGardGenomoAb) %>%
  mutate(Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2")))

# note these tables contain samples with at least one identified clade or genomospecies of gardnerella
```


# Create sample subsets
  Create subsets of one sample per subject.
  First observe sampling over trimesters 
```{r}
sgDF %>%
  ggplot(aes(x=sampleTrimester, y=SubjectID, color=Cohort)) +
  geom_jitter(alpha=0.5) +
  theme(axis.text.y = element_blank()) +
  quadColors
```
  Select one sample per subject from the second trimester
```{r}
# sampleSubset <- sgDF %>%
#   filter(sampleTrimester=="Second") %>%
#   group_by(SubjectID) %>%
#   sample_n(size=1) %>%
#   ungroup %>%
#   .$SampleID
#save list of samples in subset 
# write_lines(sampleSubset, "./sampleSubsets.txt")
sampleSubset <- read_lines("./sampleSubsets.txt")

#number of samples per cohort in subset
sgDF %>%
  filter(SampleID %in% sampleSubset) %>%
  count(Cohort) %>%
  formattable()
```


# Describe bacterial load and *Gardnerella* variant abundance and bacterial load
## Clades per sample
```{r}
gardCladeAbU <- gardCladeAb %>%
  full_join(mDF[,c("SampleID", "Cohort", "Gardnerella_vaginalis")], by=c("SampleID","Cohort")) %>%
  spread(Clade, relativeAbundance) %>%
  select(-`<NA>`) %>% # remove "NA column" created by samples that have zero Gardnerella
  mutate(Uncharacterized_Gardnerella=case_when(Gardnerella_vaginalis>=0&is.na(C1)&is.na(C2)&is.na(C3)&is.na(C4)&is.na(C5)&is.na(C6)~Gardnerella_vaginalis,
                                               Gardnerella_vaginalis>0&!is.na(clades)~0)) %>%
  gather("var", "relativeAbundance", c(clades, Gardnerella_vaginalis, Uncharacterized_Gardnerella)) %>%
  replace_na(list(relativeAbundance=0))
  
gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nClades=C1+C2+C3+C4+C5+C6,
         nClades=as.character(nClades),
         nClades=case_when(Uncharacterized_Gardnerella!="0"~"U",
                            Uncharacterized_Gardnerella=="0"~nClades)) %>%
  count(Cohort, nClades) %>%
  group_by(Cohort) %>%
  mutate(n=n/sum(n),
         isUnch=case_when(nClades=="U"~TRUE,
                          nClades!="U"~FALSE),
         nClades=case_when(nClades=="U"~"0",
                          nClades!="U"~nClades)) %>%
  ggplot(aes(x=nClades, y=n, fill=isUnch)) +
  geom_bar(stat="identity") +
  facet_wrap(~Cohort) +
  scale_fill_manual(values = c("black", "grey")) +
  labs(x="Clades per Sample", y="Proportion of Samples", fill=bquote('Uncharacterized'~italic(Gardnerella))) +
  theme(legend.position = "bottom")
#ggsave(file.path(figureOut, paste(Sys.Date(), "cladesPerSample.png", sep="_")))
```
  More clades per sample in UAB and enriched HMP2 cohorts than Stanford and HMP2 cohorts. 
  
  Number of samples that have uncharacterized Gardnerella per cohort
```{r}
gardCladeAbU %>%
  filter(var=="Uncharacterized_Gardnerella") %>%
  filter(relativeAbundance>0) %>%
  count(Cohort) %>%
  formattable()
```

## Genomospecies per sample
```{r, message=FALSE}
gardGenomoAbU <- gardGenomoAb %>%
  spread(Genomospecies, relativeAbundance) %>%
  full_join(mDF[,c("SampleID", "Cohort", "Gardnerella_vaginalis")], by=c("SampleID", "Cohort")) %>%
  mutate(Uncharacterized_Gardnerella=case_when(Gardnerella_vaginalis>=0&is.na(GS1)&is.na(GS2)&is.na(GS3)&is.na(GS4)&is.na(GS5)&is.na(GS6)&is.na(GS7)&is.na(GS8)&is.na(GS9)&is.na(GS10)&is.na(GS11)&is.na(GS12)&is.na(GS13)&is.na(GS14)~Gardnerella_vaginalis,
                                               Gardnerella_vaginalis>0&!is.na(genomos)~0)) %>%
  gather("var", "relativeAbundance", c(genomos, Gardnerella_vaginalis, Uncharacterized_Gardnerella)) %>%
  replace_na(list(relativeAbundance=0))
  
gardGenomoAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_at(genomos, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nGenomos=GS1+GS2+GS3+GS4+GS5+GS6+GS7+GS8+GS9+GS10+GS11+GS12+GS13+GS14,
         nGenomos=as.character(nGenomos),
         nGenomos=case_when(Uncharacterized_Gardnerella!="0"~"U",
                           Uncharacterized_Gardnerella=="0"~nGenomos)) %>%
  count(Cohort, nGenomos) %>%
  group_by(Cohort) %>%
  mutate(n=n/sum(n),
         isUnch=case_when(nGenomos=="U"~TRUE,
                          nGenomos!="U"~FALSE),
         nGenomos=case_when(nGenomos=="U"~"0",
                            nGenomos!="U"~nGenomos),
         nGenomos=factor(nGenomos, c("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14"))) %>% # convert to fraction of samples
  ggplot(aes(x=nGenomos, y=n, fill=isUnch)) +
  geom_bar(stat="identity") +
  facet_wrap(~Cohort) +
  scale_fill_manual(values = c("black", "grey")) +
  labs(x="Genomospecies per Sample", y="Proportion of Samples") +
  theme(legend.position = "none")
#ggsave(file.path(figureOut, paste(Sys.Date(), "genomospeciesPerSample.png", sep = "_")))
```
  More genomospecies per sample in UAB cohort.
  
  Number of samples that have uncharacterized Gardnerella per cohort
```{r}
gardGenomoAbU %>%
  filter(var=="Uncharacterized_Gardnerella") %>%
  filter(relativeAbundance>0) %>%
  count(Cohort)
```


## Variant Relative abundance
### Clades
```{r, fig.width=7, fig.height=3.5}
gardCladeAbU %>%
  mutate(var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"), labels=c(clades, "Unch.", "Total")),
         Cohort=factor(Cohort, levels=c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels = c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  #ggplot(aes(x=Cohort, y=relativeAbundance, color=var, fill=var)) +
  ggplot(aes(x=var, y=relativeAbundance, color=var, fill=var)) +
  geom_jitter(alpha=0.5, size=2) +
  geom_boxplot(color="black", alpha=0, outlier.shape = 0) +
  cladeColorsTotalUnch +
  #facet_grid(.~var) +
  facet_grid(Cohort~.) +
  theme(legend.position = "none",
        axis.text.x = element_text(size = 16),
        #axis.text.x = element_text(angle=45, hjust=1, size=16),
        axis.text.y = element_text(size=16),
        axis.title.y = element_text(size=23),
        strip.text = element_text(size=17)) +
  labs(x=NULL, y="Relative Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "cladeRelativeAbundanceByCohort.png", sep="_")))
```
  Mean abundance per cohort:
```{r, fig.width=6, fig.height=3}
gardCladeAbU %>%
  group_by(Cohort, var) %>%
  summarise(avgAbundance=mean(relativeAbundance)) %>%
  mutate(var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"), labels=c(clades, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels=c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  ungroup %>%
  #ggplot(aes(x=Cohort, y=avgAbundance, color=var, fill=var)) +
  ggplot(aes(x=var, fill=var, y=avgAbundance)) +
  geom_bar(stat="identity") +
  cladeColorsTotalUnch +
  #facet_wrap(~var,nrow=1) +
  facet_grid(Cohort~.) +
  #scale_y_continuous(limits = c(0,0.6), n.breaks = 7) +
  scale_y_continuous(limits=c(0,0.6), n.breaks = 4) +
  theme(legend.position = "none",
        #axis.text.x = element_text(angle=45, hjust=1, size=13),
        axis.text.x = element_text(size = 13),
        axis.text.y = element_text(size=13),
        axis.title.y = element_text(size=20),
        strip.text = element_text(size=13)) +
  labs(x=NULL, y="Mean Relative Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "cladeCohortAvgAb.png", sep = "_")))
```
  
  Clades differ in abundance and abundance of clades varies across cohorts. Clade 3 is less abundant in Stanford cohort but abundant in other cohorts, especially UAB. Clade 1 is more abundant in Stanford. Clade 6 is rare in Stanford but appears in UAB and HMP2 cohorts.  

  As stacked bar plot
```{r, fig.height=4, fig.width=3}
gardCladeAbU %>%
  dplyr::rename(Clade=var) %>%
  filter(Clade!="Gardnerella_vaginalis") %>%
  group_by(Cohort, Clade) %>%
  summarise(avgAbundance=mean(relativeAbundance)) %>%
  ggplot(aes(x=Cohort, y=avgAbundance, color=Clade, fill=Clade)) +
  geom_col(width=0.5) +
  cladeColorsUnch +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  labs(x=NULL, y="Mean Relative Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "meanRelAbundanceBarplot.png", sep="_")))
```

  Prevalence vs relative abundance of *Gardnerella* clades
```{r}
prevTable <- gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  mutate_if(is.numeric, ~case_when(.x>0~1,
                                   .x==0~0)) %>%
  group_by(Cohort) %>%
  summarize_if(is.numeric, ~sum(.x)/n()) %>%
  gather("Clade", "prevalence", c(clades, "Uncharacterized_Gardnerella", "Gardnerella_vaginalis"))

prevTable %>%
  mutate(prevalence=round(prevalence, 2)) %>%
  arrange(Cohort) %>%
  formattable

gardCladeAbU %>%
  dplyr::rename(Clade=var) %>%
  group_by(Cohort, Clade) %>%
  summarise(avgAbundance=mean(relativeAbundance)) %>%
  left_join(prevTable, by=c("Cohort", "Clade")) %>%
  mutate(Clade=factor(Clade, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"))) %>%
  ggplot(aes(x=prevalence, y=avgAbundance, color=Clade)) +
  geom_point(size=3) +
  scale_x_continuous(limits=c(0,1), breaks=seq(0,1,0.2)) +
  scale_y_continuous(limits =c(0,1), breaks=seq(0,1,0.2)) +
  facet_wrap(~Cohort) +
  cladeColorsTotalUnch +
  coord_fixed() +
  theme(legend.position = "left") +
  labs(x="Prevalence", y="Mean Relative Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "meanRelAbundanceVsPrevalence.png", sep="_")))
```

### Genomospecies
```{r, fig.height=3, fig.width=7}
gardGenomoAbU %>%
  mutate(var=factor(var, levels=c(genomos, "Uncharacterized_Gardnerella", "Gardnerella_vaginalis"), labels=c(genomos, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels = c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  #ggplot(aes(x=Cohort, y=relativeAbundance, color=var, fill=var)) +
  ggplot(aes(x=var, y=relativeAbundance, color=var, fill=var)) +
  geom_jitter(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  #facet_wrap(~var, nrow = 1) +
  facet_grid(Cohort~.) +
  theme(#axis.text.x = element_text(angle=45, hjust=1, size=11),
        axis.text.x = element_text(size=11),
        axis.text.y = element_text(size=12),
        axis.title.y = element_text(size=20),
        strip.text = element_text(size = 11),
        legend.position="none") +
  labs(x=NULL, y="Relative Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "genomoRelativeAbundanceByCohort.png", sep = "_")))
```
 Mean abundance per cohort
```{r, fig.height=3, fig.width=6}
gardGenomoAbU %>%
  group_by(Cohort, var) %>%
  summarise(avgAbundance=mean(relativeAbundance)) %>%
  mutate(var=factor(var, levels=c(genomos, "Uncharacterized_Gardnerella", "Gardnerella_vaginalis"), labels=c(genomoNames, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels = c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  #ggplot(aes(x=Cohort, y=avgAbundance, color=var, fill=var)) +
  ggplot(aes(x=var, y=avgAbundance, color=var, fill=var)) +
  geom_bar(stat = "identity") +
  #facet_wrap(~var, nrow = 1) +
  facet_grid(Cohort~.) +
  scale_y_continuous(limits=c(0,0.6), n.breaks = 4) +
  theme(#axis.text.x = element_text(angle=45, hjust=1, size=10),
        axis.text.x = element_text(size=10),
        axis.text.y = element_text(size=10),
        legend.position="none") +
  labs(x=NULL, y="Mean Relative Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "genomoCohortAvgAb.png", sep = "_")))
```
## Bacterial Load
  Compare bacterial load across the four cohorts:
```{r}
# density plot with outliers
sgDF %>%
  mutate(Cohort2=case_when(Cohort=="Stanford"|Cohort=="UAB"~"Stanford & UAB",
                           Cohort=="HMP2"|Cohort=="HMP2 Enriched"~"HMP2")) %>%
  ggplot(aes(x=bacLoad, fill=Cohort, color=Cohort)) +
  geom_density(alpha=0.5) +
  facet_grid(Cohort2~., scales = "free") +
  quadColors +
  #scale_x_log10() +  
  labs(x="Bacterial Load")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "bacLoadDensOutLog.png", sep="_")))

# table of descriptive statistics with outliers
sgDF %>%
  group_by(Cohort) %>%
  summarise_at("bacLoad", list(min=min, mean=mean, median=median, max=max)) %>%
  mutate_if(is.numeric, ~round(.x, 4)) %>%
  formattable()

# bacterial load of outliers
sgDF %>%
  filter(bacLoad>2) %>%
  select(SampleID, Cohort, bacLoad) %>%
  formattable

# density plot without outliers (bacload ≥ 2)
sgDF %>%
  filter(bacLoad<2) %>%
  mutate(Cohort2=case_when(Cohort=="Stanford"|Cohort=="UAB"~"Stanford & UAB",
                           Cohort=="HMP2"|Cohort=="HMP2 Enriched"~"HMP2"),
         Cohort2=factor(Cohort2, levels=c("Stanford & UAB", "HMP2"))) %>%
  ggplot(aes(x=bacLoad, fill=Cohort, color=Cohort)) +
  geom_density(alpha=0.5) +
  facet_grid(Cohort2~., scales="free") +
  quadColors +
  theme(legend.position = "bottom") +
  labs(x="Bacterial Load")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "bacLoadDensNoOut.png", sep="_")))


# table of descriptive statistics without outliers
sgDF %>%
  filter(bacLoad<2) %>%
  group_by(Cohort) %>%
  summarise_at("bacLoad", list(min=min, mean=mean, median=median, max=max)) %>%
  mutate_if(is.numeric, ~round(.x, 4)) %>%
  formattable()
```

## Absolute abundance
### Clade aboslute abundance
```{r, fig.width=7, fig.height=3.5}
gardCladeAbU %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad, 
         var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella", "Gardnerella_vaginalis"), labels = c(clades, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels=c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  #ggplot(aes(x=Cohort, y=absoluteAbundance, color=var)) +
  ggplot(aes(x=var, y=absoluteAbundance, color=var)) +
  geom_jitter(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  cladeColorsTotalUnch +
  scale_y_continuous(limits=c(0,1), n.breaks = 5) +
  #facet_wrap(~var, nrow = 1) +
  facet_grid(Cohort~.) +
  theme(legend.position = "none",
        #axis.text.x = element_text(angle=45, hjust=1, size=16),
        axis.text.x = element_text(size=16),
        axis.text.y = element_text(size=16),
        axis.title.y = element_text(size=23),
        strip.text = element_text(size=17)) +
  labs(x=NULL, y="Absolute Abundance")
#ggsave(file.path(figureOut, file.path(paste(Sys.Date(), "cladeAbsoluteAbundanceByCohort.png", sep="_"))))
```
  Clades 3 and 4 appear to show the greatest absolute abundance.
  
```{r, fig.width=6, fig.height=3}
gardCladeAbU %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad, 
         var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella", "Gardnerella_vaginalis"), labels = c(clades, "Unch.", "Total")),
         Cohort=factor(Cohort, levels=c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels = c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  group_by(Cohort,var) %>%
  summarize(avgAbundance=mean(absoluteAbundance)) %>%
  #ggplot(aes(x=Cohort, y=avgAbundance, fill=var, color=var)) +
  ggplot(aes(x=var, y=avgAbundance, fill=var, color=var)) +
  geom_bar(stat = "identity") +
  cladeColorsTotalUnch +
  #facet_wrap(~var, nrow = 1) +
  facet_grid(Cohort~.) +
  theme(legend.position = "none",
        #axis.text.x = element_text(angle=45, hjust=1, size=13),
        axis.title.x = element_text(size = 13),
        axis.text.y = element_text(size=13),
        axis.title.y = element_text(size=20),
        strip.text = element_text(size=13)) +
  labs(x=NULL, y="Mean Absolute Abundance")
#ggsave(file.path(figureOut, file.path(paste(Sys.Date(), "cladeAbsoluteAbundanceCohortAvg.png", sep="_"))))
```

### Genomospecies aboslute abundance
```{r, fig.height=3, fig.width=7}
gardGenomoAbU %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad, 
         var=factor(var, levels = c(genomos, "Uncharacterized_Gardnerella", "Gardnerella_vaginalis"), labels = c(genomoNames, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels = c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  #ggplot(aes(x=Cohort, y=absoluteAbundance, color=var)) +
  ggplot(aes(x=var, y=absoluteAbundance, color=var)) +
  geom_jitter(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  #facet_wrap(~var, nrow = 1) +
  facet_grid(Cohort~.) +  
  theme(#axis.text.x = element_text(angle=45, hjust=1, size=11),
        axis.text.x = element_text(size=11),
        axis.text.y = element_text(size=12),
        axis.title.y = element_text(size=20),
        strip.text = element_text(size = 11),
        legend.position="none") +
  labs(x=NULL, y="Absolute Abundance")
#ggsave(file.path(figureOut, file.path(paste(Sys.Date(), "genomoAbsoluteAbundanceByCohort.png", sep="_"))))
```
  Genomospecies *G. swidsinskii* and genomospecies 7 appear to show the greatest absolute abundance.
  
```{r, fig.height=3, fig.width=7}
gardGenomoAbU %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad,
         var=factor(var, levels=c(genomos, "Uncharacterized_Gardnerella", "Gardnerella_vaginalis"), labels=c(genomoNames, "Unch.", "Total")),
         Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels = c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  group_by(Cohort,var) %>%
  summarize(avgAbundance=mean(absoluteAbundance)) %>%
  #ggplot(aes(x=Cohort, y=avgAbundance, color=var, fill=var)) +
  ggplot(aes(x=var, y=avgAbundance, color=var, fill=var)) +
  geom_bar(stat = "identity") +
  #facet_wrap(~var, nrow = 1) +
  facet_grid(Cohort~.) +
  theme(#axis.text.x = element_text(angle=45, hjust=1, size=10),
        axis.text.x = element_text(size=10),
        axis.text.y = element_text(size=10),
        legend.position="none") +
  labs(x=NULL, y="Mean Absolute Abundance")
#ggsave(file.path(figureOut, file.path(paste(Sys.Date(), "genomoAbsoluteAbundanceCohortAvg.png", sep="_"))))
```

## Bacterial load by number of clades
```{r}
gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  filter(Uncharacterized_Gardnerella==0) %>%
  select(-Gardnerella_vaginalis) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  gather("var", "abundance", clades) %>%
  group_by(SampleID, Cohort) %>%
  summarize(nClades=sum(abundance)) %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  ggplot(aes(x=nClades, y=bacLoad)) +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Cohort) +
  scale_x_continuous(breaks = seq(0,6,1)) +
  labs(x="Clades per Sample", y="Bacterial Load")
#ggsave(file.path(figureOut, file.path(paste(Sys.Date(), "bacLoadByNoClades.png", sep="_"))))
```
  Is this a function of sequencing depth?
```{r}
gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  filter(Uncharacterized_Gardnerella==0) %>%
  select(-Gardnerella_vaginalis) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  gather("var", "abundance", clades) %>%
  group_by(SampleID, Cohort) %>%
  summarize(nClades=sum(abundance)) %>%
  left_join(sgDF[,c("SampleID", "bacLoad", "Sickle_pairs")], by="SampleID") %>%
  filter(bacLoad<2) %>%
  ggplot(aes(x=nClades, y=Sickle_pairs)) +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Cohort) +
  scale_x_continuous(breaks = seq(0,6,1)) +
  labs(x="Clades per Sample", y="Library Size") # library size is QC-filtered reads
#ggsave(file.path(figureOut, file.path(paste(Sys.Date(), "libSizeByNoClades.png", sep="_"))))
```
  Bacterial Reads
```{r}
gardCladeAbU %>%
  spread(var, relativeAbundance) %>%
  filter(Uncharacterized_Gardnerella==0) %>%
  select(-Gardnerella_vaginalis) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  gather("var", "abundance", clades) %>%
  group_by(SampleID, Cohort) %>%
  summarize(nClades=sum(abundance)) %>%
  left_join(sgDF[,c("SampleID", "bacLoad", "bbmapFiltered_pairs")], by="SampleID") %>%
  filter(bacLoad<2) %>%
  ggplot(aes(x=nClades, y=bbmapFiltered_pairs)) +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Cohort) +
  scale_x_continuous(breaks = seq(0,6,1)) +
  labs(x="Clades per Sample", y="Bacterial Reads")
#ggsave(file.path(figureOut, file.path(paste(Sys.Date(), "bacReadsByNoClades.png", sep="_"))))
```

# Observe communities
##  Top taxa
```{r}
avgTop20Abundance0 <- mDF %>%
  select(SampleID, Cohort, everything()) %>%
  gather("Species", "Abundance", 3:ncol(.)) %>%
  group_by(Cohort, Species) %>%
  summarise(avgAbundance=mean(Abundance)) 


top20taxa <- avgTop20Abundance0 %>%
  group_by(Cohort) %>%
  top_n(20, avgAbundance) %>%
  .$Species %>%
  unique

avgTop20Abundance0 %>%
  spread(Cohort, avgAbundance) %>%
  mutate(stanfordRank=rank(-Stanford),
         uabRank=rank(-UAB), 
         hmp2Rank=rank(-HMP2),
         hmp2eRank=rank(-`HMP2 Enriched`),
         Stanford=round(Stanford, 2),
         UAB=round(UAB,2),
         HMP2=round(HMP2, 2),
         `HMP2 Enriched`=round(`HMP2 Enriched`, 2)) %>%
  filter(Species %in% top20taxa) %>%
  select(Species, Stanford, stanfordRank, UAB, uabRank, `HMP2 Enriched`, hmp2eRank, HMP2, hmp2Rank) %>%
  arrange(stanfordRank) %>%
  formattable()
  
x <- mDF %>%
  select(SampleID, Cohort, top20taxa) %>%
  mutate_if(is.numeric, ~case_when(.x>0~1,
                                   .x==0~0))%>%
  group_by(Cohort) %>%
  summarize_if(is.numeric, ~sum(.x)/n()) %>%
  data.frame()
rownames(x) <- x$Cohort
x %>%
  select(-Cohort) %>%
  t %>%
  as.data.frame() %>%
  rownames_to_column("Species") %>%
  arrange(-Stanford) %>%
  formattable()
```

## Relative abundance of common taxa
  First create abundance table with taxa of interest
```{r}
# abundance table of taxa of interest with clades
mDFtoiClades <- gardCladeAb %>%
  spread(Clade, relativeAbundance) %>%
  full_join(mDF, by=c("SampleID", "Cohort")) %>%
  select(SampleID, Cohort, Gardnerella_vaginalis, clades, lactos, anaerobes) %>% # keep total Gardnerella abundance in table for now. Many operations are taxon vs. taxon. Will need to remove this column when analyses call for it.
  mutate_at(vars(c(Gardnerella_vaginalis, clades, lactos, anaerobes)), replace_na, 0) %>%
  mutate(Cohort=factor(Cohort, levels=c("Stanford", "UAB", "HMP2 Enriched", "HMP2")))
```
  
  Then, get the dominant taxon in each sample for organizing -- using clades. Define as the taxon with the greatest relative abundance. If no taxon has a relative abundance greater than 30%, specify as "no type"
```{r}
dominantTax <- gardCladeAb %>%
  spread(Clade, relativeAbundance) %>%
  right_join(mDF, by=c("SampleID", "Cohort")) %>%
  select(-Gardnerella_vaginalis) %>%
  mutate_at(vars(clades), replace_na,  0) %>%
  gather("dominantTax", "abundance", !one_of(c("SampleID", "Cohort"))) %>%
  group_by(SampleID) %>%
  filter(abundance==max(abundance)) %>%
  mutate(dominantTax=case_when(abundance<.3~"No type",
                               abundance>=.3~dominantTax)) %>%
  select(-abundance) %>%
  ungroup

unique(dominantTax$dominantTax)
length(unique(dominantTax$dominantTax))
```
   Create dataframe for order of dominant types to order in barplot figure
```{r}
dominantTaxOrder <- data.frame(dominantTax=c(clades, lactos, "Atopobium_vaginae", "Prevotella_bivia", "Prevotella_amnii", "Prevotella_timonensis", "Mycoplasma_hominis", "Corynebacterium_sp_HFH0082", "Bifidobacterium_longum", "Bacteroides_fragilis", "Staphylococcus_epidermidis", "Lactobacillus_delbrueckii", "Escherichia_coli", "Haemophilus_parainfluenzae", "Lactobacillus_phage_Lv_1", "Alphapapillomavirus_9", "Staphylococcus_aureus", "Lactobacillus_helveticus", "Bradyrhizobium_sp_DFCI_1", "Enterobacter_cloacae", "Alphapapillomavirus_11", "Alphapapillomavirus_14", "Alphapapillomavirus_3", "Alphapapillomavirus_5", "No type"),
                               dominantTaxOrder=c(1:33))
```

  Stacked bar plot
```{r}
# Ordered by dominant taxon 
mDFtoiClades %>%
  select(SampleID, Cohort, clades, lactos, anaerobesSL) %>%
  filter(SampleID!="6744034", 
         SampleID!="6744034_E") %>% # this sample is 100% uncharacterized gard and therefore cannot be displayed in this figure
  mutate(Other=1-C1-C2-C3-C4-C5-C6-Lactobacillus_crispatus-Lactobacillus_gasseri-Lactobacillus_jensenii-Lactobacillus_iners-Atopobium_vaginae-Mycoplasma_hominis-Prevotella_bivia-Ureaplasma_parvum) %>%
  gather("Taxon", "relativeAbundance", !contains(c("SampleID","Cohort"))) %>%
  mutate(relativeAbundance=na_if(relativeAbundance, 0),
         Taxon=factor(Taxon, levels=c("Other", clades, lactos, anaerobesSL))) %>%
  left_join(dominantTax, by=c("SampleID", "Cohort")) %>%
  left_join(dominantTaxOrder, by="dominantTax") %>%
  group_by(SampleID) %>%
  #ggplot(aes(x=SampleID, y=relativeAbundance, fill=Taxon)) +
  ggplot(aes(x=reorder(reorder(SampleID, relativeAbundance, na.rm=TRUE), dominantTaxOrder, FUN=min), y=relativeAbundance, color=Taxon, fill=Taxon)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_blank()) +
  facet_wrap(~Cohort, scales = "free") +
  labs(x="Sample", y="Relative Abundance") +
  taxColorsOther
#ggsave(file.path(figureOut, file.path(paste(Sys.Date(), "relativeAbundanceBarPlot.png", sep="_"))))
       
# ordered by bacterial load
mDFtoiClades %>%
  select(SampleID, Cohort, clades, lactos, anaerobesSL) %>%
  mutate(Other=1-(C1+C2+C3+C4+C5+C6+Lactobacillus_crispatus+Lactobacillus_gasseri+Lactobacillus_jensenii+Lactobacillus_iners+Atopobium_vaginae+Mycoplasma_hominis+Prevotella_bivia+Ureaplasma_parvum)) %>%
  gather("Taxon", "relativeAbundance", !contains(c("SampleID","Cohort"))) %>%
  left_join(bacLoad, by="SampleID") %>%
  mutate(Taxon=factor(Taxon, levels=c("Other", clades, lactos, anaerobesSL)),
         absoluteAbundance=relativeAbundance*bacLoad) %>%
  ggplot(aes(x=reorder(SampleID, absoluteAbundance, FUN=sum), y=relativeAbundance, color=Taxon, fill=Taxon)) +
  geom_bar(stat="identity") +
  theme(axis.text.x = element_blank()) +
  facet_wrap(~Cohort, scales="free_x") +
  labs(x="Sample", y="Relative Abundance") +
  taxColorsOther
#ggsave(file.path(figureOut, paste(Sys.Date(), "relativeAbundanceBarPlotByBacLoad.png", sep="_")))
```
## Absolute abundance of common taxa
```{r}
#By dominant taxon 
# aaBarPlot <- mDFtoiClades %>%
#   select(SampleID, Cohort, clades, lactos, anaerobesSL) %>%
#   mutate(Other=1-C1-C2-C3-C4-C5-C6-Lactobacillus_crispatus-Lactobacillus_gasseri-Lactobacillus_jensenii-Lactobacillus_iners-Atopobium_vaginae-Mycoplasma_hominis-Prevotella_bivia-Ureaplasma_parvum) %>%
#   gather("Taxon", "relativeAbundance", !contains(c("SampleID","Cohort"))) %>%
#   left_join(bacLoad, by="SampleID") %>%
#   mutate(relativeAbundance=na_if(relativeAbundance, 0),
#          Taxon=factor(Taxon, levels=c("Other", clades, lactos, anaerobes)),
#          absoluteAbundance=relativeAbundance*bacLoad) %>%
#   left_join(dominantTax, by="SampleID") %>%
#   left_join(dominantTaxOrder, by="dominantTax") %>%
#   ggplot(aes(x=reorder(reorder(SampleID, relativeAbundance, FUN=sum, na.rm=TRUE), dominantTaxOrder, FUN=min), y=absoluteAbundance, color=Taxon, fill=Taxon)) +
#   geom_bar(stat="identity") +
#   theme(axis.text.x = element_blank()) +
#   facet_wrap(~Cohort, scales="free_x") +
#   labs(x="Sample", y="Absolute Abundance") +
#   taxColorsOther
# aaBarPlot
# #ggsave(file.path(figureOut, "absoluteAbundanceBarPlot.png"))

# By absolute abundance
aaBarPlot <- mDFtoiClades %>%
  select(SampleID, Cohort, clades, lactos, anaerobesSL) %>%
  mutate(Other=1-C1-C2-C3-C4-C5-C6-Lactobacillus_crispatus-Lactobacillus_gasseri-Lactobacillus_jensenii-Lactobacillus_iners-Atopobium_vaginae-Mycoplasma_hominis-Prevotella_bivia-Ureaplasma_parvum) %>%
  gather("Taxon", "relativeAbundance", !contains(c("SampleID","Cohort"))) %>%
  left_join(bacLoad, by="SampleID") %>%
  mutate(Taxon=factor(Taxon, levels=c("Other", clades, lactos, anaerobesSL)),
         absoluteAbundance=relativeAbundance*bacLoad) %>%
  left_join(bacLoad, by="SampleID") %>%
  ggplot(aes(x=reorder(SampleID, absoluteAbundance, FUN=sum), y=absoluteAbundance, color=Taxon, fill=Taxon)) +
  geom_bar(stat="identity") +
  theme(axis.text.x = element_blank()) +
  facet_wrap(~Cohort, scales="free_x") +
  labs(x="Sample", y="Absolute Abundance") +
  taxColorsOther
aaBarPlot
#ggsave(file.path(figureOut, paste(Sys.Date(), "absoluteAbundanceBarPlotByBacLoad.png", sep="_")))
```

```{r}
aaBarPlot +
    coord_cartesian(ylim=c(0,1.5))
#ggsave(file.path(figureOut, paste(Sys.Date(),"absoluteAbundanceBarPlot_yaxiscutbybacload.png", sep="_")))
```
  What is in the outlier samples?
```{r}
outlierSamples <- bacLoad %>%
  filter(bacLoad > 2) %>%
  .$SampleID
outlierSamples
```
  Sample 1005601348
```{r}
mDF %>%
  filter(SampleID==outlierSamples[1]) %>%
  gather("Taxon", "Abundance", !contains("SampleID")) %>%
  filter(Abundance>=0.01) %>%
  arrange(-Abundance)
```
  Sample 4009835178
```{r}
mDF %>%
  filter(SampleID==outlierSamples[2]) %>%
  gather("Taxon", "Abundance", !contains("SampleID")) %>%
  filter(Abundance>=0.01) %>%
  arrange(-Abundance)
```

# Thresholds for presence-absence
  Determine what threshold will be used to count a taxon as present vs. absent for co-occurrence and presence absence vs. bacterial load analyses
  
  MetaPhlAn2 will report all taxa that have 10% of marker genes non-zero [based on this Google group response from Nicola Segata](https://groups.google.com/g/metaphlan-users/c/0jLFs8v9nI4?pli=1). Average number of marker genes per bacterial species is 184 ± 45. (Truong et al., 2015, *Nature Mathods* "MetaPhlAn2 for enhanced metagenomic taxonomic profiling"). About 1 million markers from >7,500 species. Minimum number of reads to be detected would be 18.4. 
```{r}
sgDF1 %>%
  mutate(ntile=ntile(bbmapFiltered_pairs, n=20)) %>%
  filter(ntile==5) %>%
  filter(bbmapFiltered_pairs==min(bbmapFiltered_pairs)) %>%
  .$bbmapFiltered_pairs

sgDF1 %>%
  mutate(over30K=bbmapFiltered_pairs>100000) %>%
  count(over30K) %>% 
  mutate(freq=n/sum(n))
```
  80% of samples have greater than 100,000 read pairs. MetaPhlAn2 uses about 10% of the reads in the library, and 18.4/10,000 yields a functional limit of detection of roughly 0.1%.

```{r}
18.4/10000 
```

  A threshold of 0.1% has been used frequently in presence-absence analyses in microbiome data and will be used as a threshold here.

# Co occurrence with Jaccard distance

## All samples
### Long List
```{r, fig.height=5, fig.width=6.5}
jaccards <- mDFtoiClades %>%
  select(SampleID, Cohort, Gardnerella_vaginalis, clades, lactos, anaerobes) %>%
  mutate_if(is.numeric, ~case_when(.x>=0.001~1, 
                                   .x<0.001~0)) %>%
  split(.$Cohort) %>%
  map(~column_to_rownames(.x, var="SampleID")) %>%
  map(~select(.x, -Cohort)) %>%
  map(t) %>%
  map(~vegdist(.x, method="jaccard")) %>%
  map(as.matrix)

jaccards$Stanford[lower.tri(jaccards$Stanford, diag = TRUE)] <- NA
jaccards$UAB[lower.tri(jaccards$UAB, diag = TRUE)] <- NA
jaccards$`HMP2 Enriched`[lower.tri(jaccards$`HMP2 Enriched`, diag = TRUE)] <- NA
jaccards$HMP2[lower.tri(jaccards$HMP2, diag = TRUE)] <- NA

jaccards %>%
  map(melt) %>%
  map2(names(.), ~mutate(.x, cohort=.y)) %>%
  purrr::reduce(full_join) %>%
  mutate(Var1=factor(Var1, levels=c("Gardnerella_vaginalis", clades, lactos, anaerobes), labels = abbrevs), 
         Var2=factor(Var2, levels=rev(c("Gardnerella_vaginalis", clades, lactos, anaerobes)), labels = rev(abbrevs)),
         cohort=factor(cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"))) %>%
  filter(!(Var1=="TG" & Var2 %in% clades),
         !is.na(value)) %>%
  ggplot(aes(x=Var1, y=Var2, fill=value)) +
  geom_tile() +
  scale_x_discrete(drop=FALSE) +
  scale_y_discrete(drop=FALSE) +
  scale_fill_gradient2(low = "#56B4E9", mid = "gray", high = "#D55E00", midpoint = 0.5) +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  facet_wrap(~cohort) +
  labs(x="", y="", fill="Jaccard Distance")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "jacFourCohort.png", sep="_")), height=6, width=7.5, units="in")
```

### Short List
```{r} 
#, fig.height=4, fig.width=6.5
jaccards <- mDFtoiClades %>%
  select(SampleID, Cohort, Gardnerella_vaginalis, clades, lactos, anaerobesSL) %>%
  mutate_if(is.numeric, ~case_when(.x>=0.001~1, 
                                   .x<0.001~0)) %>%
  split(.$Cohort) %>%
  map(~column_to_rownames(.x, var="SampleID")) %>%
  map(~select(.x, -Cohort)) %>%
  map(t) %>%
  map(~vegdist(.x, method="jaccard")) %>%
  map(as.matrix)

jaccards$Stanford[lower.tri(jaccards$Stanford, diag = TRUE)] <- NA
jaccards$UAB[lower.tri(jaccards$UAB, diag = TRUE)] <- NA
jaccards$`HMP2 Enriched`[lower.tri(jaccards$`HMP2 Enriched`, diag = TRUE)] <- NA
jaccards$HMP2[lower.tri(jaccards$HMP2, diag = TRUE)] <- NA

jaccards %>%
  map(melt) %>%
  map2(names(.), ~mutate(.x, cohort=.y)) %>%
  purrr::reduce(full_join) %>%
  mutate(Var1=factor(Var1, levels=c("Gardnerella_vaginalis", clades, lactos, anaerobesSL), labels = abbrevsSL), 
         Var2=factor(Var2, levels=rev(c("Gardnerella_vaginalis", clades, lactos, anaerobesSL)), labels = rev(abbrevsSL)),
         cohort=factor(cohort, levels=c("Stanford", "UAB", "HMP2 Enriched", "HMP2"))) %>%
  filter(!(Var1=="TG" & Var2 %in% clades),
         !is.na(value)) %>%
  ggplot(aes(x=Var1, y=Var2, fill=value)) +
  geom_tile() +
  scale_x_discrete(drop=FALSE) +
  scale_y_discrete(drop=FALSE) +
  scale_fill_gradient2(low = "#56B4E9", mid = "gray", high = "#D55E00", midpoint = 0.5) +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  facet_wrap(~cohort) +
  labs(x="", y="", fill="Jaccard Distance")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "jaccFourCohortShortList.png", sep="_")))
```

## Sample Subset
```{r, fig.height=4, fig.width=5}
jaccardsSubset <- mDFtoiClades %>%
  filter(SampleID %in% sampleSubset) %>%
  select(SampleID, Cohort, Gardnerella_vaginalis, clades, lactos, anaerobes) %>%
  mutate_if(is.numeric, ~case_when(.x>=0.001~1, 
                                   .x<0.001~0)) %>%
  split(.$Cohort) %>%
  map(~column_to_rownames(.x, var="SampleID")) %>%
  map(~select(.x, -Cohort)) %>%
  map(t) %>%
  map(~vegdist(.x, method="jaccard")) %>%
  map(as.matrix)

jaccardsSubset$Stanford[lower.tri(jaccardsSubset$Stanford, diag = TRUE)] <- NA
jaccardsSubset$UAB[lower.tri(jaccardsSubset$UAB, diag = TRUE)] <- NA
jaccardsSubset$`HMP2 Enriched`[lower.tri(jaccardsSubset$`HMP2 Enriched`, diag = TRUE)] <- NA
jaccardsSubset$HMP2[lower.tri(jaccardsSubset$HMP2, diag = TRUE)] <- NA

jaccardsSubset %>%
  map(melt) %>%
  map2(names(.), ~mutate(.x, cohort=.y)) %>%
  purrr::reduce(full_join) %>%
  mutate(Var1=factor(Var1, levels=c("Gardnerella_vaginalis", clades, lactos, anaerobes), labels = abbrevs), 
         Var2=factor(Var2, levels=rev(c("Gardnerella_vaginalis", clades, lactos, anaerobes)), labels = rev(abbrevs)),
         cohort=factor(cohort, levels=c("Stanford", "UAB", "HMP2 Enriched", "HMP2"))) %>%
  filter(!(Var1=="TG" & Var2 %in% clades),
         !is.na(value)) %>%
  ggplot(aes(x=Var1, y=Var2, fill=value)) +
  geom_tile() +
  scale_x_discrete(drop=FALSE) +
  scale_y_discrete(drop=FALSE) +
  scale_fill_gradient2(low = "#56B4E9", mid = "gray", high = "#D55E00", midpoint = 0.5) +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  facet_wrap(~cohort) +
  labs(x="", y="", fill="Jaccard Distance")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "jaccFourCohortSampleSubset.png", sep="_")), height=6, width=7.5, units="in")
```

# Presence absence vs. absolute abundance
## All samples
### Full List
```{r, fig.width=7, fig.height=2}
mDFtoiClades %>%  
  gather("Taxon", "relativeAbundance", c(clades, Gardnerella_vaginalis, lactos, anaerobes)) %>%
  select(SampleID, Cohort, Taxon, relativeAbundance) %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(relativeAbundance=case_when(relativeAbundance<0.001~"-",
                                      relativeAbundance>=0.001~"+"),
         Taxon=factor(Taxon, levels = c("Gardnerella_vaginalis", clades, lactos, anaerobes), labels = c(abbrevs)),
         Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels=c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  ggplot(aes(x=relativeAbundance, y=bacLoad)) +
  geom_violin(aes(color=relativeAbundance, fill=relativeAbundance), alpha=0.5) +
  binaryColors +
  facet_grid(Cohort~Taxon) +
  scale_y_log10() +
  theme(text = element_text(size=15),
        axis.text = element_text(size = 12),
        legend.position = "none") +
  labs(x="", y="Bacterial Load")
#ggsave(file.path(figureOut, paste(Sys.Date(), "presAbsVsBacLoad.png", sep = "_")))
```


### Short list
```{r, fig.width=5, fig.height=2}
mDFtoiClades %>%  
  gather("Taxon", "relativeAbundance", c(clades, Gardnerella_vaginalis, lactos, anaerobesSL)) %>%
  select(SampleID, Cohort, Taxon, relativeAbundance) %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(relativeAbundance=case_when(relativeAbundance<0.001~"-",
                                      relativeAbundance>=0.001~"+"),
         Taxon=factor(Taxon, levels = c("Gardnerella_vaginalis", clades, lactos, anaerobesSL), labels = c(abbrevsSL)),
         Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels=c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  ggplot(aes(x=relativeAbundance, y=bacLoad)) +
  geom_violin(aes(color=relativeAbundance, fill=relativeAbundance), alpha=0.5) +
  binaryColors +
  facet_grid(Cohort~Taxon) +
  scale_y_log10() +
  theme(text = element_text(size=15),
        axis.text = element_text(size = 12),
        legend.position = "none") +
  labs(x="", y="Bacterial Load")
#ggsave(file.path(figureOut, paste(Sys.Date(), "presAbsVsBacLoadShort.png", sep="_")))
```

## Sample Subset
```{r, fig.width=7, fig.height=2}
mDFtoiClades %>%  
  filter(SampleID %in% sampleSubset) %>%
  gather("Taxon", "relativeAbundance", c(clades, Gardnerella_vaginalis, lactos, anaerobes)) %>%
  select(SampleID, Cohort, Taxon, relativeAbundance) %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(relativeAbundance=case_when(relativeAbundance<0.001~"-",
                                      relativeAbundance>=0.001~"+"),
         Taxon=factor(Taxon, levels = c("Gardnerella_vaginalis", clades, lactos, anaerobes), labels = c(abbrevs)),
         Cohort=factor(Cohort, levels = c("Stanford", "UAB", "HMP2 Enriched", "HMP2"), labels=c("Stanford", "UAB", "HMP2 E", "HMP2"))) %>%
  ggplot(aes(x=relativeAbundance, y=bacLoad)) +
  geom_violin(aes(color=relativeAbundance, fill=relativeAbundance), alpha=0.5) +
  binaryColors +
  facet_grid(Cohort~Taxon) +
  scale_y_log10() +
  theme(text = element_text(size=15),
        axis.text = element_text(size = 12),
        legend.position = "none") +
  labs(x="", y="Bacterial Load")
#ggsave(file.path(figureOut, paste(Sys.Date(), "presAbsVsBacLoadSampleSubset.png", sep="_")))
```

# Absolute abundance correlations
## Descriptives
### Abundance distributions
```{r, fig.width=10, fig.height=4}
mDFtoiClades %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>% # remove outliers
  mutate_at(vars("Gardnerella_vaginalis", clades, lactos, anaerobes), ~.x*bacLoad) %>%
  select(-bacLoad) %>%
  gather("Taxon", "absAb", c(Gardnerella_vaginalis, clades, lactos, anaerobes)) %>%
  mutate(Taxon=factor(Taxon, levels = c("Gardnerella_vaginalis", clades, lactos, anaerobes), labels = abbrevs)) %>%
  ggplot(aes(x=absAb)) +
  geom_histogram() +
  facet_grid(Cohort~Taxon, scales = "free_y")
```


```{r}
taxCombs <- combn(c("Gardnerella_vaginalis", clades, lactos, anaerobes), 2) %>%
  t %>%
  as.data.frame() %>%
  dplyr::rename(Taxon1=V1,
                Taxon2=V2)
head(taxCombs)
nrow(taxCombs)

absAb1 <- mDFtoiClades %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>% # remove outliers
  mutate_at(vars("Gardnerella_vaginalis", clades, lactos, anaerobes), ~.x*bacLoad) %>%
  select(-bacLoad) %>%
  split(.$Cohort) %>%
  map(~gather(.x, "Taxon1", "absAb1", c(Gardnerella_vaginalis, clades, lactos, anaerobes))) %>%
  map(~right_join(.x, taxCombs, by="Taxon1"))

absAb2 <- mDFtoiClades %>%
  left_join(bacLoad, by="SampleID") %>%
  filter(bacLoad<2) %>% # remove outliers
  mutate_at(vars("Gardnerella_vaginalis", clades, lactos, anaerobes), ~.x*bacLoad) %>%
  select(-bacLoad) %>%
  split(.$Cohort) %>%
  map(~gather(.x, "Taxon2", "absAb2", c(Gardnerella_vaginalis, clades, lactos, anaerobes))) %>%
  map(~right_join(.x, taxCombs, by="Taxon2"))

# are nrow of a and b the same
map2(map(absAb1, nrow), map(absAb2, nrow), ~.x==.y)

taxTaxAbsAbDF <- absAb1 %>% # dataframe 
  map2(absAb2, ~full_join(.x, .y, by = c("SampleID", "Cohort", "Taxon1", "Taxon2"))) %>%
  reduce(full_join, by = c("SampleID", "Cohort", "Taxon1", "absAb1", "Taxon2", "absAb2")) %>%
  mutate(Taxon1=factor(Taxon1, levels = c("Gardnerella_vaginalis", clades, lactos, anaerobes), labels = abbrevs),
         Taxon2=factor(Taxon2, levels = rev(c("Gardnerella_vaginalis", clades, lactos, anaerobes)), labels = rev(abbrevs)))
```

### Scatterplots
```{r, fig.height=10, fig.width=10}
taxTaxAbsAbDF %>%
  #mutate(Taxon1=factor(Taxon1),
  #       Taxon2=factor(Taxon2)) %>%
  split(.$Cohort) %>%
  map(~ggplot(.x, aes(x=absAb1, y=absAb2)) +
    geom_point() +
    facet_grid(Taxon2~Taxon1, drop = FALSE) +
    #ylim(0,0.5) +
    #xlim(0,0.5) +
    theme(axis.text.x = element_text(angle=90)) +
    labs(title=paste(unique(.x$Cohort), "Cohort"), x="Absolute abundance: Taxon 1", y="Absolute abundance: Taxon 2"))
```
## All samples
### Long List
```{r, warning=FALSE, message=FALSE, fig.width=6, fig.height=4.5}
# create dataframes for running correlations
corrDF <- taxTaxAbsAbDF %>%
  split(list(.$Taxon1, .$Taxon2, .$Cohort)) %>%
  keep(~nrow(.x)>0) %>%
  map(~cor(.x$absAb1, .x$absAb2, method="spearman")) %>%
  map(tidy) %>%
  map2(names(.), ~mutate(.x, VARS=.y)) %>%
  purrr::reduce(full_join, by=c("VARS", "x")) %>%
  separate(VARS, c("Taxon1", "Taxon2", "cohort"),  sep="\\.") %>%
  dplyr::rename(spearman=x) %>%
  mutate(Taxon1=factor(Taxon1, levels = abbrevs),
         Taxon2=factor(Taxon2, levels=rev(abbrevs)),
         cohort=factor(cohort, levels=c("Stanford", "UAB", "HMP2 Enriched", "HMP2"))) %>%
  filter(!(Taxon1=="Gv"&Taxon2 %in% c("C1", "C2", "C3", "C4", "C5", "C6")))

corrDF %>%
  ggplot(aes(x=Taxon1, y=Taxon2, fill=spearman)) +
  geom_tile() +
  scale_x_discrete(drop=FALSE) +
  scale_y_discrete(drop=FALSE) +
  scale_fill_gradient2(high = "#56B4E9", mid = "gray", low = "#D55E00", midpoint = 0.0) +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  facet_wrap(~cohort) +
  labs(x="", y="")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "spearmanThreeCohort.png", sep="_")), height=6, width=7.5, units="in")
```

```{r}
mDFtoiClades %>%
  select(-Gardnerella_vaginalis) %>%
  split(.$Cohort) %>%
  map(~select(.x, -Cohort)) %>%
  map(~column_to_rownames(.x, "SampleID")) %>%
  map(~cor(.x, method="spearman")) %>%
  map2(., names(.), ~corrplot(.x, diag = FALSE, type = "lower", method = "circle", tl.col = "black", tl.cex = 0.7, title=.y))
```


### Short List
```{r, fig.width=5, fig.height=3.75}
corrDF %>%
  filter(Taxon1 %in% abbrevsSL,
         Taxon2 %in% abbrevsSL) %>%
  mutate(Taxon1=factor(Taxon1, levels  = abbrevsSL),
         Taxon2=factor(Taxon2, levels=rev(abbrevsSL)), 
         cohort=factor(cohort, levels=c("Stanford", "UAB", "HMP2"))) %>%
  ggplot(aes(x=Taxon1, y=Taxon2, fill=spearman)) +
  geom_tile() +
  scale_x_discrete(drop=FALSE) +
  scale_y_discrete(drop=FALSE) +
  scale_fill_gradient2(high = "#56B4E9", mid = "gray", low = "#D55E00", midpoint = 0.0) +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  facet_wrap(~cohort) +
  labs(x="", y="")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "spearmanThreeCohortShortList.png", sep="_")))
```

## Sample subset
```{r, warning=FALSE, message=FALSE, fig.width=6, fig.height=4.5}
# create dataframes for running correlations
corrDFSubset <- taxTaxAbsAbDF %>%
  filter(SampleID %in% sampleSubset) %>%
  split(list(.$Taxon1, .$Taxon2, .$Cohort)) %>%
  keep(~nrow(.x)>0) %>%
  map(~cor(.x$absAb1, .x$absAb2, method="spearman")) %>%
  map(tidy) %>%
  map2(names(.), ~mutate(.x, VARS=.y)) %>%
  purrr::reduce(full_join, by=c("VARS", "x")) %>%
  separate(VARS, c("Taxon1", "Taxon2", "cohort"),  sep="\\.") %>%
  dplyr::rename(spearman=x) %>%
  mutate(Taxon1=factor(Taxon1, levels = abbrevs),
         Taxon2=factor(Taxon2, levels=rev(abbrevs)),
         cohort=factor(cohort, levels=c("Stanford", "UAB", "HMP2"))) %>%
  filter(!(Taxon1=="Gv"&Taxon2 %in% c("C1", "C2", "C3", "C4", "C5", "C6")))

corrDFSubset %>%
  ggplot(aes(x=Taxon1, y=Taxon2, fill=spearman)) +
  geom_tile() +
  scale_x_discrete(drop=FALSE) +
  scale_y_discrete(drop=FALSE) +
  scale_fill_gradient2(high = "#56B4E9", mid = "gray", low = "#D55E00", midpoint = 0.0) +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  facet_wrap(~cohort) +
  labs(x="", y="")
#ggsave(filename = file.path(figureOut, paste(Sys.Date(), "spearmanThreeCohortSampleSubset.png", sep="_")), height=6, width=7.5, units="in")
```


# Gardnerella clades by PTB status:
## Subject average relative abundance by PTB status
```{r}
gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre")], by="SampleID") %>%
  filter(!is.na(TermPre)) %>%
  group_by(SubjectID, Cohort, TermPre, var) %>%
  summarise(meanRelativeAbundance=mean(relativeAbundance)) %>%
  mutate(var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"), labels = c(clades, "Unch.", "Total"))) %>%
  ggplot(aes(x=TermPre, y=meanRelativeAbundance, color=TermPre, fill=TermPre)) +
  #geom_violin(alpha=0.5) +
  geom_jitter(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  binaryColors +
  facet_grid(Cohort~var) +
  theme(legend.position = "none") +
  labs(x=NULL, y="Relative Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "cladeRelativeAbundancePTB.png", sep="_")))
```

  Wilcox rank-sum
```{r}
relativeWilcoxResults <- gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre")], by="SampleID") %>%
  filter(!is.na(TermPre)) %>%
  group_by(SubjectID, Cohort, TermPre, var) %>%
  summarise(meanRelativeAbundance=mean(relativeAbundance)) %>%
  split(list(.$var, .$Cohort)) %>%
  map(~wilcox.test(.x$meanRelativeAbundance~.$TermPre, alternative="greater")) %>%
  map(broom::tidy) %>%
  map2(., names(.), ~mutate(.x, cohort_var=.y)) %>%
  purrr::reduce(~full_join(.x, .y, by = c("statistic", "p.value", "method", "alternative", "cohort_var"))) %>%
  mutate(BH.adjust=p.adjust(p.value, method="BH"))

relativeWilcoxResults %>%
  arrange(p.value) %>%
  formattable()
```


## Subject average absolute abundance by PTB status
```{r}
gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre", "bacLoad")], by="SampleID") %>%
  filter(!is.na(TermPre),
         bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  group_by(SubjectID, Cohort, TermPre, var) %>%
  summarise(meanAbsoluteAbundance=mean(absoluteAbundance)) %>%
  mutate(var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"), labels = c(clades, "Unch.", "Total"))) %>%
  ggplot(aes(x=TermPre, y=meanAbsoluteAbundance, color=TermPre, fill=TermPre)) +
  #geom_violin(alpha=0.5) +
  geom_jitter(alpha=0.5) +
  geom_boxplot(color="black", alpha=0) +
  binaryColors +
  facet_grid(Cohort~var) +
  theme(legend.position = "none") +
  labs(x=NULL, y="Absolute Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "cladeAbsoluteAbundancePTB.png", sep="_")))
```

```{r}
absoluteWilcoxResults <- gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre", "bacLoad")], by="SampleID") %>%
  filter(!is.na(TermPre),
         bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  group_by(SubjectID, Cohort, TermPre, var) %>%
  summarise(meanAbsoluteAbundance=mean(absoluteAbundance)) %>%
  split(list(.$var, .$Cohort)) %>%
  map(~wilcox.test(.x$meanAbsoluteAbundance~.$TermPre, alternative="greater")) %>%
  map(broom::tidy) %>%
  map2(., names(.), ~mutate(.x, cohort_var=.y)) %>%
  purrr::reduce(~full_join(.x, .y, by = c("statistic", "p.value", "method", "alternative", "cohort_var")))

absoluteWilcoxResults %>%
  arrange(p.value) %>%
  formattable()
```


## Number of clades by PTB
```{r}
gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "TermPre")], by="SampleID") %>%
  spread(var, relativeAbundance) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  mutate(nClades=C1+C2+C3+C4+C5+C6,
         nClades=as.character(nClades),
         nClades=case_when(Uncharacterized_Gardnerella!="0"~"U",
                           Uncharacterized_Gardnerella=="0"~nClades)) %>%
  count(Cohort, TermPre, nClades) %>%
  replace_na(list(TermPre="Unknown")) %>%
  group_by(Cohort) %>%
  mutate(n=n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=nClades, y=n, fill=TermPre)) +
  geom_histogram(stat="identity") +
  scale_fill_manual(values=c("#D55E00", "#56B4E9", "grey")) +
  facet_wrap(~Cohort) +
  labs(x="Clades per Sample", y="Proportion of Samples") 
```
  Mean clades per sample in term vs. preterm birth patients
```{r}
gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre")], by="SampleID") %>%
  filter(!is.na(TermPre)) %>%
  spread(var, relativeAbundance) %>%
  mutate_at(clades, ~case_when(.x>0~1,
                               .x==0~0)) %>%
  filter(Uncharacterized_Gardnerella==0) %>%
  mutate(nClades=C1+C2+C3+C4+C5+C6) %>%
  replace_na(list(TermPre="Unknown")) %>%
  group_by(SubjectID, Cohort, TermPre) %>%
  summarise(meanNClades=mean(nClades)) %>%
  ungroup %>%
  ggplot(aes(x=TermPre, y=meanNClades, color=TermPre, fill=TermPre)) +
  geom_violin(alpha=0.5) +
  facet_wrap(~Cohort) +
  binaryColors
```
  Bacterial load by PTB
```{r}
sgDF %>%
  filter(!is.na(TermPre),
         bacLoad<2) %>%
  group_by(SubjectID, Cohort, TermPre) %>%
  summarize(meanBacLoad=mean(bacLoad)) %>%
  ggplot(aes(x=TermPre, color=TermPre, fill=TermPre, y=meanBacLoad)) +
  #geom_violin(alpha=0.5) +
  geom_jitt(alpha=0.5) +
  geom_boxplot(color="black", alpha=0.5) +
  binaryColors +
  facet_wrap(~Cohort) +
  theme(legend.position = "none") +
  labs(x=NULL, y="Bacterial Load")
#ggsave(file.path(figureOut, paste(Sys.Date(), "bacLoadByPTB.png", sep="_")))
```

# Gardnerella Clades by Race
  Subject average abundance by self reported race
```{r}
gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "Race")], by="SampleID") %>%
  group_by(SubjectID, Cohort, Race, var) %>%
  summarise(meanRelativeAbundance=mean(relativeAbundance)) %>%
  mutate(var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"), labels = c(clades, "Unch.", "Total")),
         Race=factor(Race, levels=c("Asian", "Black", "White", "Other", "Unknown"))) %>%
  ggplot(aes(x=Race, y=meanRelativeAbundance, color=Race, fill=Race)) +
  geom_point(alpha=0.5, position = position_jitterdodge()) +
  geom_boxplot(color="black", alpha=0, outlier.shape = 0) +
  facet_grid(Cohort~var) +
  scale_x_discrete(labels=c("A", "B", "W", "O")) +
  theme(legend.position = "bottom") +
  labs(x=NULL, y="Relative Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "cladeRelativeAbundanceRace.png", sep="_")))
```
  Subject average absolute abundance by race
```{r}
gardCladeAbU %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "TermPre", "Race", "bacLoad")], by="SampleID") %>%
  filter(bacLoad<2) %>%
  mutate(absoluteAbundance=relativeAbundance*bacLoad) %>%
  group_by(SubjectID, Cohort, Race, var) %>%
  summarise(meanAbsoluteAbundance=mean(absoluteAbundance)) %>%
  mutate(var=factor(var, levels = c(clades, "Uncharacterized_Gardnerella",  "Gardnerella_vaginalis"), labels = c(clades, "Unch.", "Total")),
         Race=factor(Race, levels=c("Asian", "Black", "White", "Other", "Unknown"))) %>%
  ggplot(aes(x=Race, y=meanAbsoluteAbundance, color=Race, fill=Race)) +
  geom_point(alpha=0.5, position = position_jitterdodge()) +
  geom_boxplot(color="black", alpha=0, outlier.shape = 0) +
  facet_grid(Cohort~var) +
  scale_x_discrete(labels=c("A", "B", "W", "O")) +
  theme(legend.position = "bottom") +
  labs(x=NULL, y="Absolute Abundance")
#ggsave(file.path(figureOut, paste(Sys.Date(), "cladeAbsoluteAbundanceRace.png", sep="_")))
```


# Session Info
```{r}
sessionInfo()
```