---
title: "BioLunch Figure"
author: "Hanna Berman"
date: "6/4/2019"
output: html_document
---

## Set up
### packages
```{r}
library(tidyverse)
```

### Data Paths
```{r}
repo_dir <- "/Users/hlberman/Documents/GitHub/pregnancy_metagenome/"

refDF_path <- file.path(repo_dir, "GardnerellaMetadata.csv")
refDF <- read_csv(refDF_path)

sgDF_path <- file.path(repo_dir, "shotgunMetadata.tsv")
sgDF <- read_delim(file = file.path(sgDF_path), delim = "\t")

sgGardDF_path <- "./gardRelativeAbundance.tsv"
sgGardDF <- read_tsv(sgGardDF_path)

S16gardDF_path <- "./16SgardVariants.tsv"
S16gardDF <- read_tsv(S16gardDF_path)

figure_out <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/gardMapping/clade_assignments"
```

```{r}
refDF %>%
  select(Strain, sixteenS_variant) %>%
  View()
```

As colorstrip
```{r}
varMatrix <- tibble(StrainID=refDF$Strain,
                    symbol=1,
                    size=3,
                    color=case_when(refDF$sixteenS_variant=="G2"~"#FF0000",
                                    refDF$sixteenS_variant=="G1"~"#0000FF",
                                    refDF$sixteenS_variant=="G3"~"#666666",
                                    refDF$sixteenS_variant=="G4"~"#9400d3",
                                    refDF$sixteenS_variant=="G5"~"#009e73"), 
                    fill=1,
                    position=1,
                    v4var=refDF$sixteenS_variant)

write("DATASET_SYMBOL\nSEPARATOR COMMA\nDATASET_LABEL,v4BiolunchTips\nCOLOR,#ffffff\nLEGEND_TITLE,V4 Variant\n\nDATA", 
     file.path(figure_out, "v4tipShapes.txt"))
write_delim(varMatrix, 
           file.path(figure_out, "v4tipShapes.txt"),
           delim=",", col_names=FALSE, append=TRUE)

varMatrix0 <- tibble(StrainID=refDF$Strain,
                    color=case_when(refDF$sixteenS_variant=="G2"~"#FF0000",
                                    refDF$sixteenS_variant=="G1"~"#0000FF",
                                    refDF$sixteenS_variant=="G3"~"#666666",
                                    refDF$sixteenS_variant=="G4"~"#9400d3",
                                    refDF$sixteenS_variant=="G5"~"#009e73"), 
                    v4var=refDF$sixteenS_variant)

write("DATASET_COLORSTRIP\nSEPARATOR COMMA\nDATASET_LABEL,v4BiolunchStrip\nCOLOR,#ffffff\nLEGEND_TITLE,V4 Variant\nLEGEND_SHAPES,1,1,1,1,1\nLEGEND_COLORS,#FF0000,#0000FF,#666666,#9400d3,#009e73\nLEGEND_LABELS,G2,G1,G3,G4,G5\n\nDATA", 
     file.path(figure_out, "v4colorStrip.txt"))
write_delim(varMatrix0, 
           file.path(figure_out, "v4colorStrip.txt"),
           delim=",", col_names=FALSE, append=TRUE)

varMatrix1 <- tibble(StrainID=refDF$Strain,
                    symbol=4,
                    size=3,
                    color=case_when(refDF$sixteenS_variant=="G2"~"#FF0000",
                                    refDF$sixteenS_variant=="G1"~"#0000FF",
                                    refDF$sixteenS_variant=="G3"~"#666666",
                                    refDF$sixteenS_variant=="G4"~"#9400d3",
                                    refDF$sixteenS_variant=="G5"~"#009e73"), 
                    fill=1,
                    position=1,
                    v4var=refDF$sixteenS_variant)

write("DATASET_SYMBOL\nSEPARATOR COMMA\nDATASET_LABEL,v4BiolunchTipsTri\nCOLOR,#ffffff\nLEGEND_TITLE,V4 Variant\nLEGEND_SHAPES,5,5,5,5,5\nLEGEND_COLORS,#FF0000,#0000FF,#666666,#9400d3,#009e73\nLEGEND_LABELS,G2,G1,G3,G4,G5\n\nDATA", 
     file.path(figure_out, "v4tipShapesTri.txt"))
write_delim(varMatrix1, 
           file.path(figure_out, "v4tipShapesTri.txt"),
           delim=",", col_names=FALSE, append=TRUE)
```

## Unify dataframe
```{r}
S16gardDF0 <- S16gardDF %>%
  select(SampleID, Var, propGard) %>%
  mutate(Method="V4 region 16S amplicon")

sgGardDF0 <- sgGardDF %>%
  select(-n) %>%
  dplyr::rename(Var=Clade,
                propGard=propClade) %>%
  mutate(Var=case_when(Var==1~"Clade 1",
                       Var==2~"Clade 2",
                       Var==3~"Clade 3",
                       Var==4~"Clade 4",
                       Var==5~"Clade 5",
                       Var==6~"Clade 6"),
         Method="Shotgun")


gardVarDF <- S16gardDF0 %>%
  full_join(sgGardDF0, by=c("SampleID", "Var", "propGard", "Method")) %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "GWcoll", "RNAlater")], by="SampleID") %>%
  filter(SubjectID!="10008",
         SubjectID!="10013",
         SubjectID!="10018",
         SubjectID!="10027",
         SubjectID!="10031",
         SubjectID!="10040",
         SubjectID!="19004")

S16gardDF %>%
  select(-OTU, -Abundance) %>%
  left_join(sgGardDF[,c("SampleID", "Clade", "propClade")], by="SampleID") %>%
  dplyr::rename(propGard16S=propGard,
                var16S=Var,
                propGardSG=propClade,
                cladeSG=Clade) %>%
  mutate(cladeSG=as.character(cladeSG)) %>%
  select(SampleID, SubjectID, GWcoll, RNAlater, var16S, propGard16S, cladeSG, propGardSG) -> gardVarDF0
```

## Plot
```{r}
gardVarDF0 %>%
  filter(RNAlater=="No") %>%
#  mutate(GWcoll=case_when(SubjectID=="10508"&GWcoll==32~34,
#                          SubjectID=="10508"&GWcoll==31~GWcoll,
#                          SubjectID!="10508"~GWcoll)) %>% # the two timepoints in this subject are too close together, split for visuals
ggplot() +
  geom_point(aes(x=GWcoll, y=propGard16S, color=var16S), shape=1, size=3) +
  geom_point(aes(x=GWcoll, y=propGardSG, color=cladeSG), shape=16, size=1.5) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade 1 (shotgun)", "Clade 2 (shotgun)", "Clade 3 (shotgun)", "Clade 4 (shotgun)", "Clade 5 (shotgun)", "Clade 6 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")) + # note: change labels to match Callahan 2017
  facet_wrap(~SubjectID) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", color="Variant")
```  

```{r}
gardVarDF %>%
  filter(RNAlater=="No",
         SubjectID=="10621"|SubjectID=="19024")%>%
ggplot(aes(x=GWcoll, y=propGard, color=Var, shape=Method, size=Method)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 3)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "red", "blue1", "grey48"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)")) + # note: change labels to match Callahan 2017
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~SubjectID) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3)))
#ggsave(file.path(figure_out, "biolunchV4SG2samp.png"), width=8, height=6)
```

```{r}
gardVarDF %>%
  filter(RNAlater=="No")%>%
ggplot(aes(x=GWcoll, y=propGard, color=Var, shape=Method, size=Method)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 3)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "Clade C5 (shotgun)", "Clade C6 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")) + # note: change labels to match Callahan 2017
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3)))
ggsave(file.path(figure_out, "biolunchV4SGv1.png"), width=8, height=6)
```

```{r}
gardVarDF %>%
  filter(RNAlater=="No", 
         SubjectID=="10056"| 
         SubjectID=="10508"|
         SubjectID=="10608"|
         SubjectID=="10621"|
         SubjectID=="10626") %>%
ggplot(aes(x=GWcoll, y=propGard, color=Var, shape=Method, size=Method)) +
  geom_point() +
  scale_size_manual(values=c(2, 2)) +
  scale_color_manual(values=c("salmon", "indianred", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade 1 (shotgun)", "Clade 2 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)")) + # note: change labels to match Callahan 2017
  scale_shape_manual(values = c(1, 16)) +
  scale_x_continuous(breaks = c(15,20,25,30,35)) +
  facet_grid(SubjectID~Method) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3)))
#ggsave(file.path(figure_out, "biolunchV4SGv2a.png"), width=7, height=7)
```

```{r}
gardVarDF %>%
  filter(RNAlater=="No",
         SubjectID=="10627"|
         SubjectID=="10629"|
         SubjectID=="19024"|
         SubjectID=="20505"|
         SubjectID=="40021") %>%
ggplot(aes(x=GWcoll, y=propGard, color=Var, shape=Method, size=Method)) +
  geom_point() +
  scale_size_manual(values=c(2, 2)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade 1 (shotgun)", "Clade 2 (shotgun)", "Clade 3 (shotgun)", "Clade 4 (shotgun)", "Clade 5 (shotgun)", "Clade 6 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")) + # note: change labels to match Callahan 2017
  scale_shape_manual(values = c(1, 16)) +
  facet_grid(SubjectID~Method) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3)))
#ggsave(file.path(figure_out, "biolunchV4SGv2b.png"), width=7, height=7)
```

## Session Info
```{r}
sessionInfo()
```