---
title: "Gardnerella clades, relative abundance, and preterm birth"
author: "Hanna Berman"
date: "4/25/2019"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
---

# Setup
## Packages
```{r}
library(tidyverse)
```

## File paths
```{r}
gardRelativeAbundance <- read_delim("/Users/hlberman/Documents/GitHub/pregnancy_metagenome/gard_mapping/clade_assignments/gardRelativeAbundance.tsv", delim = "\t") %>%
  mutate(SampleID=as.character(SampleID),
         Clade=as.character(Clade))
alignmentsFinal <- read_delim("/Users/hlberman/Documents/GitHub/pregnancy_metagenome/gard_mapping/clade_assignments/aglingmentsFinal.tsv", delim = "\t")
sgDF <- read_delim("/Users/hlberman/Documents/GitHub/pregnancy_metagenome/shotgunMetadata.tsv", delim = "\t")%>%
  mutate(SampleID=as.character(SampleID))

#metaphlan data
#load("/Users/hlberman/Documents/GitHub/pregnancy_metagenome/methods_comparisons/PS_RData/metaphlan_PS.RData")
mDF0 <- read_tsv("/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/humann2/out_tables/mergedMetaphlanAbundance.tsv") %>%
  filter(str_detect(ID, "s__"), !str_detect(ID, "t__")) %>%
  mutate(Species = str_extract(ID, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -ID) %>%
  as.data.frame()

#make col and row names
samps <- colnames(mDF0[2:ncol(mDF0)]) %>%
  str_extract("[0-9]{10}")
samps
colnames(mDF0) <- c("Species", samps)
rownames(mDF0) <- mDF0$Species

#Final edits to table
mDF <- mDF0 %>%
  select(-Species) %>%
  t() %>%
  as.data.frame() %>%
  mutate(SampleID=samps) %>%
  select(SampleID, everything())
mDF[1:6,1:5]
```

# Per sample
Recalculate proportions of Gardnerella as proportions of the whole microbiome. (Rethink how this should be done... reads/total reads?) Currently done by multiplying times proportion of all Gardnerella found by Metaphlan
```{r}
SampleIDlist <- rep(unique(sgDF$SampleID), 6) %>%
  .[order(.)]

cladeFrame <- tibble(SampleID=SampleIDlist,
                     Clade=rep(c("1","2", "3", "4", "5", "6"), 107),
                     n=0, 
                     propClade=0)

gardRelativeAbundance0 <- gardRelativeAbundance %>%
 full_join(cladeFrame) %>%
 group_by(SampleID, Clade) %>%
 filter(n==max(n)) %>%
 ungroup() %>%  
 left_join(unique.data.frame(sgDF[,c("SampleID", "SubjectID", "TermPre", "GWdell", "Cohort", "GWcoll")])) %>%
 mutate(SubjectID=as.character(SubjectID))
```  

Get Gardnerella relative abundances from metaphlan
```{r}
# metaphlanGard <- PS_metaphlan_all@otu_table %>%
#   t() %>% 
#   as.data.frame() %>%
#   mutate(SampleID=rownames(.)) %>%
#   select(Gardnerella_vaginalis, SampleID)
 
gardRelativeAbundance1 <- gardRelativeAbundance0 %>%
   left_join(mDF[,c("SampleID", "Gardnerella_vaginalis")], by="SampleID") %>%
   mutate(propCladeCommunity=propClade*Gardnerella_vaginalis)
```

### Clade abundance

Observe total proportion of other *Gardnerella* clades in the community vs. proportion of each clade
```{r}
ggplot(gardRelativeAbundance1, aes(x=propCladeCommunity, y=Gardnerella_vaginalis-propCladeCommunity, color=Clade)) +
  geom_point() +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73")) +
  labs(x="Proportion of Gardnerella clade in community", y="Proportion of other Gardnerella clades in the community", color="Clade") + 
  facet_grid(Cohort~Clade) +
  theme(axis.text.x = element_text(angle=45))
```  
<br />Not the best plot for visualizing the data, but it does show that when the total there is one clade of Garnerella in the community and the community is nearly 100% *Gardnerella*, that clade is clade 1. 

<br />Observe proportion of clades vs other clades
```{r}
# reformat dataframe for this
cladexclade <- gardRelativeAbundance1 %>%
  select(-propClade, -n) %>%
  spread(Clade, propCladeCommunity) %>%
  mutate(x1=`1`, 
         x2=`2`, 
         x3=`3`,
         x4=`4`,
         x5=`5`,
         x6=`6`) %>%
  gather("Clade1","propClade1", c(`1`,`2`,`3`,`4`,`5`,`6`)) %>%
  gather("Clade2", "propClade2", c(x1, x2, x3, x4, x5, x6))%>%
  mutate(Clade2=case_when(Clade2=="x1"~"1",
                          Clade2=="x2"~"2",
                          Clade2=="x3"~"3",
                          Clade2=="x4"~"4",
                          Clade2=="x5"~"5",
                          Clade2=="x6"~"6")) %>%
  filter(Clade1!=Clade2)
#Stanford
stanford <- cladexclade %>%
  filter(Cohort=="Stanford")
ggplot(stanford, aes(x=propClade1, y=propClade2))+
  geom_point() +
  facet_grid(Clade1~Clade2) + 
  coord_fixed() +
  labs(title="Stanford", x="Proportion clade x", y="Proportion clade y") +
  theme(axis.text.x = element_text(angle=45))

#UAB
stanford <- cladexclade %>%
  filter(Cohort=="UAB")
ggplot(stanford, aes(x=propClade1, y=propClade2))+
  geom_point() +
  facet_grid(Clade1~Clade2) + 
  coord_fixed() +
  labs(title="UAB", x="Proportion clade x", y="Proportion clade y") +
  theme(axis.text.x = element_text(angle=45))
```
<br \>No clear patterns. 

<br \>Observe proportion of each clade in each sample vs. term/preterm delivery
```{r}
ggplot(data=gardRelativeAbundance1, aes(x=TermPre, y=propCladeCommunity, color=TermPre)) +
  geom_jitter() +
  geom_boxplot(alpha=0) +
  scale_y_log10() +
  scale_x_discrete(labels=c("Preterm", "Term")) +
  scale_color_manual(values=c("#56B4E9", "#D55E00")) +
  facet_grid(Cohort~Clade) +
  labs(x=element_blank(), y="Proportion clade in community") +
  theme(legend.position = "none")
```
<br />Proportion of community vs preterm/term births faceted by clade. Does not suggest differences in proportion of each clade in each sample and term vs. preterm births. Perhaps the time series information is important here...

<br \>look at two specific samples
```{r}
ggplot(gardRelativeAbundance1[gardRelativeAbundance1$SubjectID=="10031",], aes(x=GWcoll, y=propCladeCommunity, color=Clade, fill=Clade))+
  geom_bar(stat = "identity")+
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73")) +
  labs(title = "Subject 10031", subtitle = "Preterm")

ggplot(gardRelativeAbundance1[gardRelativeAbundance1$SubjectID=="10621",], aes(x=GWcoll, y=propCladeCommunity, color=Clade, fill=Clade))+
  geom_bar(stat = "identity") +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73")) +
  labs(title="Subject 10621", subtitle = "Term", x="Gestation weeks at collection", y="Proportion of community")
```
<br />Not going to draw any major conclusions from these two subjects (small n), but I just wanted to take a closer look since they have the highest proportion of *Gardnerella* and specifically clade 1. 

<br />Gestational frequency vs PTB
```{r}
sgDF <- sgDF %>%
  mutate(SubjectID=as.character(SubjectID))



gardRelativeAbundance1 %>%
  mutate(pres=case_when(propCladeCommunity>0~1,
                        propCladeCommunity==0~0)) %>%
  group_by(SubjectID, Clade) %>%
  summarise(freq=mean(pres)) %>%
  left_join(sgDF[,c("SubjectID", "Cohort", "TermPre")], by="SubjectID") %>% # add delivery time info 
ggplot(., aes(x=TermPre, y=freq, color=TermPre))+
  geom_jitter() +
  geom_boxplot(alpha=0) +
  scale_x_discrete(labels=c("Preterm", "Term")) +
  scale_color_manual(values=c("#56B4E9", "#D55E00")) +
  facet_grid(Cohort~Clade) +
  theme(legend.position = "none") + 
  labs(x="Preterm vs. Term")
```

<br /> Prevalence of each clade vs ptb (proportion samples in each group with/without clade)
```{r}
gardRelativeAbundance1 %>%
  mutate(pres=case_when(propCladeCommunity>0~1,
                        propCladeCommunity==0~0)) %>%
  group_by(Cohort, TermPre, Clade) %>%
  summarise(Prevalence=mean(pres)) %>%
ggplot(., aes(x=TermPre, y=Prevalence, fill = TermPre)) +
  geom_col() +
  scale_x_discrete(labels=c("Preterm", "Term")) +
  scale_fill_manual(values=c("#56B4E9", "#D55E00")) +
  facet_grid(Cohort~Clade) + 
  labs(x="Preterm vs. Term")
```


### Clade presence
Clade presence and overall *Gardnerella* relative abundance
```{r}
ggplot(gardRelativeAbundance1, aes(x=(propCladeCommunity == 0), y=Gardnerella_vaginalis, color=(propCladeCommunity>0))) +
  geom_jitter() +
  geom_boxplot(alpha=0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Cohort~Clade) +
  labs(x="Clade presence", y="Total proportion of Gardnerella")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  theme(legend.position = "none")
```
<br /> Will need to test, but it seems that presence of clade 1 may predict relative abundance of *Gardnerella*. 

<br />Should test for odds ratio of clade 1 presence and PTB risk. (The selection of samples may hinder our ability to do this. Should we focus more on clade presence and overall relative abundance of *Gardnerella*). 

### Bar plots
Remove 0 values and replace with NA so they don't show up on the barplot
```{r}
gardRelativeAbundance2 <- gardRelativeAbundance1 %>%
  mutate(propCladeCommunity=na_if(propCladeCommunity, 0))
```

Plot
```{r}
ggplot(data=gardRelativeAbundance2, mapping=aes(x=reorder(SampleID, -propCladeCommunity, FUN=sum, na.rm=TRUE), y=propCladeCommunity, fill=Clade, color=Clade)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade 1", "Clade2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade 1", "Clade2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Preterm", "Term")) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank())
```
<br />Looks like we should test for a relationship between clade 1/2 presence and overall *Gardnerella* abundance. 

```{r}
ggplot(data=gardRelativeAbundance2, mapping=aes(x=reorder(reorder(SampleID, -propCladeCommunity, FUN=sum, na.rm=TRUE), TermPre=="T", na.rm=TRUE), y=propCladeCommunity, fill=Clade, color=Clade)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade 1", "Clade2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade 1", "Clade2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Preterm", "Term")) +
  facet_wrap(~Cohort) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank())
```
<br />A relationship with PTB might be more complicated. Consider working in methods that will include the time-series aspect of the data or subject average. 

```{r}
ggplot(data=gardRelativeAbundance2, mapping=aes(x=reorder(SampleID, TermPre=="T"), y=propCladeCommunity, fill=Clade, color=Clade)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade 1", "Clade2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade 1", "Clade2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Preterm", "Term")) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank()) +
  geom_line(aes(x=reorder(as.numeric(SubjectID), TermPre=="T", na.rm=TRUE), y=0))
```
<br />Separate bars by subject.

# Subject averages
Caclulate average proportion of each clade per subject
```{r}
cladeAverages <- gardRelativeAbundance1 %>%
  data.frame(stringsAsFactors=FALSE) %>%
  group_by(Clade, SubjectID) %>%
  mutate(cladeAvg=mean(propCladeCommunity)) %>%
  ungroup() %>%
  select(SubjectID, Clade, cladeAvg, TermPre, Cohort, GWdell) %>%
  unique.data.frame() 
```

Bar plot of average clade proportion per subject.
```{r}
cladeAverages %>%
    mutate(cladeAvg=na_if(cladeAvg, 0)) %>%
ggplot(mapping=aes(x=reorder(reorder(SubjectID, -cladeAvg, FUN=sum, na.rm=TRUE), TermPre=="T", na.rm=TRUE), y=cladeAvg, fill=Clade, color=Clade)) +
  geom_bar(stat = "identity") +
  labs(x="Subject", y="Proportion", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", size=5.5, show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", size=5.5, show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", size=5.5, show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade 1", "Clade2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade 1", "Clade2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Preterm", "Term")) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank())
```
<br />
<br />Average proportion of each clade in the entire community vs. preterm/term:
```{r}
ggplot(data=cladeAverages, aes(x=TermPre, y=cladeAvg, color=TermPre)) +
  geom_jitter() +
  geom_boxplot(alpha=0)+
  scale_x_discrete(labels=c("Preterm", "Term")) +
  scale_y_log10() +
  scale_color_manual(values=c("#D55E00", "#56B4E9")) +
  facet_grid(Cohort~Clade) +
  labs(x=element_blank(), y="Proportion of community") +
  theme(legend.position = "none")
```
<br />May not be any differences in PTB risk by proportion of each clade in the entire community. Perhaps presence is a better predictor than proportion... 
<br />If the clade was present at any time in the subject
```{r}
ggplot(cladeAverages, aes(x=(cladeAvg==0), fill=TermPre)) +
  geom_bar(position="dodge") +
  scale_fill_manual(values=c("#D55E00", "#56B4E9"), labels=c("Preterm", "Term")) +
  facet_grid(Cohort~Clade) +
  labs(x="Clade presence", fill="Delivery") +
  scale_x_discrete(labels=c("Present", "Absent"))
```
<br />Could clade 1 or 2 presence increase risk for PTB?
<br />It appears that all subjects in the UAB cohort had *Gardnerella* clades 1, 2, and 4. 

# Dominant *Gardnerella* clade over time
Look at the dominant *Gardnerella* clade in each sample (of those that have *Gardnerella*)
```{r}
gardRelativeAbundance2 <- gardRelativeAbundance1 %>%
  group_by(SampleID) %>%
  filter(propClade==max(propClade)) %>%
  ungroup()

ggplot(gardRelativeAbundance2) +
  geom_point(aes(x=floor(GWcoll), y=reorder(SubjectID, GWdell), color=Clade)) +
  geom_point(aes(y=SubjectID, x=GWdell+.5), shape=4, color="black") +
  #geom_point(aes(x=0, y=SubjectID, color=Cohort)) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999")) +
  geom_vline(xintercept = 37.5) +
  labs(x="Gestation weeks at collection", y="Subject ID", title="Most abundant Gardnerella clade")
```
<br />X marks gestation weeks at delivery. Vertical line is 37 weeks. 

<br />I think we need to test for clade 1 and/or 2 presence (especially clade 1) and total proportion of *Gardnerella* in the sample and also PTB risk. Without doing any statistics  yet, it seems that presence of clade 1 predicts total proportion of *Gardnerella*. My initial hunch is that when you have this "BV-like" microbiome, abundance in *Gardnerella*, it is likely that these "pathogenic" strains are present, but you also may have these "non-pathogenic" strains in the community as well, and they may be more abundant than the "pathogenic" strains. The genomics information will be very interesting. Metatranscriptomic data would also be *very* interesting if it is true that these strains can reshape the community even if they are not the dominant strain of *Gardnerella* in the community!!

# Add other taxa to analyses
## Start with Lactos
### Proportion Gard clade vs. proportion Lacto
```{r}
LGabundance <- gardRelativeAbundance1 %>%
  left_join(mDF[,c("SampleID", "Lactobacillus_crispatus", "Lactobacillus_iners", "Lactobacillus_jensenii", "Lactobacillus_gasseri")], by="SampleID") %>%
  gather("Lactobacillus", "LAbundance", c(Lactobacillus_crispatus, Lactobacillus_iners, Lactobacillus_jensenii, Lactobacillus_gasseri)) %>%
  mutate(LAbundance=LAbundance/100, 
         Lactobacillus=case_when(Lactobacillus=="Lactobacillus_crispatus"~"Lactobacillus crispatus", 
                                 Lactobacillus=="Lactobacillus_gasseri"~"Lactobacillus gasseri", 
                                 Lactobacillus=="Lactobacillus_iners"~"Lactobacillus iners", 
                                 Lactobacillus=="Lactobacillus_jensenii"~"Lactobacillus jensenii"))
head(LGabundance)

LGabundance %>%
  filter(Cohort=="Stanford") %>%
ggplot(aes(x=propCladeCommunity, y=LAbundance, color=Lactobacillus)) +
  geom_point() +
  facet_grid(Lactobacillus~Clade) +
  #coord_fixed() +
  labs(title="Stanford", x="Proportion Gardnerella clade", y="Proportion Lactobacillus")  
#ggsave("/Users/hlberman/Desktop/test1.pdf", height = 8, width=15)

LGabundance %>%
  filter(Cohort=="UAB") %>%
ggplot(aes(x=propCladeCommunity, y=LAbundance, color=Lactobacillus)) +
  geom_point() +
  facet_grid(Lactobacillus~Clade) +
  #coord_fixed() +
  labs(title="UAB", x="Proportion Gardnerella clade", y="Proportion Lactobacillus")  
#ggsave("/Users/hlberman/Desktop/test2.pdf", height = 8, width=15)
```
<br /> Mainly only see *L. iners* in UAB samples, as we have seen previously. Clear inverse relationship between most clades and *Lactobacillus* species, except clade 4 and *L. gasseri* in Stanford, and clades 1 and 3 and *L. iners* in UAB cohort.

### Gard clade presence vs proportion Lacto
```{r}
LGabundance %>%
  filter(Cohort=="Stanford") %>%
ggplot(aes(x=(propCladeCommunity == 0), y=LAbundance, color=(propCladeCommunity==0))) +
  geom_jitter() +
  geom_boxplot(alpha=0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Lactobacillus~Clade) +
  labs(title="Stanford", x="Clade presence", y="Proportion of Lactobacillus")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  #scale_y_log10() +
  theme(legend.position = "none")

LGabundance %>%
  filter(Cohort=="UAB") %>%
ggplot(aes(x=(propCladeCommunity == 0), y=LAbundance, color=(propCladeCommunity==0))) +
  geom_jitter() +
  geom_boxplot(alpha=0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Lactobacillus~Clade) +
  labs(title="UAB", x="Clade presence", y="Proportion of Lactobacillus")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  #scale_y_log10() +
  theme(legend.position = "none")
```
<br />Difficult to make any conclusions in the UAB cohort, because most samples do not have species other than *L. iners*. In the Stanford cohort, *L. crispatus* relative abundance may be greater without any *Gardnerella* clade. May not be a difference between *L. iners* relative abundance with and without *Gardnerlla* clades. 

### Bar plot
```{r}
LGabundance0 <- LGabundance %>%
  select(-n,-propClade) %>%
  spread(Clade, propCladeCommunity) %>%
  spread(Lactobacillus, LAbundance) %>%
  gather("Taxon", "Abundance", `1`, `2`, `3`, `4`, `5`, `6`, `Lactobacillus crispatus`, `Lactobacillus gasseri`, `Lactobacillus iners`, `Lactobacillus jensenii`) %>%
  filter(!is.na(Abundance)) 

ggplot(LGabundance0, aes(x=reorder(reorder(SampleID, -Abundance, FUN=sum, na.rm=TRUE), TermPre=="T", na.rm=TRUE), , y=Abundance, fill=Taxon, color=Taxon)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#999999"), labels=c("Gardnerella clade 1", "Gardnerella clade2", "Gardnerella clade 3", "Gardnerella clade 4", "Gardnerella clade 5", "Gardnerella clade 6", "Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus iners", "Lactobacillus jensenii", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#999999"), labels=c("Gardnerella clade 1", "Gardnerella clade2", "Gardnerella clade 3", "Gardnerella clade 4", "Gardnerella clade 5", "Gardnerella clade 6", "Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus iners", "Lactobacillus jensenii", "Preterm", "Term")) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank()) 

#  geom_line(aes(x=reorder(as.numeric(SubjectID), TermPre=="T", na.rm=TRUE), y=0))
```

# Subject averages
Caclulate average proportion of each clade per subject
```{r}
taxAverages <- LGabundance0 %>%
  data.frame(stringsAsFactors=FALSE) %>%
  group_by(Taxon, SubjectID) %>%
  mutate(taxAvg=mean(Abundance)) %>%
  ungroup() %>%
  select(SubjectID, Taxon, taxAvg, TermPre, Cohort, GWdell) %>%
  unique.data.frame() 
```

Bar plot of average clade proportion per subject.
```{r}
taxAverages %>%
    mutate(taxAvg=na_if(taxAvg, 0)) %>%
ggplot(mapping=aes(x=reorder(reorder(SubjectID, -taxAvg, FUN=sum, na.rm=TRUE), TermPre=="T", na.rm=TRUE), y=taxAvg, fill=Taxon, color=Taxon)) +
  geom_bar(stat = "identity") +
  labs(x="Subject", y="Proportion", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", size=5.5, show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", size=5.5, show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", size=5.5, show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#999999"), labels=c("Gardnerella clade 1", "Gardnerella clade2", "Gardnerella clade 3", "Gardnerella clade 4", "Gardnerella clade 5", "Gardnerella clade 6", "Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus iners", "Lactobacillus jensenii", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#999999"), labels=c("Gardnerella clade 1", "Gardnerella clade2", "Gardnerella clade 3", "Gardnerella clade 4", "Gardnerella clade 5", "Gardnerella clade 6", "Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus iners", "Lactobacillus jensenii", "Preterm", "Term")) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank())
```

# Session Info
```{r}
sessionInfo()
```
