---
title: "Gardnerella clades, relative abundance, and preterm birth"
subtitle: "Exploratory analyses"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
---

# Setup
  Packages
```{r}
library(tidyverse)
library(formattable)
`%!in%` <- Negate(`%in%`)
```

## ggplot settings
### Colors for clades
```{r}
cladeColors <-  list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73"), labels=c("C1", "C2", "C3", "C4", "C5", "C6")),
                     scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73"), labels=c("C1", "C2", "C3", "C4", "C5", "C6")))

cladeColorsTotal <- list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))),
                         scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", bquote('Total'~italic(Gardnerella)))))

cladeColorsN <-  list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", "Uncharacterized")),
                      scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", "Uncharacterized")))
                 
cladeColorsNTotal <- list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", "Uncharacterized", bquote('Total'~italic(Gardnerella)))),
                          scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000"), labels=c("C1", "C2", "C3", "C4", "C5", "C6", "Uncharacterized", bquote('Total'~italic(Gardnerella)))))
```

### Gard and Lactos Colors
```{r}
gLColors <-   list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")),
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred", "darkseagreen4"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")))

gLColorsN <-   list(scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "khaki1", "darkblue", "darkred"), labels=c("A. vaginae", "G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")),
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "khaki1", "darkblue", "darkred"), labels=c("A. vaginae", "G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")))
```

### All Taxa Colors
```{r}
taxColors <-   list(scale_color_manual(values=c("mediumpurple4", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred", "darkseagreen4"), labels=c("A. vaginae", "G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "P. bivia")),
  scale_fill_manual(values=c("mediumpurple4", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred", "darkseagreen4"), labels=c("A. vaginae", "G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "P. bivia")))

taxColorsN <-   list(scale_color_manual(values=c("mediumpurple4", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "khaki1", "darkblue", "darkred", "darkseagreen4"), labels=c("A. vaginae", "G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "P. bivia")),
  scale_fill_manual(values=c("mediumpurple4", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "khaki1", "darkblue", "darkred", "darkseagreen4"), labels=c("A. vaginae", "G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "P. bivia")))
```

### Binary colors
```{r}
binaryColors <- list(scale_color_manual(values=c("#56B4E9", "#D55E00")),
                     scale_fill_manual(values=c("#56B4E9", "#D55E00")))
```


## File paths
```{r}
gardRelativeAbundance <- file.path("gardRelativeAbundance.tsv") %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

alignmentsFinal <- file.path("aglingmentsFinal.tsv") %>%
  read_tsv()

sgDF <- file.path("..", "..", "shotgunMetadata.tsv") %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

#metaphlan data
mDF0 <- file.path("/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/humann2/merged_tables/mergedMetaphlanAbundance.tsv") %>%
  read_tsv() %>%
  filter(str_detect(ID, "s__"), !str_detect(ID, "t__")) %>%
  mutate(Species = str_extract(ID, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -ID) %>%
  as.data.frame()

#make col and row names
samps <- colnames(mDF0[2:ncol(mDF0)]) %>%
  str_extract("[0-9]{10}")
colnames(mDF0) <- c("Species", samps)
rownames(mDF0) <- mDF0$Species

#Final edits to table
mDF <- mDF0 %>%
  select(-Species) %>%
  t() %>%
  as_tibble() %>%
  mutate_all(funs(./100)) %>%
  mutate(SampleID=samps) %>% 
  select(SampleID, everything())
mDF[1:6,1:5]
```

## Other shortcuts
```{r}
cladeLevels <- c("C1", "C2", "C3", "C4", "C5", "C6", "Uncharacterized")
cladeLevelsT <- c("C1", "C2", "C3", "C4", "C5", "C6", "Uncharacterized", "Total Gardnerella")
gardAndLactos <- c("C1", "C2", "C3", "C4", "C5", "C6", "Uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")
taxaOfInterest <- c("C1", "C2", "C3", "C4", "C5", "C6", "Uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "Atopobium_vaginae", "Prevotella_bivia")
```


# Per sample
Recalculate proportions of Gardnerella as proportions of the whole microbiome. (Rethink how this should be done... reads/total reads?) Currently done by multiplying times proportion of all Gardnerella found by Metaphlan
```{r}
SampleIDlist <- rep(unique(sgDF$SampleID), 6) %>%
  .[order(.)]
```  

## Get Gardnerella relative abundances from metaphlan
```{r}
#Table with uncharacterized G. vaginalis
gardRelativeAbundance1 <- gardRelativeAbundance %>%
  full_join(mDF[,c("SampleID", "Gardnerella_vaginalis")], by="SampleID") %>%
  mutate(propCladeCommunity=propClade*Gardnerella_vaginalis) %>%
  select(-n, -propClade) %>%# Remove for column spreading
  spread(Clade, propCladeCommunity) %>%
  select(-`<NA>`) %>% # that spread creates an NA column
  replace_na(list(C1=0, C2=0, C3=0, C4=0, C5=0, C6=0)) %>% #replace NAs with zeros
  mutate(Uncharacterized=case_when((C1+C2+C3+C4+C5+C6)==0 & Gardnerella_vaginalis>0 ~ Gardnerella_vaginalis,
                                   (C1+C2+C3+C4+C5+C6)==0 & Gardnerella_vaginalis==0|
                                    (C1+C2+C3+C4+C5+C6)>0 & Gardnerella_vaginalis>0~0)) %>%
  gather("Clade", "propCladeCommunity", c("C1", "C2", "C3", "C4","C5", "C6", "Uncharacterized")) %>%
  left_join(gardRelativeAbundance) %>% # add n and proportion information and metadata
  replace_na(list(n=0, propClade=0)) %>%
  left_join(unique.data.frame(sgDF[,c("SampleID", "SubjectID", "TermPre", "GWdell", "Cohort", "GWcoll")])) %>%
  mutate(SubjectID=as.character(SubjectID))

#Table without uncharacerized G. vaginalis
gardRelativeAbundance2 <- gardRelativeAbundance1 %>%
  filter(Clade!="Uncharacterized")
```

## reads vs. metaphlan proportion
```{r}
mapVMetaphlan <- gardRelativeAbundance %>%
  select(SampleID, n) %>%
  full_join(sgDF[c("SampleID")], by="SampleID") %>% # add samples with zero gard reads by mapping method
  replace_na(list(n=0)) %>%
  group_by(SampleID) %>%
  summarise(n=sum(n)) %>%
  ungroup %>%
  left_join(sgDF[,c("SampleID", "bbmapFiltered_pairs")], by="SampleID") %>%
  left_join(mDF[,c("SampleID", "Gardnerella_vaginalis")], by="SampleID") %>%
  rename(Metaphlan=Gardnerella_vaginalis) %>%
  mutate(Mapping=n/bbmapFiltered_pairs)

mapVMetaphlan %>%
  ggplot(aes(x=Mapping, y=Metaphlan)) +
    geom_point() +
    labs(x="Mapping proportion", y="MetaPhlAn proportion")
```
  Some agreement between mapping and metaphlan methods.

## Clades per sample
```{r}
#samples with uncharacterized Gard
samplesGardUncharacterized <- gardRelativeAbundance1 %>%
  filter(Clade=="Uncharacterized",
         propCladeCommunity>0) %>%
  .$SampleID
samplesGardUncharacterized

#get clade counts
cladeCounts <- gardRelativeAbundance1 %>%
   mutate(propCladeCommunity=case_when(propCladeCommunity>0~1,
                                       propCladeCommunity==0~0)) %>%
   group_by(SampleID) %>%
   summarise(nClade=sum(propCladeCommunity)) %>%
   mutate(nClade=case_when((SampleID %in% samplesGardUncharacterized)~-1, #note uncharacterized gard+ samples, record these as -1
                           (SampleID %!in% samplesGardUncharacterized)~nClade))

sgDF0 <- sgDF %>%
  left_join(cladeCounts, by="SampleID")

#plot
## ALL
sgDF0 %>%
   ggplot(aes(x=nClade)) +
    geom_histogram(stat="count", binwidth = 1) +
    scale_x_continuous(breaks = c(-1,0,1,2,3,4,5,6), labels = c("Unch.", 0, 1, 2, 3, 4, 5, 6)) +
    facet_wrap(~Cohort)+
    labs(x="Clades per sample")
## Positive only
sgDF0 %>%
  filter(nClade>0) %>%
  ggplot(aes(x=nClade)) +  
    geom_histogram(stat="count", binwidth = 1) +
    scale_x_continuous(breaks = c(-1,0,1,2,3,4,5,6)) +
    facet_wrap(~Cohort)+
    labs(x="Clades per sample")
  
#average clades per sample per subject
# remove subjects 10604 and 10626
sgDF0 %>%
    filter(SubjectID!="10604", # remove subjects w uncharacterized gard
         SubjectID!="10626") %>%
  group_by(SubjectID) %>%
  summarise(meanNClade=mean(nClade)) %>%
  left_join(sgDF0[,c("SubjectID", "Cohort")], by="SubjectID")%>%
  filter(Cohort!="unknown") %>%
  ggplot(aes(x=round(meanNClade))) +
    geom_histogram(stat="count") +
    scale_x_continuous(breaks = c(0,1,2,3,4,5,6)) +
    facet_wrap(~Cohort)+
    labs(x="Average clades per sample (per subject)")

gardCladePositveSamples <- sgDF0 %>%
  filter(nClade>0)%>%
  .$SampleID
head(gardCladePositveSamples)
```

### Clade abundance
#### Proportions of Clades
Proportion *G. vaginalis* by Clade
```{r}
gardRelativeAbundance1 %>%
  ggplot(aes(x=Clade, y=propClade, color=Clade)) +
  geom_boxplot(outlier.alpha=0) +
  geom_jitter() +
  cladeColorsN +
  facet_grid(Cohort~.) +
  labs(x="Clade", y=bquote('Proportion of'~italic(Gardnerella)))
```
  
Proportion each clade of the community
```{r}
GardAb <- gardRelativeAbundance1 %>%
  select(-Clade,-propCladeCommunity,-propClade,-n) %>%
  unique.data.frame() %>%
  dplyr::rename(propCladeCommunity=Gardnerella_vaginalis) %>%
  mutate(Clade="Total Gardnerella")

gardRelativeAbundance1 %>%
  select(-Gardnerella_vaginalis, -propClade, -n) %>%
  full_join(GardAb) %>%
  mutate(Clade=factor(Clade, levels=cladeLevelsT)) %>%
ggplot(aes(x=Clade, y=propCladeCommunity, color=Clade)) +
  geom_boxplot(show.legend = FALSE, outlier.alpha = 0) +
  geom_jitter() + 
  cladeColorsNTotal +
  scale_x_discrete(labels=c("C1", "C2", "C3", "C4", "C5", "C6", "Uncharacterized", bquote('Total'~italic(Gardnerella)))) +
  facet_grid(Cohort~.)+
  labs(x="Clade", y="Proportion of community") +
  theme(axis.text.x = element_text(angle=45, hjust=1))
```

#### Clade:Clade Ratios
Observe total proportion of other *Gardnerella* clades in the community vs. proportion of each clade
```{r}
ggplot(gardRelativeAbundance1, aes(x=propCladeCommunity, y=Gardnerella_vaginalis-propCladeCommunity, color=Clade)) +
  geom_point() +
  #scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73"), labels=c("C1", "C2", "C3", "C4", "C5", "C6")) +
  cladeColorsN +
  labs(x="Proportion of Gardnerella clade in community", y="Proportion of other Gardnerella clades in the community", color="Clade") + 
  facet_grid(Cohort~Clade) +
  theme(axis.text.x = element_text(angle=45))
```  
  
  Not the best plot for visualizing the data, but it does show that when the total there is one clade of Garnerella in the community and the community is nearly 100% *Gardnerella*, that clade is clade 1. 
    
  Ratios of clades to other clades (clade-positive samples only)
```{r}
cladexclade <- gardRelativeAbundance1 %>%
  filter(SampleID %in% gardCladePositveSamples) %>%
  select(-propClade, -n) %>%
  spread(Clade, propCladeCommunity) %>%
  mutate(x1=`C1`,
         x2=`C2`,
         x3=`C3`,
         x4=`C4`,
         x5=`C5`,
         x6=`C6`) %>%
  gather("Clade1","propClade1", c(`C1`,`C2`,`C3`,`C4`,`C5`,`C6`)) %>%
  gather("Clade2", "propClade2", c(x1, x2, x3, x4, x5, x6))%>%
  mutate(Clade2=case_when(Clade2=="x1"~"C1",
                          Clade2=="x2"~"C2",
                          Clade2=="x3"~"C3",
                          Clade2=="x4"~"C4",
                          Clade2=="x5"~"C5",
                          Clade2=="x6"~"C6")) %>%
  filter(Clade1!=Clade2)

cladexclade %>%
  filter(Cohort=="Stanford"|Cohort=="UAB") %>%
ggplot(aes(x=Clade1, y=propClade1/propClade2, color=Clade2, fill=Clade2))+
  geom_violin(draw_quantiles = 0.5, alpha=0.5) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73"), labels=c("C1", "C2", "C3", "C4", "C5", "C6")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73"), labels=c("C1", "C2", "C3", "C4", "C5", "C6")) +
  scale_y_log10() +
  facet_wrap(~Cohort) +
  labs(title="Clade : Clade ratios", x="Clade A", y="Clade A : Clade B", color="Clade B", fill="Clade B")
```
  
  No clear patterns.
  
  Observe proportion of each clade in each sample vs. term/preterm delivery
```{r}
ggplot(data=gardRelativeAbundance2, aes(x=TermPre, y=propCladeCommunity, color=TermPre)) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0) +
  scale_y_log10(limits=c(1e-03, 1e03)) +
  scale_x_discrete(labels=c("Preterm", "Term")) +
  binaryColors +
  facet_grid(Cohort~Clade) +
  labs(x=element_blank(), y="Proportion clade in community") +
  theme(legend.position = "none")
```
  
Proportion of community vs preterm/term births faceted by clade. Does not suggest differences in proportion of each clade in each sample and term vs. preterm births. Perhaps the time series information is important here...
  
Look at two specific samples:
```{r}
ggplot(gardRelativeAbundance2[gardRelativeAbundance2$SubjectID=="10031",], aes(x=GWcoll, y=propCladeCommunity, color=Clade, fill=Clade))+
  geom_bar(stat = "identity")+
  cladeColors +
  labs(title = "Subject 10031", subtitle = "Preterm")

ggplot(gardRelativeAbundance2[gardRelativeAbundance2$SubjectID=="10621",], aes(x=GWcoll, y=propCladeCommunity, color=Clade, fill=Clade))+
  geom_bar(stat = "identity") +
  cladeColors +
  labs(title="Subject 10621", subtitle = "Term", x="Gestation weeks at collection", y="Proportion of community")
```
  
Not going to draw any major conclusions from these two subjects (small n), but I just wanted to take a closer look since they have the highest proportion of *Gardnerella* and specifically clade C1. 

Prevalence of each clade vs ptb (proportion samples in each group with/without clade)
```{r}
gardRelativeAbundance2 %>%
  mutate(pres=case_when(propCladeCommunity>0~1,
                        propCladeCommunity==0~0)) %>%
  group_by(Cohort, TermPre, Clade) %>%
  summarise(Prevalence=mean(pres)) %>%
ggplot(., aes(x=TermPre, y=Prevalence, fill = TermPre)) +
  geom_col() +
  scale_x_discrete(labels=c("Preterm", "Term")) +
  binaryColors +
  facet_grid(Cohort~Clade) + 
  labs(x="Preterm vs. Term") +
  theme(legend.position = "none")
```


### Clade presence
Clade presence and overall *Gardnerella* relative abundance
```{r}
ggplot(gardRelativeAbundance2, aes(x=(propCladeCommunity == 0), y=Gardnerella_vaginalis, color=(propCladeCommunity>0))) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Cohort~Clade) +
  labs(x="Clade presence", y="Proportion all Gv in community proportion of Gardnerella")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  theme(legend.position = "none")

gardRelativeAbundance2 %>%
  filter(Cohort=="Stanford") %>%
ggplot(aes(x=(propCladeCommunity == 0), y=Gardnerella_vaginalis, color=(propCladeCommunity>0))) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_wrap(~Clade, nrow=1) +
  labs(x="Clade presence", y="Total proportion Gv in community")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  scale_y_log10() +
  theme(legend.position = "none")
#ggsave("/Users/hlberman/Desktop/GvProps.png", height=3, width = 7)
```
  
  Does presence of clade C1 predict relative abundance of *Gardnerella*? 
  
Ratio of each clade to other *G. vaginalis* clades
```{r}
gardRelativeAbundance2 %>%
  mutate(raioGard=propCladeCommunity/(Gardnerella_vaginalis-propCladeCommunity)) %>%
ggplot(aes(x=Clade, y=raioGard, color=Clade)) +
  geom_boxplot(show.legend = FALSE, outlier.alpha = 0) +
  geom_jitter() +
  cladeColors +
  scale_y_log10(limits=c(1e-03, 1e03)) +
  facet_grid(Cohort~.) +
  labs(x= "Clade", y= bquote(italic(Gardnerella)~'Clade : Other Clades (Ratio)'))
```
  
Ratio of gard to other Gards vs gard frequency(proportion)
```{r}
gardRelativeAbundance2 %>%
  mutate(ratioGard=propCladeCommunity/(Gardnerella_vaginalis-propCladeCommunity)) %>%
ggplot(aes(y=ratioGard, x=Gardnerella_vaginalis, color=Clade)) +
  geom_jitter() +
  cladeColors +
  scale_y_log10(limits=c(1e-04,1e+04), breaks=c(1e-03, 1e-01, 1e01, 1e+03)) +
  facet_grid(Cohort~Clade) +
  #geom_smooth(method = "lm", show.legend = FALSE) +
  labs(y=bquote(italic(Gardnerella)~'Clade : Other Clades (Ratio)'), x=bquote('Total'~italic(Gardnerella)~'(Frequency)')) +
  theme(axis.text.x = element_text(angle=45))
```
  
Should test for odds ratio of clade C1 presence and PTB risk. (The selection of samples may hinder our ability to do this. Should we focus more on clade presence and overall relative abundance of *Gardnerella*). 

### Bar plots
Remove 0 values and replace with NA so they don't show up on the barplot
```{r}
gardRelativeAbundance3 <- gardRelativeAbundance1 %>%
  mutate(propCladeCommunity=na_if(propCladeCommunity, 0),
         Clade=case_when(Clade=="Uncharacterized"~"Clade uncharacterized",
                         Clade!="Uncharacterized"~Clade))
```

Plot
```{r}
gardRelativeAbundance3 %>%
  dplyr::filter(Cohort=="UAB") %>%
ggplot(mapping=aes(x=reorder(SampleID, -propCladeCommunity, FUN=sum, na.rm=TRUE), y=propCladeCommunity, fill=Clade, color=Clade)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), size=4, shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), size=4, shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), size=4, shape="square", show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#000000", "#999999"), labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6", "Preterm", "Term")) +
  theme_classic() +
  facet_wrap(~Cohort) +
  #ggtitle("UAB") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank())
  

gardRelativeAbundance3 %>%
  filter(Cohort=="Stanford") %>%
ggplot(mapping=aes(x=reorder(SampleID, -propCladeCommunity, FUN=sum, na.rm=TRUE), y=propCladeCommunity, fill=Clade, color=Clade)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), size=3, shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), size=3, shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), size=3, shape="square", show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000", "#FFFFFF"), labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6", "Clade uncharacterized", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000", "#FFFFFF"), labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6", "Clade uncharacterized", "Preterm", "Term")) +
  theme_classic() +
  facet_wrap(~Cohort) +
  #ggtitle("Stanford") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank())
```
  
  Again apparrent potential for Clade C1 and sample dominance. 

  Separate bars by subject:
```{r}
ggplot(data=gardRelativeAbundance3, mapping=aes(x=reorder(SampleID, TermPre=="T"), y=propCladeCommunity, fill=Clade, color=Clade)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000", "#FFFFFF"), labels=c("Clade 1", "Clade 2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Clade uncharacterized", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000", "#FFFFFF"), labels=c("Clade 1", "Clade 2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "Clade uncharacterized", "Preterm", "Term")) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank()) +
  geom_line(aes(x=reorder(as.numeric(SubjectID), TermPre=="T", na.rm=TRUE), y=0))
```
  
## Subject averages
Caclulate average proportion of each clade per subject
```{r}
#with uncharacterized G. vaginalis category
cladeAverages <- gardRelativeAbundance1 %>%
  data.frame(stringsAsFactors=FALSE) %>%
  group_by(Clade, SubjectID) %>%
  mutate(cladeAvg=mean(propCladeCommunity),
         gardAvg=mean(Gardnerella_vaginalis)) %>%
  ungroup() %>%
  mutate(Clade=case_when(Clade=="Uncharacterized"~"Clade uncharacterized",
                         Clade!="Uncharacterized"~Clade)) %>%
  select(SubjectID, Clade, cladeAvg, gardAvg, TermPre, Cohort, GWdell) %>%
  unique.data.frame()

#without uncharacterized G. vaginalis category
cladeAverages1 <- cladeAverages %>%
  filter(Clade!="Clade uncharacterized")
```

Bar plot of average clade proportion per subject.
```{r}
cladeAverages %>%
  mutate(cladeAvg=na_if(cladeAvg, 0)) %>%
ggplot(mapping=aes(x=reorder(reorder(SubjectID, -cladeAvg, FUN=sum, na.rm=TRUE), TermPre=="T", na.rm=TRUE), y=cladeAvg, fill=Clade, color=Clade)) +
  geom_bar(stat = "identity") +
  labs(x="Subject", y="Proportion", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", size=5.5, show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", size=5.5, show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", size=5.5, show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000", "#FFFFFF"), labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6", "Clade uncharacterized", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "#000000", "#FFFFFF"), labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6", "Clade uncharacterized", "Preterm", "Term")) +
  facet_wrap(~Cohort, scales = "free_x") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank())
```
  
  Average relative abundance of each clade through gestation vs. preterm/term:
```{r}
ggplot(data=cladeAverages1, aes(x=TermPre, y=cladeAvg, color=TermPre)) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0)+
  scale_x_discrete(labels=c("Preterm", "Term")) +
  scale_y_log10(limits=c(1e-03, 1e03)) +
  binaryColors +
  facet_grid(Cohort~Clade) +
  labs(x=element_blank(), y="Proportion of community") +
  theme(legend.position = "none")
```
  May not be any differences in PTB risk by proportion of each clade in the entire community. Perhaps presence is a better predictor than proportion... 
  If the clade was present at any time in the subject:
```{r}
ggplot(cladeAverages1, aes(x=(cladeAvg==0), fill=TermPre)) +
  geom_bar(position="dodge") +
  scale_fill_manual(values=c("#D55E00", "#56B4E9"), labels=c("Preterm", "Term")) +
  facet_grid(Cohort~Clade) +
  labs(x="Clade presence", fill="Delivery") +
  scale_x_discrete(labels=c("Present", "Absent")) +
  theme(axis.text.x = element_text(angle=90))
```
  Could clade C1 or C2 presence increase risk for PTB?
  It appears that all subjects in the UAB cohort had *Gardnerella* clades 1, 2, and 4. 
  
  Subject average proportion of other *Gardnerella* clades in the community vs. proportion of each clade
```{r}
ggplot(data=cladeAverages, aes(x=cladeAvg, y=gardAvg-cladeAvg)) +
  geom_point() +
  cladeColorsN +
  labs(x="Proportion of Gardnerella clade in community", y="Proportion of other Gardnerella clades in the community", color="Clade") + 
  facet_grid(Cohort~Clade) +
  theme(axis.text.x = element_text(angle=45))
```

## Dominant *Gardnerella* clade over time
Look at the dominant *Gardnerella* clade in each sample (of those that have *Gardnerella*)
```{r}
gardMax <- gardRelativeAbundance1 %>%
  group_by(SampleID) %>%
  filter(propClade==max(propClade)) %>%
  ungroup()

ggplot(gardMax) +
  geom_point(aes(x=floor(GWcoll), y=reorder(SubjectID, GWdell), color=Clade)) +
  geom_point(aes(y=SubjectID, x=GWdell+.5), shape=4, color="black") +
  #geom_point(aes(x=0, y=SubjectID, color=Cohort)) +
  cladeColorsN +
  geom_vline(xintercept = 37.5) +
  labs(x="Gestation weeks at collection", y="Subject ID", title="Most abundant Gardnerella clade")
```
  X marks gestation weeks at delivery. Vertical line is 37 weeks. 

  I think we need to test for clade 1 and/or 2 presence (especially clade 1) and total proportion of *Gardnerella* in the sample and also PTB risk. Without doing any statistics  yet, it seems that presence of clade 1 predicts total proportion of *Gardnerella*. My initial hunch is that when you have this "BV-like" microbiome, abundance in *Gardnerella*, it is likely that these "pathogenic" strains are present, but you also may have these "non-pathogenic" strains in the community as well, and they may be more abundant than the "pathogenic" strains. The genomics information will be very interesting. Metatranscriptomic data would also be *very* interesting if it is true that these strains can reshape the community even if they are not the dominant strain of *Gardnerella* in the community!!

# G. vaginalis clades vs. Lactobacillus spp.
## Proportion Gard clade vs. proportion Lacto
```{r}
LGabundance <- gardRelativeAbundance1 %>%
  left_join(mDF[,c("SampleID", "Lactobacillus_crispatus", "Lactobacillus_iners", "Lactobacillus_jensenii", "Lactobacillus_gasseri")], by="SampleID") %>%
  gather("Lactobacillus", "LAbundance", c(Lactobacillus_crispatus, Lactobacillus_iners, Lactobacillus_jensenii, Lactobacillus_gasseri)) %>%
  mutate(LAbundance=LAbundance, 
         Lactobacillus=case_when(Lactobacillus=="Lactobacillus_crispatus"~"L. crispatus", 
                                 Lactobacillus=="Lactobacillus_gasseri"~"L. gasseri", 
                                 Lactobacillus=="Lactobacillus_iners"~"L. iners", 
                                 Lactobacillus=="Lactobacillus_jensenii"~"L. jensenii"))

LGabundanceNoUnch <- LGabundance %>%
  filter(Clade!="Uncharacterized")  

LGabundanceNoUnch %>%
  filter(Cohort=="Stanford") %>%
ggplot(aes(x=propCladeCommunity, y=LAbundance, color=Lactobacillus)) +
  geom_point() +
  facet_grid(Lactobacillus~Clade) +
  coord_fixed() +
  scale_color_manual(values=c("darkgreen", "gold3", "darkblue", "darkred")) +
  labs(title="Stanford", x="Proportion Gardnerella clade", y="Proportion Lactobacillus")  
#ggsave("/Users/hlberman/Desktop/test1.pdf", height = 8, width=15)

LGabundanceNoUnch %>%
  filter(Cohort=="UAB") %>%
ggplot(aes(x=propCladeCommunity, y=LAbundance, color=Lactobacillus)) +
  geom_point() +
  facet_grid(Lactobacillus~Clade) +
  coord_fixed() +
  scale_color_manual(values=c("darkgreen", "gold3", "darkblue", "darkred")) +
  labs(title="UAB", x="Proportion Gardnerella clade", y="Proportion Lactobacillus")  
#ggsave("/Users/hlberman/Desktop/test2.pdf", height = 8, width=15)
```
  
Mainly only see *L. iners* in UAB samples, as we have seen previously. Clear inverse relationship between most clades and *Lactobacillus* species, except clade 4 and *L. gasseri* in Stanford, and clades 1 and 3 and *L. iners* in UAB cohort.

## Ratio Gard clade to Lactos
```{r}
LGabundanceNoUnch %>%
  filter(Cohort=="UAB") %>%
ggplot(aes(x=Clade, y=propCladeCommunity/LAbundance, color=Lactobacillus)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter() +
  facet_grid(Lactobacillus~.) +
  scale_y_log10() +
  scale_color_manual(values=c("darkgreen", "gold3", "darkblue", "darkred")) +
  labs(title="UAB", y="G. vaginalis : Lactobacillus sp.")  
#ggsave("/Users/hlberman/Desktop/test1.pdf", height = 8, width=15)

LGabundanceNoUnch %>%
  filter(Cohort=="Stanford") %>%
ggplot(aes(x=Clade, y=propCladeCommunity/LAbundance, color=Lactobacillus)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter() +
  facet_grid(Lactobacillus~.) +
  scale_color_manual(values=c("darkgreen", "gold3", "darkblue", "darkred")) +
  scale_y_log10(limits=c(1e-02, 1e02)) +
  labs(title="Stanford", y="G. vaginalis : Lactobacillus sp.")    
#ggsave("/Users/hlberman/Desktop/test2.pdf", height = 8, width=15)
```

Ratio Gard clades to Lcjg
```{r}
LGabundanceNoUnch %>%
  spread(Lactobacillus, LAbundance) %>%
  mutate(Lcjg=`L. crispatus`+`L. gasseri`+`L. jensenii`) %>%
 ggplot(aes(x=Clade, y=propCladeCommunity/Lcjg)) +
  geom_boxplot(outlier.alpha = 0)+
  geom_jitter()+
  facet_grid(Cohort~.) +
  scale_y_log10()
```

## Gard clade presence vs proportion Lactobacillus spp.
```{r}
LGabundanceNoUnch %>%
  filter(Cohort=="Stanford") %>%
ggplot(aes(x=(propCladeCommunity == 0), y=LAbundance, color=(propCladeCommunity==0))) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Lactobacillus~Clade) +
  labs(title="Stanford", x="Clade presence", y="Proportion of Lactobacillus")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  scale_y_log10() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle=90))

LGabundanceNoUnch %>%
  filter(Cohort=="UAB") %>%
ggplot(aes(x=(propCladeCommunity == 0), y=LAbundance, color=(propCladeCommunity==0))) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Lactobacillus~Clade) +
  labs(title="UAB", x="Clade presence", y="Proportion of Lactobacillus")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  scale_y_log10() +
  theme(legend.position = "none", 
        axis.text.x = element_text(angle=90))
```
  
Difficult to make any conclusions in the UAB cohort, because most samples do not have species other than *L. iners*. In the Stanford cohort, *L. crispatus* relative abundance may be greater without any *Gardnerella* clade. May not be a difference between *L. iners* relative abundance with and without *Gardnerlla* clades. 

## Gard presence vs. all Lactobacillus
```{r}
LGabundanceNoUnch %>%
  group_by(SampleID, Clade) %>%
  mutate(TLAbundance=sum(LAbundance)) %>%
  ungroup %>%
ggplot(aes(x=(propCladeCommunity == 0), y=TLAbundance, color=(propCladeCommunity==0))) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Clade~Cohort) +
  labs(title="Proportion all Lactobacillus vs. Clade Presence", x="Clade presence", y="Proportion of Lactobacillus")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  #scale_y_log10() +
  theme(legend.position = "none")

LGabundanceNoUnch %>%
  spread(Lactobacillus, LAbundance) %>%
  mutate(Lcjg=`L. crispatus`+`L. gasseri`+`L. jensenii`) %>%
ggplot(aes(x=(propCladeCommunity == 0), y=Lcjg, color=(propCladeCommunity==0))) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Cohort~Clade) +
  labs(title="Proportion L. crispatus, jensenii, gasserii vs. Clade Presence", x="Clade presence", y="Proportion of Lactobacillus cristpatus, jensenii, gasseri")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  scale_y_log10() +
  theme(legend.position = "none")

LGabundanceNoUnch %>%
  spread(Lactobacillus, LAbundance) %>%
ggplot(aes(x=(propCladeCommunity == 0), y=`L. iners`, color=(propCladeCommunity==0))) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Cohort~Clade) +
  labs(title="Proportion L. iners vs. Clade Presence", x="Clade presence", y="Proportion of L. iners")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  scale_y_log10() +
  theme(legend.position = "none")

Gv <- LGabundanceNoUnch %>%
  filter(Cohort=="Stanford") %>%
  select(SampleID, Gardnerella_vaginalis, Clade, propCladeCommunity) %>%
  dplyr::rename(Proportion=Gardnerella_vaginalis) %>%
  mutate(Taxon="Gardnerella vaginalis") %>%
  unique.data.frame()

LGabundanceNoUnch %>%
  spread(Lactobacillus, LAbundance) %>%
  mutate(Proportion=`L. crispatus`+`L. gasseri`+`L. jensenii`) %>%
  filter(Cohort=="Stanford") %>%
  select(SampleID, Clade, Proportion, propCladeCommunity) %>%
  mutate(Taxon="Lactobacillus") %>%
  full_join(Gv, by=c("SampleID", "Proportion", "Taxon", "Clade", "propCladeCommunity")) %>%
ggplot(aes(x=(propCladeCommunity == 0), y=Proportion, color=(propCladeCommunity==0))) +
  geom_jitter() +
  geom_boxplot(alpha=0, outlier.alpha = 0) + 
  scale_color_manual(values=c("#56B4E9", "#D55E00"), labels=c("Present", "Absent")) +
  facet_grid(Taxon~Clade) +
  labs(x="Clade presence", y="Proportion of community")+
  scale_x_discrete(labels=c("Present", "Absent")) +
  scale_y_log10() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle=90))
#ggsave("/Users/hlberman/Desktop/cladevlacto.png", width=7, height = 3)  
```
  
  Ratio of *G. vaginalis* clades to *L. crispatus*, *L. jensenii*, and *L. gasseri* vs. the sum of *G. vaginalis* and these *Lactobacillus* spp.
```{r, fig.height = 2, fig.width=6}
ratioFreqTable <- LGabundanceNoUnch %>%
  select(-n, -propClade) %>%
  spread(Lactobacillus, LAbundance) %>%
  spread(Clade, propCladeCommunity) %>%
  dplyr::rename(`Total Gardnerella`=Gardnerella_vaginalis) %>%
  gather("Clade", "propCladeCommunity", c(`Total Gardnerella`, C1, C2, C3, C4, C5, C6))

ratioFreqTable %>%
  mutate(Y=propCladeCommunity/(`L. crispatus`+`L. jensenii`+`L. gasseri`),
         X=propCladeCommunity+`L. crispatus`+`L. jensenii`+`L. gasseri`,
         Clade=factor(Clade, levels=cladeLevelsT)) %>%
  ggplot(aes(y=Y, x=X, color=Clade)) +
    geom_point() +
    facet_grid(Cohort~Clade) +
    scale_y_log10(limits=c(1e-04, 1e04), breaks=c(1e-04, 1e-02, 1, 1e02, 1e04), labels= c(expression(1:10^4), expression(1:10^2), "1:1", expression(10^2:1), expression(10^4:1))) +
    cladeColorsTotal+  
    labs(x=bquote(italic(Gardnerella)~'+'~italic(L.)~italic(crisp.)~italic(jens.)~italic(gas.)~'(Frequency)'), y=bquote(italic(Gardnerella)~':'~italic(L.)~italic(crisp.)~italic(jens.)~italic(gas.)~'(Ratio)')) +
    theme(axis.text.x = element_text(angle=45, size=2))
#ggsave("/Users/hlberman/Desktop/test.jpg", height = 3, width=6)
```
  Ratios vs. sum of these relationships by individual *Lactobacillus* sp.
  *L. crispatus*:
```{r, fig.height = 2, fig.width=6}
ratioFreqTable %>%
  mutate(Clade=factor(Clade, levels=cladeLevelsT)) %>%
 ggplot(aes(y=(propCladeCommunity/`L. crispatus`), x=(propCladeCommunity+`L. crispatus`), color=Clade)) +
    geom_point() +
    facet_grid(Cohort~Clade) +
    scale_y_log10(limits=c(1e-04, 1e04), breaks=c(1e-04, 1e-02, 1, 1e02, 1e04), labels=c(expression(1:10^4), expression(1:10^2), "1:1", expression(10^2:1), expression(10^4:1))) +
    cladeColorsTotal +
    labs(x=bquote(italic(Gardnerella)~'+'~italic(L.)~italic(crispatus)~'(Frequency)'), y=bquote(italic(Gardnerella)~':'~italic(L.)~italic(crispatus)~'(Ratio)')) +
    theme(axis.text.x = element_text(angle=45))
```
  *L. jensenii*:
```{r, fig.height = 2, fig.width=6}
ratioFreqTable %>%
  mutate(Clade=factor(Clade, levels=cladeLevelsT)) %>%
 ggplot(aes(y=(propCladeCommunity/`L. jensenii`), x=(propCladeCommunity+`L. jensenii`), color=Clade)) +
    geom_point() +
    facet_grid(Cohort~Clade) +
    scale_y_log10(limits=c(1e-04, 1e04), breaks=c(1e-04, 1e-02, 1, 1e02, 1e04), labels=c(expression(1:10^4), expression(1:10^2), "1:1", expression(10^2:1), expression(10^4:1))) +
    cladeColorsNTotal +  
    labs(x=bquote(italic(Gardnerella)~'+'~italic(L.)~italic(jensenii)~'(Frequency)'), y=bquote(italic(Gardnerella)~':'~italic(L.)~italic(jensenii)~'(Ratio)')) +
    theme(axis.text.x = element_text(angle=45))
```
  *L. gasseri*:
```{r, fig.height = 2, fig.width=6}
ratioFreqTable %>%
  mutate(Clade=factor(Clade, levels=cladeLevelsT)) %>%
 ggplot(aes(y=(propCladeCommunity/`L. gasseri`), x=(propCladeCommunity+`L. gasseri`), color=Clade)) +
    geom_point() +
    facet_grid(Cohort~Clade) +
    scale_y_log10(limits=c(1e-04, 1e04), breaks=c(1e-04, 1e-02, 1, 1e02, 1e04), labels=c(expression(1:10^4), expression(1:10^2), "1:1", expression(10^2:1), expression(10^4:1))) +
    cladeColorsTotal +  
    labs(x=bquote(italic(Gardnerella)~'+'~italic(L.)~italic(gasseri)~'(Frequency)'), y=bquote(italic(Gardnerella)~':'~italic(L.)~italic(gasseri)~'(Ratio)')) +
    theme(axis.text.x = element_text(angle=45))
```
  *L. iners*:
```{r, fig.height = 2, fig.width=6}
ratioFreqTable %>%
  mutate(Clade=factor(Clade, levels=cladeLevelsT)) %>%
 ggplot(aes(y=(propCladeCommunity/`L. iners`), x=(propCladeCommunity+`L. iners`), color=Clade)) +
    geom_point() +
    facet_grid(Cohort~Clade) +
    scale_y_log10(limits=c(1e-04, 1e04), breaks=c(1e-04, 1e-02, 1, 1e02, 1e04), labels=c(expression(1:10^4), expression(1:10^2), "1:1", expression(10^2:1), expression(10^4:1))) +
    cladeColorsTotal +
    labs(x=bquote(italic(Gardnerella)~'+'~italic(L.)~italic(iners)~'(Frequency)'), y=bquote(italic(Gardnerella)~':'~italic(L.)~italic(iners)~'(Ratio)'))
    theme(axis.text.x = element_text(angle=45))
```

## Bar plot
  With Term/Preterm
```{r}
LGabundance0 <- LGabundance %>%
  select(-n,-propClade) %>%
  spread(Clade, propCladeCommunity) %>%
  spread(Lactobacillus, LAbundance) %>%
  gather("Taxon", "Abundance", `C1`, `C2`, `C3`, `C4`, `C5`, `C6`, `Uncharacterized`, `L. crispatus`, `L. gasseri`, `L. iners`, `L. jensenii`) %>%
  filter(!is.na(Abundance)) %>%
  mutate(Taxon=case_when(Taxon=="Uncharacterized"~"Clade uncharacterized",
                         Taxon!="Uncharacterized"~Taxon))


LGabundance0 %>%
  filter(Cohort=="Stanford") %>%
ggplot(aes(x=reorder(reorder(SampleID, -Abundance, FUN=sum, na.rm=TRUE), TermPre=="T", na.rm=TRUE), y=Abundance, fill=Taxon, color=Taxon)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", size=2.5, show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", size=2.5, show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", size=2.5, show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#FFFFFF"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "Preterm", "Term")) +
 scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#FFFFFF"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "Preterm", "Term")) +
  #ggtitle("Stanford") +
  facet_wrap(~Cohort) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank()) 


LGabundance0 %>%
  filter(Cohort=="UAB") %>%
ggplot(aes(x=reorder(reorder(SampleID, -Abundance, FUN=sum, na.rm=TRUE), TermPre=="T", na.rm=TRUE), y=Abundance, fill=Taxon, color=Taxon)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Legend", fill="Legend") +
  geom_point(aes(y=-.03, color=TermPre, fill=TermPre), shape="square", size=3.5, show.legend = FALSE) +
  geom_point(aes(y=-.04, color=TermPre, fill=TermPre), shape="square", size=3.5, show.legend = FALSE) +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", size=3.5, show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#FFFFFF"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "Preterm", "Term")) +
 scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#FFFFFF"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "Preterm", "Term")) +
  #ggtitle("UAB") +
  facet_wrap(~Cohort) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank()) 
```

## Subject averages
  Caclulate average proportion of each clade per subject
```{r}
taxAverages <- LGabundance0 %>%
  data.frame(stringsAsFactors=FALSE) %>%
  group_by(Taxon, SubjectID) %>%
  mutate(taxAvg=mean(Abundance)) %>%
  ungroup() %>%
  select(SubjectID, Taxon, taxAvg, TermPre, Cohort, GWdell) %>%
  unique.data.frame() %>%
  mutate(Taxon=factor(Taxon, levels = c("C1", "C2", "C3", "C4", "C5", "C6", "Clade uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii")))
```

  Bar plot of average clade proportion per subject.
```{r}
taxAverages %>%
    mutate(taxAvg=na_if(taxAvg, 0)) %>%
    filter(Cohort=="Stanford") %>%
ggplot(mapping=aes(x=reorder(reorder(SubjectID, -taxAvg, FUN=sum, na.rm=TRUE), TermPre=="T", na.rm=TRUE), y=taxAvg, fill=Taxon, color=Taxon)) +
  geom_bar(stat = "identity") +
  labs(x="Subject", y="Proportion", color="Legend", fill="Legend") +
  geom_point(aes(y=-.05, color=TermPre, fill=TermPre), shape="square", size=8.25, show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#FFFFFF"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "#A9A9A9", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#FFFFFF"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "G. vaginalis uncharacterized", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "Preterm", "Term")) +
  #ggtitle("Stanford") +
  facet_wrap(~Cohort) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank())

taxAverages %>%
    mutate(taxAvg=na_if(taxAvg, 0)) %>%
    filter(Cohort=="UAB") %>%
ggplot(mapping=aes(x=reorder(reorder(SubjectID, -taxAvg, FUN=sum, na.rm=TRUE), TermPre=="T", na.rm=TRUE), y=taxAvg, fill=Taxon, color=Taxon)) +
  geom_bar(stat = "identity") +
  labs(x="Subject", y="Proportion", color="Legend", fill="Legend") +
  geom_point(aes(y=-.065, color=TermPre, fill=TermPre), shape="square", size=10.5, show.legend = FALSE) +
  scale_color_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#FFFFFF"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "Preterm", "Term")) +
  scale_fill_manual(values=c("#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73", "darkgreen", "#F0E442", "darkblue", "darkred", "#000000", "#FFFFFF"), labels=c("G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "Preterm", "Term")) +
  #ggtitle("UAB") +
  facet_wrap(~Cohort) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust=1),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank())
```
  
# Key Taxa Barplot
## Create vagitype df for ordering barplot
  Set up dataframes for vagitype classification (max taxon. If maximum taxon is <30%, then vagitype is "no type")
```{r}
gard4metaphlan <- gardRelativeAbundance1 %>%
  select(SampleID, Clade, propCladeCommunity) %>%
  mutate(Clade=paste0("Gardnerella_vaginalis_", Clade)) %>%
  spread(Clade, propCladeCommunity)
head(gard4metaphlan)
dim(gard4metaphlan)

mDFG <- mDF %>%
  select(-Gardnerella_vaginalis) %>%
  left_join(gard4metaphlan, by="SampleID")
```

Get types and add to sgDF
```{r}
mDFG0 <- mDFG %>%
  select(-SampleID)

vType <- mDFG0 %>%
  apply(1, which.max)
vTypeName <- colnames(mDFG0)[vType]
maxAb <- mDFG0 %>%
  apply(1, max)
  
vtypeDF <- data.frame(SampleID=mDFG[,"SampleID"],
                      Vagitype=vTypeName,
                      maxAb=maxAb) %>%
  mutate(Vagitype=case_when(maxAb<0.3~"No type",
                            maxAb>=0.3~as.character(Vagitype)),
         Vagitype=case_when(Vagitype=="Atopobium_vaginae"~"A. vaginae",
                        Vagitype=="Gardnerella_vaginalis_C1"~"G. vaginalis C1",
                        Vagitype=="Gardnerella_vaginalis_C2"~"G. vaginalis C2",
                        Vagitype=="Gardnerella_vaginalis_C3"~"G. vaginalis C3", 
                        Vagitype=="Gardnerella_vaginalis_C4"~"G. vaginalis C4",
                        Vagitype=="Gardnerella_vaginalis_C5"~"G. vaginalis C5", 
                        Vagitype=="Gardnerella_vaginalis_C6"~"G. vaginalis C6",
                        Vagitype=="Lactobacillus_crispatus"~"L. crispatus",
                        Vagitype=="Lactobacillus_gasseri"~"L. gasseri",
                        Vagitype=="Lactobacillus_iners"~"L. iners",
                        Vagitype=="Lactobacillus_jensenii"~"L. jensenii", 
                        Vagitype=="Prevotella_bivia"~"P. bivia",
                        Vagitype=="No type"~"No type"))
#join data with sgDF 
sgDF1 <- sgDF0 %>%
  left_join(vtypeDF, by="SampleID")
colnames(sgDF0)
```

Descriptives on vagitypes
```{r}
ggplot(data=sgDF1, aes(x=Vagitype)) +
  geom_histogram(stat="count") +
  facet_wrap(~Cohort) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=45, hjust=1))
```

Gard type df
```{r}
gardTypeDF <- gardRelativeAbundance1 %>%
  mutate(propClade=case_when(propCladeCommunity==0~0, #1050801318 has one read in one clade but shows up as zero Gard based on metaphlan results
                             propCladeCommunity>0&propClade==0~1, #4 samples with uncharacterized Gardnerella
                             propCladeCommunity>0&propClade>0~propClade)) %>%
  group_by(SampleID) %>%
  filter(propClade==max(propClade)) %>%
  ungroup() %>%
  mutate(Clade=case_when(propClade==0~"No Gardnerella",
                         propClade==0.50~"No type",
                         propClade>0&propClade<0.30~"No type",
                         propClade>0.30&propClade!=0.50~Clade)) %>%
  select(SampleID, SubjectID, Clade, Cohort) %>% 
  unique.data.frame() %>%
  dplyr::rename(gardType=Clade) %>%
  mutate(gardType=case_when(gardType=="No Gardnerella"~"No Gardnerella",
                            gardType=="No type"~"No type",
                            gardType=="C1"~"G. vaginalis C1",
                            gardType=="C2"~"G. vaginalis C2",
                            gardType=="C3"~"G. vaginalis C3",
                            gardType=="C4"~"G. vaginalis C4",
                            gardType=="C5"~"G. vaginalis C5",
                            gardType=="C6"~"G. vaginalis C6",
                            gardType=="Uncharacterized"~"G. vaginalis uncharacterized"))
```

Let's plot
```{r}
ggplot(data=gardTypeDF, aes(x=gardType)) +
  geom_histogram(stat="count") +
  facet_wrap(~Cohort) +
  #theme_classic() +
  theme(axis.text.x = element_text(angle=45, hjust=1))
```

## Key taxa barplot figure
```{r}
LGabundanceX <- LGabundance %>%
  select(-n,-propClade) %>%
  spread(Clade, propCladeCommunity) %>%
  spread(Lactobacillus, LAbundance) %>%
  left_join(mDF[,c("SampleID", "Atopobium_vaginae", "Prevotella_bivia")], by="SampleID") %>%
  gather("Taxon", "Abundance", `C1`, `C2`, `C3`, `C4`, `C5`, `C6`, `Uncharacterized`, `L. crispatus`, `L. gasseri`, `L. iners`, `L. jensenii`, `Atopobium_vaginae`, `Prevotella_bivia`) %>%
  left_join(vtypeDF[,c("SampleID", "Vagitype")], by="SampleID") %>%
  filter(!is.na(Abundance)) %>%
  mutate(vTypeOrder=case_when(Vagitype=="A. vaginae"~1,
                              Vagitype=="G. vaginalis C1"~2,
                              Vagitype=="G. vaginalis C2"~3,
                              Vagitype=="G. vaginalis C3"~4,
                              Vagitype=="G. vaginalis C4"~5,
                              Vagitype=="G. vaginalis C5"~6,
                              Vagitype=="G. vaginalis C6"~7,
                              Vagitype=="L. crispatus"~8,
                              Vagitype=="L. gasseri"~9,
                              Vagitype=="L. iners"~10,
                              Vagitype=="L. jensenii"~11,
                              Vagitype=="P. bivia"~12,
                              Vagitype=="No type"~13),
         Taxon=case_when(Taxon=="Uncharacterized"~"Clade uncharacterized",
                         Taxon!="Uncharacterized"~Taxon))

LGabundanceX$Vagitype <- factor(LGabundanceX$Vagitype, levels=c("A. vaginae", "G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. jensenii", "L. iners", "P. bivia", "No type"))

LGabundanceX %>%
ggplot(aes(x=reorder(reorder(SampleID, Abundance, FUN=sum, na.rm=TRUE), vTypeOrder, FUN=min), y=Abundance, fill=Taxon, color=Taxon, order=Vagitype)) +
  geom_bar(stat="identity") +
  labs(x="Sample", y="Proportion of community", color="Taxon", fill="Taxon") +
  taxColorsN+
  #scale_color_manual(values=c("mediumpurple4", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred", "darkseagreen4"), labels=c("A. vaginae", "G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "P. bivia")) +
  #scale_fill_manual(values=c("mediumpurple4", "#CC79A7", "#D55E00", "#56B4E9", "#0072B2", "#E69F00", "#009E73","darkgreen", "khaki1", "darkblue", "darkred", "darkseagreen4"), labels=c("A. vaginae", "G. vaginalis C1", "G. vaginalis C2", "G. vaginalis C3", "G. vaginalis C4", "G. vaginalis C5", "G. vaginalis C6", "L. crispatus", "L. gasseri", "L. iners", "L. jensenii", "P. bivia")) +
  facet_wrap(~Cohort, scales="free_x") +
  theme_grey() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank(),
        aspect.ratio = 1/1.5,
        legend.position = "bottom")
```

# Save Files
  Vagitype Dataframe
```{r}
vTypesClades <- vtypeDF %>%
  left_join(cladeCounts, by="SampleID") %>%
  left_join(gardTypeDF, by="SampleID")
#write_tsv(vTypesClades, "../../vagitypesCladeCounts.tsv")
```

  save Rdata of Metaphlan relative abundances
```{r}
gardAb <- gardRelativeAbundance1 %>%
  select(SampleID, Clade, n, propClade, propCladeCommunity, Gardnerella_vaginalis) %>%
  dplyr::rename(total_Gardnerella_vaginalis=Gardnerella_vaginalis)
#save(mDF, mDFG, gardAb, file=file.path("metaphlanAbundances.Rdata"))
```

# Session Info
```{r}
sessionInfo()
```
