---
title: "Build Bowtie DB"
author: "Hanna Berman"
date: "2/26/2019"
output: html_document
---

# Setup
##Load packages
```{r libraries, warning=FALSE}
library(tidyverse); packageVersion("tidyverse")
library(ShortRead);packageVersion("ShortRead")
library(Biostrings);packageVersion("Biostrings")
```

## Define file paths
```{r}
repo_dir <- "/Users/hlberman/Documents/GitHub/pregnancy_metagenome/"
gard_tree_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/Gard WGS"
roary_path <- file.path(gard_tree_dir, "20181008_GvRoary95strains", "gene_presence_absence.csv")
ref_ffns <- file.path(gard_tree_dir, "annotated_genomes_byext", "ffn")
ref_tsvs <- file.path(gard_tree_dir, "annotated_genomes_byext", "tsv")
output_dir <- "/Users/hlberman/Documents/GitHub/pregnancy_metagenome/gard_mapping/clade_assignments/ref_genes"
df_path <- file.path(repo_dir, "shotgunMetadata.tsv")
```

# Get Reference Genes
```{r}
strains <- str_extract(list.files(ref_ffns, full.names = FALSE, all.files = FALSE, pattern = ".ffn"), 
                       ".*(?=\\.ffn)")

#roary gene presence absence file and select core genes that are definitely unique
gnpa <- read_csv(roary_path)
refGenes <- gnpa %>%
  add_count(Annotation) %>%
  filter(`No. isolates`==72,
         n==1) %>%
  select(-n)

refSeqNames <- refGenes %>% #make a vector of all the names of reference genes for which we want to extract sequences
  select(strains) %>%
  c() %>%
  unlist
names(refSeqNames) <- NULL

#fasta seqs
allSeqs <- DNAStringSetList(lapply(list.files(ref_ffns, full.names = TRUE, all.files = FALSE, pattern = ".ffn"), 
            readDNAStringSet))
names(allSeqs) <- strains
allSeqs <- unlist(allSeqs) #make one DNA string set
allSeqs@ranges@NAMES <- str_extract(allSeqs@ranges@NAMES, "(?<=\\.).*_[0-9]*(?=\\ .*)") #remove extra strain name label and annotation from each DNA string name


##GET STRING DETECTION WORKING 
cladeRefSeqs <- allSeqs[allSeqs@ranges@NAMES %in% refSeqNames]
length(cladeRefSeqs) == nrow(refGenes)*length(strains) #check that there are the appropriate amount of reference sequences
```

# Save sequences and reference gene information
```{r}
write_csv(refGenes, file.path(output_dir, "refGenesDF.csv"))
writeFasta(cladeRefSeqs, file.path(output_dir, "refGenes.fasta"))
```

# Session Info
```{r}
sessionInfo()
```