---
title: "Make bam Rdata files"
author: "Hanna Berman"
date: "2/28/2019"
output: html_document
---

```{r libraries, warning=FALSE}
library(tidyverse); packageVersion("tidyverse")
library(GenomicAlignments);packageVersion("GenomicAlignments")
```

## Define file paths
```{r}
repo_dir <- "/Users/hlberman/Documents/GitHub/pregnancy_metagenome/"
output_dir <- file.path(repo_dir, "gard_mapping/clade_assignments")
bam_path <-file.path("/Users/hlberman/mnt/cluster/home/hlberman/VMMG/gard_map/filtered_out")
```

# Load and Save Bam files
```{r}
#list bam files
bamFiles <- list.files(bam_path, pattern = "_FgardRef_s.bam$", full.names = TRUE)

#make list of bam file objects
bamList <- map(bamFiles, BamFile)
bams <- bplapply(bamList, scanBam)
names(bams) <- list.files(bam_path, pattern = "_FgardRef_s.bam$", full.names = FALSE) %>%
  str_sub(1,10)

alns <- map(bamList, readGAlignmentPairs)
```

```{r}
#save(bamList, file=file.path(output_dir, "gardBams.RData"))
save(bams, file=file.path(output_dir, "gardBams.RData"))
save(alns, file=file.path(output_dir, "gardBamsAlns.RData"))
```

#SessionInfo
```{r}
sessionInfo()
```