---
title: "Shotgun Vs. 16S Gardnerella variants"
author: "Hanna Berman"
date: "12/10/2019"
output: html_document
---
# Set up
## packages
```{r}
library(tidyverse)
library(phyloseq)
library(Biostrings)
```

## Paths
```{r}
repo_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/pregnancy_metagenome"
google_drive_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/"

#shotgun metadata
sgDF <- file.path(repo_dir, "shotgunMetadata.tsv") %>%
  read_delim(delim = "\t") %>%
  mutate(SampleID=as.character(SampleID),
         SampleID_a=as.character(SampleID_a))

#ASVs from Callahan et al., 2017
S16_path <- file.path(google_drive_dir, "16S_data/RepRefine_Scripts/input/processed.rda")
load(S16_path)

#Gardnerella 16S fasta sequences from reference WGS strains
rRNAs16S <- readDNAStringSet(file.path(repo_dir, "gard_mapping", "clade_assignments", "16SrRNAs.fasta"))

#gardnerella metadata
refDF <- file.path(repo_dir, "GardnerellaMetadata.csv") %>%
  read_csv

#Gardnerella abundance from shotgun results
sgGardDF <- "./gardRelativeAbundance.tsv" %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

#figure_out <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/gardMapping/clade_assignments/figures"
```

Set up metaphlan table
```{r}
#metaphlan data
mDF0 <- file.path(google_drive_dir, "/humann2/merged_tables/mergedMetaphlanAbundance.tsv") %>%
  read_tsv() %>%
  filter(str_detect(ID, "s__"), !str_detect(ID, "t__")) %>%
  mutate(Species = str_extract(ID, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -ID) %>%
  as.data.frame()

#make col and row names
samps <- colnames(mDF0[2:ncol(mDF0)]) %>%
  str_extract("[0-9]{10}")
samps
colnames(mDF0) <- c("Species", samps)
rownames(mDF0) <- mDF0$Species

#Final edits to table
mDF <- mDF0 %>%
  select(-Species) %>%
  t() %>%
  as_tibble() %>%
  mutate_all(funs(./100)) %>%
  mutate(SampleID=samps) %>% 
  select(SampleID, everything())
mDF[1:6,1:5]
```

# Get variants
```{r}
studyIDs <- sgDF$SampleID_a

PS17 <- phyloseq(otu_table(st, taxa_are_rows=FALSE), sample_data(df), tax_table(tax))
  #transform_sample_counts(function(x) {x/sum(x)}) #make read counts into relative abundances

has.shotgun <- PS17@sam_data$SampleID %in% studyIDs
PS16S_170 <- PS17 %>%
  subset_samples(has.shotgun) #keep only samples of interest in OTU table

# sgDF0 <- sgDF %>%
#  filter(PaperCohort==2017) %>%
#  mutate(SampleID=as.character(SampleID),
#         SampleID_a=as.character(SampleID_a))

gardVarTable <- PS16S_170 %>%
  psmelt() %>% 
  dplyr::rename(SampleID_a = SampleID) %>%
  filter(Genus=="Gardnerella", 
         Abundance>0) %>%
  left_join(., sgDF[,c("SampleID_a", "SampleID", "RNAlater")], by="SampleID_a")
```
Note: there are 9 variants among the entire sample set for 2017, but only 5 within these 81 samples

## Function to get the variants among all sequences
```{r}
varSequences <- function(label, sequences){
  sequenceVariants <- unique(sequences)
  names(sequenceVariants) <- paste(label, c(1:length(sequenceVariants)), sep =  "_")
  print(sequenceVariants)
}
```

## Relate Variant assignments to clades
```{r}
vars <- as.list(varSequences("G", gardVarTable$OTU)) # get list of variants from 16S march of dimes

strains <- names(rRNAs16S) %>%
  str_extract(., ".*(?=_[0-9]{5} 16S ribosomal RNA$)")
varClades <- vars %>%
  map(.,~vcountPattern(.x, rRNAs16S)) %>%
  data.frame(Strain=strains) %>%
  mutate(Var=case_when(G_1==1~"G1",
                       G_2==1~"G2",
                       G_3==1~"G3",
                       G_4==1~"G4",
                       G_5==1~"G5")) %>%
  left_join(refDF[,c("Strain", "Clade")], by="Strain") %>%
  select(Var, Clade) %>%
  mutate(Clade=as.character(Clade)) %>%
  unique()
varClades
```

**NOTE: THESE LABELS DO NOT MATCH Callahan 2017 G1/G2 ARE FLIPPED**

## classify samples
```{r}
S16gardDF <- gardVarTable %>%
  mutate(Var=case_when(OTU==vars$G_1~"G2: Clades 1/2", #change G1/G2 to match Callahan 2017 
                        OTU==vars$G_2~"G1: Clades 3/4",
                        OTU==vars$G_3~"G3: indetereminant",
                        OTU==vars$G_4~"G4: Clade 5",
                        OTU==vars$G_5~"G5: Clade 6")) %>%
  group_by(SampleID) %>%
  mutate(propGard=Abundance/sum(Abundance)) %>%
  ungroup() %>%
  select(OTU, SampleID, SubjectID, Abundance, propGard, GWcoll, RNAlater, Var) 
S16gardDF$Var <- factor(S16gardDF$Var, levels = c("G2: Clades 1/2", "G1: Clades 3/4", "G3: indetereminant", "G4: Clade 5", "G5: Clade 6")) #Keep order G2, G1, G3, G4, G5

ggplot(data=S16gardDF, aes(x=GWcoll, y=propGard, color=Var, shape=RNAlater)) +
  geom_point() +
  facet_wrap(~SubjectID) +
  scale_color_manual(values=c("indianred2", "steelblue2", "black", "darkorchid2", "yellowgreen")) + # labels=c("G2: Clades 1/2", "G1: Clades 3/4", "G3: indetereminant", "G4: Clade 5", "G5: Clade 6")) +
  labs(title = "Clade assignments by 16S Amplicon Sequencing", x= "GW collection", y="Proportion of Gardnerella", shape="RNA Later", color="V4 variant")
#ggsave(filename = file.path(figure_out, "cladesBy16S.pdf"))
```

## Save table
```{r}
#write_delim(S16gardDF, "./16SgardVariants.tsv", delim="\t")
```

# Comparison to shotgun methods:

## Unify dataframe
```{r}
S16gardDF0 <- S16gardDF %>%
  select(SampleID, Var, propGard) %>%
  mutate(Method="V4 region 16S amplicon")

sgGardDF0 <- sgGardDF %>%
#  select(-n) %>%
  dplyr::rename(Var=Clade,
                propGard=propClade) %>%
  mutate(Var=case_when(Var=="C1"~"Clade 1",
                       Var=="C2"~"Clade 2",
                       Var=="C3"~"Clade 3",
                       Var=="C4"~"Clade 4",
                       Var=="C5"~"Clade 5",
                       Var=="C6"~"Clade 6"),
         Method="Shotgun")

gardVarDF <- S16gardDF0 %>%
  full_join(sgGardDF0, by=c("SampleID", "Var", "propGard", "Method")) %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "GWcoll", "RNAlater", "PaperCohort")], by="SampleID")

gardVarDF$Var <- factor(gardVarDF$Var, levels=c("Clade 1", "Clade 2", "Clade 3", "Clade 4", "Clade 5", "Clade 6", "G2: Clades 1/2", "G1: Clades 3/4", "G3: indetereminant", "G4: Clade 5", "G5: Clade 6"))
```

## Plot 
Subset of two samples  
```{r}
gardVarDF %>%
  filter(RNAlater=="No",
         SubjectID=="10621"|SubjectID=="19024")%>%
ggplot(aes(x=GWcoll, y=propGard, color=Var, shape=Method, size=Method)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 3)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "red", "blue1", "grey48"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)")) +
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~SubjectID) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3)))
#ggsave(file.path(figure_out, "biolunchV4SG2samp.png"), width=8, height=6)
```
  
All samples 
```{r}
gardVarDF %>%
  filter(RNAlater=="No",
         PaperCohort==2017)%>%
ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 3)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "Clade C5 (shotgun)", "Clade C6 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")) + 
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3))) +
  xlim(0,40)
#ggsave(file.path(figure_out, "biolunchV4SGv1.png"), width=8, height=6)
```

Four Subjects:
```{r}
gardVarDF %>%
  filter(RNAlater=="No",
         PaperCohort==2017, 
         SubjectID %in% c("10621", "10627", "10629", "19024"))%>%
ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 3)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "Clade C5 (shotgun)", "Clade C6 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")) + 
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3))) +
  theme(plot.background = element_rect(fill = "transparent", color = NA),
        legend.background = element_rect(fill = "transparent"))
#ggsave("/Users/hlberman/Desktop/figure1.png", width=7, height=5, bg = "transparent")
```

# Add thresholds and test:
```{r, fig.height=6, fig.width=9}
# create data frames
thresholds <- c(1:49) %>% # set the threshold range
  map(~filter(gardVarDF, n>.x)) %>% # filter out clades below read threshold 
  map(~group_by(.x, SampleID)) %>%
  map(~mutate(.x, propGard=propGard/sum(propGard))) %>% # recalculate proportions after threshold filtering
  map(ungroup) %>%
  map2(c(1:49), ~mutate(.x, Threshold=.y)) %>% # add threshold column
  purrr::reduce(full_join, by=c(colnames(gardVarDF), "Threshold")) # reduce to one data frame

# add threshold information to data frame
gardVarDF0 <- gardVarDF %>%
  mutate(Threshold=0) %>% # set zero threshold for baseline
  rbind(thresholds) %>%
  mutate(Threshold=Threshold+1) # make threshold as >=n 

# plot - thresholds 1 to 50
gardVarDF0 %>%
  filter(RNAlater=="No", PaperCohort=="2017") %>%
  ggplot(aes(x=Threshold, y=propGard, shape=Method, size=Method, color=Var)) + 
  geom_point() +
  scale_size_manual(values=c(2, 4), guide=FALSE) +
  scale_shape_manual(values = c(16, 5), guide=FALSE) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "Clade C5 (shotgun)", "Clade C6 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")) +
  ylim(0,1) +
  scale_x_continuous(breaks = seq(0,50, 5)) +
  facet_wrap(~SampleID) +
  guides(color=guide_legend(override.aes=list(shape=c(rep(16, 6), rep(5, 5)), size=c(rep(2, 6), rep(4, 5))))) +
  labs(x="Threshold (reads ≥ n)", y="Proportion of Gardnerella")

# plot - thresholds 1 to 5
gardVarDF0 %>%
  filter(RNAlater=="No", PaperCohort=="2017", Threshold<=5) %>%
  ggplot(aes(x=Threshold, y=propGard, shape=Method, size=Method, color=Var)) + 
  geom_point() +
  scale_size_manual(values=c(2, 4), guide=FALSE) +
  scale_shape_manual(values = c(16, 5), guide=FALSE) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "Clade C5 (shotgun)", "Clade C6 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")) +
  ylim(0,1) +
  scale_x_continuous(breaks = seq(0,5, 1)) +
  facet_wrap(~SampleID) +
  guides(color=guide_legend(override.aes=list(shape=c(rep(16, 6), rep(5, 5)), size=c(rep(2, 6), rep(4, 5))))) +
  labs(x="Threshold (reads ≥ n)", y="Proportion of Gardnerella")
```

 
```{r}
#fig.height=5, fig.width=6}
V4Vars <- gardVarDF0 %>%
  filter(RNAlater=="No",
         PaperCohort=="2017",
         Method=="V4 region 16S amplicon") %>%
  select(SampleID, Var, propGard) %>%
  group_by(SampleID) %>%
  summarise(V4vars=n())

sum(V4Vars$V4vars)

varCounts <- gardVarDF0 %>%
  filter(RNAlater=="No",
         PaperCohort=="2017",
         Method=="Shotgun") %>%
  select(SampleID, Var, propGard, Threshold) %>%
  group_by(SampleID, Threshold) %>%
  summarise(shotgunVars=n()) %>%
  full_join(map(1:50, ~mutate(V4Vars, Threshold=.x)) %>%
            purrr::reduce(full_join, by=c("SampleID", "V4vars", "Threshold")), 
            by=c("SampleID", "Threshold")) %>%
  replace_na(list(shotgunVars=0))
# Thresholds 1:50
varCounts %>%
  gather("Method", "Vars", c(shotgunVars, V4vars)) %>%
ggplot(aes(x=Threshold, y=Vars, group=Method, color=Method, linetype=Method)) +
  geom_line() +
  facet_wrap(~SampleID) +
  scale_y_continuous(breaks=seq(0,6,1)) +
  scale_color_manual(values=c("blue", "red"), labels=c("Shotgun variants", "V4 variants")) +
  scale_linetype(guide=FALSE) +
  guides(color=guide_legend(override.aes=list(linetype=c(1,2)))) +
  labs(x="Threshold (reads≥n)", y="Variants", color="Method")
# Thresholds 1:5
varCounts %>%
  gather("Method", "Vars", c(shotgunVars, V4vars)) %>%
ggplot(aes(x=Threshold, y=Vars, group=Method, color=Method, linetype=Method)) +
  geom_line() +
  facet_wrap(~SampleID) +
  scale_y_continuous(breaks=seq(0,6,1)) +
  xlim(1,5) +
  scale_color_manual(values=c("blue", "red"), labels=c("Shotgun variants", "V4 variants")) +
  scale_linetype(guide=FALSE) +
  guides(color=guide_legend(override.aes=list(linetype=c(1,2)))) +
  labs(x="Threshold (reads≥n)", y="Variants", color="Method")
```

**Sensitivity**: Percent of cases where a variant is lost
At threshold ≥1: 8 V4 variants are unrepresented
At threshold ≥2: 10 V4 variants are unrepresented
At threshold ≥3: 10 V4 variants are unrepresented
At threshold ≥4: 14 V4 variants are unrepresented
At threshold ≥5: 14 V4 variants are unrepresented

With 56 V4 variants total in these 26 samples, percent missing at each threshold:
```{r}
tribble(
  ~Threshold, ~nMissing, ~total,
  1,8,56,
  2,10,56,
  3,10,56,
  4,14,56,
  5,14,56
) %>%
  mutate(`Percent Missing`=round(nMissing/total*100, 1)) %>%
  select(-nMissing, -total) %>%
  formattable::formattable()
```

**Specificity**: how many clades have no support
At threshold ≥1: 2 clades are unsupported
At threshold ≥2: 1 clade is unsupported
At threshold ≥3: 0 clades are unsupported
```{r}
# Total number of clades found in at threshold >= 1 read matching
gardVarDF0 %>%
  select(SampleID, Var, propGard, RNAlater, PaperCohort, Method, Threshold) %>%
  filter(Method=="Shotgun", Threshold==1, RNAlater=="No", PaperCohort=="2017") %>%
  nrow
```
60 clades are found among the 28 samples being compated with 16S results

```{r}
tribble(
  ~Threshold, ~nUnsupported, ~total,
  1,2,60,
  2,1,60,
  3,0,60,
) %>%
  mutate(`Percent Supported`=round((total-nUnsupported)/total*100, 1)) %>%
  select(-nUnsupported, -total) %>%
  formattable::formattable()
```
  However, one of of these "unsupported clades" is clade 6 and G3 appears in that sample, so G3 *could* represent clade 6. Also, both of these have a match by 16S in another sample of the same subject.

  Clades that appear in addition and are supported (ie a V4 variant appears as two clades):
  At threshold ≥1: 9 clades are added
  At threshold ≥2: 7  clade is added
  At threshold ≥3: 5 clades are added
  At threshold ≥4: 4 clades are added
  At threshold ≥5: 4 clades are added
```{r}
tribble(
  ~Threshold, ~added, ~total,
  1,9,56,
  2,7,56,
  3,5,56,
  4,4,56,
  5,4,56
) %>%
  mutate(`Percent Supported`=round(added/total*100, 1)) %>%
  select(-added, -total) %>%
  formattable::formattable()
```


Look at distribution of clades per sample after threshold filtering
```{r}
# 1 to 5
add0s <- tibble(SampleID=rep(sgDF$SampleID, 5),
               Threshold=c(rep(1, 107), rep(2, 107), rep(3, 107), rep(4, 107), rep(5, 107)))

gardVarDF0 %>%
  filter(Method == "Shotgun") %>%
  filter(Threshold %in% c(1:5)) %>%
  group_by(Threshold, SampleID) %>%
  summarise(nClades=n()) %>%
  full_join(add0s, by=c("SampleID", "Threshold")) %>% #each sample has 5 entries (1 per threshold)
  mutate(nClades=replace_na(nClades, 0)) %>% #replace NAs with 0's
  left_join(sgDF[,c("SampleID", "Cohort")]) %>%
  ggplot(aes(x=nClades, group=Threshold, fill=Threshold)) +
  geom_bar(stat="count", position="dodge") +
  facet_wrap(~Cohort) +
  scale_x_continuous(breaks=seq(0,6,1)) +
  scale_y_continuous(breaks=seq(0,20,2)) +
  labs(x="Clades per sample", y="Count", fill="Threshold ≥")

add0s <- tibble(SampleID=rep(sgDF$SampleID, 5),
               Threshold=c(rep(6, 107), rep(7, 107), rep(8, 107), rep(9, 107), rep(10, 107)))

#6 to 10
gardVarDF0 %>%
  filter(Method == "Shotgun") %>%
  filter(Threshold %in% c(6:10)) %>%
  group_by(Threshold, SampleID) %>%
  summarise(nClades=n()) %>%
  full_join(add0s, by=c("SampleID", "Threshold")) %>% #each sample has 5 entries (1 per threshold)
  mutate(nClades=replace_na(nClades, 0)) %>% #replace NAs with 0's
  left_join(sgDF[,c("SampleID", "Cohort")]) %>%
  ggplot(aes(x=nClades, group=Threshold, fill=Threshold)) +
  geom_bar(stat="count", position="dodge") +
  facet_wrap(~Cohort) +
  scale_x_continuous(breaks=seq(0,6,1)) +
  scale_y_continuous(breaks=seq(0,20,2)) +
  labs(x="Clades per sample", y="Count", fill="Threshold ≥")
```

# Proportion of reads before and after filtering
```{r}
sapply(list(ReadDepth=sgDF$TotalPairs, inputReads=sgDF$upload_pairs, bbmapFiltered=sgDF$bbmapFiltered_pairs), summary)
meanDepth <- mean(sgDF$TotalPairs)
meanDepth
meanInput <- mean(sgDF$bbmapFiltered_pairs)
meanInput

sgDF0 <- sgDF %>%
  select(TotalPairs, bbmapFiltered_pairs) %>%
  gather("Reads", "Count", c(1:2))
sgDF0$Reads <- factor(sgDF0$Reads, levels=c("TotalPairs", "bbmapFiltered_pairs"))

ggplot(data=sgDF0, aes(y=Count, x=Reads)) +
  geom_boxplot() +
  geom_jitter()
```

## Samples where we lose all Gardnerella
In subject 10608: No Gardnerella was detected. Shotgun sequencing can demonstrate less sensitivity than amplicon sequencing.  
In subject 10626: Some Gardnerella was detected (by metaphlan), but we do not see any after filtering the reads with our criteria.   
**Sample 10608**
```{r}
#gardnerella found by metaphlan in subject 10608
gardVarDF %>%
  filter(SubjectID==10608,
         RNAlater=="No") %>%
  select(SampleID, Var, propGard, Method)

samples10608 <- gardVarDF %>%
  filter(SubjectID==10608,
         RNAlater=="No") %>%
  .$SampleID %>%
  unique
samples10608

#percent Gardnerella in these samples
mDF %>%
  select(SampleID, Gardnerella_vaginalis) %>%
  filter(SampleID %in% samples10608)

#read depth of these samples
sgDF %>%
  filter(SampleID %in% samples10608) %>%
  select(SampleID, TotalPairs, bbmapFiltered_pairs, finalPctHuman_po)

ggplot(data=sgDF) +
  geom_density(aes(x=TotalPairs)) +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID==samples10608[1]], color="red") +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID==samples10608[2]], color="blue")

ggplot(data=sgDF) +
  geom_density(aes(x=bbmapFiltered_pairs)) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID==samples10608[1]], color="red") +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID==samples10608[2]], color="blue")
```

**Subject 10626**
```{r}
#gardnerella found by metaphlan in subject 10626
gardVarDF %>%
  filter(SubjectID==10626,
         RNAlater=="No") %>%
  select(SampleID, Var, propGard, Method)

samples10626 <- gardVarDF %>%
  filter(SubjectID==10626,
         RNAlater=="No") %>%
  .$SampleID %>%
  unique
samples10626

#percent Gardnerella in these samples
mDF %>%
  select(SampleID, Gardnerella_vaginalis) %>%
  filter(SampleID %in% samples10626)

#read depth of these samples
sgDF %>%
  filter(SampleID %in% samples10626) %>%
  select(SampleID, TotalPairs, bbmapFiltered_pairs, PctHuman, finalPctHuman_po)

ggplot(data=sgDF) +
  geom_density(aes(x=TotalPairs)) +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID==samples10626[1]], color="red") +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID==samples10626[2]], color="blue") +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID==samples10626[3]], color="orange")

ggplot(data=sgDF) +
  geom_density(aes(x=bbmapFiltered_pairs)) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID==samples10626[1]], color="red") +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID==samples10626[2]], color="blue") +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID==samples10626[3]], color="orange")
```
Vertical lines represent reads in subject 10608’s samples. Fewer bacterial reads in these samples.  

#### Samples with no Gardnerella by clade mapping but Gardnerella found by Metaphlan
```{r}
sgGardDF %>%
  full_join(mDF[,c("SampleID", "Gardnerella_vaginalis")], by="SampleID") %>%
  filter(is.na(n), Gardnerella_vaginalis>0)
```
**Subject 10626** noted above. Additional Sample **1060435348**. This sample previously had one read of Clade 4, and now has no reads. There is only 0.2% *Gardnerella vaginalis* in the sample. 

```{r}
#proportion Gardnerella vaginalis from metaphlan results
mDF %>%
  filter(SampleID=="1060435348") %>%
  select(SampleID, Gardnerella_vaginalis)

#read depth of this sample
sgDF %>%
  filter(SampleID=="1060435348") %>%
  select(SampleID, TotalPairs, bbmapFiltered_pairs, PctHuman, finalPctHuman_po)

ggplot(data=sgDF) +
  geom_density(aes(x=TotalPairs)) +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID=="1060435348"], color="red")

ggplot(data=sgDF) +
  geom_density(aes(x=bbmapFiltered_pairs)) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID=="1060435348"], color="red")
```
Relatively low read count in this sample. 

### Samples where we lose one variant 
First 10508 sample (1050801308) see only C2 with shotgun. G3 (ambiguous) and G1 were see. –we lose G1  
```{r}
#gardnerella found by metaphlan in sample 1050801308
gardVarDF %>%
  filter(SampleID=="1050801308") %>%
  select(SampleID, Var, propGard, Method)

#percent Gardnerella in this sample
mDF %>%
  select(SampleID, Gardnerella_vaginalis) %>%
  filter(SampleID=="1050801308")

#read depth of these samples
sgDF %>%
  filter(SampleID=="1050801308") %>%
  select(SampleID, TotalPairs, bbmapFiltered_pairs, PctHuman, finalPctHuman_po)

ggplot(data=sgDF) +
  geom_density(aes(x=TotalPairs)) +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID=="1050801308"], color="red")

ggplot(data=sgDF) +
  geom_density(aes(x=bbmapFiltered_pairs)) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID=="1050801308"], color="red")
```

Third sample 10629 (1062901318) We lose G5 (ie don’t see C6, but G5 was the least abundant variant in this sample.  
```{r}
#gardnerella found by metaphlan in sample 1062901318
gardVarDF %>%
  filter(SampleID=="1062901318") %>%
  select(SampleID, Var, propGard, Method)

#percent Gardnerella in this sample
mDF %>%
  select(SampleID, Gardnerella_vaginalis) %>%
  filter(SampleID=="1062901318")

#read depth of these samples
sgDF %>%
  filter(SampleID=="1062901318") %>%
  select(SampleID, TotalPairs, bbmapFiltered_pairs, PctHuman, finalPctHuman_po)

ggplot(data=sgDF) +
  geom_density(aes(x=TotalPairs)) +
  geom_vline(xintercept = sgDF$TotalPairs[sgDF$SampleID=="1062901318"], color="red")

ggplot(data=sgDF) +
  geom_density(aes(x=bbmapFiltered_pairs)) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID=="1062901318"], color="red")
```

Second 10627 sample (1062701298) G3 looks like C1, C2, and C6 – we haven’t seen a G3 variant in C6 before, but it is not impossible as we only have 1 reference strain in this clade at present.  

## Clades found by *Gardnerella* mapping method but not by MetaPhlAn
```{r}
sgGardDF %>%
  full_join(mDF[,c("SampleID", "Gardnerella_vaginalis")], by="SampleID") %>%
  filter(Gardnerella_vaginalis==0, n>0)
```

Sample 1050801318
```{r}
#No gardnerella found by metaphlan in sample 1050801318
gardVarDF %>%
  filter(SubjectID=="10508") %>%
  select(SampleID, Var, propGard, Method)

#read depth of all sub samples
sgDF %>%
  filter(SampleID=="1050801318") %>%
  select(SampleID, TotalPairs, bbmapFiltered_pairs, PctHuman, finalPctHuman_po)

#Bacterial reads in all subject 10508 samples
ggplot(data=sgDF) +
  geom_density(aes(x=bbmapFiltered_pairs)) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID=="1050801318"], color="blue", show.legend = TRUE) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID=="1050801308"], color="red", show.legend = TRUE) +
  geom_vline(xintercept = sgDF$bbmapFiltered_pairs[sgDF$SampleID=="1050835178"], color="green", show.legend = TRUE)
```

# Reads cutoff?:
```{r}
sgGardDF$n %>%
  summary

sgGardDF %>%
  group_by(SampleID) %>%
  summarise(nClades=n()) %>%
  right_join(sgGardDF, by="SampleID") %>%
  ggplot(aes(x=n, fill=nClades)) +
  geom_histogram(binwidth = 50)

sgGardDF %>%
  group_by(SampleID) %>%
  summarise(nClades=n()) %>%
  right_join(sgGardDF, by="SampleID") %>%
  ggplot(aes(x=n, fill=nClades)) +
  geom_bar() +
  scale_x_continuous(limits=c(0,50), breaks=seq(0,50,by=5))
```
range of reads per clade is 1 to ~16,000. A large portion are below 5 reads per clade.

Cases of one read per clade:
```{r}
sgGardDF %>%
  group_by(SampleID) %>%
  summarise(nClades=n()) %>%
  right_join(sgGardDF, by="SampleID") %>%
  ungroup() %>%
  filter(n==1) %>%
  ggplot(aes(x=nClades)) +
  geom_bar() +
  scale_y_continuous(limits = c(0,10), breaks = seq(0,10,by=1), minor_breaks = NULL)
```
Only 3 samples have one read of one clade. 


# Session Info
```{r}
sessionInfo()
```

