---
title: "Shotgun Genomospecies Vs. 16S Gardnerella variants"
subtitle: "Reference Databse with 1 Strain per Genomospecies"
output: html_document
---
# Set up
## packages
```{r}
library(tidyverse)
library(phyloseq)
library(Biostrings)
library(formattable)
```

## Paths
```{r}
repo_dir <- "../../"
google_drive_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/"

#shotgun metadata
sgDF <- file.path(repo_dir, "shotgunMetadata.tsv") %>%
  read_delim(delim = "\t") %>%
  mutate(SampleID=as.character(SampleID),
         SampleID_a=as.character(SampleID_a))

#ASVs from Callahan et al., 2017
S16_path <- file.path(google_drive_dir, "16S_data/RepRefine_Scripts/input/processed.rda")
load(S16_path)

#Gardnerella 16S fasta sequences from reference WGS strains
rRNAs16S <- readDNAStringSet(file.path(repo_dir, "gard_mapping", "clade_assignments", "16SrRNAs.fasta"))

#gardnerella metadata
refDF <- file.path(repo_dir, "GardnerellaMetadata.csv") %>%
  read_csv

#Genomospecies data frame
gsDF <- file.path(repo_dir, "72GardStrainGenomospecies.csv") %>%
  read_csv

#Gardnerella abundance from shotgun results
sgGenomoDF <- "./gardGenomospeciesRelativeAbundance_ref1strain1genomospecies.tsv" %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

sgGenomoDF_all72 <- "./gardGenomospeciesRelativeAbundance.tsv" %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

#Gardnerella abundance from shotgun results
sgCladeDF <- "./gardRelativeAbundance_ref1strain1genomospecies.tsv" %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

sgCladeDF_all72 <- "./gardRelativeAbundance.tsv" %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

sgCladeDF %>%
  select(SampleID) %>%
  unique

#figure_out <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/gardMapping/clade_assignments/figures"
```

Set up metaphlan table
```{r}
#metaphlan data
mDF0 <- file.path(google_drive_dir, "/humann2/merged_tables/mergedMetaphlanAbundance.tsv") %>%
  read_tsv() %>%
  filter(str_detect(ID, "s__"), !str_detect(ID, "t__")) %>%
  mutate(Species = str_extract(ID, "(?<=s__)\\w+$")) %>%
  select(Species, everything(), -ID) %>%
  as.data.frame()

#make col and row names
samps <- colnames(mDF0[2:ncol(mDF0)]) %>%
  str_extract("[0-9]{10}")
samps
colnames(mDF0) <- c("Species", samps)
rownames(mDF0) <- mDF0$Species

#Final edits to table
mDF <- mDF0 %>%
  select(-Species) %>%
  t() %>%
  as_tibble() %>%
  mutate_all(funs(./100)) %>%
  mutate(SampleID=samps) %>% 
  select(SampleID, everything())
mDF[1:6,1:5]
```

# Get variants
```{r}
studyIDs <- sgDF$SampleID_a

PS17 <- phyloseq(otu_table(st, taxa_are_rows=FALSE), sample_data(df), tax_table(tax))
  #transform_sample_counts(function(x) {x/sum(x)}) #make read counts into relative abundances

has.shotgun <- PS17@sam_data$SampleID %in% studyIDs
PS16S_170 <- PS17 %>%
  subset_samples(has.shotgun) #keep only samples of interest in OTU table

# sgDF0 <- sgDF %>%
#  filter(PaperCohort==2017) %>%
#  mutate(SampleID=as.character(SampleID),
#         SampleID_a=as.character(SampleID_a))

gardVarTable <- PS16S_170 %>%
  psmelt() %>% 
  dplyr::rename(SampleID_a = SampleID) %>%
  filter(Genus=="Gardnerella", 
         Abundance>0) %>%
  left_join(., sgDF[,c("SampleID_a", "SampleID", "RNAlater")], by="SampleID_a")
```
Note: there are 9 variants among the entire sample set for 2017, but only 5 within these 81 samples

## Function to get the variants among all sequences
```{r}
varSequences <- function(label, sequences){
  sequenceVariants <- unique(sequences)
  names(sequenceVariants) <- paste(label, c(1:length(sequenceVariants)), sep =  "_")
  print(sequenceVariants)
}
```

## Relate Variant assignments to clades
```{r}
vars <- as.list(varSequences("G", gardVarTable$OTU)) # get list of variants from 16S march of dimes

strains <- names(rRNAs16S) %>%
  str_extract(., ".*(?=_[0-9]{5} 16S ribosomal RNA$)")
varClades <- vars %>%
  map(.,~vcountPattern(.x, rRNAs16S)) %>%
  data.frame(Strain=strains) %>%
  mutate(Var=case_when(G_1==1~"G1",
                       G_2==1~"G2",
                       G_3==1~"G3",
                       G_4==1~"G4",
                       G_5==1~"G5")) %>%
  left_join(refDF[,c("Strain", "Clade")], by="Strain") %>%
  select(Var, Clade) %>%
  mutate(Clade=as.character(Clade)) %>%
  unique() %>%
  filter(!is.na(Var))
varClades
```

**NOTE: THESE LABELS DO NOT MATCH Callahan 2017 G1/G2 ARE FLIPPED**

## classify samples
```{r}
S16gardDF <- gardVarTable %>%
  mutate(Var=case_when(OTU==vars$G_1~"G2: Clades 1/2", #change G1/G2 to match Callahan 2017 
                        OTU==vars$G_2~"G1: Clades 3/4",
                        OTU==vars$G_3~"G3: indetereminant",
                        OTU==vars$G_4~"G4: Clade 5",
                        OTU==vars$G_5~"G5: Clade 6")) %>%
  group_by(SampleID) %>%
  mutate(propGard=Abundance/sum(Abundance)) %>%
  ungroup() %>%
  select(OTU, SampleID, SubjectID, Abundance, propGard, GWcoll, RNAlater, Var) 
S16gardDF$Var <- factor(S16gardDF$Var, levels = c("G2: Clades 1/2", "G1: Clades 3/4", "G3: indetereminant", "G4: Clade 5", "G5: Clade 6")) #Keep order G2, G1, G3, G4, G5

ggplot(data=S16gardDF, aes(x=GWcoll, y=propGard, color=Var, shape=RNAlater)) +
  geom_point() +
  facet_wrap(~SubjectID) +
  scale_color_manual(values=c("indianred2", "steelblue2", "black", "darkorchid2", "yellowgreen")) + # labels=c("G2: Clades 1/2", "G1: Clades 3/4", "G3: indetereminant", "G4: Clade 5", "G5: Clade 6")) +
  labs(title = "Clade assignments by 16S Amplicon Sequencing", x= "GW collection", y="Proportion of Gardnerella", shape="RNA Later", color="V4 variant")
#ggsave(filename = file.path(figure_out, "cladesBy16S.pdf"))
```

## Save table
```{r}
#write_delim(S16gardDF, "./16SgardVariants.tsv", delim="\t")
```

# Comparison to shotgun methods:
## Unify dataframe
```{r}
S16gardDF0 <- S16gardDF %>%
  select(SampleID, Var, propGard) %>%
  mutate(Method="V4 region 16S amplicon")

sgGenomoDF0 <- sgGenomoDF %>%
  select(-SpeciesNumber) %>%
  dplyr::rename(Var=Genomospecies,
                propGard=propGenomospecies) %>%
  mutate(Var=as.character(Var),
          Var=case_when(Var=="1"~"Genomospecies 1",
                        Var=="2"~"Genomospecies 2",
                        Var=="3"~"Genomospecies 3",
                        Var=="4"~"Genomospecies 4",
                        Var=="5"~"Genomospecies 5",
                        Var=="6"~"Genomospecies 6",
                        Var=="7"~"Genomospecies 7",
                        Var=="8"~"Genomospecies 8",
                        Var=="9"~"Genomospecies 9",
                        Var=="10"~"Genomospecies 10",
                        Var=="11"~"Genomospecies 11",
                        Var=="12"~"Genomospecies 12",
                        Var=="13"~"Genomospecies 13"),
         Method="Shotgun Genomospecies")

gardVarDF <- S16gardDF0 %>%
  full_join(sgGenomoDF0, by=c("SampleID", "Var", "propGard", "Method")) %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "GWcoll", "RNAlater", "PaperCohort")], by="SampleID")

gardVarDF$Var <- factor(gardVarDF$Var, 
                        levels=c("Genomospecies 1", 
                                 "Genomospecies 2", 
                                 "Genomospecies 3",
                                 "Genomospecies 4",
                                 "Genomospecies 11",                                 
                                 "Genomospecies 5", 
                                 "Genomospecies 6", 
                                 "Genomospecies 7", 
                                 "Genomospecies 8", 
                                 "Genomospecies 9", 
                                 "Genomospecies 10", 
                                 "Genomospecies 12", 
                                 "Genomospecies 13", 
                                 "G2: Clades 1/2", 
                                 "G1: Clades 3/4", 
                                 "G3: indetereminant", 
                                 "G4: Clade 5", 
                                 "G5: Clade 6"), 
                        labels=c("Genomospecies 1, G. vaginalis", 
                                 "Genomospecies 2", 
                                 "Genomospecies 3", 
                                 "Genomospecies 4, G. piotti",
                                 "Genomospecies 11",
                                 "Genomospecies 5, G. leopoldii", 
                                 "Genomospecies 6, G. swidsinskii", 
                                 "Genomospecies 7", 
                                 "Genomospecies 8", 
                                 "Genomospecies 9", 
                                 "Genomospecies 10", 
                                 "Genomospecies 12", 
                                 "Genomospecies 13", 
                                 "G2: Clades 1/2", 
                                 "G1: Clades 3/4", 
                                 "G3: indetereminant", 
                                 "G4: Clade 5", 
                                 "G5: Clade 6"))
```

## Plot 
### Genomospecies vs. V4 variants only
Subset of two samples  
```{r}
gardVarDF %>%
  filter(RNAlater=="No",
         SubjectID=="10621"|SubjectID=="19024")%>%
ggplot(aes(x=GWcoll, y=propGard, color=Var, shape=Method, size=Method)) +
  geom_point() +
  #scale_size_manual(values=c(1.5, 3)) +
  #scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "red", "blue1", "grey48"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)")) +
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~SubjectID) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3)))
#ggsave(file.path(figure_out, "biolunchV4SG2samp.png"), width=8, height=6)
```
  
All samples 
```{r, fig.height=2.1, fig.width=4.5}
gardVarDF %>%
  filter(RNAlater=="No",
         PaperCohort==2017)%>%
ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 3)) +
  scale_color_manual(values=c("salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "blue4", "blue", "lightblue1", "lightblue2", "lightblue3", "lightblue4",   "darkorchid2", "yellowgreen",
                              "red", "blue1", "grey48", "darkorchid2", "yellowgreen"),
                     labels=c("Genomospecies 1, G. vaginalis (shotgun)",
                              "Genomospecies 2 (shotgun)",
                              "Genomospecies 3 (shotgun)",
                              "Genomospecies 4, G. piotti (shotgun)",
                              "Genomospecies 11 (shotgun)",
                               "Genomospecies 5, G. leopoldii (shotgun)",
                               "Genomospecies 6, G. swidsinskii (shotgun)",
                               "Genomospecies 7 (shotgun)",
                               "Genomospecies 8 (shotgun)",
                               "Genomospecies 9 (shotgun)",
                               "Genomospecies 10 (shotgun)",
                               "Genomospecies 12 (shotgun)",
                               "Genomospecies 13 (shotgun)",
                               "G2: Clades 1/2 (V4 amplicon)",
                               "G1: Clades 3/4 (V4 amplicon)",
                               "G3: indeterminant (V4 amplicon)",
                               "G4: Clade 5 (V4 amplicon)",
                               "G5: Clade 6 (V4 amplicon)")) +
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3),
                            ncol=2)) +
  xlim(0,40)
#ggsave(file.path(figure_out, "biolunchV4SGv1.png"), width=8, height=6)
```

  Table for comparing percent missing using clades or genomospecies
```{r}
tribble(
        ~` `, ~Clades, ~Genomospeices,
        "Percent V4 variants not represented", "", "") %>%
  formattable
```
### Genomospecies vs. V4 variants and clades
```{r}
gardVarDFall <- sgCladeDF %>%
  dplyr::rename(Var=Clade,
                propGard=propClade) %>%
  mutate(Var=case_when(Var=="C1"~"Clade C1",
                       Var=="C2"~"Clade C2",
                       Var=="C3"~"Clade C3", 
                       Var=="C4"~"Clade C4",
                       Var=="C5"~"Clade C5", 
                       Var=="C6"~"Clade C6"),
         Method="Shotgun Clade") %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "GWcoll", "RNAlater", "PaperCohort")], by="SampleID") %>%
  full_join(gardVarDF) %>%
   mutate(Var=factor(Var, levels = c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6",
                                     "Genomospecies 1, G. vaginalis", "Genomospecies 2", "Genomospecies 3", "Genomospecies 4, G. piotti", "Genomospecies 11", "Genomospecies 5, G. leopoldii", "Genomospecies 6, G. swidsinskii", "Genomospecies 7", "Genomospecies 8", "Genomospecies 9", "Genomospecies 10",  "Genomospecies 12", "Genomospecies 13",
                                     "G2: Clades 1/2", "G1: Clades 3/4", "G3: indetereminant", "G4: Clade 5", "G5: Clade 6")))
head(gardVarDFall)
unique(gardVarDF$Var)
unique(gardVarDFall$Var)
```

```{r, fig.height=2.1, fig.width=4.5}
gardVarDFall %>%
  filter(RNAlater=="No",
         PaperCohort==2017)%>%
ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 2, 3)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen",
                              "salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "blue4", "blue", "lightblue1", "lightblue2", "lightblue3", "lightblue4",   "darkorchid2", "yellowgreen",
                              "red", "blue1", "grey48", "darkorchid2", "yellowgreen"),
                     labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "Clade C5 (shotgun)", "Clade C6 (shotgun)",
                              "Genomospecies 1, G. vaginalis (shotgun)",
                              "Genomospecies 2 (shotgun)",
                              "Genomospecies 3 (shotgun)",
                              "Genomospecies 4, G. piotti (shotgun)",
                              "Genomospecies 11 (shotgun)",
                              "Genomospecies 5, G. leopoldii (shotgun)",
                              "Genomospecies 6, G. swidsinskii (shotgun)",
                              "Genomospecies 7 (shotgun)",
                              "Genomospecies 8 (shotgun)",
                              "Genomospecies 9 (shotgun)",
                              "Genomospecies 10 (shotgun)",
                              "Genomospecies 12 (shotgun)",
                              "Genomospecies 13 (shotgun)",
                              "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indetereminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")) +
  scale_shape_manual(values = c(16, 5, 1)) +
  facet_wrap(~SubjectID, nrow=2) +
  #coord_fixed(ratio=0) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3))) +
  xlim(0,40)
```
  Shotgun methods only
```{r, fig.height=2.1, fig.width=4.5}
gardVarDFall %>%
  filter(RNAlater=="No",
         PaperCohort==2017,
         Method!="V4 region 16S amplicon")%>%
ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 2)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen",
                              "salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "blue4", "blue", "lightblue1", "lightblue2", "lightblue3", "lightblue4",   "darkorchid2", "yellowgreen"),
                     labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6",
                              "Genomospecies 1, G. vaginalis",
                              "Genomospecies 2",
                              "Genomospecies 3",
                              "Genomospecies 4, G. piotti",
                              "Genomospecies 11",
                              "Genomospecies 5, G. leopoldii",
                              "Genomospecies 6, G. swidsinskii",
                              "Genomospecies 7",
                              "Genomospecies 8",
                              "Genomospecies 9",
                              "Genomospecies 10",
                              "Genomospecies 12",
                              "Genomospecies 13")) +
  scale_shape_manual(values = c(16, 5)) +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3), ncol=2)) +
  xlim(0,40)
```

  Shotgun methods only
```{r, fig.height=4, fig.width=5}
gardVarDFall %>%
  filter(RNAlater=="No",
         PaperCohort==2017,
         Method!="V4 region 16S amplicon")%>%
ggplot(aes(x=GWcoll, y=propGard, shape=Method, size=Method, color=Var)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 2)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen",
                              "salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "blue4", "blue", "lightblue1", "lightblue2", "lightblue3", "lightblue4",   "darkorchid2", "yellowgreen"),
                     labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6",
                              "Genomospecies 1, G. vaginalis",
                              "Genomospecies 2",
                              "Genomospecies 3",
                              "Genomospecies 4, G. piotti",
                              "Genomospecies 11",
                              "Genomospecies 5, G. leopoldii",
                              "Genomospecies 6, G. swidsinskii",
                              "Genomospecies 7",
                              "Genomospecies 8",
                              "Genomospecies 9",
                              "Genomospecies 10",
                              "Genomospecies 12",
                              "Genomospecies 13")) +
  scale_shape_manual(values = c(16, 5)) +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3))) +
  xlim(0,40)
```

  Shotgun methods only: n not proportion
```{r, fig.height=2.1, fig.width=4.5}
gardVarDFall %>%
  filter(RNAlater=="No",
         PaperCohort==2017,
         Method!="V4 region 16S amplicon")%>%
ggplot(aes(x=GWcoll, y=n, shape=Method, size=Method, color=Var)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 2)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen",
                              "salmon1", "salmon3", "indianred1", "indianred2", "indianred3", "blue4", "blue", "lightblue1", "lightblue2", "lightblue3", "lightblue4",   "darkorchid2", "yellowgreen"),
                     labels=c("Clade C1", "Clade C2", "Clade C3", "Clade C4", "Clade C5", "Clade C6",
                              "Genomospecies 1, G. vaginalis",
                              "Genomospecies 2",
                              "Genomospecies 3",
                              "Genomospecies 4, G. piotti",
                              "Genomospecies 11",
                              "Genomospecies 5, G. leopoldii",
                              "Genomospecies 6, G. swidsinskii",
                              "Genomospecies 7",
                              "Genomospecies 8",
                              "Genomospecies 9",
                              "Genomospecies 10",
                              "Genomospecies 12",
                              "Genomospecies 13")) +
  scale_shape_manual(values = c(16, 5)) +
  scale_y_log10() +
  facet_wrap(~SubjectID, nrow=2, scales = "free_y") +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3), ncol=2)) +
  xlim(0,40)
```

# Genomospecies per sample
```{r}
a <- sgGenomoDF0 %>%
  group_by(SampleID) %>%
  summarise(n=n()) %>%
  mutate(n=as.character(n))

b <- sgGenomoDF0 %>%
  full_join(mDF[,c("SampleID", "Gardnerella_vaginalis")], by="SampleID") %>%
  select(-n) %>%
  filter(is.na(Var)) %>%
  mutate(Gardnerella_vaginalis=as.character(Gardnerella_vaginalis),
         Gardnerella_vaginalis=case_when(Gardnerella_vaginalis=="0"~"0",
                                         Gardnerella_vaginalis!="0"~"Uncharacterized")) %>%
  dplyr::rename(n=Gardnerella_vaginalis) %>%
  select(SampleID, n) %>%
  full_join(a, by=c("SampleID", "n")) %>%
  mutate(n=factor(n, levels=c("Uncharacterized", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13"), labels=c("Unch.", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13"))) %>%
  left_join(sgDF[c("SampleID", "Cohort")])

ggplot(b, aes(x=n)) +
  geom_bar() +
  facet_wrap(~Cohort) +
  scale_y_continuous(breaks = seq(1,13,1)) +
  theme_bw() +
  theme(panel.grid.major.x = element_blank()) +
  labs(x="Genomospecies per sample")
```
# Clades per sample
```{r}
c <- sgCladeDF %>%
  group_by(SampleID) %>%
  summarise(n=n()) %>%
  mutate(n=as.character(n))

d <- sgCladeDF %>%
  full_join(mDF[,c("SampleID", "Gardnerella_vaginalis")], by="SampleID") %>%
  select(-n) %>%
  filter(is.na(Clade)) %>%
  mutate(Gardnerella_vaginalis=as.character(Gardnerella_vaginalis),
         Gardnerella_vaginalis=case_when(Gardnerella_vaginalis=="0"~"0",
                                         Gardnerella_vaginalis!="0"~"Uncharacterized")) %>%
  dplyr::rename(n=Gardnerella_vaginalis) %>%
  select(SampleID, n) %>%
  full_join(c, by=c("SampleID", "n")) %>%
  mutate(n=factor(n, levels=c("Uncharacterized", "0", "1", "2", "3", "4", "5", "6"))) %>%
  left_join(sgDF[c("SampleID", "Cohort")])

ggplot(d, aes(x=n)) +
  geom_bar() +
  facet_wrap(~Cohort) +
  theme_bw() +
  theme(panel.grid.major.x = element_blank()) +
  labs(x="Clades per sample")
```
# Reads assigning to variants
## reads assigning to clades
### 1 Strain 1 Genomospecies
```{r}
gardVarDFall %>%
  filter(Method=="Shotgun Clade") %>%
  ggplot(aes(x=n)) +
  geom_histogram() +
  theme_classic()

gardVarDFall %>%
  filter(Method=="Shotgun Clade") %>%
  ggplot(aes(x=n)) +
  geom_histogram() +
  scale_x_continuous(limits = c(0,10), breaks = seq(0,10,1)) +
  theme_classic()
```
### All 72 reference genes
```{r}
sgCladeDF_all72 %>%
  ggplot(aes(x=n)) +
  geom_histogram() +
  theme_classic()

sgCladeDF_all72 %>%
  ggplot(aes(x=n)) +
  geom_histogram() +
  scale_x_continuous(limits = c(0,10), breaks = seq(0,10,1)) +
  theme_classic()
```

## reads assigning to genomospecies
### 1 Strain per genomospecies
```{r}
gardVarDFall %>%
  filter(Method=="Shotgun Genomospecies") %>%
  ggplot(aes(x=n)) +
  geom_histogram() +
  theme_classic()

gardVarDFall %>%
  filter(Method=="Shotgun Genomospecies") %>%
  ggplot(aes(x=n)) +
  geom_histogram() +
  scale_x_continuous(limits = c(0,10), breaks = seq(0,10,1)) +
  theme_classic()
```

### All 72 reference strains
```{r}
sgGenomoDF_all72 %>%
  ggplot(aes(x=n)) +
  geom_histogram() +
  theme_classic()

sgGenomoDF_all72 %>%
  ggplot(aes(x=n)) +
  geom_histogram() +
  scale_x_continuous(limits = c(0,10), breaks = seq(0,10,1)) +
  theme_classic()
```


# Session Info
```{r}
sessionInfo()
```

