---
title: "16S Data to Phyloseq Objects"
author: "Hanna Berman"
date: "January 2019"
output: html_notebook
---

The following is to obtain phyloseq objects from the 16S data from DiGiulio et al., 2015 and Callahan et al., 2017 to compare to relative abundance profiles obtained by analyzing shotgun metagenomic sequencing data with Metaphlan2 and Kraken2/Bracken

# Setup
##Load packages
```{r libraries, warning=FALSE}
library(phyloseq); packageVersion("phyloseq")
library(tidyverse); packageVersion("tidyverse")
```

## Define file paths
```{r}
main_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/"
path.in15 <- file.path(main_dir, "16S_data", "digiulio2015")
path.in17 <- file.path(main_dir, "16S_data", "RepRefine_Scripts", "input")
DFmg_path <- file.path(main_dir, "metadata", "shotgunMetadata.tsv")
output_dir <- file.path(main_dir, "abundanceMethodsComparisons", "PS_RData")
```

## Get IDs of samples that we will need
Create a list of sample IDs that will have corresponding shotgun metagenome samples
```{r}
DFmg <- DFmg_path %>%
  read.table(header=TRUE, sep="\t", quote="",stringsAsFactors=FALSE) %>%
  as.data.frame()
ii <- substr(DFmg$SampleID, 8, 9)
studyIDs <- paste0(DFmg$SubjectID,"01", ii,"8") #make study IDs look identical to old IDs
```

# Create phyloseq object with all OTUs

data from 2017 paper
```{r}
load(file.path(path.in17, "processed.rda")) #load Rdata file with phyloseq object of samples from 2017 paper

PS17 <- phyloseq(otu_table(st, taxa_are_rows=FALSE), sample_data(df), tax_table(tax)) %>%
  transform_sample_counts(function(x) {x/sum(x)}) #make read counts into relative abundances

has.shotgun <- rownames(PS17@sam_data) %in% studyIDs
PS16S_170 <- PS17 %>%
  subset_samples(has.shotgun) #keep only samples of interest in OTU table and sam table

PS16S_170@sam_data <- PS16S_170@sam_data %>%   #keep only columns of interest from sample data frame
  select(SampleID, SubjectID, GWcoll)
PS16S_170@sam_data[,"PaperCohort"] <- "2017"

PS16S_17L <- PS16S_170 %>%
  subset_taxa(Genus=="Lactobacillus") %>%
  tax_glom(taxrank="Species")
PS16S_17G <- PS16S_170 %>%
  subset_taxa(Genus!="Lactobacillus") %>%
  tax_glom(taxrank="Genus")  
PS16S_17 <- merge_phyloseq(PS16S_17L, PS16S_17G) 
summary(sample_sums(PS16S_17))

PS16S_17@tax_table <- PS16S_17@tax_table[,1:7] #make tax table uniform with 2015 PS object
PS16S_17 <- phyloseq(PS16S_17@otu_table, PS16S_17@tax_table, PS16S_17@sam_data) #remove tree
```

data from 2015 paper
```{r}
load(file.path(path.in15, "PregnancyClosed15.Rdata")) #load Rdata file with phyloseq object of samples from 2015 paper
PS15 <- PS %>%
  transform_sample_counts(function(x) {x/sum(x)}) #make read counts into relative abundances
remove(PS, PSbs, PSPost, PSPreg)

PS15@sam_data$SampleID <- replace(x <- PS15@sam_data$SampleID, x=="1001301318.rs", "1001301318") #remove ".rs" designation -- was causing an item to be missed (in sam_table)
rownames(PS15@sam_data) <- PS15@sam_data$SampleID
rownames(PS15@otu_table) <- replace(x <- rownames(PS15@otu_table), x=="1001301318.rs", "1001301318") #remove ".rs" designation -- was causing an item to be missed (in OTU_table)

has.shotgun <- PS15@sam_data$SampleID %in% studyIDs
PS16S_15 <- PS15 %>%
  subset_samples(has.shotgun) #keep only samples of interest in OTU table and sam table

PS16S_15@sam_data <- PS16S_15@sam_data %>%
  select(SampleID, SubjectID, GWColl)
colnames(PS16S_15@sam_data) <- c("SampleID", "SubjectID", "GWcoll")
PS16S_15@sam_data[,"PaperCohort"] <- "2015"  

summary(sample_sums(PS16S_15))

#EDIT TAX TABLE TO MATCH 2017 TAX TABLE
tax15 <- as.data.frame(PS16S_15@tax_table, stringsAsFactors=FALSE)

#Put NAs for taxa classification levels below most descriptive calssification (make match with 2017)
tax15$Class <- sub("(P:).*$", NA, tax15$Class)
tax15$Order <- sub("(P:|C:).*$", NA, tax15$Order)
tax15$Family <- sub("(P:|C:|O:).*$", NA, tax15$Family)
tax15$Genus <- sub("(P:|C:|O:|F:).*$", NA, tax15$Genus)
tax15$Species <- sub("(P:|C:|O:|F:|G:).*$", NA, tax15$Species)

#remove "P:", "C:"...
tax150 <- tax15 %>% 
  mutate(Phylum=str_extract(Phylum, "(?<=P:).*$")) %>% 
  mutate(Class=str_extract(Class, "(?<=C:).*$")) %>%
  mutate(Order=str_extract(Order, "(?<=O:).*$")) %>%
  mutate(Family=str_extract(Family, "(?<=F:).*$")) %>%
  mutate(Species=str_extract(Species, "(?<=Lactobacillus.).*$")) #remove extra lactobacillus label

#prep for adding to phyloseq object
tax151 <- as.matrix(tax150)
rownames(tax151) <- rownames(tax15)

#add tax table back into the phyloseq object, remove tree, and remove variants
PS16S_150 <- phyloseq(PS16S_15@otu_table, tax_table(tax151), PS16S_15@sam_data)
PS16S_15L <- PS16S_150 %>%
  subset_taxa(Genus=="Lactobacillus") %>%
  tax_glom(taxrank="Species")
PS16S_15G <- PS16S_150 %>%
  subset_taxa(Genus!="Lactobacillus") %>%
  tax_glom(taxrank="Genus")  
PS16S_15 <- merge_phyloseq(PS16S_15L, PS16S_15G)
summary(sample_sums(PS16S_15))
```

MERGE PHYLOSEQ OBJECTS
```{r}
PS_16S <- merge_phyloseq(PS16S_15, PS16S_17)
to.add.space <- c(is.na(PS_16S@tax_table[,"Species"])) #select rows with a genus classification but no species classification
PS_16S@tax_table[to.add.space,"Species"] <- " " #change those NAs to a space (" ")
```

```{r}
#what sample is missing?
w16S <- rownames(PS_16S@sam_data)
setdiff(studyIDs, w16S)
```

Get taxa of interest and make psmelt tables
```{r}
PS_16S_toi_o <- PS_16S %>%
  subset_taxa((Genus=="Gardnerella"|Genus=="Lactobacillus" | Genus=="Mycoplasma" | Genus=="Atopobium" | Genus=="Prevotella") &
              (Species=="crispatus" | Species=="iners" | Species=="gasseri" | Species=="jensenii" | Species==" "))
summary(sample_sums(PS_16S_toi_o))

PS_16S_toi <- PS_16S_toi_o %>%
  transform_sample_counts(function(x) {x/sum(x)})

otherCounts <- data.frame(OTU = "Other",
                     Abundance = 1 - sample_sums(PS_16S_toi_o),
                     SampleID = names(sample_sums(PS_16S_toi_o)),
                     Genus="Other",
                     Species=" ",
                     stringsAsFactors = FALSE) %>%
                left_join(as.data.frame(PS_16S@sam_data), by="SampleID")
  
#make tables for subsequent plotting:
sixteenS_toi_o_df <- PS_16S_toi_o %>%
                        psmelt() %>%
                        filter(Abundance > 0) %>%
                        select(OTU, Abundance, SampleID, SubjectID, GWcoll, PaperCohort, Genus, Species) %>%
                        full_join(otherCounts) %>% # add other
                        mutate(Taxon=paste(Genus, Species))

sixteenS_toi_df <- PS_16S_toi %>%
                        psmelt() %>%
                        filter(Abundance > 0) %>%
                        select(OTU, Abundance, SampleID, SubjectID, GWcoll, PaperCohort, Genus, Species) %>%
                        mutate(Taxon=paste(Genus, Species))
```

# Save phyloseq objects to an RData file
```{r}
save(PS_16S, sixteenS_toi_o_df, sixteenS_toi_df, file = file.path(output_dir, "16S_PS.RData"))
```

# Session Info
```{r}
sessionInfo()
```
