---
title: "Gardnerella Gene Deconvolution"
author: "Hanna Berman"
date: "8/9/2019"
output: html_document
---

# Setup
## Packages
```{r}
library(tidyverse)
library(nnls)
library(glmnet)
```

## File paths
```{r}
#metadata
sgDF <- file.path("..", "shotgunMetadata.tsv") %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

#metaphlan2 data
load(file.path("/Users/hlberman/Documents/GitHub/pregnancy_metagenome/gard_mapping/clade_assignments/metaphlanAbundances.Rdata"))

#subset
sampleSubset <- read_lines(file.path("..", "sampleSubset.txt"))
```

## Humann2 Output tables
```{r}
#humann2 outs
dataDir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/humann2/out_tables"
GF_path <- file.path(dataDir, "VM_genefamilies_names_cpm.tsv")
# GFP_path <- file.path(dataDir, "VM_genefamilies_paths.tsv")
# GOnames_path <- file.path(dataDir, "map_go_name.txt")
# GFGO_path <- file.path(dataDir, "map_go_uniref90.txt")
# PWnames_path <- file.path(dataDir, "map_metacyc-pwy_name.txt") 
# PC_path <- file.path(dataDir, "VM_pathcoverage.tsv")
# PA_path <- file.path(dataDir, "VM_pathabundance.tsv")

# read in data frames
GF <- read_delim(GF_path, delim = "\t") %>% # Gene family abundances
  separate(`# Gene Family`, into=c("geneFamily", "org"), sep="\\|", fill="right") %>%
  separate(org, into=c("Genus", "Species"), sep="\\.") %>%
  mutate(Genus=str_extract(Genus,"[^g__].*$"),
         Species=str_extract(Species, "[^s__].*"))
head(GF)

#Species that metaphlan says have no Gard
noGard <- mDF %>%
  filter(Gardnerella_vaginalis==0) %>%
  select(SampleID) %>%
  .[[1]]

gardSamples <- mDF %>%
  filter(Gardnerella_vaginalis!=0) %>%
  select(SampleID) %>%
  .[[1]]

mDFG0 <- mDFG %>%
  filter(!is.element(SampleID, noGard)) %>%
  mutate(SubjectID=substr(SampleID, 1, 5)) %>%
  select(SampleID, SubjectID, contains("Gardnerella")) #%>%
  #mutate(gardSum=Gardnerella_vaginalis_C1+Gardnerella_vaginalis_C2+Gardnerella_vaginalis_C3+Gardnerella_vaginalis_C4+Gardnerella_vaginalis_C5+Gardnerella_vaginalis_C6) %>%
  #mutate_at(vars(Gardnerella_vaginalis_C1,Gardnerella_vaginalis_C2,Gardnerella_vaginalis_C3,Gardnerella_vaginalis_C4,Gardnerella_vaginalis_C5,Gardnerella_vaginalis_C6), ~.x/gardSum)
head(mDFG0)

class(mDFG0$Gardnerella_vaginalis_C6)

mDFG0 %>%
  select(contains("Gardnerella")) %>%
  rowSums()

# Obtain Gardnerella-only genes
gvGF <- GF %>%
  dplyr::rename_all(~str_extract(.x, "geneFamily|Genus|Species|[0-9]{10}")) %>%
  filter(Genus=="Gardnerella") %>%
  select(-Genus, -Species, -one_of(noGard))
head(gvGF)

# GFP <- read_delim(GFP_path, delim = "\t") # pathways for each gene family
# PWnames <- read_delim(PWnames_path, delim = "\t", col_names = c("Pathway", "pathwayAnnotation"))
# PC <- read_delim(PC_path, delim = "\t") # pathway coverage 
# PA <- read_delim(PA_path, delim = "\t") # pathway abundnance
# GOnames <- read_delim(GOnames_path, delim = "\t", col_names = c("GO", "Function")) # names and annotations of GO categories
# GFGO <- read_lines(GFGO_path) %>% # read in the file as a list
#   str_split("\t")
```

# Choose constant genomic elements
Ban et al. chose 11 ribosomal proteins from a starting set of 31. The 11 that they chose were present in at least 98% of the bacteria and archaea in KEGG v60.  
I am starting with 19 robosomal proteins from the Mende et al., 2013 set (where I was certain the match between the COGs listed in Mende et al. matched the uniprot genes in the humann2 output) and chose the genes that were present in ~98% of samples (~94 samples)  
```{r}
#19 ribosomal proteins from Mende et al., 2013 list of universal single copy reference genes
univSingleCopy <- c("UniRef90_R6G9Q3: Ribosomal protein S7",
                    "UniRef90_D2RB42: 30S ribosomal protein S2",
                    "UniRef90_D2R9N2: 30S ribosomal protein S3",
                    "UniRef90_E3D7X1: 30S ribosomal protein S5",
                    "UniRef90_A1A058: 30S ribosomal protein S9",
                    "UniRef90_B8DW37: 30S ribosomal protein S11",
                    "UniRef90_B8DTV4: 30S ribosomal protein S12",
                    "UniRef90_A1A093: 30S ribosomal protein S13",
                    "UniRef90_D2RC68: 50S ribosomal protein L1",
                    "UniRef90_B8DW16: 50S ribosomal protein L2",
                    "UniRef90_M4RGM7: 50S ribosomal protein L3",
                    "UniRef90_E3D7Y7: 50S ribosomal protein L4",
                    "UniRef90_A1A081: 50S ribosomal protein L5",
                    "UniRef90_A0ZZZ5: 50S ribosomal protein L11",
                    "UniRef90_E3D7Z2: 50S ribosomal protein L13",
                    "UniRef90_B8DW23: 50S ribosomal protein L14",
                    "UniRef90_D2R9P2: 50S ribosomal protein L15",
                    "UniRef90_D2R9N9: 50S ribosomal protein L18",
                    "UniRef90_D2R9N1: 50S ribosomal protein L22")

univSingleCopy0 <- paste(univSingleCopy, collapse ="|")

#choose reference genes that are present in 98% of samples (~94 samples)
Ggenes <- gvGF %>%
  filter(str_detect(geneFamily, univSingleCopy0)) %>%
  gather("Sample", "rpk", gardSamples) %>%
  mutate(rpk=case_when(rpk>0~1,
                       rpk==0~0)) %>%
  group_by(geneFamily) %>%
  summarise(nSamples=sum(rpk)) %>%
  filter(nSamples>=94) %>%
  .$geneFamily
Ggenes
Ggenes0 <- paste(Ggenes, collapse="|")
```
11 genes tota chosen as constant genomic elements 

Function for deconvoluting gene presence in taxa:    
```{r}
deconGene <- function(Gene){
gvGF0 <- gvGF %>%
  filter(str_detect(geneFamily, Gene)) %>%
  rbind(colnames(gvGF)) %>%
  t() %>%
  as.data.frame() %>%
  filter(V2!="geneFamily") %>%
  dplyr::rename(geneE=V1, 
                SampleID=V2) %>%
  mutate(geneE=as.numeric(as.character(geneE)),
         SampleID=as.character(SampleID)) %>%
  select(SampleID, geneE)
head(gvGF0)

GDF <- gvGF %>%
  filter(str_detect(geneFamily,  Ggenes0)) %>%
  rbind(colnames(gvGF)) %>%
  t() %>%
  as.data.frame() %>%
  slice(2:nrow(.)) %>%
  gather("refGene", "rpk", c(1:length(Ggenes))) %>%
  mutate(rpk=as.numeric(as.character(rpk))) %>%
  dplyr::rename(SampleID=paste0("V", length(Ggenes)+1)) %>%
  group_by(SampleID) %>%
  summarise(refG=sum(rpk))

regDF <- mDFG0 %>%
  left_join(gvGF0, by="SampleID") %>%
  left_join(GDF, by="SampleID") %>%
  mutate(Ginv=refG/length(Ggenes),
         Greads=geneE/refG)

A <- glm(data=regDF, geneE~SubjectID+Gardnerella_vaginalis_C1*refG+Gardnerella_vaginalis_C2*refG+Gardnerella_vaginalis_C3*refG+Gardnerella_vaginalis_C4*refG+Gardnerella_vaginalis_C5*refG+Gardnerella_vaginalis_C6*refG) %>%
  summary()

B <- glm(data=regDF, Greads~SubjectID+Gardnerella_vaginalis_C1+Gardnerella_vaginalis_C2+Gardnerella_vaginalis_C3+Gardnerella_vaginalis_C4+Gardnerella_vaginalis_C5+Gardnerella_vaginalis_C6) %>%
  summary()
return(list(A, B))
}
```

# Test case with alpha-L-Fucosidase
Uniref gene E3D7F6
```{r}
deconGene("E3D7F6")
```

# Test case with LicD
Uniref gene D2RC48  
Previously found in Clades 3, 4, 5, and 6, and in one strain in clade 2
```{r}
deconGene("D2RC48")
```
Results from each of the two genes are giving a 1 significant coefficient that is less than 1. This may reflect a chance of one copy appearing in strains of that clade.  Will assess regression methods further.  

# Session Info
```{r}
sessionInfo()
```

