---
title: "Simulation of Gardnerella Gene Content Deconvolution"
author: "Hanna Berman"
date: "8/28/2019"
output: html_document
---

# Setup
## Packages
```{r}
library(tidyverse)
library(broom)
#library(nnls)
#library(glmnet)
#library(formattable)
set.seed(100)
```

## File paths
```{r}
# #metadata
# sgDF <- file.path("..", "shotgunMetadata.tsv") %>%
#   read_tsv() %>%
#   mutate(SampleID=as.character(SampleID))
# 
# #metaphlan2 data
# load(file.path("/Users/hlberman/Documents/GitHub/pregnancy_metagenome/gard_mapping/clade_assignments/metaphlanAbundances.Rdata"))
# 
# #subset
# sampleSubset <- read_lines(file.path("..", "sampleSubset.txt"))

#humann2 outs
dataDir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/humann2/out_tables"
GF_path <- file.path(dataDir, "VM_genefamilies_names_cpm.tsv")
# GFP_path <- file.path(dataDir, "VM_genefamilies_paths.tsv")
# GOnames_path <- file.path(dataDir, "map_go_name.txt")
# GFGO_path <- file.path(dataDir, "map_go_uniref90.txt")
# PWnames_path <- file.path(dataDir, "map_metacyc-pwy_name.txt") 
# PC_path <- file.path(dataDir, "VM_pathcoverage.tsv")
# PA_path <- file.path(dataDir, "VM_pathabundance.tsv")

# read in data frames
sgDF <- file.path("..", "shotgunMetadata.tsv") %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

#gardnerella reference sequences
roaryDir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/Gard WGS/20181114_reAnnotatedGvRoary70_set1"

geneAnnos <- file.path(roaryDir, "gene_presence_absence.csv") %>%
  read_csv() %>%
  select(Gene, Annotation)
print(geneAnnos)

GvRoary <- file.path(roaryDir, "gene_presence_absence.Rtab") %>%
  read_delim(delim="\t") %>%
  left_join(geneAnnos, by="Gene") %>%
  select(Gene, Annotation, everything())
print(GvRoary)

geneAnnos <- file.path(roaryDir, "gene_presence_absence.csv") %>%
  read_csv() %>%
  select(Gene, Annotation)
print(geneAnnos)

#Roary output for core genome
GvRoaryCore <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/Gard WGS/20181008_GvRoary95strains/gene_presence_absence.csv" %>%
  read_csv()
print(GvRoaryCore)

#Reference dataframe
gvDF <- file.path("..", "GardnerellaMetadata.csv") %>%
  read_csv()
print(gvDF)
```

## Check Gardnerella read counts in samples
```{r}
# GF <- read_delim(GF_path, delim = "\t") %>% # Gene family abundances
#   separate(`# Gene Family`, into=c("geneFamily", "org"), sep="\\|", fill="right") %>%
#   separate(org, into=c("Genus", "Species"), sep="\\.") %>%
#   mutate(Genus=str_extract(Genus,"[^g__].*$"),
#          Species=str_extract(Species, "[^s__].*"))
# 
# GF %>%
#   dplyr::filter(Genus=="Gardnerella") %>%
#   select(-geneFamily, -Genus, -Species) %>%
#   colSums() %>%
#   as.data.frame() %>%
#   filter(.!=0) %>%
#   summary(.)
```  
values for read counts: 10,000 (~1st q) 400,000 (~max)

# Basic Simulation with Gene Presence Absence
First simulate with gene presence absence information
## Create Samples
```{r}
sampleCreate <- function(genePA){
  x <- sample(c(1:6), 1)
  if(x==1){
    Sample <- genePA %>%
    dplyr::select(Gene, sample(c(3:ncol(.)), 1, replace=FALSE))}
  if(x==2){
    Sample <- genePA %>%
    select(Gene, sample(c(3:ncol(.)), 2, replace=FALSE))}
  if(x==3){
    Sample <- genePA %>%
    select(Gene, sample(c(3:ncol(.)), 3, replace=FALSE))}
  if(x==4){
    Sample <- genePA %>%
    select(Gene, sample(c(3:ncol(.)), 4, replace=FALSE))}
  if(x==5){
    Sample <- genePA %>%
    select(Gene, sample(c(3:ncol(.)), 5, replace=FALSE))}
  if(x==6){
    Sample <- genePA %>%
    select(Gene, sample(c(3:ncol(.)), 6, replace=FALSE))}
  return(Sample)
}

#subset reference strains
sampleStrains <- rerun(100, sampleCreate(GvRoary))
sampleStrains[1]

#assign abundnaces of each strain and its genes by multiplying all of the genes of each strain by a constant
sampleGeneReads <- sampleStrains %>%
  map(., ~mutate_if(.x, is.numeric, ~.x*round(runif(1, 2000, 70000))))
sampleGeneReads[1]
```

## Estimate relative abundance with core genes
```{r}
# Get core genes
coreGenes <- GvRoaryCore %>%
  add_count(Annotation) %>%
  filter(`No. isolates`==72,
         n==1) %>%
  .$Annotation %>%
  {dplyr::filter(GvRoary, is.element(Annotation, .))} %>%
  .$Gene
coreGenes

#calculate relative abundance of each strain using marker genes
strainAb <- sampleGeneReads %>%
  map(~filter(.x, is.element(Gene, coreGenes))) %>%
  map(~summarise_if(.x, is.numeric, sum))
strainAb[1]
samCoreTot <- sampleGeneReads %>%
  map(~filter(.x, is.element(Gene, coreGenes))) %>%
  map(~summarise_if(.x, is.numeric, sum)) %>%
  map(~summarise(.x, sums=rowSums(.x))) %>%
  map(~.x$sums) %>%
  unlist
samCoreTot[1]

strainAbundances <- strainAb0 %>%
  map2(., samCoreTot, ~mutate(.x, Totals=.y)) %>%
  map(., ~mutate_all(.x, ~.x/Totals)) %>%
  map(., ~select(.x, -Totals)) %>%
  map(., ~gather(.x, "Strain", "Abundance"))
strainAbundances[1]

#turn strain abundances into clade abundances and reduce to one table
cladeAbundances <- strainAbundances %>%
  map(., ~left_join(.x, gvDF[,c("Strain", "Clade")], by="Strain")) %>%
  map(., ~group_by(.x, Clade)) %>%
  map(., ~summarise(.x, Abundance=sum(Abundance))) %>%
  map(., ~ungroup(.x)) %>%
  map2(., c(1:100), ~mutate(.x, Sample=.y)) %>% # add sample labels
  map(., ~select(.x, Sample, Clade, Abundance)) %>%
  purrr::reduce(dplyr::full_join, by=c("Clade", "Abundance", "Sample")) %>% # reduce table
  spread(Clade, Abundance) %>%
  replace_na(list(C1=0, C2=0, C3=0, C4=0, C5=0, C6=0))
cladeAbundances

#Create unlabeled gene reads for testing
testSamples <- sampleGeneReads %>%
  map(., ~gather(.x, "Strain", "Reads", 2:ncol(.x))) %>%
  map(., ~group_by(.x, Gene)) %>%
  map(., ~summarise(.x, Reads=sum(Reads)))
testSamples[1]
```

Steps:
1) Compute G inverse
  a) get the total number of reads in each sample
  b) get the number of secY genes
4) bind the G column and normalize read counts
3) Get gene name
4) Compute glm

**1)** Compute G inverse
  a) get the total number of reads in each sample
  b) get the number of secY genes
```{r}
#G inverse (copy number/SecY reads):
#get the number of secY (universal single copy) reads  
Gs <- testSamples %>%
  map(., ~filter(.x, Gene=="secY")) %>%
  map(., ~rename(.x, secYreads=Reads)) %>%
  map2(., c(1:100), ~mutate(.x, Sample=.y)) %>%
  purrr::reduce(full_join, by=c("Sample", "secYreads")) %>%
  #left_join(samTot, by="Sample") %>%
  mutate(G=1/secYreads,
         Ginv=secYreads/1) %>%
  select(Sample, secYreads, G, Ginv) # samTots
Gs
```

**3)** Get gene names
```{r}
#Test genes:
#make a table of the genes for testing
exampleGenes <- tribble(
  ~mech, ~gene, ~name, 
  "AMR", "HMPREF0421_20446", "Methicillin resistance protein",
  "AMR", "HMPREF0421_20507", "Aminoglycoside phosphotransferase",
  "Biofilm formation", "HMPREF0421_20407", "Family 2 glycosyltransferase",
  "Biofilm formation", "HMPREF0421_20405", "Family 2 glycosyltransferase",
  "Biofilm formation", "HMPREF0421_20412", "Family 2 glycosyltransferase",
  "Biofilm formation", "HMPREF0421_20413", "Family 2 glycosyltransferase",
  "Biofilm formation", "HMPREF0424_1180", "Family 2 glycosyltransferase",
  "Biofilm formation", "HMPREF0424_1189", "Family 2 glycosyltransferase",
  "Biofilm formation", "HMPREF0424_1181", "Family 1 glycosyltransferase",
  "Epithelial adhesion", "HMPREF0424_1026", "Type-I fimbrial major subunit precursor",
  "Mucin Degredation", "Sialidase$", "Sialidase",
  "Mucin Degredation", "HMPREF0421_20101", "a-L-fucosidase",
  "Mucin Degredation", "Sialidase A$", "Sialidase A",
  "Virulence", "HMPREF0424_1186", "LicD protein",
)
head(exampleGenes)
```

```{r}
#names of genes from example genes
geneNames <- exampleGenes$gene %>%
  map(~filter(geneAnnos, str_detect(Annotation, .x))) %>%
  map(~.$Gene)
geneNames
```

**4** Compute glm
```{r}
#bind G column and normalize read counts
computeGLM <- function(g, n){
  testSamples %>%
    map(., ~filter(.x, Gene %in% g)) %>%
    map2(., c(1:100), ~mutate(.x, Sample=.y)) %>%
    purrr::reduce(full_join, by=c("Gene", "Reads", "Sample")) %>%
    group_by(Sample) %>%
    summarise(Reads=sum(Reads)) %>%
    mutate(Gene=n) %>%
    select(Gene, Sample, Reads) %>%
    left_join(cladeAbundances, by="Sample") %>%
    left_join(Gs[,c("Sample", "G", "Ginv")], by="Sample") %>%
    #mutate(GReads=Reads*G) %>%
  glm(data = ., Reads~C1*Ginv+C2*Ginv+C3*Ginv+C4*Ginv+C5*Ginv+C6*Ginv) %>%
  tidy -> x
  return(x)
}

map2(geneNames, exampleGenes$name, ~computeGLM(.x, .y))
```

# Assess actual average copy number by clades in sample set 
```{r}
calculateCopies <- function(g) {
mC <- sampleStrains %>%
  map(., ~filter(.x, Gene %in% g)) %>% # select genes
  map(., ~gather(.x, "Strain", "Reads", 2:ncol(.x))) %>%
  map2(., c(1:100), ~mutate(.x, Sample=.y)) %>%
  purrr::reduce(full_join, by=c("Gene", "Sample", "Strain", "Reads")) %>% 
  group_by(Sample, Strain) %>%
  summarise(Copies=sum(Reads)) %>%
  left_join(gvDF[,c("Strain", "Clade")], by="Strain") %>%
  group_by(Clade) %>%
  summarise(avgCopies=mean(Copies))
return(mC)
}

map(geneNames, ~calculateCopies(.x)) %>%
  map2(exampleGenes$name, ~mutate(.x, gene=.y))
```
The glm predicted 1.68 copies per strain in clade 2, in the sample set, there were an average of 1.24 sialidase copies per strain in clade 2. 

## Actual coefficients
Calculated by actual read counts  
Total reads = Clade1 reads + Clade2 reads + Clade3 reads + Clade4 reads + Clade5 reads + Clade6 reads
```{r}
#reads of gene per clade
calcGenesPerClade <- function(g) {
  x <- sampleGeneReads %>% 
    map(., ~filter(.x, Gene %in% g)) %>%
    map(., ~gather(.x, "Strain", "Reads", 2:ncol(.x))) %>%
    map2(., c(1:100), ~mutate(.x, Sample=.y)) %>%
    purrr::reduce(full_join, by=c("Gene", "Strain", "Reads", "Sample")) %>%
    left_join(gvDF[,c("Strain", "Clade")], by="Strain") %>%
    group_by(Clade, Sample) %>%
    summarise(cladeReads=sum(Reads))
  return(x)
}

#reads of gene per sample
calcGenesPerSample <- function(g){
  y <- sampleGeneReads %>%
    map(., ~filter(.x, Gene %in% g)) %>%
    map(., ~gather(.x, "Strain", "Reads", 2:ncol(.x))) %>%
    map2(., c(1:100), ~mutate(.x, Sample=.y)) %>%
    purrr::reduce(full_join, by=c("Gene", "Strain", "Reads", "Sample")) %>%
    left_join(gvDF[,c("Strain", "Clade")], by="Strain") %>%
    group_by(Sample) %>%
    summarise(sampleReads=sum(Reads))
  return(y)
}

# Compute exact glm
computeGLMexact <- function(x,y){
  z <- x %>%
    left_join(y, by="Sample") %>%
    spread(Clade, cladeReads) %>%
    replace_na(list(C1=0, C2=0, C3=0, C4=0, C5=0, C6=0)) %>%
  glm(data=., sampleReads~C1+C2+C3+C4+C5+C6) %>%
    tidy
return(z)
}

step1 <- map(geneNames, ~calcGenesPerClade(.x))
step2 <- map(geneNames, ~calcGenesPerSample(.x))
map2(step1, step2, ~computeGLMexact(.x, .y))
```

Session Info
```{r}
sessionInfo()
```