---
title: "Simulation of Gardnerella Gene Content Deconvolution"
author: "Hanna Berman"
date: "1/31 /2019"
output: html_document
---

# Setup
## Packages
```{r}
library(tidyverse)
library(broom)
library(glmnet)
#library(nnls)
#library(glmnet)
#library(formattable)
set.seed(999)
```

## File paths
```{r}
# #metadata
# sgDF <- file.path("..", "shotgunMetadata.tsv") %>%
#   read_tsv() %>%
#   mutate(SampleID=as.character(SampleID))
# 
# #metaphlan2 data
# load(file.path("/Users/hlberman/Documents/GitHub/pregnancy_metagenome/gard_mapping/clade_assignments/metaphlanAbundances.Rdata"))
# 
# #subset
# sampleSubset <- read_lines(file.path("..", "sampleSubset.txt"))

#humann2 outs
dataDir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/humann2/out_tables"
GF_path <- file.path(dataDir, "VM_genefamilies_names_cpm.tsv")
# GFP_path <- file.path(dataDir, "VM_genefamilies_paths.tsv")
# GOnames_path <- file.path(dataDir, "map_go_name.txt")
# GFGO_path <- file.path(dataDir, "map_go_uniref90.txt")
# PWnames_path <- file.path(dataDir, "map_metacyc-pwy_name.txt") 
# PC_path <- file.path(dataDir, "VM_pathcoverage.tsv")
# PA_path <- file.path(dataDir, "VM_pathabundance.tsv")

# read in data frames
sgDF <- file.path("..", "shotgunMetadata.tsv") %>%
  read_tsv() %>%
  mutate(SampleID=as.character(SampleID))

#gardnerella reference sequences
roaryDir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/Gard WGS/20181114_reAnnotatedGvRoary70_set1"

geneAnnos <- file.path(roaryDir, "gene_presence_absence.csv") %>%
  read_csv() %>%
  select(Gene, Annotation)
print(geneAnnos)

GvRoary <- file.path(roaryDir, "gene_presence_absence.Rtab") %>%
  read_delim(delim="\t") %>%
  left_join(geneAnnos, by="Gene") %>%
  select(Gene, Annotation, everything())
print(GvRoary)

geneAnnos <- file.path(roaryDir, "gene_presence_absence.csv") %>%
  read_csv() %>%
  select(Gene, Annotation)
print(geneAnnos)

#Roary output for core genome
GvRoaryCore <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/Gard WGS/20181008_GvRoary95strains/gene_presence_absence.csv" %>%
  read_csv()
print(GvRoaryCore)

#Reference dataframe
gvDF <- file.path("..", "GardnerellaMetadata.csv") %>%
  read_csv()
print(gvDF)
```

## Check Gardnerella read counts in samples
```{r}
# GF <- read_delim(GF_path, delim = "\t") %>% # Gene family abundances
#   separate(`# Gene Family`, into=c("geneFamily", "org"), sep="\\|", fill="right") %>%
#   separate(org, into=c("Genus", "Species"), sep="\\.") %>%
#   mutate(Genus=str_extract(Genus,"[^g__].*$"),
#          Species=str_extract(Species, "[^s__].*"))
# 
# GF %>%
#   dplyr::filter(Genus=="Gardnerella") %>%
#   select(-geneFamily, -Genus, -Species) %>%
#   colSums() %>%
#   as.data.frame() %>%
#   filter(.!=0) %>%
#   summary(.)
```  
values for read counts: 10,000 (~1st q) 400,000 (~max)

# Genes for testing
```{r}
#Test genes:
#make a table of the genes for testing
exampleGenes <- tribble(
  ~mech, ~gene, ~name, 
  "AMR", "HMPREF0421_20446", "Methicillin resistance protein",
  "AMR", "HMPREF0421_20507", "Aminoglycoside phosphotransferase",
  "Biofilm formation", "HMPREF0421_20407", "Family 2 glycosyltransferase HMPREF0421_20407",
  "Biofilm formation", "HMPREF0421_20405", "Family 2 glycosyltransferase HMPREF0421_20405",
  "Biofilm formation", "HMPREF0421_20412", "Family 2 glycosyltransferase HMPREF0421_20412",
  "Biofilm formation", "HMPREF0421_20413", "Family 2 glycosyltransferase HMPREF0421_20413",
  "Biofilm formation", "HMPREF0424_1180", "Family 2 glycosyltransferase HMPREF0424_1180",
  "Biofilm formation", "HMPREF0424_1189", "Family 2 glycosyltransferase HMPREF0424_1189",
  "Biofilm formation", "HMPREF0424_1181", "Family 1 glycosyltransferase HMPREF0424_1181",
  "Epithelial adhesion", "HMPREF0424_1026", "Type-I fimbrial major subunit precursor",
  "Mucin Degredation", "Sialidase$", "Sialidase",
  "Mucin Degredation", "HMPREF0421_20101", "a-L-fucosidase",
  "Mucin Degredation", "Sialidase A$", "Sialidase A",
  "Virulence", "HMPREF0424_1186", "LicD protein",
)
head(exampleGenes)
```


```{r}
#names of genes from example genes
geneNames <- exampleGenes$gene %>%
  map(~filter(geneAnnos, str_detect(Annotation, .x))) %>%
  map(~.$Gene)
geneNames
```

# Create Samples
**1)** Randomly sample 0 or 1 strains per clade for 100 samples.
**2)** Multiply each strain by an integer to simulate read counts
```{r}
# add clade information to gene presence absence dataframe. 
geneCladesPA <- GvRoary %>%
  gather("Strain", "Reads", 3:ncol(.)) %>%
  left_join(gvDF[,c("Strain", "Clade")], by="Strain")

sampleCreate <- function(genePA){
  nClades <- sample(1:6, 1, replace = FALSE) # determine the number of clades to sample 
  whichClades <- sample(c("C1", "C2", "C3", "C4", "C5", "C6"), nClades, replace=FALSE) # determine which clades to sample
  SAMPLE <- genePA %>%
  filter(Clade %in% whichClades) %>%
  split(.$Clade) %>% #split by clade
  map(., ~filter(.x, Strain==sample(unique(.x$Strain), 1))) %>% # select one strain per clade
  purrr::reduce(full_join, by = c("Gene", "Annotation", "Strain", "Reads", "Clade")) # collapse data frames
  return(SAMPLE)
}
#subset reference strains
sampleStrains <- rerun(100, sampleCreate(geneCladesPA))
sampleStrains[[1]]

#assign abundnaces of each strain and its genes by multiplying all of the genes of each strain by a constant
sampleGeneReads <- sampleStrains %>%
  map(., ~split(.x, .x$Strain)) %>% #separate by strain
  map(~map(.x, ~mutate(.x, Reads=Reads*round(runif(1, 2000, 70000))))) %>% #multiply each strain
  map(~purrr::reduce(.x, full_join, by = c("Gene", "Annotation", "Strain", "Reads", "Clade"))) # regroup
sampleGeneReads[[1]]
```

# Testing
## Basic Simulation with Gene Presence Absence
First simulate with gene presence absence information

Steps:
1) Estimate abundances with core genes
2) Compute G inverse for normalizing reads
  a) Normalize to the number of secY genes in the sample and the number of secY copies
3) Select genes, normalize, and compute glm

### 1) Estimate relative abundance with core genes
```{r}
# Get core genes
coreGenes <- GvRoaryCore %>%
  add_count(Annotation) %>%
  filter(`No. isolates`==72,
         n==1) %>%
  .$Annotation %>%
  {dplyr::filter(GvRoary, is.element(Annotation, .))} %>%
  .$Gene
coreGenes

#calculate relative abundance of each strain using marker genes
#Total number of core genes per strain
cladeAbundances <- sampleGeneReads %>%
  map(~filter(.x, Gene %in% coreGenes)) %>%
  map(~group_by(.x, Clade)) %>%
  map(~summarise(.x, strainCoreTotals=sum(Reads))) %>%
  map(~mutate(.x, sampleCoreTotals=sum(strainCoreTotals))) %>%
  map(~mutate(.x, cladeAbundance=strainCoreTotals/sampleCoreTotals)) %>%
  map2(., c(1:100), ~mutate(.x, Sample=.y)) %>%
  purrr::reduce(full_join, by=c("Clade", "strainCoreTotals", "sampleCoreTotals", "cladeAbundance", "Sample"))
head(cladeAbundances)

#Create unlabeled gene reads for testing
testSamples <- sampleGeneReads %>%
  map(., ~group_by(.x, Gene)) %>%
  map(., ~summarise(.x, Reads=sum(Reads))) %>%
  map2(., c(1:100), ~mutate(.x, Sample=.y))
testSamples[1]
```

### 2) Compute G inverse for normalizing reads
Normalize to the number of secY genes in the sample and the number of secY copies
```{r}
#G inverse (copy number/SecY reads):
#get the number of secY (universal single copy) reads  
Gs <- testSamples %>%
  map(., ~filter(.x, Gene=="secY")) %>%
  map(., ~rename(.x, secYreads=Reads)) %>%
  map2(., c(1:100), ~mutate(.x, Sample=.y)) %>%
  purrr::reduce(full_join, by=c("Sample", "secYreads")) %>%
  #left_join(samTot, by="Sample") %>%
  mutate(G=1/secYreads,
         Ginv=secYreads/1) %>%
  select(Sample, secYreads, G, Ginv) # samTots
Gs
```
### Select genes, normalize, and compute glm
```{r}
#bind G column and normalize read counts
computeGLM <- function(g, n){
 lm <- testSamples %>%
    map(., ~filter(.x, Gene %in% g)) %>%
    purrr::reduce(full_join, by=c("Gene", "Reads", "Sample")) %>%
    group_by(Sample) %>% 
    summarise(Reads=sum(Reads)) %>% # collapse when copies have multiple names
    select(Sample, Reads) %>%
    left_join(cladeAbundances[,c("Clade", "cladeAbundance", "Sample")], by="Sample") %>%
    spread(Clade, cladeAbundance) %>%
    mutate_at(vars(C1, C2, C3, C4, C5, C6), ~replace_na(.x, 0)) %>% #create a column per gene
    left_join(Gs[,c("Sample", "G", "Ginv")], by="Sample") %>%
    mutate(GReads=Reads*G) %>%
    glm(data=., GReads~C1+C2+C3+C4+C5+C6)
  coefs <- tidy(lm) %>%
    mutate(Gene=n)
  confints <- confint.default(lm) %>%
    data.frame() %>%  
    rownames_to_column(var="term") %>%
      mutate(Gene=n)
   out <-  coefs %>%
      full_join(confints, by=c("term", "Gene"))
    return(out)
}

estimates <- map2(geneNames, exampleGenes$name, ~computeGLM(.x, .y)) %>%
  purrr::reduce(full_join, by = c("term", "estimate", "std.error", "statistic", "p.value", "Gene", "X2.5..", "X97.5..")) %>%
  dplyr::rename(`2.5pct`=`X2.5..`, `97.7pct`=`X97.5..`)
View(estimates)
```

## Assess actual average copy number by clades in sample set 
```{r}
calculateCopies <- function(g, n) {
  mC <- sampleStrains %>%
    map(., ~filter(.x, Gene %in% g)) %>% # select genes
    map2(., c(1:100), ~mutate(.x, Sample=.y)) %>%
    purrr::reduce(full_join, by=c("Gene", "Annotation", "Strain", "Reads", "Clade", "Sample")) %>% 
    group_by(Clade, Sample) %>%
    summarise(Reads=sum(Reads)) %>% #collapse 
    group_by(Clade) %>%
    summarise(avgCopies=mean(Reads)) %>%
    mutate(Gene=n)
return(mC)
}

averageCopies <- map2(geneNames, exampleGenes$name, ~calculateCopies(.x, .y)) %>%
  purrr::reduce(full_join, by=c("Clade", "avgCopies", "Gene")) %>%
  dplyr::rename(term=Clade)
```


## Compare the results
```{r, fig.height=6, fig.width=7}
compare <- estimates %>%
  left_join(averageCopies, by=c("Gene", "term")) %>%
  select(Gene, term, avgCopies, estimate, `2.5pct`, `97.7pct`, std.error, statistic, p.value)
head(compare)

compare %>%
  filter(term!="(Intercept)") %>%
  ggplot(aes(x=avgCopies, y=estimate, color=term, shape=(p.value<0.05))) +
  geom_point() +
  facet_wrap(~Gene) +
  xlim(0,1.5) +
  ylim(0,1.5) +
  labs(x="Average Copies Per Clade", y="Estimate", color="Clade")
```


Session Info
```{r}
sessionInfo()
```