---
title: "Community Profiling with Humann2"
author: "HLB"
date: "4/11/2019"
output: html_notebook
---

# Setup
## Packages
```{r}
library(tidyverse)
```

## File paths
```{r}
data_path <- "/Users/hlberman/mnt/cluster/home/hlberman/VMMG/humann2/merged_tables"
gene_families_path <- file.path(data_path, "VM_genefamilies_cpm.tsv")
path_abundance_path <- file.path(data_path, "VM_pathabundance_cpm.tsv")
path_coverage_path <- file.path(data_path, "VM_pathcoverage_cpm.tsv")
sgDF_path <- file.path("/Users/hlberman/Documents/GitHub/pregnancy_metagenome/shotgunMetadata.tsv")
#figures_path
```

## Load Files
```{r}
gene_families <- read_delim(gene_families_path, delim="\t")
head(gene_families)
path_abundance <- read_delim(path_abundance_path, delim="\t")
head(path_abundance)
path_coverage <- read_delim(path_coverage_path, delim="\t")
head(path_coverage)
sgDF <- read_delim(sgDF_path, delim="\t")
sgDF$SampleID <- as.character(sgDF$SampleID)
head(sgDF)
```

# 
```{r}
colnames(gene_families)
colnames(path_abundance)
colnames(path_coverage)
```

# Adding metadata
## Rename colnames
```{r}
#gene families table
gf_names <- colnames(gene_families)[2:ncol(gene_families)] %>%
  map_chr(., ~str_extract(.x, "[0-9]{10}")) %>%
  c("FEATURE", .)
colnames(gene_families) <- gf_names

#path abundace table
pa_names <- colnames(path_abundance)[2:ncol(path_abundance)] %>%
  map_chr(., ~str_extract(.x, "[0-9]{10}")) %>%
  c("FEATURE", .)
colnames(path_abundance) <- pa_names

#path coverage table
pc_names <- colnames(path_coverage)[2:ncol(path_coverage)] %>%
  map_chr(., ~str_extract(.x, "[0-9]{10}")) %>%
  c("FEATURE", .)
colnames(path_coverage) <- pc_names
```

## Add metadata
```{r}
path_abundance0 <- rbind(c("SubjectID", sgDF$SubjectID[order(sgDF$SampleID)]),
                        c("GWcoll", sgDF$GWcoll[order(sgDF$SampleID)]),
                        c("GDcoll", sgDF$GDcoll[order(sgDF$SampleID)]),
                        c("GWdell", sgDF$GWdell[order(sgDF$SampleID)]),
                        c("RNAlater", sgDF$RNAlater[order(sgDF$SampleID)]),
                        c("UAB", sgDF$UAB[order(sgDF$SampleID)]),
                        c("PaperCohort", sgDF$PaperCohort[order(sgDF$SampleID)]),
                        c("TermPre", sgDF$TermPre[order(sgDF$SampleID)]),
                        c("topGardClade", sgDF$topGardClade[order(sgDF$SampleID)]),
                        path_abundance
                        )
write_delim(path_abundance0, file.path(data_path, "VM_pathabundance_cpm_md.tsv"), delim="\t")

path_coverage0 <- rbind(c("SubjectID", sgDF$SubjectID[order(sgDF$SampleID)]),
                        c("GWcoll", sgDF$GWcoll[order(sgDF$SampleID)]),
                        c("GDcoll", sgDF$GDcoll[order(sgDF$SampleID)]),
                        c("GWdell", sgDF$GWdell[order(sgDF$SampleID)]),
                        c("RNAlater", sgDF$RNAlater[order(sgDF$SampleID)]),
                        c("UAB", sgDF$UAB[order(sgDF$SampleID)]),
                        c("PaperCohort", sgDF$PaperCohort[order(sgDF$SampleID)]),
                        c("TermPre", sgDF$TermPre[order(sgDF$SampleID)]),
                        c("topGardClade", sgDF$topGardClade[order(sgDF$SampleID)]),
                        path_coverage
                        )
write_delim(path_coverage0, file.path(data_path, "VM_pathcoverage_cpm_md.tsv"), delim="\t")

gene_families0 <- rbind(c("SubjectID", sgDF$SubjectID[order(sgDF$SampleID)]),
                        c("GWcoll", sgDF$GWcoll[order(sgDF$SampleID)]),
                        c("GDcoll", sgDF$GDcoll[order(sgDF$SampleID)]),
                        c("GWdell", sgDF$GWdell[order(sgDF$SampleID)]),
                        c("RNAlater", sgDF$RNAlater[order(sgDF$SampleID)]),
                        c("UAB", sgDF$UAB[order(sgDF$SampleID)]),
                        c("PaperCohort", sgDF$PaperCohort[order(sgDF$SampleID)]),
                        c("TermPre", sgDF$TermPre[order(sgDF$SampleID)]),
                        c("topGardClade", sgDF$topGardClade[order(sgDF$SampleID)]),
                        gene_families
                        )
write_delim(gene_families0, file.path(data_path, "VM_genefamilies_cpm_md.tsv"), delim="\t")
```

## Some descriptives
### Numbers of genes/pathways
```{r}
gene_counts <- colSums(gene_families!=0) %>%
  -2 %>%
  .[-1]
pathway_counts <- colSums(path_abundance!=0) %>%
  -2 %>%
  .[-1]
sapply(list(Genes=gene_counts, Pathways=pathway_counts), summary)
ggplot(as.data.frame(gene_counts), aes(gene_counts)) +
  geom_histogram() +
  labs(x="Genes")
ggplot(as.data.frame(pathway_counts), aes(pathway_counts)) + 
  geom_histogram() +
  labs(x="Pathways")
```

```{r}
df1 <- tibble(SampleID=names(gene_counts),
              GeneCount=gene_counts,
              PathwayCount=pathway_counts)
sgDF0 <- sgDF %>%
  left_join(df1, by="SampleID")

ggplot(data=sgDF0) +
  geom_point(aes(x=GeneCount, y=GWdell, color=UAB)) + 
  geom_hline(yintercept=37)

ggplot(data=sgDF0) +
  geom_point(aes(x=PathwayCount, y=GWdell, color=UAB)) + 
  geom_hline(yintercept=37)
```

# Session Info
```{r}
sessionInfo()
```
