---
title: "R Notebook"
output: html_notebook
---

# Set up
## Packages
```{r}
library(tidyverse)
library(rMSA)
library(msa)
library(Biostrings)
```

## File paths
```{r}
repo_dir <- "/Users/hlberman/Documents/GitHub/pregnancy_metagenome/gard_mapping/clade_assignments"
ref_fastas_path <- file.path(repo_dir, "ref_genes", "refGenes.fasta")
ref_genes_df_path <- file.path(repo_dir, "ref_genes", "refGenesDF.csv")
ref_df_path <- "/Users/hlberman/Documents/GitHub/pregnancy_metagenome/GardnerellaMetadata.csv"
out_path <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/gardMapping/clade_assignments/multiples_testing"
```

## Import
```{r}
ref_fastas <- readDNAStringSet(ref_fastas_path)
ref_genes_df <- read_csv(ref_genes_df_path)
ref_df <- read_csv(ref_df_path)
```

#
## Split by annotation
```{r}
## make list of annos for each
ref_genes0 <- ref_genes_df %>%
  select(Annotation, ref_df$Strain[ref_df$In_Directory==TRUE])
rownames(ref_genes0) <- ref_genes0$Annotation
ref_genes0 <- ref_genes0 %>%
  select(-Annotation) %>%
  t() %>%
  as_tibble() %>%
  as.list() 

## make separate DNA string sets for each annotation
ref_fastas_byanno <- map(ref_genes0, ~ref_fastas[is.element(ref_fastas@ranges@NAMES, .x)])
head(ref_fastas_byanno)
```

## Perform MSA
```{r}
alns <- map(ref_fastas_byanno, rMSA::mafft)
test <- map(alns, ~consensusViews(.x))
test1 <- map(alns, ~consensusMatrix(.x))
test2 <- map(alns, ~alphabetFrequency(.x))
test2[1]
msa::msaPrettyPrint(test, output = "pdf", file=file.path(out_path, "msa1.pdf"))
```

## Assess to find 150nt region with 100% ID
```{r}
rMSA::plot.DNAMultipleAlignment(alns[1])
class(alns[1])
```

# Session Info
```{r}
sessionInfo()
```

