---
title: "Assess Kraken Reads"
author: "Hanna Berman"
date: "1/9/2019"
output: html_document
---
# Setup

## Load packages
```{r}
library(tidyverse);packageVersion("tidyverse")
```

## Load data
```{r}
krakenTable <- "/Users/hlberman/mnt/cluster/home/hlberman/VMMG/kraken2_bracken/upload_bracken_out/merged_braken_KrakenReads_pe.tsv" %>%
  read_tsv() 
rownames(krakenTable) <- krakenTable$name
output_dir <- getwd()
sgDF <- "/Users/hlberman/Documents/GitHub/pregnancy_metagenome/shotgunMetadata.tsv" %>%
  read_tsv()
```

# Kraken human reads 
```{r}
#make separate reads number and abundance tables
knumTable <- krakenTable %>%
  select(name, contains("_num")) %>%
  rename_if(is.numeric, ~substr(.x, 1, 10))

KabTable <- krakenTable %>%
  select(name, contains("_frac")) %>%
  rename_if(is.numeric, ~substr(.x, 1, 10)) 

sapply(list(KnumTable=colSums(knumTable[,2:ncol(knumTable)]), kabTable=colSums(KabTable[2:ncol(knumTable)])), summary)
```

## Assess Human DNA Contamination and compare to previous human contamination results
```{r}
humanTable <- krakenTable %>%
  filter(name=="Homo sapiens") %>%
  select(-name, -taxonomy_id, -taxonomy_lvl) %>%
  gather("SampleID", "temp", everything()) %>%
  mutate(Ab_frac=case_when(str_detect(SampleID, "num")~"krakenHuReads_pe",
                           str_detect(SampleID, "frac")~"krakenHuAbundance_pe"),
         SampleID=substr(SampleID, 1,10)) %>%
  spread(Ab_frac, temp)
  
write_delim(humanTable, file.path(output_dir, "krakenHumanReads.tsv"), delim = "\t")
```


print session info
```{r}
sessionInfo()
```