---
title: "Compare Relative Abundance Assignment Methods"
author: "Hanna Berman"
date: "2/04/2019"
output: html_notebook
---

Compare microbiome relative abundance measures from 16S, Metaphlan2, and Kraken2/Bracken. Taxa of interest (TOI) including *Lactobacillus crispatus*, *Lactobacillus iners*, *Lactobacillus gasseri*, *Lactobacillus jensenii*, *Gardnerella vaginalis*, *Mycoplasma hominis*, *Prevotella bivia*, *Atopobium vaginae* (or only *Lactobacillus* at species level and the four others at genus level)

# Setup
##Load packages
```{r libraries, warning=FALSE}
library(tidyverse); packageVersion("tidyverse")
library(phyloseq); packageVersion("phyloseq") 
```

## Define file paths
```{r}
main_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/abundanceMethodsComparisons"
PS_path <- file.path(main_dir, "PS_RData")
metaphlanPS <- file.path(PS_path, "metaphlan_PS.RData")
brackenPS <- file.path(PS_path, "bracken_PS.RData")
sixteenSPS <- file.path(PS_path, "16S_PS.RData")
output_dir <- file.path(main_dir, "analysisOutput")
df_path <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/metadata/shotgunMetadata.tsv"
```

## load PS objects and dataframe
```{r}
load(metaphlanPS, envir = parent.frame(), verbose = FALSE)
load(brackenPS, envir = parent.frame(), verbose = FALSE)
load(sixteenSPS, envir = parent.frame(), verbose = FALSE)
df <- read.table(df_path, sep="\t", header=TRUE, stringsAsFactors = FALSE)
```

## Plot subject traces
### 16S vs metaphlan w. other
```{r}
ggplot() + 
  geom_line(data=sixteenS_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=metaphlLG_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella  ", "Prevotella  ", "Atopobium  ", "Mycoplasma  ", "Other  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#777777", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Abundance of taxa of interest\n16S (solid) vs. Metaphlan (dashed)", x="Gestation Weeks at Collection", "Relative Abundance", color = "Taxon") +
  ggsave(file.path(output_dir, "Abundance_metaphlV16S_other.png"), width = 20, height = 15)
```

### 16S vs metaphlan w.o other
```{r}
ggplot() + 
  geom_line(data = sixteenS_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data = metaphlLG_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella  ", "Prevotella  ", "Atopobium  ", "Mycoplasma  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Abundance of taxa of interest\n16S (solid) vs. Metaphlan (dashed)", x="Gestation Weeks at Collection", "Relative Abundance", color = "Taxon") +
  ggsave(file.path(output_dir, "Abundance_metaphlV16S.png"), width = 20, height = 15)
```

### 16S vs bracken w. other
```{r}
ggplot() + 
  geom_line(data= sixteenS_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=brackenLG_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella  ", "Prevotella  ", "Atopobium  ", "Mycoplasma  ", "Other  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#777777", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Abundance of taxa of interest\n16S (solid) vs. Bracken (dashed)", x="Gestation Weeks at Collection", "Relative Abundance", color = "Taxon") +
  ggsave(file.path(output_dir, "Abundance_brackenV16S_other.png"), width = 20, height = 15)
```

### 16S vs bracken w.o other
```{r}
ggplot() + 
  geom_line(data= sixteenS_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=brackenLG_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella  ", "Prevotella  ", "Atopobium  ", "Mycoplasma  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Abundance of taxa of interest\n16S (solid) vs. Bracken (dashed)", x="Gestation Weeks at Collection", "Relative Abundance", color = "Taxon") +
  ggsave(file.path(output_dir, "Abundance_brackenV16S.png"), width = 20, height = 15)
```

### metaphlan vs bracken w. other
```{r}
ggplot() + 
  geom_line(data= metaphlA_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=brackenA_toi_o_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella vaginalis", "Prevotella bivia", "Atopobium vaginae", "Mycoplasma hominis", "Other  "),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#777777", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Abundance of taxa of interest\nBracken (solid) vs. Metaphlan (dashed)", x="Gestation Weeks at Collection", "Relative Abundance", color = "Taxon") +
  ggsave(file.path(output_dir, "Abundance_brackenVmetaphlan_other.png"), width = 20, height = 15)
```

### metaphlan vs bracken w.o other
```{r}
ggplot() + 
  geom_line(data= metaphlA_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity") +
  geom_line(data=brackenA_toi_df, aes(x = GWcoll, y = Abundance, color = Taxon), stat = "identity", linetype="dashed") +
  scale_color_manual(breaks=c("Lactobacillus crispatus", "Lactobacillus gasseri", "Lactobacillus jensenii", "Lactobacillus iners", "Gardnerella vaginalis", "Prevotella bivia", "Atopobium  vaginae", "Mycoplasma hominis"),
                     values = c("#F8766D", "#D39200", "#00BA38", "#93AA00", "#00B9E3", "#00C19F", "#DB72FB", "#FF61C3"))+
  facet_wrap(~SubjectID, nrow = 7) +
  theme(text =element_text(size=30)) +
  labs(title = "Abundance of taxa of interest\nBracken (solid) vs. Metaphlan (dashed)", x="Gestation Weeks at Collection", "Relative Abundance", color = "Taxon") +
  ggsave(file.path(output_dir, "Abundance_brackenVmetaphlan.png"), width = 20, height = 15)
```

## Scatterplot by taxa

### Make merged data frame 
first edit the data frames to make columns necessary for table merging and further analyses
```{r}

prepTable16S <- function(dafr_in, abname){
  dafr_out <- dafr_in %>%
    select(SampleID, Taxon, Abundance)
  colnames(dafr_out) <- c("SampleID_a", "Taxon", abname)
  return(dafr_out)}

prepTableShotgun <- function(dafr_in, abname){
  dafr_out <- dafr_in %>%
    select(SampleID_a, Taxon, Abundance) %>%
    mutate(SampleID_a=as.character(SampleID_a)) 
  colnames(dafr_out) <- c("SampleID_a", "Taxon", abname)
  return(dafr_out)}

sixteenS_toi_df0 <- prepTable16S(sixteenS_toi_df, "Abundance_16S")
sixteenS_toi_o_df0 <- prepTable16S(sixteenS_toi_o_df, "Abundance_16S")
metaphlA_toi_df0 <- prepTableShotgun(metaphlA_toi_df, "Abundance_metaphl")
metaphlA_toi_o_df0 <- prepTableShotgun(metaphlA_toi_o_df, "Abundance_metaphl")
metaphlLG_toi_df0 <- prepTableShotgun(metaphlLG_toi_df, "Abundance_metaphl")
metaphlLG_toi_o_df0 <- prepTableShotgun(metaphlLG_toi_o_df, "Abundance_metaphl")
brackenA_toi_df0 <- prepTableShotgun(brackenA_toi_df, "Abundance_brack")
brackenA_toi_o_df0 <- prepTableShotgun(brackenA_toi_o_df, "Abundance_brack")
brackenLG_toi_df0 <- prepTableShotgun(brackenLG_toi_df, "Abundance_brack")
brackenLG_toi_o_df0 <- prepTableShotgun(brackenLG_toi_o_df, "Abundance_brack")
```

Merge into one data frame
```{r}
RNA.later <- df$SampleID_a[df$RNAlater=="Yes"]

#vs. 16S with other:
meta_brack_16S_other <- full_join(sixteenS_toi_o_df0,
                                  metaphlLG_toi_o_df0, 
                                  by=c("SampleID_a", "Taxon")) %>%
                        full_join(brackenLG_toi_o_df0, 
                                   by=c("SampleID_a", "Taxon")) %>%
                        replace_na(list(Abundance_metaphl=0, Abundance_brack=0, Abundance_16S=0)) %>%
                        mutate(RNAlater=NA)
meta_brack_16S_other$RNAlater[meta_brack_16S_other$SampleID_a %in% RNA.later] <- "Yes"
meta_brack_16S_other$RNAlater[!(meta_brack_16S_other$SampleID_a %in% RNA.later)] <- "No"

meta_brack_16S <- full_join(sixteenS_toi_df0,
                            metaphlLG_toi_df0, 
                            by=c("SampleID_a", "Taxon")) %>%
                  full_join(brackenLG_toi_df0, 
                            by=c("SampleID_a", "Taxon")) %>%
                  replace_na(list(Abundance_metaphl=0, Abundance_brack=0, Abundance_16S=0)) %>%
                  mutate(RNAlater=NA)
meta_brack_16S$RNAlater[meta_brack_16S$SampleID_a %in% RNA.later] <- "Yes"
meta_brack_16S$RNAlater[!(meta_brack_16S$SampleID_a %in% RNA.later)] <- "No"

meta_brack_other <- full_join(metaphlA_toi_o_df0,
                              brackenA_toi_o_df0,
                              by=c("SampleID_a", "Taxon")) %>%
                    replace_na(list(Abundance_metaphl=0, Abundance_brack=0, Abundance_16S=0)) %>%
                    mutate(RNAlater=NA)
meta_brack_other$RNAlater[meta_brack_other$SampleID_a %in% RNA.later] <- "Yes"
meta_brack_other$RNAlater[!(meta_brack_other$SampleID_a %in% RNA.later)] <- "No"

meta_brack <- full_join(metaphlA_toi_df0,
                        brackenA_toi_df0, 
                        by=c("SampleID_a", "Taxon")) %>%
              replace_na(list(Abundance_metaphl=0, Abundance_brack=0, Abundance_16S=0)) %>%
              mutate(RNAlater=NA)
meta_brack$RNAlater[meta_brack$SampleID_a %in% RNA.later] <- "Yes"
meta_brack$RNAlater[!(meta_brack$SampleID_a %in% RNA.later)] <- "No"
```

### 16 vs. Metaphlan w. other
```{r}
ggplot(data= meta_brack_16S_other) + 
  geom_point(aes(x = Abundance_metaphl, y = Abundance_16S, color=RNAlater), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Relative Abundances:\nMetaphlan vs. 16S", x="Relative Abundance (Metaphlan)", y="Relative Abundance (16S)", color = "RNA Later ") +
  ggsave(file.path(output_dir, "Metaphlan16S_byTaxon_o.png"))
```

### 16 vs. Bracken w. other
```{r}
ggplot(data= meta_brack_16S_other) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_16S, color=RNAlater), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Relative Abundances:\nBracken vs. 16S", x="Relative Abundance (Bracken)", y="Relative Abundance (16S)", color = "RNA Later ") +
  ggsave(file.path(output_dir, "Bracken16S_byTaxon_o.png"))
```

### Metaphlan vs. Bracken w. other
```{r}
ggplot(data= meta_brack_16S_other) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_metaphl), color="#00C19F", stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Relative Abundances:\nBracken vs. Metaphlan", x="Relative Abundance (Bracken)", y="Relative Abundance (Metaphlan)") +
  ggsave(file.path(output_dir, "BrackenMetaphl_byTaxon_o.png"))
```

### 16 vs. Metaphlan w.o other
```{r}
ggplot(data= meta_brack_16S) + 
  geom_point(aes(x = Abundance_metaphl, y = Abundance_16S, color=RNAlater), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Relative Abundances:\nMetaphlan vs. 16S", x="Relative Abundance (Metaphlan)", y="Relative Abundance (16S)", color = "RNA Later ") +
  ggsave(file.path(output_dir, "Metaphlan16S_byTaxon.png"))
```

### 16 vs. Bracken w.o other
```{r}
ggplot(data= meta_brack_16S) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_16S, color=RNAlater), stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Relative Abundances:\nBracken vs. 16S", x="Relative Abundance (Bracken)", y="Relative Abundance (16S)", color = "RNA Later ") +
  ggsave(file.path(output_dir, "Bracken16S_byTaxon.png"))
```

### Metaphlan vs. Bracken w.o other
```{r}
ggplot(data= meta_brack_16S) + 
  geom_point(aes(x = Abundance_brack, y = Abundance_metaphl), color="#00C19F", stat = "identity") + 
  facet_wrap(~Taxon) + 
  coord_fixed() + 
  labs(title = "Relative Abundances:\nBracken vs. Metaphlan", x="Relative Abundance (Bracken)", y="Relative Abundance (Metaphlan)") +
  ggsave(file.path(output_dir, "BrackenMetaphl_byTaxon.png"))
```

```{r}
#get ratio comparison conditions
taxaLG <- c("Lactobacillus iners", "Lactobacillus gasseri", "Lactobacillus crispatus", "Lactobacillus jensenii", "Gardnerella  ", "Mycoplasma  ", "Prevotella  ", "Atopobium  ")
pairLG <- combn(taxaLG, 2)

taxaS <- c("Lactobacillus iners", "Lactobacillus gasseri", "Lactobacillus crispatus", "Lactobacillus jensenii", "Gardnerella vaginalis", "Mycoplasma hominis", "Prevotella bivia", "Atopobium vaginae")
pairS <- combn(taxaS, 2)

#make list of 16S taxa ratios
ratios16S <- c(NULL)
for(i in c(1:length(df$SampleID_a))){
  sam <- df$SampleID_a[i]
  for (j in c(1:dim(pairLG)[2])){
    a <- meta_brack_16S_other$Abundance_16S[meta_brack_16S_other$SampleID_a==sam & meta_brack_16S_other$Taxon==pairLG[1,j]]
    b <- meta_brack_16S_other$Abundance_16S[meta_brack_16S_other$SampleID_a==sam & meta_brack_16S_other$Taxon==pairLG[2,j]] 
    c <- a/b
    ifelse((length(c)>0), 
           d <- log10(c),
           d <- NA)
    names(d) <- paste0(sam, "_", pairLG[1,j], ":", pairLG[2,j])
    ratios16S <- c(ratios16S, d)
    }
}

#make list of metaphlan taxa ratios (lactobacillus at species level, only)
#16S
ratios16S <- matrix(nrow = 0, ncol = 3)
colnames(ratios16S) <- c("SampleID_a", "Taxa", "logRatio_16S")
for(i in c(1:length(df$SampleID_a))){
  sam <- df$SampleID_a[i]
  for (j in c(1:dim(pairLG)[2])){
    a <- meta_brack_16S_other$Abundance_16S[meta_brack_16S_other$SampleID_a==sam & meta_brack_16S_other$Taxon==pairLG[1,j]]
    b <- meta_brack_16S_other$Abundance_16S[meta_brack_16S_other$SampleID_a==sam & meta_brack_16S_other$Taxon==pairLG[2,j]] 
    c <- a/b
    ifelse((length(c)>0), 
           d <- log10(c),
           d <- NA)
    e <- c(sam, paste0(pairLG[1,j], ":", pairLG[2,j]), d)
    ratios16S <- rbind(ratios16S, e)
    }
}
ratios16S[ratios16S=="Inf"|ratios16S=="-Inf"|ratios16S=="NaN"] <- NA

#metaphlan
ratiosMetaphlan <- matrix(nrow = 0, ncol = 3)
colnames(ratiosMetaphlan) <- c("SampleID_a", "Taxa", "logRatio_metaphlan")
for(i in c(1:length(df$SampleID_a))){
  sam <- df$SampleID_a[i]
  for (j in c(1:dim(pairLG)[2])){
    a <- meta_brack_16S_other$Abundance_metaphl[meta_brack_16S_other$SampleID_a==sam & meta_brack_16S_other$Taxon==pairLG[1,j]]
    b <- meta_brack_16S_other$Abundance_metaphl[meta_brack_16S_other$SampleID_a==sam & meta_brack_16S_other$Taxon==pairLG[2,j]] 
    c <- a/b
    ifelse((length(c)>0), 
           d <- log10(c),
           d <- NA)
    e <- c(sam, paste0(pairLG[1,j], ":", pairLG[2,j]), d)
    ratiosMetaphlan <- rbind(ratiosMetaphlan, e)
    }
}
ratiosMetaphlan[ratiosMetaphlan=="Inf"|ratiosMetaphlan=="-Inf"|ratiosMetaphlan=="NaN"] <- NA

#bracken
ratiosBracken <- matrix(nrow = 0, ncol = 3)
colnames(ratiosBracken) <- c("SampleID_a", "Taxa", "logRatio_bracken")
for(i in c(1:length(df$SampleID_a))){
  sam <- df$SampleID_a[i]
  for (j in c(1:dim(pairLG)[2])){
    a <- meta_brack_16S_other$Abundance_brack[meta_brack_16S_other$SampleID_a==sam & meta_brack_16S_other$Taxon==pairLG[1,j]]
    b <- meta_brack_16S_other$Abundance_brack[meta_brack_16S_other$SampleID_a==sam & meta_brack_16S_other$Taxon==pairLG[2,j]] 
    c <- a/b
    ifelse((length(c)>0), 
           d <- log10(c),
           d <- NA)
    e <- c(sam, paste0(pairLG[1,j], ":", pairLG[2,j]), d)
    ratiosBracken <- rbind(ratiosBracken, e)
    }
}
ratiosBracken[ratiosBracken=="Inf"|ratiosBracken=="-Inf"|ratiosBracken=="NaN"] <- NA

#ratios with species for metaphlan vs bracken
#metaphlan
ratiosMetaphlanS <- matrix(nrow = 0, ncol = 3)
colnames(ratiosMetaphlanS) <- c("SampleID_a", "Taxa", "logRatio_metaphlan")
for(i in c(1:length(df$SampleID_a))){
  sam <- df$SampleID_a[i]
  for (j in c(1:dim(pairLG)[2])){
    a <- meta_brack_other$Abundance_metaphl[meta_brack_other$SampleID_a==sam & meta_brack_other$Taxon==pairS[1,j]]
    b <- meta_brack_other$Abundance_metaphl[meta_brack_other$SampleID_a==sam & meta_brack_other$Taxon==pairS[2,j]] 
    c <- a/b
    ifelse((length(c)>0), 
           d <- log10(c),
           d <- NA)
    e <- c(sam, paste0(pairS[1,j], ":", pairS[2,j]), d)
    ratiosMetaphlanS <- rbind(ratiosMetaphlanS, e)
    }
}
ratiosMetaphlanS[ratiosMetaphlanS=="Inf"|ratiosMetaphlanS=="-Inf"|ratiosMetaphlanS=="NaN"] <- NA

#bracken
ratiosBrackenS <- matrix(nrow = 0, ncol = 3)
colnames(ratiosBrackenS) <- c("SampleID_a", "Taxa", "logRatio_bracken")
for(i in c(1:length(df$SampleID_a))){
  sam <- df$SampleID_a[i]
  for (j in c(1:dim(pairLG)[2])){
    a <- meta_brack_other$Abundance_brack[meta_brack_other$SampleID_a==sam & meta_brack_other$Taxon==pairS[1,j]]
    b <- meta_brack_other$Abundance_brack[meta_brack_other$SampleID_a==sam & meta_brack_other$Taxon==pairS[2,j]] 
    c <- a/b
    ifelse((length(c)>0), 
           d <- log10(c),
           d <- NA)
    e <- c(sam, paste0(pairS[1,j], ":", pairS[2,j]), d)
    ratiosBrackenS <- rbind(ratiosBrackenS, e)
    }
}
ratiosBrackenS[ratiosBrackenS=="Inf"|ratiosBrackenS=="-Inf"|ratiosBrackenS=="NaN"] <- NA

#merge tables
ratiosLG <- as.data.frame(ratios16S) %>%
  full_join(as.data.frame(ratiosMetaphlan), by = c("SampleID_a", "Taxa")) %>%
  full_join(as.data.frame(ratiosBracken), by = c("SampleID_a", "Taxa")) %>%
  mutate(logRatio_16S=as.numeric(as.character(logRatio_16S)),
         logRatio_metaphlan=as.numeric(as.character(logRatio_metaphlan)),
         logRatio_bracken=as.numeric(as.character(logRatio_bracken)))


ratiosS <- as.data.frame(ratiosMetaphlanS) %>%
  full_join(as.data.frame(ratiosBrackenS), by = c("SampleID_a", "Taxa")) %>%
  mutate(logRatio_metaphlan=as.numeric(as.character(logRatio_metaphlan)),
         logRatio_bracken=as.numeric(as.character(logRatio_bracken)))

#add RNA Later information
is.rna.later <-  ratiosLG$SampleID_a %in% df$SampleID_a[df$RNAlater=="Yes"]
ratiosLG$RNAlater[is.rna.later] <- "Yes"
ratiosLG$RNAlater[!is.rna.later] <- "No"

is.rna.laterS <-  ratiosS$SampleID_a %in% df$SampleID_a[df$RNAlater=="Yes"]
ratiosS$RNAlater[is.rna.laterS] <- "Yes"
ratiosS$RNAlater[!is.rna.laterS] <- "No"
```

# Taxa ratios: plot separately
```{r}
ggplot(data=ratiosLG) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_16S, color=RNAlater)) +
        labs(title = "Log ratios of taxa:\nMetaphlan vs. 16S Amplicon", x="Metaphlan", y="16S Amplicon") + 
        theme(text = element_text(size=10),
              axis.title = element_text(size = 15),
              title=element_text(size=20)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) +
ggsave(file.path(output_dir, "logRatio_metaphlV16S.png"), height = 20, width = 20)

ggplot(data=ratiosLG) +
        geom_point(aes(x=logRatio_bracken, y=logRatio_16S, color=RNAlater)) +
        labs(title = "Log ratios of taxa:\nBracken vs. 16S Amplicon", x="Bracken", y="16S Amplicon") + 
        theme(text = element_text(size=10),
              axis.title = element_text(size = 15),
              title=element_text(size=20)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) +
ggsave(file.path(output_dir, "logRatio_brackenV16S.png"), height = 20, width = 20)

ggplot(data=ratiosLG) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_bracken)) +
        labs(title = "Log ratios of taxa:\nMetaphlan vs. Bracken", x="Metaphlan", y="Bracken") + 
        theme(text = element_text(size=10),
              axis.title = element_text(size = 15),
              title=element_text(size=20)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) +
ggsave(file.path(output_dir, "logRatio_metaphlVbrack.png"), height = 20, width = 20)

ggplot(data=ratiosS) +
        geom_point(aes(x=logRatio_metaphlan, y=logRatio_bracken)) +
        labs(title = "Log ratios of taxa:\nMetaphlan vs. Bracken", x="Metaphlan", y="Bracken") + 
        theme(text = element_text(size=10),
              axis.title = element_text(size = 15),
              title=element_text(size=20)) +
        coord_fixed() +
        xlim(-5,5)+
        ylim(-5,5) +
        facet_wrap(~Taxa) +
ggsave(file.path(output_dir, "logRatio_metaphlVbrack_S.png"), height = 20, width = 20)

```

```{r}
#p1 <- ggplot(data=ratiosLG) +
#        geom_point(aes(x=logRatio_metaphlan, y=logRatio_16S, color=RNAlater), stat = "identity") +
#        coord_fixed() +
#        facet_grid(Taxa~.)
#p2 <- ggplot(data=ratiosLG) +
#        geom_point(aes(x=logRatio_bracken, y=logRatio_16S, color=RNAlater), stat = "identity") +
#        coord_fixed() +
#        facet_grid(Taxa~.)
#p3 <- ggplot(data=ratiosLG) +
#        geom_point(aes(x=logRatio_bracken, y=logRatio_metaphlan), color="#00C19F", stat="identity") +
#        coord_fixed() +
#        facet_grid(Taxa~.)
#ratioPlot <- grid.arrange(p1, p2, p3, ncol=3)
#ggsave(file.path(output_dir, "test.png"), plot=ratioPlot, width = 30, height = 50, limitsize = FALSE, units="in")
```


# Session Info
```{r}
sessionInfo()
```