---
title: "gardPresenceVsMetadata"
author: "Hanna Berman"
date: "2/14/2019"
output: html_document
---

# Setup
##Load packages
```{r libraries, warning=FALSE}
library(tidyverse); packageVersion("tidyverse")
library(phyloseq); packageVersion("phyloseq")
library(ggpubr) ; packageVersion("ggpubr")
library(tables); packageVersion("tables")
```

## Define file paths
```{r}
main_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/abundanceMethodsComparisons"
PS_path <- file.path(main_dir, "PS_RData")
metaphlanPS <- file.path(PS_path, "metaphlan_PS.RData")
brackenPS <- file.path(PS_path, "bracken_PS.RData")
sixteenSPS <- file.path(PS_path, "16S_PS.RData")
output_dir <- file.path(main_dir, "analysisOutput")
df_path <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/metadata/shotgunMetadata.tsv"
```

## load PS objects and dataframe
```{r}
load(file=metaphlanPS)
load(file=brackenPS)
load(file=sixteenSPS)
df <- read.table(df_path, sep="\t", header=TRUE, stringsAsFactors = FALSE)
```

# Gard Dominance vs. PTB (by sample)
```{r}
tableGardMax <- function(PS){
  PS0 <- PS %>%
    psmelt()
  PS1 <- PS0 %>%
    aggregate(data=., Abundance~SampleID, max) %>%
    left_join(PS0[,c("SampleID", "SubjectID", "Genus", "Abundance", "TermPre", "UAB", "PaperCohort")], by=c("Abundance", "SampleID")) %>% 
    mutate(GardMax=case_when(Genus=="Gardnerella"~"Yes",
                             Genus!="Gardnerella"~"No"))
  tab <- table(GardMax=PS1$GardMax, Preterm=PS1$TermPre)
  PC <- xtabs(~ GardMax + TermPre + PaperCohort, PS1)
  UAB <- xtabs(~ GardMax + TermPre + UAB, PS1)
  return(list(`All`=tab, `By Paper Cohort`=PC, `By Cohort`=UAB))
}
```

## Based on 16S Data
```{r}
tableBy16S <- tableGardMax(PS_16S)
tableBy16S
```

## Based on Metaphlan Data
```{r}
tableByMetaphlan <- tableGardMax(PS_metaphlan_all)
tableByMetaphlan
```

## Based on Bracken Data
```{r}
tableByBracken <- tableGardMax(PS_bracken_all)
tableByBracken
```

# Max taxon per sample by method
```{r}
A <- PS_16S %>%
  psmelt()
B <- PS_metaphlan_all %>%
  psmelt()
C <- PS_bracken_all %>%
  psmelt()

A0 <- A %>%
    aggregate(data=., Abundance~SampleID_a, max) %>%
    left_join(A[,c("SampleID_a", "Genus", "Abundance")], by=c("Abundance", "SampleID_a")) %>%
    rename(AmplicGenus=Genus, AmplicAbundance=Abundance)
B0 <- B %>%
    aggregate(data=., Abundance~SampleID_a, max) %>%
    left_join(B[,c("SampleID_a", "Genus", "Abundance")], by=c("Abundance", "SampleID_a")) %>%
    rename(MetaphlGenus=Genus, MetaphlAbundance=Abundance)
C0 <- C %>%
    aggregate(data=., Abundance~SampleID_a, max) %>%
    left_join(C[,c("SampleID_a", "Genus", "Abundance")], by=c("Abundance", "SampleID_a")) %>%
    rename(BrackGenus=Genus, BrackAbundance=Abundance)

DominantTaxa <- A0 %>%
  full_join(B0, by="SampleID_a") %>%
  full_join(C0, by="SampleID_a")
```

Session Info
```{r}
sessionInfo()
```