---
title: "Metaphlan phyloseq objects for comparison of relative abundance methods"
author: "Hanna Berman"
date: "1/9/2019"
output:
  html_document: default
  html_notebook: default
---

The following is to create phyloseq objects from a merged metaphlan output from running metaphlan2 with the paired end files ONLY from files uploaded by D Goltsman. Only pe files were used as Kraken cannot handle and odd number of file inputs. (Single reads files were less than 1% of reads) 

# Setup

##Load packages
```{r libraries, warning=FALSE}
library(tidyverse); packageVersion("tidyverse")
library(phyloseq); packageVersion("phyloseq")
library(metaphlanr); packageVersion("metaphlanr")
```

## Define file paths
```{r}
main_dir <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/"
abundance_path <- file.path(main_dir, "RelativeAbundanceOutputs", "upload_metaphlan_merged_abundance_table_PE.tsv")
output_dir <- file.path(main_dir, "abundanceMethodsComparisons", "PS_RData")
df <- file.path(main_dir, "metadata", "shotgunMetadata.tsv")
```

# Make phyloseq object with all OTUs
A phyloseq object will merge sample metadata, OTU data, and taxonomy data 

**1)** Sample metadata
```{r}
sam <- df %>%
  read.table(header=TRUE, sep="\t", quote="", stringsAsFactors=FALSE) %>%
  data.frame()
rownames(sam) <- sam$SampleID
SAM <- sample_data(sam)
head(SAM)
```

**2)** Prepare abundance table
Abundance Table can be found: `./RelativeAbundanceOutputs/upload_metaphlan_merged_abundance_table_PE.tsv`  upload here refers to this data being the reads uploaded by D Goltsman, PE refers to only paired end reads files used as input (for the purpose of preparing with kraken/Bracken which cannot handle 3 files as input.
```{r}
abundanceTable <- readr::read_tsv(abundance_path)
#rename sample names
colnames(abundanceTable) <-c("Taxon", substr(colnames(abundanceTable[,2:ncol(abundanceTable)]), 1, 10))

# Filter to species and get species name
at0 <- abundanceTable %>%
    filter(str_detect(Taxon, "s__"), !str_detect(Taxon, "t__")) %>%
    mutate(Species = str_extract(Taxon, "(?<=s__)\\w+$")) %>%
    select(Species, everything())
summary(colSums(at0[,3:ncol(at0)]))

#remove Taxon column and species column and make rownames species
at1 <- at0 %>%
  select(-Taxon, -Species) %>%
  as.matrix()
rownames(at1) <- at0$Species
#make phyloseq OTU table object
OTU <- otu_table(at1, taxa_are_rows = TRUE)
head(OTU[,1:7])
```

**3)** Prepare taxonomy
```{r}
#Get the tax table
tax <- metaphlanr::parse_taxonomy(at0$Taxon)

#Remove Genus names from Species 
tax0 <- tax %>%
  select(-Strain, -Clade) %>%
  mutate(Species=str_extract(Species, paste0("(?<=",Genus,"_).*$"))) 

#MAKE MATRIX
taxmat <- tax0 %>%
    as.matrix()
rownames(taxmat) <- at0$Species
TAX <- tax_table(taxmat)
head(TAX)
```

**3)** Make phyloseq object with all otus
```{r}
PS_metaphlan_all <- phyloseq(SAM, OTU, TAX) %>%
  transform_sample_counts(function(x) {x/100})
```

# Lactobacillus at Species and all others at genus level
```{r}
PS_metaphlan_L <- PS_metaphlan_all %>%
  subset_taxa(Genus=="Lactobacillus") %>%
  tax_glom(taxrank="Species")
PS_metaphlan_G <- PS_metaphlan_all %>%
  subset_taxa(Genus!="Lactobacillus") %>%
  tax_glom(taxrank="Genus")  
PS_metaphlan_LG <- merge_phyloseq(PS_metaphlan_L, PS_metaphlan_G)
summary(sample_sums(PS_metaphlan_LG))
PS_metaphlan_LG@tax_table[is.na(PS_metaphlan_LG@tax_table[,"Species"]),"Species"] <- " " #change those NAs to a space (" ")
```

# Pull out Taxa of Interest and make data frames using psmelt
```{r}
#all species
metaphlan_a_toi0 <- PS_metaphlan_all %>%
  subset_taxa((Genus=="Gardnerella"|Genus=="Lactobacillus" | Genus=="Mycoplasma" | Genus=="Atopobium" | Genus=="Prevotella") & 
                (Species=="crispatus" | Species=="iners" | Species=="gasseri" | Species=="jensenii" | Species=="vaginalis" | Species=="hominis" | Species == "bivia" | Species=="vaginae") & 
                !(Genus=="Lactobacillus"&Species=="vaginalis")) # get samples of interest
summary(sample_sums(metaphlan_a_toi0))

metaphlan_a_toi <- metaphlan_a_toi0 %>%
  transform_sample_counts(function(x) {x/sum(x)})
summary(sample_sums(metaphlan_a_toi))

otherCounts <- data.frame(OTU = "Other",
                     Abundance = 1 - sample_sums(metaphlan_a_toi0),
                     SampleID = as.numeric(names(sample_sums(metaphlan_a_toi0))),
                     Genus="Other",
                     Species=" ",
                     stringsAsFactors = FALSE) %>%
                left_join(as.data.frame(PS_metaphlan_all@sam_data), by="SampleID")
  
#make tables for subsequent plotting:
metaphlA_toi_o_df <- metaphlan_a_toi0 %>%
                        psmelt() %>%
                        filter(Abundance > 0) %>%
                        full_join(otherCounts) %>%  # add "other"
                        mutate(Taxon=paste(Genus, Species))

metaphlA_toi_df <- metaphlan_a_toi %>%
                        psmelt() %>%
                        filter(Abundance > 0) %>%
                        mutate(Taxon=paste(Genus, Species))

#Lactobacillus at species level and all others at genus level
metaphlan_lg_toi0 <- PS_metaphlan_LG %>%
  subset_taxa((Genus=="Gardnerella"|Genus=="Lactobacillus" | Genus=="Mycoplasma" | Genus=="Atopobium" | Genus=="Prevotella") & (Species=="crispatus" | Species=="iners" | Species=="gasseri" | Species=="jensenii" | Species==" ")) #get samples of interest
summary(sample_sums(metaphlan_lg_toi0))

metaphlan_lg_toi <- metaphlan_lg_toi0 %>%
  transform_sample_counts(function(x) {x/sum(x)})
summary(sample_sums(metaphlan_lg_toi))

otherCounts1 <- data.frame(OTU = "Other",
                     Abundance = 1 - sample_sums(metaphlan_lg_toi0),
                     SampleID = as.numeric(names(sample_sums(metaphlan_lg_toi0))),
                     Genus="Other",
                     Species=" ",
                     stringsAsFactors = FALSE) %>%
                left_join(as.data.frame(PS_metaphlan_all@sam_data), by="SampleID")
  
#make tables for subsequent plotting:
metaphlLG_toi_o_df <- metaphlan_lg_toi0 %>%
                        psmelt() %>%
                        filter(Abundance > 0) %>%
                        full_join(otherCounts1) %>% # add "other" 
                        mutate(Taxon=paste(Genus, Species))

metaphlLG_toi_df <- metaphlan_lg_toi %>%
                        psmelt() %>%
                        filter(Abundance > 0) %>%
                        mutate(Taxon=paste(Genus, Species))

```

# Save phyloseq objects to an RData file
```{r}
save(PS_metaphlan_all, PS_metaphlan_LG, metaphlA_toi_o_df, metaphlA_toi_df, metaphlLG_toi_o_df, metaphlLG_toi_df, file = file.path(output_dir, "metaphlan_PS.RData"))
```

Session Info:
```{r}
sessionInfo()
```
