---
title: "BioLunch Figure"
author: "Hanna Berman"
date: "6/4/2019"
output: html_document
---

## Set up
### packages
```{r}
library(tidyverse)
```

### Data Paths
```{r}
repo_dir <- "/Users/hlberman/Documents/GitHub/pregnancy_metagenome/"

refDF_path <- file.path(repo_dir, "GardnerellaMetadata.csv")
refDF <- read_csv(refDF_path)

sgDF_path <- file.path(repo_dir, "shotgunMetadata.tsv")
sgDF <- read_delim(file = file.path(sgDF_path), delim = "\t")

sgGardDF <- "./gardRelativeAbundance.tsv" %>%
  read_tsv

S16gardDF_path <- "./16SgardVariants.tsv" %>%
S16gardDF <- read_tsv

#figure_out <- "/Volumes/GoogleDrive/My Drive/Callahan Lab/vagMicrobiomeMetagenome/gardMapping/clade_assignments"
```

```{r}
refDF %>%
  select(Strain, sixteenS_variant) %>%
  View()
```

## Unify dataframe
```{r}
S16gardDF0 <- S16gardDF %>%
  select(SampleID, Var, propGard) %>%
  mutate(Method="V4 region 16S amplicon")

sgGardDF0 <- sgGardDF %>%
  select(-n) %>%
  dplyr::rename(Var=Clade,
                propGard=propClade) %>%
  mutate(Var=case_when(Var==1~"Clade 1",
                       Var==2~"Clade 2",
                       Var==3~"Clade 3",
                       Var==4~"Clade 4",
                       Var==5~"Clade 5",
                       Var==6~"Clade 6"),
         Method="Shotgun")


gardVarDF <- S16gardDF0 %>%
  full_join(sgGardDF0, by=c("SampleID", "Var", "propGard", "Method")) %>%
  left_join(sgDF[,c("SampleID", "SubjectID", "GWcoll", "RNAlater")], by="SampleID") %>%
  filter(SubjectID!="10008",
         SubjectID!="10013",
         SubjectID!="10018",
         SubjectID!="10027",
         SubjectID!="10031",
         SubjectID!="10040",
         SubjectID!="19004")

S16gardDF %>%
  select(-OTU, -Abundance) %>%
  left_join(sgGardDF[,c("SampleID", "Clade", "propClade")], by="SampleID") %>%
  dplyr::rename(propGard16S=propGard,
                var16S=Var,
                propGardSG=propClade,
                cladeSG=Clade) %>%
  mutate(cladeSG=as.character(cladeSG)) %>%
  select(SampleID, SubjectID, GWcoll, RNAlater, var16S, propGard16S, cladeSG, propGardSG) -> gardVarDF0
```

```{r}
gardVarDF %>%
  filter(RNAlater=="No",
         SubjectID=="10621"|SubjectID=="19024")%>%
ggplot(aes(x=GWcoll, y=propGard, color=Var, shape=Method, size=Method)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 3)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "red", "blue1", "grey48"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)")) + # note: change labels to match Callahan 2017
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~SubjectID) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3)))
#ggsave(file.path(figure_out, "biolunchV4SG2samp.png"), width=8, height=6)
```

```{r}
gardVarDF %>%
  filter(RNAlater=="No")%>%
ggplot(aes(x=GWcoll, y=propGard, color=Var, shape=Method, size=Method)) +
  geom_point() +
  scale_size_manual(values=c(1.5, 3)) +
  scale_color_manual(values=c("salmon", "indianred", "lightblue", "blue3", "darkorchid2", "yellowgreen", "red", "blue1", "grey48", "darkorchid2", "yellowgreen"), labels=c("Clade C1 (shotgun)", "Clade C2 (shotgun)", "Clade C3 (shotgun)", "Clade C4 (shotgun)", "Clade C5 (shotgun)", "Clade C6 (shotgun)", "G2: Clades 1/2 (V4 amplicon)", "G1: Clades 3/4 (V4 amplicon)", "G3: indeterminant (V4 amplicon)", "G4: Clade 5 (V4 amplicon)", "G5: Clade 6 (V4 amplicon)")) + # note: change labels to match Callahan 2017
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~SubjectID, nrow=2) +
  labs(x="Gestation weeks at collection", y="Proportion of Gardnerella", shape="Method", color="Variant") +
  guides(color=guide_legend(override.aes=list(shape=15, size=3)))
#ggsave(file.path(figure_out, "biolunchV4SGv1.png"), width=8, height=6)
```

## Session Info
```{r}
sessionInfo()
```