---
title: "Assess Human Filtering"
author: "Hanna L. Berman"
output: html_document
---

Determine how well host reads were filtered from shotgun metagenomic sequencing data.

# Set up
## Packages
```{r}
library(tidyverse)
```

## File paths 
```{r}
#set paths
getwd()
countsDFbowtie <- file.path("fastq_counts.txt") %>%
  read_delim(delim="\ ")

#countsDFbbmapOLD <- file.path("fastq_counts_bbmapX.txt") %>%
#  read_delim(delim="\ ")

countsDFbbmap <- file.path("fastq_counts_bbmap.txt") %>%
  read_delim(delim="\ ")

#bbmapCompare <- countsDFbbmapOLD %>%
#  left_join(countsDFbbmap, by="Sample") %>%
#  select(Sample, filteredCount1.x, filteredCount1.y) %>%
#  dplyr::rename(old=filteredCount1.x,
#                new=filteredCount1.y) %>%
#  mutate(new=new/4,
#         old=old/4,
#    diff=new-old)
#summary(bbmapCompare$diff)

sgDF <- "../shotgunMetadata.tsv" %>%
  read_tsv()

krakenReads <- file.path("..", "methods_comparisons", "assess_kraken", "krakenHumanReads.tsv") %>%
  read_tsv()
```


# Compare amount of human reads filtered
```{r}
#bowtie by HLB counts
countsDhbBowtieF0 <- countsDFbowtie %>%
  mutate(uploadCount1=uploadCount1/4,
         uploadCount2=uploadCount2/4,
         uploadCountS=uploadCountS/4,
         filteredCount1=filteredCount1/4,
         filteredCount2=filteredCount2/4,
         filteredCountS=filteredCountS/4,
         uploadTotal=uploadCount1+uploadCount2, #+uploadCountS,
         filteredTotalBowtie=filteredCount1+filteredCount2, #+filteredCountS,
         pctFilteredBowtie=filteredTotalBowtie/uploadTotal,
         pctHumanBowtie=(uploadTotal-filteredTotalBowtie)/uploadTotal,
         SampleID=Sample) 

#bbmap by HLB counts
countsDFbbmap0 <- countsDFbbmap %>%
    mutate(uploadCount1=uploadCount1/4,
         uploadCount2=uploadCount2/4,
         uploadCountS=uploadCountS/4,
         filteredCount1=filteredCount1/4,
         filteredCount2=filteredCount2/4,
         filteredCountS=filteredCountS/4,
         uploadTotal=uploadCount1+uploadCount2,  #+uploadCountS,
         filteredTotalBbmap=filteredCount1+filteredCount2,  #+filteredCountS,
         pctFilteredBbmap=filteredTotalBbmap/uploadTotal,
         pctHumanBbmap=(uploadTotal-filteredTotalBbmap)/uploadTotal,
         SampleID=Sample) 

sgDF0 <- sgDF %>%
  select(SampleID, PctHuman) %>%
  dplyr::rename(Upload=PctHuman) %>%
  left_join(., countsDhbBowtieF0[,c("SampleID", "pctHumanBowtie")], by="SampleID") %>%
  left_join(., countsDFbbmap0[,c("SampleID", "pctHumanBbmap")], by="SampleID") %>%
  left_join(., krakenReads[,c("SampleID", "krakenHuAbundance")], by ="SampleID") %>%
  dplyr::rename(Kraken=krakenHuAbundance, BBmap=pctHumanBbmap, Bowtie=pctHumanBowtie) %>%
  gather("Method", "PctHuman", 2:ncol(.))
  
ggplot(sgDF0) +
  geom_line(aes(x=as.character(SampleID), y=PctHuman, color=Method, group=Method), show.legend = TRUE) +
  theme(axis.text.x = element_blank())+
  labs(x="Sample", y="Proportion Human Reads") +
  scale_color_manual(values=c("black", "red", "blue", "darkgreen"))
#ggsave("./filterHuman.pdf")
```
  
Add information to shotgun data frame
```{r}
sgDF1 <- countsDFbbmap0 %>%
  mutate(bbmapFiltered_pairs=filteredCount1,
         bbmapFiltered_sngl=filteredCountS,
         upload_pairs=uploadCount1,
         upload_sngl=uploadCountS) %>%
  select(SampleID, bbmapFiltered_pairs, bbmapFiltered_sngl, upload_pairs, upload_sngl) %>%
  left_join(sgDF, by="SampleID") %>%
  mutate(finalPctHuman_po=((Sickle_pairs-bbmapFiltered_pairs)/Sickle_pairs), #pairs only
         finalPctHuman_all=((Sickle_pairs+Sickle_sngl)-(bbmapFiltered_pairs+bbmapFiltered_sngl))/(Sickle_pairs+Sickle_sngl)) %>% 
  select(SampleIndex, SampleType, SubjectID, SampleID, GDcoll, GWcoll, GWdell, TermPre, Operator, TotalReads, TotalPairs, Sickle100_pe, Sickle_pairs, Sickle_sngl, PctTrim, Human_pairs, Human_sngl, PctHuman, upload_pairs, upload_sngl, bbmapFiltered_pairs, bbmapFiltered_sngl, finalPctHuman_po, finalPctHuman_all, SampleID_a, RNAlater, Cohort, PaperCohort, sampleOrder, sampleTrimester)

#write_tsv(sgDF1, "../shotgunMetadata.tsv")
```

```{r}
# bbmapReads <- sgDF %>%
#   select(SampleID, bbmapFiltered_pe, bbmapFiltered_s) %>%
#   mutate(totalBBmap=bbmapFiltered_pe+bbmapFiltered_s,
#          pctSingles=bbmapFiltered_s/totalBBmap)
# View(bbmapReads)
# summary(bbmapReads$pctSingles)
# 
# ggplot(bbmapReads, aes(x=pctSingles)) +
#   geom_histogram()
# 
# ggplot(bbmapReads, aes(x=1, y=pctSingles)) +
#   geom_boxplot() +
#   geom_jitter()
# 
# bbmapReads %>%
#   gather("p_s", "count", c(bbmapFiltered_pe, bbmapFiltered_s)) %>%
#   ggplot(., aes(x=reorder(as.character(SampleID), count), y=count, fill=p_s)) +
#   geom_bar(stat="identity")
# 
# 
# bbmapReads %>%
#   ggplot(aes(x=pctSingles, y=bbmapFiltered_pe))+
#   geom_jitter()
```

# Session Info
```{r}
sessionInfo()
```